<html>
<H2  ALIGN=CENTER><FONT COLOR="#DC143C">B.pl.html</H2></FONT>
<br>
<H6 ALIGN=RIGHT>Fri Jul 10 17:10:08 EDT 1998
Created by  <A href="./headbox2html.pl.html">./headbox2html.pl</A></H6> <hr><H3><A href="B
.pl.html">B
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//B
.pl">B
.pl</A>
<pre>
<b>Author    </b> : J. Park, Jason Johnson, Sarah Teichmann, Alex Bateman,
               Astrid Reinhardt, and anybody contributed.  jong@salt2.med.harvard.edu
<b>Example   </b> : require "B.pl"; BUT, I recommand you take subroutines out and
             use it directly or modify in your programs.
<b>Function  </b> : This is a comprehensive perl subroutine library developed
               under Bioperl project and others.
             URL: http://cyrah.med.harvard.edu/Bioperlsub/

             This serves as the depository database for various perl subroutines
              or algorithms developed in Bioinformatics and Genome projects.

             You can copy any of the sub routines in this file, modify, use
             in yours...
             PLEASE MODIFY AS FREELY AS YOU WANT !! All has the same PERL copyright

             All the subroutines are tested in small files
             If you want to have such single example program
             to see how they really work, pls contact me( A Biomatic )
             For example, a file called  'handle_arguments.pl' exists to
             test the subroutine 'handle_arguments'. Usually you can find them
             in  http://cyrah.med.harvard.edu/B.pl.html

<b>Keywords  </b> : Biology, perl library, sequence handling lib
<b>Options   </b> : nothing (used as subroutine library)

<b>Usage     </b> : require "B.pl"; ##<-- This is very slow, so you'd better
             copy the subroutines in your prog. or make a smaller lib files
               which are classified according to functions(like, Bio_Seq.pl
               for sequence handling, Bio_Array.pl for various array
               subroutines..), or make your own module out of this, do whatever
               you want....

<b>Version   </b> : 1.8    (April/27/1998)
<b>Warning   </b> : CopyLEFTed, for the enhancement of Biology, Biomatics, and Science.
             This is a development companion, nothing else.
             Class is for classification of my subroutines. If it is B, it can
             be useful for biological sequence data handling. If it's Utility,
             it can also be used for general purpose file handling stuff.
             File, Array, Hash,... are my classification items.
</pre><hr>

<H3><A href="handle_arguments
.pl.html">handle_arguments
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//handle_arguments
.pl">handle_arguments
.pl</A>
<pre>
<b>Argument  </b> : any type, any amount
<b>Category  </b> : general programming
<b>Example   </b> : 'handle_arguments(\@array, $string, \%hash, 8, 'any_string')
<b>Function  </b> : Sorts input arguments going into subroutines and returns default
             arrays of references for various types (file, dir, hash, array,,,,)
             If you give (\@out, @file), it will put @out into @array as a ref
             and also the contents of @out will be dereferenced and put to
             raw_string regardless what is in it).

<b>Keywords  </b> : handling arguments, parsing arguments,
<b>Returns   </b> : Following GLOBAL variables

             $num_opt,    @num_opt     @file          @dir
             $char_opt,   @char_opt    %vars          @array,
             @hash        @string,     @raw_string    @range,

             $num_opt has 10,20
             @num_opt has (10, 20)
             @file has  xxxx.ext
             @dir has  dir  or /my/dir
             $char_opt has 'A,B'
             @char_opt has (A, B)
             @array has  (\@ar1, \@ar2)
             @hash has (\%hash1, \%hash2)
             @string  ('sdfasf', 'dfsf')
             @raw_string (file.ext, dir_name, 'strings',,)
             @range has values like  10-20
             %vars deals with x=2, y=3 stuff.

<b>Usage     </b> : Just put the whole box delimited by the two '###..' lines below
             to inside of your subroutines. It will call 'handle_arguments'
             subroutine and parse all the given input arguments.
             To use, claim the arguments, just use the variable in the box.
             For example, if you had passed 2 file names for files existing
             in your PWD(or if the string looks like this: xxxx.ext),
             you can claim them by $file[0], $file[1] in
             your subroutine.
<b>Version   </b> : 4.8
</pre><hr>

<H3><A href="sort_by_E_values
.pl.html">sort_by_E_values
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//sort_by_E_values
.pl">sort_by_E_values
.pl</A>
<pre>
<b>Function  </b> : it sorts by the 2nd column(E-value, in msp file), small comes top
<b>Keywords  </b> : sort_by_2nd_column, sort_by_second_column, sort_by_e_values,
             sort_by_evalues,
<b>Usage     </b> : @out=@{&sort_by_E_values(\@input_line_array)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="sort_hash_value_by_column
.pl.html">sort_hash_value_by_column
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//sort_hash_value_by_column
.pl">sort_hash_value_by_column
.pl</A>
<pre>
<b>Example   </b> : Above will sort the file xxxx.msp by its 3rd column(numerically)
               small numbers will come to the top.
<b>Function  </b> : it sorts values of hash by the given column , small comes top. Unless number is
             is given, it sorts by the first column.
             It returnns ARRAY of the keys of the input HASH!!!

             It can handle gzipped file. It called gunzip to open and sort.

<b>Keywords  </b> : sort_by_2nd_column, sort_by_second_column, sort_by_e_values,
             sort_by_evalues, sort_hash_by_column, sort_value_by_column,
<b>Options   </b> : 
      s  for sorting stringwise
      d  for sorting by digit
      n  for sorting by digit(numerically)
   numerically  an alias of n

<b>Usage     </b> : @out=@{&sort_by_column(\%input_line_hash, <column num>)};
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="sort_by_column
.pl.html">sort_by_column
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//sort_by_column
.pl">sort_by_column
.pl</A>
<pre>
<b>Example   </b> : sort_by_column.pl 3 xxxx.msp
               Above will sort the file xxxx.msp by its 3rd column(numerically)
               small numbers will come to the top.
<b>Function  </b> : it sorts by the given column , small comes top. Unless number is
             is given, it sorts by the first column.

             It can handle gzipped file. It called gunzip to open and sort.

<b>Keywords  </b> : sort_by_2nd_column, sort_by_second_column, sort_by_e_values,
             sort_by_evalues,
<b>Options   </b> : 
      s  for sorting stringwise
      d  for sorting by digit
      n  for sorting by digit(numerically)
<b>Usage     </b> : @out=@{&sort_by_column(\@input_line_array, <column num>)};
<b>Version   </b> : 1.4
</pre><hr>

<H3><A href="sort_by_cluster_size
.pl.html">sort_by_cluster_size
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//sort_by_cluster_size
.pl">sort_by_cluster_size
.pl</A>
<pre>
<b>Function  </b> : it sorts by the 1st digit before '-'  as in 2-183_cluster, 2-140_cluster,
               etc.
<b>Keywords  </b> : sort_by_columns, sort_by_text_columns, sort_by_column_numerically
             sort_by_pattern
<b>Usage     </b> : @out=@{&sort_by_cluster_size(\@input_line_array)};
<b>Version   </b> : 1.2
</pre><hr>

<H3><A href="sort_by_column_bigger_first
.pl.html">sort_by_column_bigger_first
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//sort_by_column_bigger_first
.pl">sort_by_column_bigger_first
.pl</A>
<pre>
<b>Function  </b> : it sorts by the 2nd column(E-value, in msp file), small comes top
             by the help of  ts <decoux@moulon.inra.fr>
<b>Keywords  </b> : sort_by_columns, sort_by_text_columns, sort_by_column_numerically

<b>Usage     </b> : @out=@{&sort_by_column_bigger_first(\@input_line_array, 1)};
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="make_scrambled_seq_database
.pl.html">make_scrambled_seq_database
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//make_scrambled_seq_database
.pl">make_scrambled_seq_database
.pl</A>
<pre>
<b>Keywords  </b> : scramble_seq_database, create_scrambled_seq_database
<b>Usage     </b> : &make_reverse_seq_database(\@input_database_fasta_file);
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="make_2D_identity_matrix_array
.pl.html">make_2D_identity_matrix_array
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//make_2D_identity_matrix_array
.pl">make_2D_identity_matrix_array
.pl</A>
<pre>
<b>Function  </b> : @matrix is like  $matrix[1][2]=1;
             This assigns number 1 to array element
             If one array is given, it makes self to self matrix.
             When 2 are given, make matrix for the 2
<b>Keywords  </b> : make_matrix
<b>Options   </b> : 
    $skip_gap_char = g  for skipping gap char (any special char)
<b>Usage     </b> : @matrix=@{&make_2D_identity_matrix(\@seq1, \@seq2)};
<b>Version   </b> : 1.2
</pre><hr>

<H3><A href="make_2D_aa_residue_matrix_array
.pl.html">make_2D_aa_residue_matrix_array
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//make_2D_aa_residue_matrix_array
.pl">make_2D_aa_residue_matrix_array
.pl</A>
<pre>
<b>Function  </b> : @matrix is like  $matrix[1][2]='A'; when aa residue is identical
             This assigns identical residue to array element
             If one array is given, it makes self to self matrix.
             When 2 are given, make matrix for the 2
<b>Keywords  </b> : make_matrix
<b>Usage     </b> : @matrix=@{&make_2D_aa_residue_matrix_array(\@seq)};
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="make_2D_identity_matrix
.pl.html">make_2D_identity_matrix
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//make_2D_identity_matrix
.pl">make_2D_identity_matrix
.pl</A>
<pre>
<b>Function  </b> : @matrix is like  $matrix[1][2]=1;
             This assigns number 1 to array element
<b>Keywords  </b> : make_matrix, make_identity_matrix
<b>Options   </b> : 
        s  for show axis
<b>Usage     </b> : @matrix=@{&make_2D_identity_matrix(\$seq, [\$seq2] )};
<b>Version   </b> : 1.2
</pre><hr>

<H3><A href="amino_acid_homology_matrix
.pl.html">amino_acid_homology_matrix
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//amino_acid_homology_matrix
.pl">amino_acid_homology_matrix
.pl</A>
<pre>
<b>Keywords  </b> : are_they_homologous, amino_acid_homology_table, compare_amino_acid_homology
<b>Usage     </b> : $yes_no=${&amino_acid_homology_matrix('E', 'D')};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="write_reverse_seq_files
.pl.html">write_reverse_seq_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//write_reverse_seq_files
.pl">write_reverse_seq_files
.pl</A>
<pre>
<b>Author    </b> : jong@salt2.med.harvard.edu
<b>Keywords  </b> : write_rev_seq_files, write_reverse_msf_files
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="make_hmm_from_alignment
.pl.html">make_hmm_from_alignment
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//make_hmm_from_alignment
.pl">make_hmm_from_alignment
.pl</A>
<pre>
<b>Keywords  </b> : 
: HMM, hidden markov model, make_HMM_from_alignment,
             make_hmm_from_msf_file, create_hmm_from_alignment,
             create_hmm_from_msf_file,
<b>Usage     </b> : &make_reverse_seq_database(\@input_database_fasta_file);
: @out_hmm_file_names=@{&make_hmm_from_alignment(\@file, "$over_write")};
<b>Version   </b> : 1.2
: 1.1
</pre><hr>

<H3><A href="make_sequence_match_table
.pl.html">make_sequence_match_table
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//make_sequence_match_table
.pl">make_sequence_match_table
.pl</A>
<pre>
<b>Example   </b> : 

  OUTPUT looks like the following;
 d1dvh__=d1fcdc1     7.1e-08
 d1fcdc1=d1dvh__     7.1e-08
 d5cytr_=d351c__     5.3e-08
 d351c__=d5cytr_     5.3e-08
 d1cyi__=d2mtac_     9.1e-06
 d2mtac_=d1cyi__     9.1e-06
 d1cyi__=d5cytr_     0.00045
 d5cytr_=d1cyi__     0.00045

: 

  INPUT looks like this: (the iss file format), first column is key

   d1ten__(110)(0.00031)     d1fna__    d1fna___1-91(578)(6.9e-37)       d1ten__(110)(0.00031)
   d1cfb_2(255)(7.8e-16)     d1cfb_2    HSU55258_741-838(255)(5.6e-12)   d1cfb_2(255)(7.8e-16)

  OUTPUT looks like the following;
   d1dvh__=d1fcdc1    Correct: 7.1e-08
 d1fcdc1=d1dvh__    Correct: 7.1e-08
 d5cytr_=d351c__    Correct: 5.3e-08
 d351c__=d5cytr_    Correct: 5.3e-08
 d1cyi__=d2mtac_    Wrong:   9.1e-06

<b>Function  </b> : gets sequences which are wrongly matched from intermediate seq search
: makes a table of match with the values for E values.
<b>Keywords  </b> : 
: make_sequence_match_Evalue_table, Evalue_table, make_Evalue_table
             make_iss_sequence_match_table
<b>Options   </b> : _  for debugging.
             #  for debugging.

: _  for debugging.
             #  for debugging.
             s  for skip SELF to SELF match entries
             w  for Smith-Waterman score result out than E value out
             r  for reflexive output

<b>Reference </b> : http://sonja.acad.cai.cam.ac.uk/perl_for_bio.html
<b>Usage     </b> : %seq=%{&get_false_positive_seq_matches(\%msp_1, \%msp2)};
: %sequence_match_table=%{&make_sequence_match_table(\%msp_1, \%msp2)};
<b>Version   </b> : 1.0
: 1.5
<b>Warning   </b> : The default is to show the best E value(lowest that is)

</pre><hr>

<H3><A href="write_iss_file
.pl.html">write_iss_file
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//write_iss_file
.pl">write_iss_file
.pl</A>
<pre>
<b>Function  </b> : writes the intermediate sequence search file.
<b>Keywords  </b> : write_interm_seq_search_file
             v  for showing the output in STDOUT
<b>Reference </b> : http://sonja.acad.cai.cam.ac.uk/perl_for_bio.html
<b>Usage     </b> : &write_iss_file(\%msp1, \%msp2);  ## for 2 msp_x file input
<b>Version   </b> : 1.2
<b>Warning   </b> : You MUST NOT delete '# options : ..' entry
              as it is read  by various subroutines.
</pre><hr>

<H3><A href="get_overlapping_seq_match_size
.pl.html">get_overlapping_seq_match_size
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_overlapping_seq_match_size
.pl">get_overlapping_seq_match_size
.pl</A>
<pre>
<b>Keywords  </b> : CF: get_overlapping_range, get_overlapping_seq_match
<b>Usage     </b> : $ovlapsize=${&get_overlapping_seq_match_size($st1, $en1, $st2, $en2)
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="get_unix_shell_name
.pl.html">get_unix_shell_name
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_unix_shell_name
.pl">get_unix_shell_name
.pl</A>
<pre>
<b>Author    </b> : jong@salt2.med.harvard.edu, On commercial use issue, Email me.
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_stat_FASTA_search_result_in_msp_0_files
.pl.html">get_stat_FASTA_search_result_in_msp_0_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_stat_FASTA_search_result_in_msp_0_files
.pl">get_stat_FASTA_search_result_in_msp_0_files
.pl</A>
<pre>
<b>Category  </b> : statistics, search, bio
: 
<b>Keywords  </b> : get_stat00_result, get_stat_msp0_files, get_stat_single_search_result
<b>Options   </b> : 

  $E_value= by e=
  $verbose=v by v
  $show_options=o by o
  $step   =  by s=
  $score_thresh1=   by t1=
  $score_thresh2=   by t2=
  $E_mult_factor1 = by m1=
  $E_mult_factor2 = by m2=

<b>Usage     </b> : &get_stat_FASTA_search_result_in_msp_0_files(\@file);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_scop_correcting_pairs
.pl.html">get_scop_correcting_pairs
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_scop_correcting_pairs
.pl">get_scop_correcting_pairs
.pl</A>
<pre>
<b>Keywords  </b> : get_pdb_correcting_pairs ,
<b>Usage     </b> : %correct=%{&get_scop_correcting_pairs()};
<b>Version   </b> : 1.3
</pre><hr>

<H3><A href="get_isearch_result_stat
.pl.html">get_isearch_result_stat
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_isearch_result_stat
.pl">get_isearch_result_stat
.pl</A>
<pre>
<b>Author    </b> : A Scientist
<b>Example   </b> : Following input (hash eg: %stat2, input with the first word as key)
              will become columnar output.

    d1ash__ d1bam__ d1mba__ d2lhb__
    d1baba_ d1flp__ d1hbg__ d1hlb__ d1mba__ d1mbd__ d2lhb__ d3aaha_ d3sdha_
    d1cpca_ d1cpcb_ d1gof_1 d2ts1_1

    Will become:
      ....
      d1ash__ d2lhb__ Homolog: G1   98 0.012
      d1baba_ d1flp__ Homolog: G1   82 0.072
      d1baba_ d1hbg__ Homolog: G1   79 0.13
      d1baba_ d2lhb__ Homolog: G1   228 8e-12
      d1baba_ d3aaha_ Nomolog: G1   74 2
      d1baba_ d3sdha_ Homolog: G1   92 0.012
      d1cola_ d1hbg__ Nomolog: G1   79 0.59
      d1cpca_ d1cpcb_ Homolog: G1   176 4.9e-08
      ....

<b>Keywords  </b> : get_stat_interm_search, get_intermediate_search_stat
<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Package   </b> : Bio
<b>Reference </b> : http://sonja.acad.cai.cam.ac.uk/perl_for_bio.html
<b>Returns   </b> : [$av_correct, $num_enq_seq]
<b>Usage     </b> : &get_self_isearch_stat(\%stat2, \@pdbg_seqs, \$evalue);
<b>Version   </b> : 2.1
</pre><hr>

<H3><A href="open_sequence_index_files
.pl.html">open_sequence_index_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_sequence_index_files
.pl">open_sequence_index_files
.pl</A>
<pre>
<b>Example   </b> : 
: %index=%{&open_sequence_index_files(\@INDEX_FILE, \@input_seq_names)};
<b>Function  </b> : 
: returns seqname with its seek pos in fasta sequence db file.
<b>Keywords  </b> : remove_sequence_ranges, remove_sequence_name_ranges,
             remove_ranges_in_sequences, strip_sequence_name_ranges,
: open_seq_index_files, open_seq_idx_files, open_idx_files,
             get_sequence_index, get_seq_index, get_sequence_with_index
<b>Options   </b> : _  for debugging.
             #  for debugging.
: _ or # for debugging
<b>Usage     </b> : 
: open_sequence_index_files(<indexfilename>, <sequencename>);
<b>Version   </b> : 1.0
: 1.2
<b>Warning   </b> : You MUST NOT delete '# options : ..' entry
              as it is read  by various subroutines.
: You MUST NOT delete '# options : ..' entry
              as it is read  by various subroutines.
</pre><hr>

<H3><A href="do_intermediate_sequence_search
.pl.html">do_intermediate_sequence_search
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//do_intermediate_sequence_search
.pl">do_intermediate_sequence_search
.pl</A>
<pre>
<b>Example   </b> : &do_intermediate_sequence_search(\%pdb_seq, $owl_db_fasta, $ARGV[0], $single_msp, $over_write,
                    "u=$upper_expect_limit", "l=$lower_expect_limit", "k=$k_tuple" );

<b>Options   </b> : 
             Query_seqs=  for enquiry sequences eg)  "Query_seqs=$ref_of_hash"
             DB=   for target DB  "DB=$DB_used"
             File= to get file base(root) name.  "File=$file[0]"
             m  for MSP format directly from FASTA or Ssearch result than through sso_to_msp to save mem
             s  for the big single output (msp file output I mean)
             o  for overwrite existing xxxx.fa files for search
             c  for create SSO file (sequence search out file)
             R  for adding ranges to the enquiry sequences as well.
             k= for k-tuple value. default is 1 (ori. FASTA prog. default is 2)
             u= for $upper_expect_limit
             l= for $lower_expect_limit
             a= for choosing either fasta or ssearch algorithm

<b>Returns   </b> : the names of files created (xxxxx.msp, yyy.msp,,)
<b>Usage     </b> : &do_intermediate_sequence_search(\%pdb_seq, $owl_db_fasta, $ARGV[0], $single_msp, $over_write,
                    "u=$upper_expect_limit", "l=$lower_expect_limit", "k=$k_tuple" );

<b>Version   </b> : 1.0
<b>Warning   </b> : You MUST NOT delete '# options : ..' entry
              as it is read  by various subroutines.
</pre><hr>

<H3><A href="do_sequence_search
.pl.html">do_sequence_search
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//do_sequence_search
.pl">do_sequence_search
.pl</A>
<pre>
<b>Example   </b> : &do_sequence_search(\%pdb_seq, $owl_db_fasta, $ARGV[0], $single_msp, $over_write,
                    "u=$upper_expect_limit", "l=$lower_expect_limit", "k=$k_tuple" );

<b>Function  </b> : do FASTA, SSEARCH or BLASTPGP(psi-blast) search
<b>Keywords  </b> : sequence_search
<b>Options   </b> : 
             Query_seqs=  for enquiry sequences eg)  "Query_seqs=$ref_of_hash"
             DB=   for target DB  "DB=$DB_used"
             File= to get file base(root) name.  "File=$file[0]"
             m  for MSP format directly from FASTA or Ssearch result than through sso_to_msp to save mem
             s  for the big single output (msp file output I mean)
             s= for the single big msp file name
             o  for overwrite existing xxxx.fa files for search
             c  for create SSO file (sequence search out file)
             d  for very simple run and saving the result in xxxx.gz format in sub dir starting with one char
             r  for reverse the query sequence
             R  for attaching ranges of sequences
             k= for k-tuple value. default is 1 (ori. FASTA prog. default is 2)
             u= for $upper_expect_limit
             l= for $lower_expect_limit
             a= for choosing either fasta or ssearch algorithm
             d= for defining the size of subdir made. 2 means it creates
                    eg, DE while 1 makes D
             d  for $make_gz_in_sub_dir_opt, putting resultant sso files in gz format and in single char subdir
             D  for $make_msp_in_sub_dir_opt, convert sso to msp and put in sub dir like /D/, /S/
             n  for new format to create new msp file format with sso_to_msp routine
          PVM=  for PVM run of FASTA (FASTA only)
             M  for machine readable format -m 10 option
             M= for machine readable format -m 10 option
             N  for 'NO' do not do any processing but, do the searches only.
       FILE_AGE for defining the age of file in days to be overwritten.
             L  for Lean output(removes xxxx.fa query seq file)

<b>Returns   </b> : the names of files created (xxxxx.msp, yyy.msp,,)
<b>Usage     </b> : &do_sequence_search("Query_seqs=\%pdb_seq", "DB=$sequence_db_fasta",
             "File=$ARGV[0]", $single_msp, $over_write,
            "u=$upper_expect_limit", "l=$lower_expect_limit",
          "k=$k_tuple", $No_processing );
<b>Version   </b> : 5.1
</pre><hr>

<H3><A href="do_hmm_sequence_search
.pl.html">do_hmm_sequence_search
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//do_hmm_sequence_search
.pl">do_hmm_sequence_search
.pl</A>
<pre>
<b>Function  </b> : does hmm sequence search using Sean Eddy's HMMER (hmmls, hmmfs)
<b>Keywords  </b> : do_seq_search_with_hmm, do_hmmt_sequence_search
<b>Options   </b> : 
    "method=ls"  for turning hmmls search option on (default)
    "method=fs"  for turning hmmfs search option on
    method= by method=
   o  for overwriting existint xxxxx.hmm files
   E=Enguiry_name    for specifying enquiry seq name rather than 'HMM', the default
   t=20  for score thresh at the level of hmmls. Default of hmmls is 0. example showed has 15
   $evalue_cutoff= by e=
$over_write = o by -o o
<b>Usage     </b> : &do_hmm_sequence_search(\@file, "method=$default_search_method",
        $over_write, "DB=$pdbd40_seq_fasta");

<b>Version   </b> : 1.6
</pre><hr>

<H3><A href="divide_clusters
.pl.html">divide_clusters
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//divide_clusters
.pl">divide_clusters
.pl</A>
<pre>
<b>Example   </b> : &divide_clusters(\@file, $verbose, $range, $merge, $sat_file,
                  $dindom, $indup, "T=$length_thresh", "e=$evalue", $over_write,
                   $optimize, "s=$score", "f=$factor");

<b>Function  </b> : This is the main funciton for divclus.pl
               divides complex single linkage cluster into smaller duplication
               module level sub clusters.
<b>Keywords  </b> : divicl, divclus, div_clus, divide clusters
<b>Options   </b> : _  for debugging.
   f=<digit>   for determing the factor in filtering out non-homologous
                  regions, 7 = 70% now!!
   l=<digit>   for seqlet(duplication module) length threshold
   t=<digit>   for seqlet(duplication module) length threshold
                  (same as l opt, confusing, huh? )
   s=<digit>   for score threshold
   e=<digit>   for evalue threshold
   z           for activating remove_similar_sequences, rather than remove_dup....
   o           for overwriting
   v           for verbose printout (infor)
   D           for dynamic factor
   S  $short_region=  S by S -S  # taking shorter region overlap in removing similar reg
   L  $large_region=  L by L -L  # taking larger  region overlap in removing similar reg
   A  $average_region=A by A -A  # taking average region overlap in removing similar reg

<b>Usage     </b> : &divide_clusters(\@file);
<b>Version   </b> : 2.8
</pre><hr>

<H3><A href="remove_similar_seqlets
.pl.html">remove_similar_seqlets
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//remove_similar_seqlets
.pl">remove_similar_seqlets
.pl</A>
<pre>
<b>Example   </b> : @seqlets=@{&remove_similar_seqlets(\@mrg1, $mrg2, \@mrg3)};
               while @mrg1=qw(M_2-100 M_2-110 M_8-105 M_4-108 N_10-110 N_12-115);
                     $mrg2='Z_3-400 Z_2-420';
                     @mrg3=('X_2-300 X_3-300', 'X_2-300', 'X_5-300 X_2-301' );
<b>Function  </b> : merges(gets average starts and ends ) of similar
             seqlets to reduce them into smaller numbers. This can also handle
              names like XLBGLO2R_8-119_d1hlm__.

<b>Keywords  </b> : merge_sequence_names, merge_seq_names, merge_sequence_ranges
             merge_seq_ranges
<b>Options   </b> : _  for debugging.
             #  for debugging.
             f= for factor
             S  for shorter region matched is used
             A  for average region matched is used
             L  for larger region matched is used

<b>Usage     </b> : @seqlets=@{&remove_similar_seqlets(\@split)};
<b>Version   </b> : 2.0
<b>Warning   </b> : You MUST NOT delete '# options : ..' entry
              as it is read  by various subroutines.
</pre><hr>

<H3><A href="show_subclusterings
.pl.html">show_subclusterings
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//show_subclusterings
.pl">show_subclusterings
.pl</A>
<pre>
<b>Example   </b> : @temp_show_sub=&show_subclusterings(\@out, $file, $sat_file, $dindom, $indup);
<b>Function  </b> : This is the very final sub of divclus.pl
<b>Keywords  </b> : print_subclusterings, sum_subclusterings, write_subclustering
             show_clusterings, display_subclusterings
<b>Options   </b> : 
             f  for file output, eg: xxxxxxx.sat

<b>Usage     </b> : &show_subclusterings(\@out);
<b>Version   </b> : 2.6
<b>Warning   </b> : You MUST NOT delete '# options : ..' entry
              as it is read  by various subroutines.
</pre><hr>

<H3><A href="exchange_query_with_match_in_msp
.pl.html">exchange_query_with_match_in_msp
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//exchange_query_with_match_in_msp
.pl">exchange_query_with_match_in_msp
.pl</A>
<pre>
<b>Keywords  </b> : swap_query_with_match_in_msp, invert_query_with_match_in_msp,
             swap_query_seq_with_match_seq_in_msp,
<b>Usage     </b> : @exchanged_msp=@{&exchange_query_with_match_in_msp(\@file)};
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="get_internal_dup_in_a_cluster
.pl.html">get_internal_dup_in_a_cluster
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_internal_dup_in_a_cluster
.pl">get_internal_dup_in_a_cluster
.pl</A>
<pre>
<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Version   </b> : 1.1
<b>Warning   </b> : You MUST NOT delete '# options : ..' entry
              as it is read  by various subroutines.
</pre><hr>

<H3><A href="get_domain_inside_domain
.pl.html">get_domain_inside_domain
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_domain_inside_domain
.pl">get_domain_inside_domain
.pl</A>
<pre>
<b>Keywords  </b> : find_dindoms, domain_inside_domain, domain_in_domain
<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Version   </b> : 1.0
<b>Warning   </b> : You MUST NOT delete '# options : ..' entry
              as it is read  by various subroutines.
</pre><hr>

<H3><A href="scale_for_horizontal_histogram
.pl.html">scale_for_horizontal_histogram
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//scale_for_horizontal_histogram
.pl">scale_for_horizontal_histogram
.pl</A>
<pre>
<b>Function  </b> : used to make things like:

<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Version   </b> : 1.0
<b>Warning   </b> : You MUST NOT delete '# options : ..' entry
              as it is read  by various subroutines.
</pre><hr>

<H3><A href="get_added_matched_regions_in_msp
.pl.html">get_added_matched_regions_in_msp
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_added_matched_regions_in_msp
.pl">get_added_matched_regions_in_msp
.pl</A>
<pre>
<b>Function  </b> : This reads MSP file regions matched for a target seq
             and adds things up to plot horizontally.
<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Version   </b> : 1.0
<b>Warning   </b> : You MUST NOT delete '# options : ..' entry
              as it is read  by various subroutines.
</pre><hr>

<H3><A href="cluster_merged_seqlet_sets
.pl.html">cluster_merged_seqlet_sets
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//cluster_merged_seqlet_sets
.pl">cluster_merged_seqlet_sets
.pl</A>
<pre>
<b>Options   </b> : _  for debugging.
             #  for debugging.
  $short_region=  S by S -S  # taking shorter region overlapped in removing similar regions
  $large_region=  L by L -L  # taking larger  region overlapped in removing similar regions
  $average_region=A by A -A # taking average region overlapped in removing similar regions

<b>Usage     </b> : @out=@{&cluster_merged_seqlet_sets(\@lines)};
<b>Version   </b> : 1.5
<b>Warning   </b> : You MUST NOT delete '# options : ..' entry
              as it is read  by various subroutines.
</pre><hr>

<H3><A href="check_linkage_of_2_similar_seqlet_sets
.pl.html">check_linkage_of_2_similar_seqlet_sets
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//check_linkage_of_2_similar_seqlet_sets
.pl">check_linkage_of_2_similar_seqlet_sets
.pl</A>
<pre>
<b>Function  </b> : connects two clusters of seqlets if they share
              identical or near identical seqlets
<b>Keywords  </b> : check_link, check_relation, check_relatedness
<b>Options   </b> : _  for debugging.
  $factor = by f=  # eg)  "f=$factor" in the higher level sub

<b>Version   </b> : 1.7
<b>Warning   </b> : You MUST NOT delete '# options : ..' entry
              as it is read  by various subroutines.
</pre><hr>

<H3><A href="merge_arrays_by_common_elements
.pl.html">merge_arrays_by_common_elements
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//merge_arrays_by_common_elements
.pl">merge_arrays_by_common_elements
.pl</A>
<pre>
<b>Function  </b> : merges arrays if there are common array elements.
             if @A has (1,2,3) and @B has (2, 4, 5), they share 2, so
             they are merged to be (1,2,3,4,5)
<b>Keywords  </b> : cluster_arrays_by_common_elements, merge_arrays_if_common_elements
             merge_array_if_common_elements, merge_arrays_when_common_elements_occur
             merge_arrays
<b>Usage     </b> : @out=@{&merge_arrays_by_common_elements(\@ref_of_arrays)}
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="check_parf_files
.pl.html">check_parf_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//check_parf_files
.pl">check_parf_files
.pl</A>
<pre>
<b>Example   </b> : 
    PARF file looks like this>
   d1nsca_   d3nn9__   Homolog -664.92 2.43.1.1.3  2.43.1.1.2
   d1dppa_   d2olba_   Homolog -617.41 3.68.1.1.6  3.68.1.1.1
   d2ach.1a1 d9api.1a1 Homolog -556.38 5.2.1.1.3   5.2.1.1.4
<b>Function  </b> : checks if given file(s) is a parf file and returns the number of
              identified parf file. If you check 2 files and both are parf, you
               will get (\$num_of_parf_file) value of 2.
<b>Usage     </b> : $number_of_parf=${&check_parf_files(@input)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="check_common_elements_in_array
.pl.html">check_common_elements_in_array
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//check_common_elements_in_array
.pl">check_common_elements_in_array
.pl</A>
<pre>
<b>Function  </b> : accepts 1 or 2 refs of arrays and checks if there is any
             common(repeating) elements between the two (or inside one)
             The result is either ref of 1, or 0
<b>Keywords  </b> : is_there_common_element, if_common_elements
<b>Usage     </b> : &check_common_elements_in_array($mother_array[$i], $mother_array[$i+1]));
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="link_ranges
.pl.html">link_ranges
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//link_ranges
.pl">link_ranges
.pl</A>
<pre>
<b>Example   </b> : INPUT:

   @input=( '1-30 1-40 1-50',
            '2-49 4-40 2-99'....)

<b>Function  </b> : merges ranges(10-20, 11-21 etc) when there is any overlap
              is present
             If you put a reverse range like '2000-20', it will
              complain and reverse the order and do the job after correction.

<b>Keywords  </b> : connect_ranges, link_overlapping_ranges, connect_overlapping_ranges
<b>Options   </b> : _  for debugging.
<b>Usage     </b> : @all_ranges = @{&link_ranges(@all_ranges)};
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="merge_similar_ranges
.pl.html">merge_similar_ranges
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//merge_similar_ranges
.pl">merge_similar_ranges
.pl</A>
<pre>
<b>Example   </b> : INPUT:

   @input=( '1-30 1-40 1-50',
            '2-49 4-40 2-99'....)

<b>Function  </b> : merges ranges(10-20, 11-21 etc) when there is any overlap
              is present (resulting in average start and end at each level)
             If you put a reverse range like '2000-20', it will
              complain and reverse the order and do the job after correction.

<b>Keywords  </b> : merge_similar_regions, merge_ranges, merge_regions,
              merge_sequence_ranges, merge_overlap_ranges, connect_ranges
              connect_overlapping_ranges, connect_similar_ranges,
              remove_similar_ranges
<b>Options   </b> : f=   for setting factor (0.7 for 70% overlap minimum)

<b>Usage     </b> : @all_ranges = @{&merge_similar_seqlets(@all_ranges)};
<b>Version   </b> : 1.3
</pre><hr>

<H3><A href="merge_similar_seqlets
.pl.html">merge_similar_seqlets
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//merge_similar_seqlets
.pl">merge_similar_seqlets
.pl</A>
<pre>
<b>Example   </b> : INPUT:

   @input=( 'seq1_1-30 seq2_1-40 seq3_1-50',
            'seq1_2-49 seq3_4-40 seq4_2-99'....)

   @output=('seq1_1-30 seq2_1-45 seq3_2-45 seq4_2-99');

<b>Function  </b> : merges seqlet sets which have identical
             sequences and share similar regions by connection factor of 30%
             This means, if any two seqlets from the same sequences which
             share more than 70% seqlet regions overlapping are merged
             This only sees the very first sequence in the seqlets line!!!
             (so, PARTIAL MERGE !!)
<b>Keywords  </b> : merge_similar_sequences, merge_sequence_names, merge_sequences,
              merge_sequence_ranges, merge_similar_sequences_with_ranges,
              merge_seqlets, merge_duplication_modules
<b>Options   </b> : 

   f=<digit>   for determing the factor in filtering out non-homologous
                  regions, 7 = 70% now!!
   l=<digit>   for seqlet(duplication module) length threshold
   z           for activating remove_similar_sequences, rather than remove_dup....
   S  $short_region=  S by S -S  # taking shorter region overlap in removing similar reg
   L  $large_region=  L by L -L  # taking larger  region overlap in removing similar reg
   A  $average_region=A by A -A  # taking average region overlap in removing similar reg

<b>Usage     </b> : @all_seqlets = @{&merge_similar_seqlets(@all_seqlets)};
<b>Version   </b> : 2.0
</pre><hr>

<H3><A href="sort_by_digits_in_string
.pl.html">sort_by_digits_in_string
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//sort_by_digits_in_string
.pl">sort_by_digits_in_string
.pl</A>
<pre>
<b>Function  </b> : sorts arrays of strings like

   MJ0228_314-573 MJ1197_348-601
   MJ0228_451-576 sll0078_502-594 sll1425_489-611
   MJ0228_479-572 sll0078_502-594

   According to the digits after seq names _314-, _451-, _479-
    in the above
   This only looks at the very first sequence in the string

<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Version   </b> : 1.4
<b>Warning   </b> : You MUST NOT delete '# options : ..' entry
              as it is read  by various subroutines.
</pre><hr>

<H3><A href="sort_words_in_string
.pl.html">sort_words_in_string
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//sort_words_in_string
.pl">sort_words_in_string
.pl</A>
<pre>
<b>Function  </b> : sort words in strings sperated by ' ' or "\n"
<b>Keywords  </b> : sort_words_in_sequences, sort_sequences_in_string,
             sort_strings_in_string, sort_string_by_words, sort_elements_in_string
<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Version   </b> : 1.1
<b>Warning   </b> : You MUST NOT delete '# options : ..' entry
              as it is read  by various subroutines.
</pre><hr>

<H3><A href="convert_hmmls_to_msp_files
.pl.html">convert_hmmls_to_msp_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//convert_hmmls_to_msp_files
.pl">convert_hmmls_to_msp_files
.pl</A>
<pre>
<b>Keywords  </b> : convert_hmmls_to_msp
<b>Options   </b> : 
   S=$single_out_file_name   for producing single msp file with all the hmmls contents
   E=Enguiry_name    for specifying enquiry seq name rather than 'HMM', the default
   $bit_score_threshold= by t=
<b>Usage     </b> : @out=@{&convert_hmmls_to_msp_files(\@file)};
<b>Version   </b> : 1.4
</pre><hr>

<H3><A href="convert_mmp_to_mrg
.pl.html">convert_mmp_to_mrg
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//convert_mmp_to_mrg
.pl">convert_mmp_to_mrg
.pl</A>
<pre>
<b>Example   </b> : 
  Example OUT as string

   slr1950 sll1920 sll0672 sll1076 sll1614 slr0797 slr0798 slr0822 slr1729
   slr1729 sll1076 sll0672 sll1614 sll1920 slr0797 slr0798 slr0822 slr1950

<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Version   </b> : 1.1
<b>Warning   </b> : You MUST NOT delete '# options : ..' entry
              as it is read  by various subroutines.
</pre><hr>

<H3><A href="add_ranges_in_msp_line
.pl.html">add_ranges_in_msp_line
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//add_ranges_in_msp_line
.pl">add_ranges_in_msp_line
.pl</A>
<pre>
<b>Function  </b> : this adds ranges to the seqnames of msp files
             mmp line is msp line with additional sequences at the end
<b>Keywords  </b> : convert_msp_to_mmp, convert_msp, convert_msp_2_mmp
             change_msp_to_mmp, add_range_in_msp, convert_msp_line_to_mmp_line
<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Version   </b> : 1.5
</pre><hr>

<H3><A href="convert_msp_line_to_mmp_line
.pl.html">convert_msp_line_to_mmp_line
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//convert_msp_line_to_mmp_line
.pl">convert_msp_line_to_mmp_line
.pl</A>
<pre>
<b>Function  </b> : this adds ranges to the seqnames of msp files
             mmp line is msp line with additional sequences at the end
<b>Keywords  </b> : convert_msp_to_mmp, convert_msp, convert_msp_2_mmp
             change_msp_to_mmp, add_range_in_msp
<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Version   </b> : 1.5
</pre><hr>

<H3><A href="merge_sequence_alignments
.pl.html">merge_sequence_alignments
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//merge_sequence_alignments
.pl">merge_sequence_alignments
.pl</A>
<pre>
<b>Keywords  </b> : combine_sequence_alignment, merge_sequence_alignment_pairs
             merge_seq_alignment, make_interm_alignment, make_3_way_alignment
             merge_alignment, combine_alignment
<b>Options   </b> : 
    l=  for sequence block length by print_seq_in_block subroutine
    t=  for specifying the length of seq names shown.
    t   for truncating the seq names in printing out.
    s   for sorting the final output lines (default anyway for print_seq_in_block)

<b>Usage     </b> : &merge_sequence_alignments(@seq);  while @seq has
              @seq=(\%hash1, \%hash2);  while %hash1 and %hash2 have
    %hash1=qw(seq1 ANN-NTMQQRRQQQRKRRRQQQSSSSTTST seq2 --NNN--QQ--QQQ--RRRR--SSSS--);
    %hash2=qw(seq2 NN-QQQQQ--RRRR----SS--SS---    seq3 -NNXQQQXQRTRRRXTTSTSSMMSSTTT);

<b>Version   </b> : 1.3
</pre><hr>

<H3><A href="merge_sequence_in_msp_file
.pl.html">merge_sequence_in_msp_file
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//merge_sequence_in_msp_file
.pl">merge_sequence_in_msp_file
.pl</A>
<pre>
<b>Example   </b> : INPUT: (MSP file) ===>
  59     2.6        47    64     d2pia_3        10    30     d1erd___10-30
  161    1.1e-07    24    91     d2pia_3        16    85     d1frd___16-85

  722    0          1     106    d1put__        1     106    d1put___1-106
  66     4.9        2     68     d1put__        43    106    d2lbp___43-106
  69     1.3        12    49     d1put__        81    120    d1cgo___81-120

  60     3.3        13    38     d1frd__        32    57     d1orda1_32-57
  65     1.7        21    58     d1frd__        40    69     d2mtac__40-69

   ==== OUTPUT ===>
    d1frd___1-98 d1frd___1-98_1-98 d1frd___16-85 d2pia_3_24-91_24-91
    d1frd___16-85_16-85 d2pia_3_24-91
    d1put___1-106 d1put___1-106_1-106
    d2pia_3_1-98 d2pia_3_1-98_1-98

<b>Keywords  </b> : mergr_seq_in_msp_file, merge_sequence_in_msp, merge_sequences_in_msp_file
<b>Options   </b> : 
$dynamic_factor =  y by y -y   # adjusting factor value dynamically(more seq higher factor)
$short_region   =  S by S -S  # taking shorter region overlapped in removing similar regions
$large_region   =  L by L -L  # taking larger  region overlapped in removing similar regions
$average_region =  A by A -A # taking average region overlapped in removing similar regions

<b>Version   </b> : 2.7
</pre><hr>

<H3><A href="merge_sequence_in_msp_chunk
.pl.html">merge_sequence_in_msp_chunk
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//merge_sequence_in_msp_chunk
.pl">merge_sequence_in_msp_chunk
.pl</A>
<pre>
<b>Function  </b> : merges sequences which are linked by common regions
             This filters the sequences by evalue and ssearch score
             This is the main algorithm of merging similar sequences.
<b>Keywords  </b> : connect_sequence_in_msp, link_sequence_in_msp_chunk
             connect_sequence_in_msp_chunk, link_sequence_in_msp
             merge_sequence, link_sequence, connect_sequence
<b>Options   </b> : _  for debugging.
             #  for debugging.
             m  for merge file output format (.mrg)
             t= for threshold of seqlet length eg)  "t=30"
             f= for overlap factor (usually between 2 to 7 )
                 2 means, if the two regions are not overlapped
                  by more than HALF of of the smaller region
                  it will not regard as common seqlet block
             s= for ssearch score minimum
             e= for ssearch e value maximum
             S  for S -S  # taking shorter region overlapped in removing similar regions
             L  for L -L  # taking larger  region overlapped in removing similar regions
             A  for A -A # taking average region overlapped in removing similar regions

<b>Version   </b> : 2.4
<b>Warning   </b> : You MUST NOT delete '# options : ..' entry
              as it is read  by various subroutines.
</pre><hr>

<H3><A href="get_overlapping_range
.pl.html">get_overlapping_range
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_overlapping_range
.pl">get_overlapping_range
.pl</A>
<pre>
<b>Keywords  </b> : get_overlapping_range_in_msp, get_overlapping_range_in_msp_file,
             get_overlapping_seq_match_range, get_overlap_seq_match_range
<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Usage     </b> : @n1=@{&get_overlapping_range(\@ranges1, \@ranges2)};
<b>Version   </b> : 1.1
<b>Warning   </b> : You MUST NOT delete '# options : ..' entry
              as it is read  by various subroutines.
</pre><hr>

<H3><A href="find_source_perl_library
.pl.html">find_source_perl_library
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//find_source_perl_library
.pl">find_source_perl_library
.pl</A>
<pre>
<b>Function  </b> : gets the default perl sub source library from ENV setenv
<b>Usage     </b> : $source_library=${&find_source_perl_library};
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="find_central_seq_msp_chunk
.pl.html">find_central_seq_msp_chunk
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//find_central_seq_msp_chunk
.pl">find_central_seq_msp_chunk
.pl</A>
<pre>
<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Usage     </b> : This finds the correct msp chunk with given seq name
             and big original or any msp chunk
<b>Version   </b> : 1.0
<b>Warning   </b> : You MUST NOT delete '# options : ..' entry
              as it is read  by various subroutines.
</pre><hr>

<H3><A href="find_central_sequence
.pl.html">find_central_sequence
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//find_central_sequence
.pl">find_central_sequence
.pl</A>
<pre>
<b>Function  </b> : accepts msp file and finds the central sequence.
             central sequence is in the centre of all the member
             sequences in a group or cluster
<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Version   </b> : 1.1
<b>Warning   </b> : You MUST NOT delete '# options : ..' entry
              as it is read  by various subroutines.
</pre><hr>

<H3><A href="write_dof_files
.pl.html">write_dof_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//write_dof_files
.pl">write_dof_files
.pl</A>
<pre>
<b>Function  </b> : write Alex's domfam file. it prints out tilde lines
             if the seqlet matched are below threshold defined.
<b>Options   </b> : _  for debugging.
             #  for debugging.
             v  for verbose STDOUT
             n  for NO seq start and end number display
             t= for teshold (eg, t=40  for Blastp(or ssearch) score 40 threshold)

<b>Usage     </b> : &write_dof_files(\@msps);
             while @msps means msp file names
<b>Version   </b> : 1.2
<b>Warning   </b> : You MUST NOT delete '# options : ..' entry
              as it is read  by various subroutines.
</pre><hr>

<H3><A href="make_filtered_list
.pl.html">make_filtered_list
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//make_filtered_list
.pl">make_filtered_list
.pl</A>
<pre>
<b>Function  </b> : this is the core of check_genome_cluster.pl
             finds good linkage seqlets in msp files
<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="follow_seqlet_link
.pl.html">follow_seqlet_link
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//follow_seqlet_link
.pl">follow_seqlet_link
.pl</A>
<pre>
<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Version   </b> : 1.0
<b>Warning   </b> : You MUST NOT delete '# options : ..' entry
              as it is read  by various subroutines.
</pre><hr>

<H3><A href="clu_to_sso_to_msp
.pl.html">clu_to_sso_to_msp
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//clu_to_sso_to_msp
.pl">clu_to_sso_to_msp
.pl</A>
<pre>
<b>Function  </b> : reads in a big single linkage cluster file(or normal cluster file)
              and creates a big msp file which contains all the entries in the
              cluster file (usually with the extension of sclu or clu)
             This normally reads in xxxx.mso, xxxx.sso like files, but if the
              corresponding  xxx.msp file already exists, it concatenates them to
              make a bigger one.
<b>Keywords  </b> : clu_2_sso_2_msp, cluster_to_msp, cluster_to_sso_to_msp
              convert_clu_to_sso_to_msp
<b>Options   </b> : USE, convert_clu_to_sso_to_msp, this is obsolute now
<b>Usage     </b> : &clu_to_sso_to_msp(\$clu);
<b>Version   </b> : 1.7
</pre><hr>

<H3><A href="convert_clu_to_sso_to_msp
.pl.html">convert_clu_to_sso_to_msp
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//convert_clu_to_sso_to_msp
.pl">convert_clu_to_sso_to_msp
.pl</A>
<pre>
<b>Function  </b> : reads in a big single linkage cluster file(or normal cluster file)
              and creates a big msp file which contains all the entries in the
              cluster file (usually with the extension of sclu or clu)
             This normally reads in xxxx.mso, xxxx.sso like files, but if the
              corresponding  xxx.msp file already exists, it concatenates them to
              make a bigger one.
<b>Keywords  </b> : clu_2_sso_2_msp, cluster_to_msp, cluster_to_sso_to_msp
              clu_to_sso_to_msp
<b>Usage     </b> : &clu_to_sso_to_msp(\$clu);
<b>Version   </b> : 1.8
</pre><hr>

<H3><A href="convert_sso_to_msp
.pl.html">convert_sso_to_msp
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//convert_sso_to_msp
.pl">convert_sso_to_msp
.pl</A>
<pre>
<b>Example   </b> : &convert_sso_to_msp(@ARGV, 'OUT.msp', $single_out_opt);
<b>Function  </b> : This takes sso file(s) and produces MSP file. It
             concatenate sso file contents when more than one
             sso file is given.
<b>Options   </b> : _  for debugging.
             #  for debugging.
             v  for showing the MSP result to screen
             s  for making single MSP file for each sso file
                    as well as big MSP file which has all sso
             u= for upper expectation value limit
             l= for lower expect val limit
             s= for single file name input eg. "s=xxxxx.msp"
             n  for new format (msp2 format)
             r  for adding range
             r2 for adding ranges in all sequence names

<b>Returns   </b> : the file names created (xxxx.msp, yyyy.msp,,,,)
<b>Usage     </b> : &convert_sso_to_msp(@ARGV, $single_out_opt);
<b>Version   </b> : 2.6
<b>Warning   </b> : This capitalize all the input file names when
              producing xxxxx.msp. xxxxx.sso -> XXXX.sso
</pre><hr>

<H3><A href="sso_to_msp
.pl.html">sso_to_msp
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//sso_to_msp
.pl">sso_to_msp
.pl</A>
<pre>
<b>Example   </b> : &sso_to_msp(@ARGV, 'OUT.msp', $single_out_opt);
<b>Function  </b> : This takes sso file(s) and produces MSP file. It
             concatenate sso file contents when more than one
             sso file is given.
<b>Keywords  </b> : sso_file_to_msp_file, convert_sso_to_msp,
<b>Options   </b> : _  for debugging.
             #  for debugging.
             v  for showing the MSP result to screen
             s  for making single MSP file for each sso file
                    as well as big MSP file which has all sso
             u= for upper expectation value limit
             l= for lower expect val limit
             s= for single file name input eg. "s=xxxxx.msp"
             n  for new format (msp2 format)
             r  for adding range
             r2 for adding ranges in all sequence names

<b>Returns   </b> : the file names created (xxxx.msp, yyyy.msp,,,,)
<b>Usage     </b> : &sso_to_msp(@ARGV, $single_out_opt);
<b>Version   </b> : 2.6
<b>Warning   </b> : This capitalize all the input file names when
              producing xxxxx.msp. xxxxx.sso -> XXXX.sso
</pre><hr>

<H3><A href="convert_sso_to_msp
.pl.html">convert_sso_to_msp
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//convert_sso_to_msp
.pl">convert_sso_to_msp
.pl</A>
<pre>
<b>Example   </b> : &convert_sso_to_msp(@ARGV, 'OUT.msp', $single_out_opt);
<b>Function  </b> : This takes sso file(s) and produces MSP file. It
             concatenate sso file contents when more than one
             sso file is given.
<b>Keywords  </b> : sso_file_to_msp_file, convert_sso_to_msp,
<b>Options   </b> : _  for debugging.
             #  for debugging.
             v  for showing the MSP result to screen
             s  for making single MSP file for each sso file
                    as well as big MSP file which has all sso
             u= for upper expectation value limit
             l= for lower expect val limit
             s= for single file name input eg. "s=xxxxx.msp"
             n  for new format (msp2 format)
             r  for adding range
             r2 for adding ranges in all sequence names

<b>Returns   </b> : the file names created (xxxx.msp, yyyy.msp,,,,)
<b>Usage     </b> : &convert_sso_to_msp(@ARGV, $single_out_opt);
<b>Version   </b> : 2.6
<b>Warning   </b> : This capitalize all the input file names when
              producing xxxxx.msp. xxxxx.sso -> XXXX.sso
</pre><hr>

<H3><A href="bla_to_msf
.pl.html">bla_to_msf
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//bla_to_msf
.pl">bla_to_msf
.pl</A>
<pre>
<b>Function  </b> : matched each query seq name and if the E value is lower than
             my arbitrary threshold, I put the subject and target pair
             alignment into a hash.
             In later iterations, the latest is replaced
<b>Keywords  </b> : convert_bla_to_msf
<b>Usage     </b> : @msf_file_made=@{&bla_to_msf(\@bla_file)};
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="convert_bla_to_msf
.pl.html">convert_bla_to_msf
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//convert_bla_to_msf
.pl">convert_bla_to_msf
.pl</A>
<pre>
<b>Function  </b> : matched each query seq name and if the E value is lower than
             my arbitrary threshold, I put the subject and target pair
             alignment into a hash.
             In later iterations, the latest is replaced
<b>Keywords  </b> : convert_bla_to_msf
<b>Usage     </b> : @msf_file_made=@{&convert_bla_to_msf(\@bla_file)};
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="convert_bla_to_msp
.pl.html">convert_bla_to_msp
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//convert_bla_to_msp
.pl">convert_bla_to_msp
.pl</A>
<pre>
<b>Author    </b> : Sarah Teichmann and Jong Park, jong@salt2.med.harvard.edu
<b>Example   </b> : %hash_out=%{&convert_bla_to_msp(\$file)};
<b>Function  </b> : reads in PSI blast output and produces MSP file format.
             Takes all the good hits below certain threshold in multiple iteration
             Reports the best evalue with a given sequence name
<b>Keywords  </b> : pbla_to_msp, blast_to_msp, bla_2_msp, blastp_to_msp_format,
             blast_to_msp_format, convert_bla_to_msp, convert_bla_to_msp_files
             bla_to_msp
<b>Options   </b> : 
$pdbd_seq_only  d   for getting dxxxx_ like seq names only(pdb40d names for examp)
$all_seq  a         for forcing all seq conversion
   $which_iteration= by i=    # choose which iteration result you want to take
$which_iteration   as just a digit
   $report_only_the_best=b by b -b
   $take_only_the_last_iteration=l by l
   $accumulative_hits_eval_thresh= by e=
   $genome_seq_only=g
   $nrdb_seq_only=n
   $evalue_thresh= by E=
   $Accumulate_matches=A   by A -A

<b>Usage     </b> : %hash_out_final=%{&convert_bla_to_msp(\$file)};
<b>Version   </b> : 3.7
</pre><hr>

<H3><A href="convert_bla_multaln_to_msf
.pl.html">convert_bla_multaln_to_msf
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//convert_bla_multaln_to_msf
.pl">convert_bla_multaln_to_msf
.pl</A>
<pre>
<b>Example   </b> : @msf_file_made=@{&convert_bla_multaln_to_msf(\@bla_file,
                                              $verbose, "i=$iteration")};
<b>Function  </b> : matched each query seq name and if the E value is lower than
             my arbitrary threshold, I put the subject and target pair
             alignment into a hash.
             In later iterations, the latest is replaced,
              when you use m6 option for PSI blast
             this adds '00x' extensions to the repeatedly occurring seq names

<b>Keywords  </b> : psi_blast_to_msf, psi_blast_multaln_to_msf
<b>Options   </b> : 
   i=$iteration
   v  for verbose
<b>Usage     </b> : @msf_file_made=@{&convert_bla_multaln_to_msf(\@bla_file, [i=2])};
<b>Version   </b> : 1.6
</pre><hr>

<H3><A href="convert_bla_multaln_to_msf
.pl.html">convert_bla_multaln_to_msf
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//convert_bla_multaln_to_msf
.pl">convert_bla_multaln_to_msf
.pl</A>
<pre>
<b>Example   </b> : @msf_file_made=@{&convert_bla_multaln_to_msf(\@bla_file, "i=$iteration")};
<b>Function  </b> : matched each query seq name and if the E value is lower than
             my arbitrary threshold, I put the subject and target pair
             alignment into a hash.
             In later iterations, the latest is replaced,
              when you use m6 option for PSI blast
             this adds '00x' extensions to the repeatedly occurring seq names

<b>Keywords  </b> : psi_blast_to_msf, psi_blast_multaln_to_msf,
             bla_multaln_to_msf
<b>Usage     </b> : @msf_file_made=@{&convert_bla_multaln_to_msf(\@bla_file, [i=2])};
<b>Version   </b> : 1.4
</pre><hr>

<H3><A href="get_sub_hash
.pl.html">get_sub_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_sub_hash
.pl">get_sub_hash
.pl</A>
<pre>
<b>Function  </b> : fetches hash keys and values by giving keys to
             a hash
<b>Keywords  </b> : subhash, sub_hash, get_hash_elements, fetch_sub_hash
             take_sub_hash, get_hash_by_keys, get_sub_hash_by_keys
<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Usage     </b> : %sub_hash=%{&get_sub_hash(\%FASTA, \@list)};
<b>Version   </b> : 1.1
<b>Warning   </b> : You MUST NOT delete '# options : ..' entry
              as it is read  by various subroutines.
</pre><hr>

<H3><A href="get_smallest_file
.pl.html">get_smallest_file
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_smallest_file
.pl">get_smallest_file
.pl</A>
<pre>
<b>Function  </b> : checks the size of files and returns the smallest
             one's name. If a file is not present in pwd or
             specified absolute path, it ignores it.
<b>Keywords  </b> : choose_smallest_file, smallest_file, find_smallest_file
             get_the_smallest_file, choose_the_smallest_file,
             fetch_smallest_file, take_smallest_file, get_smaller_file,
<b>Options   </b> : _  for debugging.
             #  for debugging.
             e  for extract the smallest from the input array
                       leaving it one element less, in this case
                       there will be two returning refs.
<b>Usage     </b> : $smallest_file_name=${&get_largest_file(@ARGV)};
<b>Version   </b> : 1.3
</pre><hr>

<H3><A href="get_largest_file
.pl.html">get_largest_file
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_largest_file
.pl">get_largest_file
.pl</A>
<pre>
<b>Function  </b> : checks the size of files and returns the largest
             one's name. If a file is not present in pwd or
             specified absolute path, it ignores it.
<b>Keywords  </b> : choose_largest_file, largest_file, find_largest_file
             get_the_largest_file, choose_the_largest_file, get_biggest_file
             fetch_largest_file, take_largest_file, get_bigger_file, get_larger_file
<b>Options   </b> : _  for debugging.
             #  for debugging.
             e  for extract the largest from the input array
                       leaving it one element less, in this case
                       there will be two returning refs.
<b>Usage     </b> : $largest_file_name=${&get_largest_file(@ARGV)};
<b>Version   </b> : 1.4
</pre><hr>

<H3><A href="get_sequence_complexity
.pl.html">get_sequence_complexity
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_sequence_complexity
.pl">get_sequence_complexity
.pl</A>
<pre>
<b>Argument  </b> : ref. of string.
<b>Example   </b> : ${&get_sequence_complexity(\$seq)};
             while $seq='TTTTTACDEFGHIKLMNPQRSTVWYAAAAACCCADFADFA'
<b>Function  </b> : caculates the single sequence's sequence complexity
             If the seq given is larger than 20, it divides it into
             frags of 20 aa and gets the average of it.
<b>Keywords  </b> : sequence_complexity, calc_sequence_complexity,
             calc_seq_complexity, get_seq_complexity,
<b>Options   </b> : _  for debugging.
             #  for debugging.
             'w=' for window size as the first arg
<b>Returns   </b> : Ref. for a scalar digit.
<b>Usage     </b> : print "\n", ${&get_sequence_complexity(\$seq)};
<b>Version   </b> : 1.3
</pre><hr>

<H3><A href="make_swiss_index
.pl.html">make_swiss_index
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//make_swiss_index
.pl">make_swiss_index
.pl</A>
<pre>
<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="fetch_sequence_from_db
.pl.html">fetch_sequence_from_db
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//fetch_sequence_from_db
.pl">fetch_sequence_from_db
.pl</A>
<pre>
<b>Argument  </b> : gets names of sequences
             eg) \@array, \%hash, \$seq, while @array=(seq1, seq2), $seq='seq1 seq1'
                                               %hash=(seq1, xxxx, seq2, yyyy);

<b>Example   </b> : %seq=%{&fetch_sequence_from_db(\@input, seq.fa, seq.fa.idx)};
              while @input=qw( 11S3_HELAN_11-31 A1AB_CANFA A1AT_PIG )
<b>Function  </b> : accept seq names (with or without ranges like _10-111 )
              and produces hash ref.
             As an option, you can write(xxxx.fa) the sequences in pwd
              with the file names with sequence names.
             The default database used is FASTA format OWL database.
              You can change this by S (for Swissprot either fasta
              or full format), P for PDB fasta format data.
             If you give the path name of DB, it will look for the
              DB given.

             This automatically checks sequence family number as
               in >d1bpi___7.6.1
               and attaches the number in final %sequence output

<b>Keywords  </b> : fetch_seq_from_db, fetch_sequence_from_database
<b>Options   </b> : _  or #  for debugging.
     w       for write fasta file
     s=      for putting source DB file name manually
     d=p100  for PDB100 fasta database from ENV
     d=p40   for PDB40  fasta database from ENV
     d=p     for PDB database (usually p100) from ENV
     d=s     for Swissprot database from ENV
     d=o     for OWL database from ENV
     i=      for index filename. If not specified, this looks for it in the same dir as fast     
     t=      for msp_threshold
  msp_threshold=0.0005  # when MSP file is given as input for getting seq names

<b>Returns   </b> : ref of hash
<b>Usage     </b> : %sequence=%{&fetch_sequence_from_db($input_file, \@string)};
<b>Version   </b> : 3.5
</pre><hr>

<H3><A href="fetch_seq
.pl.html">fetch_seq
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//fetch_seq
.pl">fetch_seq
.pl</A>
<pre>
<b>Argument  </b> : swissprot seqname
<b>Example   </b> : &fetch_swiss_seq(@ARGV);
<b>Function  </b> : fetches swissprot entry or fasta format seq with
             given seq name(like  SAA_HORSE, SA*HORSE, SAA,..)
             you can give multi files(SAA*, SAU*) at the same
             time. This uses ENV setting of 'SWDIR'
<b>Keywords  </b> : fetch_swissprot_sequence, fetch_sequence,
             find_swiss_sequence, find_sequence
<b>Options   </b> : _  for debugging.
             #  for debugging.
             -f for fasta format file output
             -a is for ALL matched seq. (same as using glob=> *YEAST)
             -c is for Creating seq.idx file
             -h is for HELP!
             -g is for GDF file format output
             -l is for list of match entries(in 1 column)
             -s is for species option (input name mst be species (YEAST, RAT, HUMAN..)
             n= is for Number of seq you want to get from swissprot
             s= is for Size limit. Min seq size in swiss, s=10  -> minimum 11 aa seq.
             S= is for Size limit. Max seq size in swiss, s=1000 -> get less than 1000

<b>Usage     </b> : &fetch_seq(@ARGV);
<b>Version   </b> : 1.6
</pre><hr>

<H3><A href="fetch_swiss_seq
.pl.html">fetch_swiss_seq
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//fetch_swiss_seq
.pl">fetch_swiss_seq
.pl</A>
<pre>
<b>Argument  </b> : swissprot seqname
<b>Example   </b> : &fetch_swiss_seq(@ARGV);
<b>Function  </b> : fetches swissprot entry or fasta format seq with
             given seq name(like  SAA_HORSE, SA*HORSE, SAA,..)
             you can give multi files(SAA*, SAU*) at the same
             time. This uses ENV setting of 'SWDIR'
<b>Keywords  </b> : fetch_swissprot_sequence, fetch_sequence,
             find_swiss_sequence, find_sequence, fetch
<b>Options   </b> : _  for debugging.
             #  for debugging.
             -f for fasta format file output
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_sequence_number
.pl.html">get_sequence_number
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_sequence_number
.pl">get_sequence_number
.pl</A>
<pre>
<b>Function  </b> : reads database and tells how many sequences are there
             fasta format db is only accepted for now.
<b>Keywords  </b> : count_number_of_sequence, get_number_of_sequence
             get_sequence_number_in_fasta
<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Version   </b> : 1.2
</pre><hr>

<H3><A href="write_msp_files
.pl.html">write_msp_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//write_msp_files
.pl">write_msp_files
.pl</A>
<pre>
<b>Example   </b> : &write_msp_files(@sso, 's', $out_file);
<b>Function  </b> : Writes input which is already in msp file format to
              files either the name is given or generated
              If more than one ref of hash is given, this will
              concatenate all the hashes to one big one to
              make one file.
             When NO output xxx.msp file name is given, it creates
              with the query sequence name.
<b>Keywords  </b> : write_msp,
<b>Options   </b> : _  for debugging.
             #  for debugging.
             s  for each single file output for each hash input
      filename  for putting output to the specified filename, should be xxx.msp

<b>Returns   </b> : if 's' option is set, it will make say,
               HI001.msp HI002.msp HI003.msp  rather than

               HI001HI002HI003.msp
  eg of one output(single file case)

   1027     0.0     1     154   HI0004     1     154   HI0004
   40       0.0     84    132   HI0004     63    108   HI0001
   31       0.0     79    84    HI0004     98    103   HI0003

<b>Usage     </b> : &write_msp_files(\%in1, \%in2, ['s'], [$filename],,)
<b>Version   </b> : 2.8
<b>Warning   </b> : When NO output xxx.msp file name is given, it creates
              with the query sequence name.
</pre><hr>

<H3><A href="write_aln_files
.pl.html">write_aln_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//write_aln_files
.pl">write_aln_files
.pl</A>
<pre>
<b>Example   </b> : &write_aln(\%hash, \$out_file_name);
  CLUSTAL W (1.74) multiple sequence alignment


  MMAF6040_1           -----MATDD--SIIVLDD----DDEDEA-AAQP-GPSNLPPN-PASTGPGPGLSQQATG
  AF015956_1           -----MATAN--SIIVLDD----DDEDEA-AAQP-GPSHPLPN-AASPGAG---------
  HSAB2381_80-900      KQRLLSVTSDEGSMNAFTGRGSPDTEIKINIKQESADVNVIGNKDVVTEEDLDVFKQAQE
                             .* :  *: .: .    * * :    *  .  :   *  . .  .

<b>Function  </b> : writes multiple seqs. in msf format (takes one or more than one seq.!!)
<b>Options   </b> : 
     $first_sequence_name= by f=  # to put a certain seq at the first in writing
<b>Usage     </b> : two argments:  $seq_hash_reference  and $output_file_name
             takes a hash which has got names keys and sequences values.
             uses Perl5 pointers(references).
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="write_msf
.pl.html">write_msf
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//write_msf
.pl">write_msf
.pl</A>
<pre>
<b>Example   </b> : &write_msf(\%hash, \$out_file_name, ["o=$seq_order"]);
             eg) $seq_order='asdf seq2 seq3 seq5';
    - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    PileUp

       MSF: 1205  Type: P    Check:  9937   ..

     Name: PYC1_YEAST      oo  Len: 1205  Check:  7954  Weight:  1.00
     Name: PYC2_YEAST      oo  Len: 1205  Check:  5807  Weight:  1.00
     Name: PYC_MOUSE       oo  Len: 1205  Check:  6176  Weight:  1.00

    //

    PYC1_YEAST      MSQ.RKFAGL RDNFNLLGEK N......... .......... .KILVANRGE
    PYC2_YEAST      MSSSKKLAGL RDNFSLLGEK N......... .......... .KILVANRGE
    PYC_MOUSE       ...MLKFQTV RGGLRLLGVR RSSSAPVASP NVRRLEYKPI KKVMVANRGE

    PYC1_YEAST      IPIRIFRTAH ELSMQTVAIY SHEDRLSTHK QKADEAYVIG EVGQYTPVGA
    PYC2_YEAST      IPIRIFRSAH ELSMRTIAIY SHEDRLSMHR LKADEAYVIG EEGQYTPVGA
    PYC_MOUSE       IAIRVFRACT ELGIRTVAVY SEQDTGQMHR QKADEAYLIG R..GLAPVQA

    PYC1_YEAST      YLAIDEIISI AQKHQVDFIH PGYGFLSENS EFADKVVKAG ITWIGPPAEV
    PYC2_YEAST      YLAMDEIIEI AKKHKVDFIH PGYGFLSENS EFADKVVKAG ITWIGPPAEV
    PYC_MOUSE       YLHIPDIIKV AKENGVDAVH PGYGFLSERA DFAQACQDAG VRFIGPSPEV

<b>Function  </b> : writes multiple seqs. in msf format (takes one or more than one seq.!!)
<b>Keywords  </b> : write_msf_files, save_msf_files
<b>Usage     </b> : two argments:  $seq_hash_reference  and $output_file_name
            takes a hash which has got names keys and sequences values.
            uses Perl5 pointers(references).
<b>Version   </b> : 2.2
</pre><hr>

<H3><A href="get_seqblock
.pl.html">get_seqblock
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_seqblock
.pl">get_seqblock
.pl</A>
<pre>
<b>Example   </b> : @blocks_in_hash=@{&get_seqblock(\%msf, 30)};
<b>Keywords  </b> : find_sequence_block, get_sequence_block,
             make_seq_block, make_seqblock, find_seqblock
<b>Options   </b> : _  for debugging.
             #  for debugging.
             m=  for margin length of the seqblock
             t=  for threshold
             l=  for min seqlet length

<b>Version   </b> : 1.3
</pre><hr>

<H3><A href="add_columns
.pl.html">add_columns
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//add_columns
.pl">add_columns
.pl</A>
<pre>
<b>Keywords  </b> : add_seq_columns, add_sequence_columns,
<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Version   </b> : 1.2
<b>Warning   </b> : if the attached name is too long(over 12 char),
             it changes to 'Added_upX' while X is a numb.
</pre><hr>

<H3><A href="get_high_score_blocks
.pl.html">get_high_score_blocks
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_high_score_blocks
.pl">get_high_score_blocks
.pl</A>
<pre>
<b>Argument  </b> : accepts one single ref. of hash
<b>Example   </b> : %block_start_end=%{&get_high_score_blocks(\%input_numb_block)};
             %out=%{&get_high_score_blocks(\%inp_numbs, 'v', 'b')};
<b>Function  </b> : gets hash of key and number string and filters out the
              number string region which is below certain threshold
              determined inside this sub and returns a selected high
              number regions
<b>Keywords  </b> : high_scoring_regions
             get_high_scoring_blocks, find_blocks, get_blocks
<b>Options   </b> : _  for debugging.
             #  for debugging.
             b  for best_block_opt, returns best block only
             v  for showing the final range hash output
             c  for connect close blocks
             c= for connect close blocks with specific closing gap size
             m=  for margin length of the seqblock
             t=  for threshold
             l=  for min seqlet length

<b>Usage     </b> : get_high_score_blocks(<ref. of hash for number string>)
<b>Version   </b> : 1.4
<b>Warning   </b> : This assumes that the inputs are multiply aligned seq
</pre><hr>

<H3><A href="delbut
.pl.html">delbut
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//delbut
.pl">delbut
.pl</A>
<pre>
<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Usage     </b> : delbut *.zip  (delete files except xxxx.zip)
<b>Version   </b> : 1.2
</pre><hr>

<H3><A href="get_msp_range
.pl.html">get_msp_range
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_msp_range
.pl">get_msp_range
.pl</A>
<pre>
<b>Keywords  </b> : get_msp_file_ranges
<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Usage     </b> : @range=@{&get_msp_range($seqlet)};
             @temp=&get_msp_range($seqlet);

<b>Version   </b> : 1.5
</pre><hr>

<H3><A href="get_msp_enquiry_sequence
.pl.html">get_msp_enquiry_sequence
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_msp_enquiry_sequence
.pl">get_msp_enquiry_sequence
.pl</A>
<pre>
<b>Function  </b> : gets the name of sequence used as enquiry(target)
<b>Keywords  </b> : get_msp_target_sequence, get_msp_enquiry_sequence_name
<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_msp_matched_sequence
.pl.html">get_msp_matched_sequence
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_msp_matched_sequence
.pl">get_msp_matched_sequence
.pl</A>
<pre>
<b>Function  </b> : gets the name of sequence used as enquiry(target)
<b>Keywords  </b> : get_msp_matched_sequence_name
<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_linked_sequence
.pl.html">get_linked_sequence
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_linked_sequence
.pl">get_linked_sequence
.pl</A>
<pre>
<b>Example   </b> : seq1 ------------------------------
                            |||||||||||
             seq2        --------------------------------
             OUT  000000000011111111111000000000000000000

<b>Function  </b> : opens msp file and links the sequences according
             to the matches.
<b>Keywords  </b> : link_sequence_from_msp_file, linked_sequenced_length
             get_clustered_sequence_length, get_annexed_sequence_length
             connect_sequences, merge_sequences, combine_sequences
<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Returns   </b> : A ref. of an array
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_averaged_prediction
.pl.html">get_averaged_prediction
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_averaged_prediction
.pl">get_averaged_prediction
.pl</A>
<pre>
<b>Author    </b> : jong@salt2.med.harvard.edu sat@mrc-lmb.cam.ac.uk
<b>Function  </b> : The content of out %average is
               $averaged{$position}=[$residue1, $sec_str2, $dif_reliability];
<b>Keywords  </b> : get_average_predator_prediction, average_predator_prediction
             get_averaged_sec_prediction, get_average_prediction
<b>Options   </b> : 
   $reverse_order_of_one_hash=r by r
   $give_weight_with_good_match=w by w # this is to give preference to well
   $weight_factor= by w=
                                        matching sec. str. I add '0.1'
<b>Usage     </b> : %av_for_back_pred=%{&get_averaged_prediction(\%sec1, \%sec_rv)};
<b>Version   </b> : 1.2
</pre><hr>

<H3><A href="get_average_sequence_size
.pl.html">get_average_sequence_size
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_average_sequence_size
.pl">get_average_sequence_size
.pl</A>
<pre>
<b>Keywords  </b> : get_av_sequence_size, get_average_seq_size
             get_av_seq_size, average_seq_size, av_seq_size
<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_linux_kernel_version
.pl.html">get_linux_kernel_version
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_linux_kernel_version
.pl">get_linux_kernel_version
.pl</A>
<pre>
<b>Keywords  </b> : get_kernel_version, kernel_version,
<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="load_mount_info
.pl.html">load_mount_info
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//load_mount_info
.pl">load_mount_info
.pl</A>
<pre>
<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="plot_vertically
.pl.html">plot_vertically
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//plot_vertically
.pl">plot_vertically
.pl</A>
<pre>
<b>Function  </b> : This is a sub used for plot_domains.pl for
             genome_analysis
<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Usage     </b> : &plot_vertically(\@query);
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="plot_histogram_horizontally
.pl.html">plot_histogram_horizontally
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//plot_histogram_horizontally
.pl">plot_histogram_horizontally
.pl</A>
<pre>
<b>Input     </b> : $input= '00001111111113333333333444444444111111111111111';

<b>Keywords  </b> : plot_horizontally, plot_numbers_horizontally, plot,
             plot_numbers,
<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Output    </b> : 
   00001111111113333333333444444444111111111111111
   1-------------------------------------------47
  |
  |
  |                       *********
  |             *******************
  |             *******************
  |    *******************************************
  |-----------------------------------------------

<b>Usage     </b> : &plot_horizontally(\@query);
<b>Version   </b> : 1.2
</pre><hr>

<H3><A href="condense_number_string
.pl.html">condense_number_string
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//condense_number_string
.pl">condense_number_string
.pl</A>
<pre>
<b>Example   </b> : @output=@{&condense_number_string(\@input, $factor)};
             with @input=qw(1 2 4 10 10 22 2 3 44 2 3); and $factor=3
<b>Function  </b> : condenses the numbers by making an average with
             given factor. If the factor is 2 on number seq
              1334284425 , result will be 23543
              133428442  ,                23541 <-- preserved end
             Factor 3 =>
              133428442  , (1+3+3)/3 = 2
                           (4+2+8)/3 = 4,,,
<b>Keywords  </b> : compact_number_string, compact_digits, condense
             condense_string
<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="get_seq_fragments
.pl.html">get_seq_fragments
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_seq_fragments
.pl">get_seq_fragments
.pl</A>
<pre>
<b>Example   </b> : 
  %test=('seq1', '1234AAAAAAAAAAAaaaaa', 'seq2', '1234BBBBBBB');
  @range = ('1-4', '5-8');

  %out = %{&get_seq_fragments(\%test, \@range)};
  %out => (seq1_5-8   AAAAA
           seq2_5-8   BBBBB
           seq1_1-4    1234
           seq2_1-4    1234 )

<b>Function  </b> : gets sequence(string) segments with defined
             ranges.
<b>Keywords  </b> : get_sequence_fragments,
<b>Options   </b> : _  for debugging.
             #  for debugging.
             l=  for min seqlet length
             r  for adding ranges in the seq names

<b>Usage     </b> : @seq_frag=&get_seq_fragments(\%msf, @RANGE);
<b>Version   </b> : 1.8
</pre><hr>

<H3><A href="make_standalone_subroutines
.pl.html">make_standalone_subroutines
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//make_standalone_subroutines
.pl">make_standalone_subroutines
.pl</A>
<pre>
<b>Author    </b> : jong@salt2.med.harvard.edu
<b>Class     </b> : Utility
<b>Example   </b> : &make_standalone_subroutines(@ARGV);
<b>Function  </b> : Creates each subroutine derived xxx.pl file from B.pl or any
             given library file. If there is a file for a sub already, it
             skips.
<b>Usage     </b> : &make_standalone_subroutines(@ARGV);
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="is_html
.pl.html">is_html
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//is_html
.pl">is_html
.pl</A>
<pre>
<b>Example   </b> : $html=&is_html(\@test);
<b>Function  </b> : Checks if it is an html file.
<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_column
.pl.html">get_column
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_column
.pl">get_column
.pl</A>
<pre>
<b>Argument  </b> : Ref of Hash, Array or just filename, and wanted column numbers.
<b>Example   </b> : For getting only necessary columns
             Input: %Hash=(1, 'col1 col2 col3',
                           2, 'col1 col2 col3',
                           3, 'col1 col2 col3');
             input format: &get_column(\%Hash, 3,2,1, 'k'); # k is opt
             Ouput format: STDOUT as

                1     col3 col2 col1
                2     col3 col2 col1
                3     col3 col2 col1

<b>Function  </b> : Prints any specified columns, can change order of them,
             can filter values of columns to filter (max or min value)
             Skipps blank line.
<b>Keywords  </b> : columns, column.pl, column, get_columns, take_columns,
<b>Options   </b> : #  for debugging.
             _  for debugging.
             k  for Key print when hash input is given.
             n  for no first line display(Handy when you have title line
                                          and wanna remove it)
             ?max?=xxx for filtering column numbers by maximum of xxx
             ?min?=yyy for filtering column numbers by minimum of yyy
                      (eg, min4=100000 means 4th column minimum is 100000)
                      (eg, 1min4=10, 2min3=10, means get 4th column values
                           below 10 as the first output column. Get 3rd
                           column values below 10 as the second out column.

$combine = 1 by -c c   # c is for combining columns in different files
$ignore  = 1 by -i i   # i is for ignoring leng diff in columns over 1 input

<b>Returns   </b> : Ref of
<b>Usage     </b> : &get_column(\@ar, 1,2 ,3);
             &get_column(\%ha, 1,2 ,3);
             &get_column(@ARGV);
             # where prompt is like: column.pl temp.txt 1 2 3 4
<b>Version   </b> : 1.5
</pre><hr>

<H3><A href="set_debug_option
.pl.html">set_debug_option
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//set_debug_option
.pl">set_debug_option
.pl</A>
<pre>
<b>Example   </b> : set_debug_option #    <-- at prompt.
<b>Function  </b> : If you put '#' or  '##' at the prompt of any program which uses
             this sub you will get verbose printouts for the program if the program
             has a lot of comments.
<b>Options   </b> : #   for 1st level of verbose printouts
             ##  for even more verbose printouts
$debug  becomes 1 by '#'  or '_'
$debug2 becomes 1 by '##'  or '__'

<b>Returns   </b> : $debug
<b>Usage     </b> : &set_debug_option;
<b>Version   </b> : 1.8
</pre><hr>

<H3><A href="write_sdb_file
.pl.html">write_sdb_file
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//write_sdb_file
.pl">write_sdb_file
.pl</A>
<pre>
<b>Argument  </b> : \%ref_of_seq
<b>Example   </b> : @out=@{&write_sdb_file(\%seq, 'v')};  ## for STDOUT as well
    ___________________________________________________________________________
    Title      : EST_YEAST.sdb
    Full Name  : Telomerase_yeast_699aa
    Nicknames  :
    EMBL       :
    PDB        :
    Swissprot  :

<b>Function  </b> : gets a hash ref. and writes the SDB file with 'sprintf'
<b>Keywords  </b> : write_sdb
<b>Options   </b> : v  for verbose representation. This will print boxes on STDOUT
            n  for no '#' leader.
            e  for Endline( '-----------------------------..' )
<b>Usage     </b> : @out=@{&write_sdb_file(\%seq)};
<b>Version   </b> : 1.0
<b>Warning   </b> : if version no. is null, it automatically puts '1.0'
</pre><hr>

<H3><A href="push_if_not_already
.pl.html">push_if_not_already
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//push_if_not_already
.pl">push_if_not_already
.pl</A>
<pre>
<b>Argument  </b> : two references. The first should be an array ref. The 2nd can be either
             scalar or array reference.
<b>Function  </b> : returns ref. of an array for a list of non-repetitive entry.
<b>Keywords  </b> : add_if_not_already, add_element_if_not_already, if_not_already
             add_element_if_not_already, push_element_if_not_already,
             if_no_already_push, put_element_if_not_already, add_new_element
             add_new_items_only, push_new_items_only, push_new_elements_only
             put_if_not_already,
<b>Returns   </b> : a ref. of an array.
<b>Usage     </b> : @out=@{&push_if_not_already(\@mother_array, \@adding_array )};
             @out=@{&push_if_not_already(\@mother_array, $adding_scalar)};
<b>Version   </b> : 1.3
</pre><hr>

<H3><A href="compare_sec_template_with_db
.pl.html">compare_sec_template_with_db
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//compare_sec_template_with_db
.pl">compare_sec_template_with_db
.pl</A>
<pre>
<b>Keywords  </b> : sec structure mapping, map sec str, map_sec_structure
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_peptide_occurance
.pl.html">get_peptide_occurance
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_peptide_occurance
.pl">get_peptide_occurance
.pl</A>
<pre>
<b>Argument  </b> : eg=> (\%ref_hash, 4)
<b>Example   </b> : %stat=%{&get_peptide_occurance(\%pro_sequence, $size)};
              while %pro_sequence has one or more sequences like
              seq1 AAAAAAAAAAAA, seq2 BBBBBBBBBBBBBB, ...
$size is number. For dipeptide=2, tripeptide=3, tetrapep=4...
<b>Function  </b> : gets the number of occurances of peptide(with given size) for
             any number of sequences given.
<b>Version   </b> : 1.2
</pre><hr>

<H3><A href="open_lottery_file
.pl.html">open_lottery_file
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_lottery_file
.pl">open_lottery_file
.pl</A>
<pre>
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="get_probable_half
.pl.html">get_probable_half
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_probable_half
.pl">get_probable_half
.pl</A>
<pre>
<b>Argument  </b> : \@array
<b>Function  </b> : This produces a hash ref. which is supposed to be most probable
             according to the given array. It divides array into halves
             gets the more probable half until it gets one single number.
<b>Keywords  </b> : get_frequent_halves,
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="divide_array
.pl.html">divide_array
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//divide_array
.pl">divide_array
.pl</A>
<pre>
<b>Function  </b> : divides any array to the denominator given.
             If you give array of 100 elem, with 5, you will
             get 5 arrays with 20 elem each.
<b>Keywords  </b> : split_array_into_pieces, split_array, chop_array,
             fragment_array,
<b>Options   </b> : s=  for dividing the array with sub array size
                 eg) to get 20 elem length sub arrays from
                     a big array
                     @ar_ref=@{&divide_array(\@array, 's=20')};
<b>Usage     </b> : &show_array(&divide_array(\@input, 6));
<b>Version   </b> : 1.4
</pre><hr>

<H3><A href="split_fasta_files
.pl.html">split_fasta_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//split_fasta_files
.pl">split_fasta_files
.pl</A>
<pre>
<b>Keywords  </b> : divide_fasta_files, split_fasta_db_files, divide_fasta_db_files
             make_single_fasta_files, write_single_fasta, write_single_fasta_files
<b>Usage     </b> : @names_of_single_files=@{&split_fasta_files(\@files)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="split_sequence
.pl.html">split_sequence
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//split_sequence
.pl">split_sequence
.pl</A>
<pre>
<b>Example   </b> : &show_array( &divide_string(\%input, 3) );
              while  $input is 'seq', '12345789ABCDEFHIJKLMN'
              The output will be 'seq_1_half', '1234578'
                                 'seq_2_half', '9ABCDEF'
                                 'seq_3_half', 'HIJKLMN'
<b>Function  </b> : divides any string to the denominator given.
<b>Keywords  </b> : divide_string, split_string, chop_string, divide_sequence
             split_sequence(look at separate split_sequence sub),
<b>Options   </b> : 
  $reverse_second_half=S by S -S
$reverse_first_half =F by F -F
$reverse_rest   =R by R -R  ## reversing all except the first
$reverse_all  =A by A -A # reverse all the fragments
<b>Usage     </b> : %out=%{&split_sequence(\%input, 2 )};
<b>Version   </b> : 1.3
</pre><hr>

<H3><A href="divide_string
.pl.html">divide_string
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//divide_string
.pl">divide_string
.pl</A>
<pre>
<b>Example   </b> : &show_array( &divide_string(\$input, 3) );
              while  $input is '12345789ABCDEFHIJKLMN'
              The output will be '1234578 9ABCDEF HIJKLMN'
<b>Function  </b> : divides any string to the denominator given.
<b>Keywords  </b> : divide_string, split_string, chop_string, divide_sequence
             split_sequence(look at separate split_sequence sub),
<b>Usage     </b> : &show_array(&divide_string(\$input, 6));
<b>Version   </b> : 1.4
</pre><hr>

<H3><A href="write_html_headbox
.pl.html">write_html_headbox
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//write_html_headbox
.pl">write_html_headbox
.pl</A>
<pre>
<b>Function  </b> : write html format headbox explanation with
              given hashes of headbox content.
<b>Keywords  </b> : write_headbox_html, write headbox in html,
               write_headbox_files
<b>Options   </b> : 'd' for date inclusion at the top of the page
  f=   for default ftp dir name

<b>Usage     </b> : &write_html_headbox($outfilename, \%entries);
<b>Version   </b> : 1.7
<b>Warning   </b> : It takes off the last '/' when $URL has it

</pre><hr>

<H3><A href="open_sdb_files
.pl.html">open_sdb_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_sdb_files
.pl">open_sdb_files
.pl</A>
<pre>
<b>Argument  </b> : One or None. If you give an argu. it should be a ref. of an ARRAY
              or a filename, or ref. of a filename.
             If no arg is given, it reads SELF, ie. the program itself.
<b>Example   </b> : Output is something like
             ('Title', 'read_head_box', 'Tips', 'Use to parse doc', ...)
<b>Keywords  </b> : read_sdb_files,read_sdb,
<b>Options   </b> : 'b' for remove blank lines. This will remove all the entries
             with no descriptions
<b>Returns   </b> : A hash ref.
<b>Usage     </b> : %entries = %{&open_sdb_files(\$file_to_read )};
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="open_stride_dat_files
.pl.html">open_stride_dat_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_stride_dat_files
.pl">open_stride_dat_files
.pl</A>
<pre>
<b>Author    </b> : jong@salt2.med.harvard.edu
<b>Usage     </b> : @out=@{&open_stride_dat_files(@ARGV)};
<b>Version   </b> : 1.2
</pre><hr>

<H3><A href="get_pdb_file_start_number
.pl.html">get_pdb_file_start_number
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_pdb_file_start_number
.pl">get_pdb_file_start_number
.pl</A>
<pre>
<b>Keywords  </b> : start_number_of_pdb, startnumber, start number of PDB,
             get_start_number_of_pdb_file,
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="write_modeller_top_file
.pl.html">write_modeller_top_file
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//write_modeller_top_file
.pl">write_modeller_top_file
.pl</A>
<pre>
<b>Argument  </b> : 1 hash ref which has model name and template name -> (\%hash)
             while %hash is (modelname, tempalatename)
<b>Example   </b> : 
     $modelname = 'gfct';
     $template = '1ovt';
     %hash=($modelname, $template);
     &write_modeller_top_file(\%hash);
<b>Function  </b> : Writes Modeller command file format.
<b>Options   </b> : v  for verbose. You will get STDOUT of the result as well as file
<b>Returns   </b> : a file of xxxx.top form.
<b>Usage     </b> : &write_modeller_top_file(\%hash, [v]);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="write_modeller_ali_file
.pl.html">write_modeller_ali_file
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//write_modeller_ali_file
.pl">write_modeller_ali_file
.pl</A>
<pre>
<b>Argument  </b> : 2 ref. of hash for seq. and optional output.name and option(s).
             If second input hash (for template) has 3rd and 4th element which are
             numbers they are regarded as the starting and ending number of the
             template(i.e. pdb file seq)
<b>Example   </b> : 
             $out = 'test.ali';
             %model =    qw(model AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAccccccccccc);
             %template = qw(templ CCAAAAAAAACCCCCCCCCCCCCCCCCCCCCCCCCCCCC 3 42);
             &write_modeller_ali_file(\%model, \%template, \$out);
<b>Function  </b> : Writes Modeller alignment format.
<b>Options   </b> : You can put 2 numbers for the second set of key and element for
             the second hash input as the starting and ending points of
             template(i.e. pdb file seq). Unless I calculate the size of seq.
             By default, it reads PDB file defined by ENV setting of 'PDB' and
             gets the starting number of pdb. If starting number is defined
             explicitly at input hash, the given starting number is used instead
             of PDB's.
             v  for verbose. You will get STDOUT of the result as well as file
<b>Returns   </b> : a file of xxxx.ali form.
<b>Usage     </b> : &write_modeller_ali_file(\%model, \%template, [\$outfilename], [v]);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="make_template_from_sec_str
.pl.html">make_template_from_sec_str
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//make_template_from_sec_str
.pl">make_template_from_sec_str
.pl</A>
<pre>
<b>Function  </b> : makes template of sec. str. like: 'H5 E4 E2' out of '__HHHHH__EEEE__EE__'
<b>Usage     </b> : %target   = %{&make_template_from_sec_str(\%seq)};
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="calculate_protein_volume
.pl.html">calculate_protein_volume
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//calculate_protein_volume
.pl">calculate_protein_volume
.pl</A>
<pre>
<b>Usage     </b> : %volumes=%{&calculate_protein_volume(\%seq)}
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="extract_words
.pl.html">extract_words
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//extract_words
.pl">extract_words
.pl</A>
<pre>
<b>Usage     </b> : @words = @{&extract_words(\$string)};
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="replace_subroutines
.pl.html">replace_subroutines
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//replace_subroutines
.pl">replace_subroutines
.pl</A>
<pre>
<b>Function  </b> : replaces subroutines of given file(s) with supplied subs.
             Doesn't care version
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="write_subroutines
.pl.html">write_subroutines
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//write_subroutines
.pl">write_subroutines
.pl</A>
<pre>
<b>Author    </b> : jong@salt2.med.harvard.edu
<b>Function  </b> : Writes subroutine file xxxx.psub with given headbox including
              hash
<b>Usage     </b> : @out_file=@{&write_subroutines(\%head_box)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="read_subroutines
.pl.html">read_subroutines
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//read_subroutines
.pl">read_subroutines
.pl</A>
<pre>
<b>Function  </b> : retunrns ALL subroutines with the keys as subroutine names
             with version like ('show_array2.2' => 'subroutine in one string')
             It reports the subroutines not found in searched file(s)
<b>Options   </b> : 'nv' for no version attachment in the keys of returning hash of subroutines
             'r'  for getting remnant file content rather than the sub routines
             't'  for leaving the original file without the sub routines taken.
     $separate_hash_entry_opt=s by s
<b>Usage     </b> : @out_subs=@{&read_subroutines(\@file, $separate_hash_entry_opt)}; or
             %out_subs=%{&read_subroutines(\@file)};
<b>Version   </b> : 1.2
</pre><hr>

<H3><A href="fetch_subroutines
.pl.html">fetch_subroutines
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//fetch_subroutines
.pl">fetch_subroutines
.pl</A>
<pre>
<b>Function  </b> : retunrns subroutines with the keys as subroutine names with version
             like in the form( 'show_array2.2' => 'subroutine in one string')
             It reports the subroutines not found in searched file(s). This
             requires the names of sub you want while read_subroutines will
             read any subroutines with their headbox to a hash.
<b>Options   </b> : 'nv' for no version attachment in the keys of returning hash of subroutines
             'r'  for getting remnant file content rather than the sub routines
             't'  for leaving the original file without the sub routines taken.
             'h'  for headbox only output.
<b>Version   </b> : 2.5
</pre><hr>

<H3><A href="update_subroutines
.pl.html">update_subroutines
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//update_subroutines
.pl">update_subroutines
.pl</A>
<pre>
<b>Example   </b> : &update_subroutines($file, \%fetched_subs);
<b>Function  </b> : replaces subroutines of given file(s) with supplied subs.
             If the given subroutine versions are not higher than the
             ones in the program, no upgrade would happen.
             This can read version information from '# Version  : 1.0' line
              or sub xxxxx{  # Version : 1.0   line
<b>Keywords  </b> : upgrade_subroutines,
<b>Usage     </b> : &update_subroutines(\@file, \%fetched_subs);
<b>Version   </b> : 2.8
</pre><hr>

<H3><A href="takeout_subroutines
.pl.html">takeout_subroutines
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//takeout_subroutines
.pl">takeout_subroutines
.pl</A>
<pre>
<b>Function  </b> : retunrns subroutines with the keys as subroutine names with version
             like in the form( 'show_array2.2' => 'subroutine in one string')
             It reports the subroutines not found in searched file(s)
             fetch_subroutines  also has this feature.
<b>Keywords  </b> : take_out_subroutines, take_subroutines, cut_subroutines,
             cutout_subroutines, remove_subroutines
<b>Options   </b> : 'nv' for no version attachment in the keys of returning hash of subroutines
             'r'  for getting remnant file content rather than the sub routines
<b>Version   </b> : 1.5
<b>Warning   </b> : If there is no headbox and version no. It thinks the version
             is 1.0
</pre><hr>

<H3><A href="get_subroutine_calls
.pl.html">get_subroutine_calls
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_subroutine_calls
.pl">get_subroutine_calls
.pl</A>
<pre>
<b>Function  </b> : gets all the subroutine calls( like &show_hash ) in the given
             file name or array of lines which is the content of a file,
             text etc. If there is no input arg, it reads the running
             program as default input
<b>Keywords  </b> : get_sub_names,get_subroutine_names, get_sub_calls,
             get_subroutine_calls, find_sub_calls, find_subroutine_calls
<b>Usage     </b> : @sub_name_array= @{&get_subroutine_calls(\@AR))};
<b>Version   </b> : 2.2
</pre><hr>

<H3><A href="set_special_options
.pl.html">set_special_options
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//set_special_options
.pl">set_special_options
.pl</A>
<pre>
<b>Argument  </b> : Nothing in a program.
<b>Example   </b> : &set_special_options.pl  ##    <-- at prompt.
<b>Function  </b> : If you put special chars like '#' or  '##', '###..' at the
             prompt of any program which uses
             this sub you will get verbose printouts for the program if
             the program has a lot of comments.
<b>Options   </b> : #   for 1st level of debugging printouts
             ##  for even more debugging printouts
             +   for more outputs(more calculations are shown, like statistics)
             ++  even more outputs.(
$DEBUG    becomes 1 by '#'
$DEBUG2   becomes 1 by '##'
$VERBOSE  becomes 1 by '+'
$VERBOSE2 becomes 1 by '++'

<b>Returns   </b> : $debug, $verbose
<b>Usage     </b> : &set_special_options;
<b>Version   </b> : 1.0
             generalized debug var is added for more verbose printouts.
</pre><hr>

<H3><A href="set_debug
.pl.html">set_debug
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//set_debug
.pl">set_debug
.pl</A>
<pre>
<b>Example   </b> : set_debug #    <-- at prompt.
<b>Function  </b> : If you put '#' or  '##' at the prompt of any program which uses
             this sub you will get verbose printouts for the program if the program
             has a lot of comments.
<b>Options   </b> : #   for 1st level of verbose printouts
             ##  for even more verbose printouts
$debug  becomes 1 by '#'  or '_'
$debug2 becomes 1 by '##'  or '__'

<b>Returns   </b> : $debug
<b>Usage     </b> : &set_debug;
<b>Version   </b> : 1.8
             generalized debug var is added for more verbose printouts.
</pre><hr>

<H3><A href="open_self
.pl.html">open_self
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_self
.pl">open_self
.pl</A>
<pre>
<b>Keywords  </b> : read self, read_self, open self, open itself
<b>Returns   </b> : one array
<b>Usage     </b> : @lines =  &open_self;
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="tell_seq_length
.pl.html">tell_seq_length
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//tell_seq_length
.pl">tell_seq_length
.pl</A>
<pre>
<b>Function  </b> : tells the sequence sizes of given sequences
<b>Usage     </b> : %hash_out = %{&tell_seq_length(\%hash_in)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="do_window_scan
.pl.html">do_window_scan
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//do_window_scan
.pl">do_window_scan
.pl</A>
<pre>
<b>Function  </b> : This is the core part of any window (of sequences)
             scanning function.
<b>Keywords  </b> : scan_sequence, scan_window
<b>Usage     </b> : @out_array = @{&do_window_scan(\@input_array, $win_size)};
             Often, bioters(Bio Computer Scientists) need to scan a long sequences
             of DNA or Protein like(ABADFAFASDFASFASDFDFA or 109384717817947) to
             caculate something out of them.
             This routine is providing such scanning
             function in perl.
<b>Version   </b> : 1.3
</pre><hr>

<H3><A href="scan_window_and_calc_something
.pl.html">scan_window_and_calc_something
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//scan_window_and_calc_something
.pl">scan_window_and_calc_something
.pl</A>
<pre>
<b>Function  </b> : scans any given length window of sequence and computes something.
<b>Options   </b> : average for getting average of given window size.
             sum for getting sum of given window size.
<b>Version   </b> : 1
</pre><hr>

<H3><A href="scan_window_and_calc_average
.pl.html">scan_window_and_calc_average
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//scan_window_and_calc_average
.pl">scan_window_and_calc_average
.pl</A>
<pre>
<b>Usage     </b> : %out_hash_final = %{scan_window_and_calc_average(\@hash, \$win_size)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="read_blast_hits
.pl.html">read_blast_hits
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//read_blast_hits
.pl">read_blast_hits
.pl</A>
<pre>
<b>Example   </b> : 
      - - - - -  EXample of blastp file  - - - - - - - - - - - - - - - - - - - - - - - - -
      BLASTP 1.4.8 [19-Dec-94] [Build 16:06:14 Jul 26 1995]
      Reference:  Altschul, Stephen F., Warren Gish, Webb Miller, Eugene W. Myers,
      and David J. Lipman (1990).  Basic local alignment search tool.  J. Mol. Biol.
      215:403-10.
      Query=  1mbs
      (153 letters)
      Database:  /nfs/ind4/ccpe1/people/A Biomatic /jpo/align/all_in_fasta.fas
      406 sequences; 77,134 total letters.
      Searching..................................................done
      WARNING:  -hspmax 100 was exceeded with 13 of the database sequences, with as
      many as 173 HSPs being found at one time.
      Smallest
      Sum
      High  Probability
      Sequences producing High-scoring Segment Pairs:              Score  P(N)      N
      1mbs                                                          804  2.0e-109  1
      1pmb                                                          718  1.4e-97   1
      1ymb                                                          707  4.7e-96   1
      2xxx                                                           31  0.55      1
<b>Function  </b> : This reads the output of blastp program(xxxx.bla or whatever file extension
             you attatched). And produces the names of found sequences which are
             above(smaller in probability) a certain threshold in the blast result.
             For example, it will produce a reference of an array (@hits, in the code)
             which contains (1mbs, 1pmb, 1ymb) from the example in this header box(down the
             lines) with the given (you give!) threshold of, say, 0.0001.
<b>Keywords  </b> : bla2fasta, take_blast_hits
<b>Usage     </b> : @array_of_names = @{&read_blast_hits(\$file_name, \$threshold)};
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="put_gaps_every_x_position_in_string.pl.html">put_gaps_every_x_position_in_string</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//put_gaps_every_x_position_in_string.pl">put_gaps_every_x_position_in_string.pl</A>
<pre>
<b>Argument  </b> : 3 arg. One is the string, second is the interval number, third is
             the gap separater
<b>Example   </b> : "1234567890123456789012345678901234567890"  will be
             "1234567890 1234567890 1234567890 1234567890"
             with
                &put_gaps_every_x_position_in_string(\$test, 10, ' ')
<b>Keywords  </b> : put_space_in_sequence, put_gaps_in_sequence, put_gaps,
             put_space
<b>Returns   </b> : 
             every char.
<b>Version   </b> : 1.1
<b>Warning   </b> : it does not returns reference
</pre><hr>

<H3><A href="transform_values
.pl.html">transform_values
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//transform_values
.pl">transform_values
.pl</A>
<pre>
<b>Argument  </b> : hash(es) and Matrix or table for conversion.
<b>Example   </b> : 
             IN =>  to transform E and H to 9 and 4

             1cdg_6taa      -------EEE-----------HH--HHHH------EE---------EEE-
             1cdg_2aaa      -------EEE-----------HH--HHHH------EE---------EEE-
             2aaa_6taa      -------EEEEE------EE-HHHHHHHH----EEEE-------EEEEE-

             OUT
             1cdg_6taa      -------999-----------44--4444------99---------999-
             1cdg_2aaa      -------999-----------44--4444------99---------999-
             2aaa_6taa      -------99999------99-44444444----9999-------99999-

<b>Function  </b> : transform any value to another value with given table, matrix..
             This is used to transform Amino Acid to its various propensities
             If you feed a sequence 'ACDEDA', this transforms it to '
             '124741' if the table given is 'A->1, C->2, D->4, E->7'
<b>Returns   </b> : hash(es)
             Sheraga_alpha_matrix
             Richardson_alpha_matrix  or any conversion table made in a hash.

<b>Usage     </b> : Used in predict_secondary_structure
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="Sheraga_alpha_matrix
.pl.html">Sheraga_alpha_matrix
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//Sheraga_alpha_matrix
.pl">Sheraga_alpha_matrix
.pl</A>
<pre>
<b>Function  </b> : an alpha matrix propensity table.
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="Richardson_alpha_matrix
.pl.html">Richardson_alpha_matrix
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//Richardson_alpha_matrix
.pl">Richardson_alpha_matrix
.pl</A>
<pre>
<b>Function  </b> : an alpha matrix propensity table.
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_segment_shift_rate
.pl.html">get_segment_shift_rate
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_segment_shift_rate
.pl">get_segment_shift_rate
.pl</A>
<pre>
<b>Argument  </b> : Two references of hashes. One for error rate the other for sec.
             assignment.
<b>Example   </b> : <input example> First block is for the first hash input
                             and Second is for the second hash input.

             1cdg_6taa      00000442222222222242222222222777700000007000000000
             1cdg_2aaa      00000442222222222242222222222777700000007000000000
             2aaa_6taa      00000000000000000000000000000000000000000000000000

             1cdg_6taa      -------EEE-----------EE--EEEE------EE---------EEE-
             1cdg_2aaa      -------EEE-----------EE--EEEE------EE---------EEE-
             2aaa_6taa      -------EEEEE------EE-EEEEEEEE----EEEE-------EEEEE-

             <intermediate output example>
             2aaa_6taa      -------00000---------00000000----0000-------00000-
             1cdg_6taa      -------442---------------2222-----------------000-
             1cdg_2aaa      -------222---------------2222-----------------000-

             <Final output>
             2aaa_6taa      0%
             1cdg_6taa      67%
             1cdg_2aaa      67%

<b>Function  </b> : calculates the secondary structure segment shift rate.
<b>Options   </b> : 'p' or 'P' for percentage term(default)
             'r' or 'R' for ratio term (0.0 - 1.0), where 1 means all the
              segments were wrongly aligned.
             's' or 'S' for Shift rate (it actually caculates the position shift
              rate for the secondary structure segment.
             'h' or 'H' for position Shift rate (it actually caculates the position
              shift rate for helical segments). If this is the only option, it
              will show the default percentage term rate for helical segments.
              If used with 'r', it will give you ratio (0.0 - 1.0) for helical
              segment. If used with 's' option, it will give you position shift
              rate for only helical segments.
             'e' or 'E' for position Shift rate (it actually caculates the position
              shift rate for beta segments). If this is the only option, it will
              show the default percentage term rate for beta segments. If used
              with 'r', it will give you ratio (0.0 - 1.0) for beta. If used
              with 's' option, it will give you position shift rate for only
              beta segments.
<b>Usage     </b> : &get_segment_shift_rate(\%hash_for_errors, \%hash_for_sec_str);
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="get_wrong_segment_rate
.pl.html">get_wrong_segment_rate
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_wrong_segment_rate
.pl">get_wrong_segment_rate
.pl</A>
<pre>
<b>Example   </b> : <input example> hash of 3 keys and values.
             2aaa_6taa      -------00000---------00000000----0000-------00000-
             1cdg_6taa      -------442---------------2222-----------------000-
             1cdg_2aaa      -------222---------------2222-----------------000-

             In the above there are two segments wrong in 3 segment blocks = 2/3
             <output example> hash of 3 percentage rates.

             2aaa_6taa      0 %
             1cdg_6taa      66.6666666666667 %
             1cdg_2aaa      66.6666666666667 %

<b>Function  </b> : Treats the segment as one single big error.
             calculates the wrong segment number compared to the correct ones.
<b>Usage     </b> : print_seq_in_block( &get_wrong_segment_rate(\%superposed_hash) );
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="tidy_secondary_structure_segments
.pl.html">tidy_secondary_structure_segments
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//tidy_secondary_structure_segments
.pl">tidy_secondary_structure_segments
.pl</A>
<pre>
<b>Argument  </b> : hashes and [options]. No options result in default of 'H3', 'E3'
<b>Example   </b> : print_seq_in_block(&tidy_secondary_structure_segments(\%hash, 'e4', 'h4'), 's');
             <makes following into the next block>

             1cdg_2aaa      -------EEE-----------EE--EEEE------EE---------EEE-
             1cdg_6taa      -------EEE-----------EE--EEEE------EE---------EEE-
             2aaa_6taa      -------EEEEE------EE-EEEEEEEE----EEEE-------EEEEE-

             <example output>

             1cdg_6taa      -------------------------EEEE---------------------
             1cdg_2aaa      -------------------------EEEE---------------------
             2aaa_6taa      -------EEEEE---------EEEEEEEE----EEEE-------EEEEE-

<b>Function  </b> : receives any secondary structure assignment hashes and
             tidys up them. That is removes very shoft secondary structure
             regions like( --HH--, -E-, -EE- ) according to the given minimum
             lengths(threshold) of segments by you.
<b>Options   </b> : something like 'H3' or 'E3' for minimum segment length set to 3 positions.
<b>Returns   </b> : array of references of hashes.
<b>Usage     </b> : print_seq_in_block(&tidy_secondary_structure_segments(\%hash, 'e4', 'h4'), 's');

<b>Version   </b> : 1.0.0
</pre><hr>

<H3><A href="define_secondary_structure_segments
.pl.html">define_secondary_structure_segments
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//define_secondary_structure_segments
.pl">define_secondary_structure_segments
.pl</A>
<pre>
<b>Argument  </b> : hashes and [options]. No options result in default of 'H3', 'E3'
<b>Example   </b> : print_seq_in_block(&define_secondary_structure_segments(\%hash, 'e4', 'h4'), 's');
             <makes following into the next block>

             1cdg_2aaa      -------EEE-----------EE--EEEE------EE---------EEE-
             1cdg_6taa      -------EEE-----------EE--EEEE------EE---------EEE-
             2aaa_6taa      -------EEEEE------EE-EEEEEEEE----EEEE-------EEEEE-

             <example output>

             1cdg_6taa      -------------------------EEEE---------------------
             1cdg_2aaa      -------------------------EEEE---------------------
             2aaa_6taa      -------EEEEE---------EEEEEEEE----EEEE-------EEEEE-

<b>Function  </b> : receives any secondary structure assignment hashes and
             tidys up them. That is removes very shoft secondary structure
             regions like( --HH--, -E-, -EE- ) according to the given minimum
             lengths of segments.
<b>Options   </b> : something like 'H3' or 'E3' for minimum segment length set to 3 positions.
<b>Returns   </b> : array of references of hashes.
<b>Usage     </b> : print_seq_in_block(&define_secondary_structure_segments(\%hash, 'e4', 'h4'), 's');

<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="overlay_seq_by_certain_chars
.pl.html">overlay_seq_by_certain_chars
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//overlay_seq_by_certain_chars
.pl">overlay_seq_by_certain_chars
.pl</A>
<pre>
<b>Argument  </b> : 2 ref for hash of identical keys and value length.
<b>Example   </b> : %out =%{&overlay_seq_by_certain_chars(\%hash1, \%hash2, 'E')};
             output> with 'E' option >>> "name1     --HHH--1232-"
<b>Function  </b> : (name1 000000112324)+(name1  ABC..AD..EFDK ) => (name1 000..00..12324)
             (name2 000000112324)+(name2  --HHH--EEEE-- ) => (name1 ---000--1123--)
             uses the second hash a template for the first sequences. gap_char is
             '-' or '.' or any given char or symbol.
             To insert gaps rather than overlap, use insert_gaps_in_seq_hash
<b>Keywords  </b> : Overlap, superpose hash, overlay, superpose_seq_hash
<b>Options   </b> : E for replacing All 'E' occurrances in ---EEEE--HHHH----, etc.
             : H for replacing all 'H'  "     " "
<b>Returns   </b> : one hash ref.
<b>Usage     </b> : %out =%{&overlay_seq_by_certain_chars(\%hash1, \%hash2, 'HE')};
<b>Version   </b> : 1.0
<b>Warning   </b> : If gap_chr ('H',,,) is not given, it replaces all the
             non-gap chars (normal alphabet), ie,
             it becomes 'superpose_seq_hash'
</pre><hr>

<H3><A href="rev_lines_pdb
.pl.html">rev_lines_pdb
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//rev_lines_pdb
.pl">rev_lines_pdb
.pl</A>
<pre>
<b>Argument  </b> : one pdb coordinate file reference
<b>Example   </b> : 
             The INPUT example >

             ATOM    191  CA  ALA   195      -2.566   8.099  42.827  1.00 12.42      1ENG 256
             ATOM    192  CA  ARG   196      -1.401  11.546  41.629  1.00  8.63      1ENG 257
             ATOM    193  CA  THR   197      -4.073  13.846  43.107  1.00  9.93      1ENG 258

             The OUTPUT example >             <first file, called  xxxx1.atm >

             ATOM      1  CA  ALA     1      -2.566   8.099  42.827  1.00 12.42      1ENG 256
             ATOM      2  CA  ARG     2      -1.401  11.546  41.629  1.00  8.63      1ENG 257
             ATOM      3  CA  THR     3      -4.073  13.846  43.107  1.00  9.93      1ENG 258

                                           <2nd file, called  xxxx2.atm >
             ATOM      1  CA  THR     1      -4.073  13.846  43.107  1.00  9.93      1ENG 258
             ATOM      2  CA  ARG     2      -1.401  11.546  41.629  1.00  8.63      1ENG 257
             ATOM      3  CA  ALA     3      -2.566   8.099  42.827  1.00 12.42      1ENG 256

<b>Function  </b> : reorders the lines of any pdb files, but takes only C alpha positions.
<b>Options   </b> : None
<b>Returns   </b> : directly writes two output files  xxxx1.atm  xxxx2.atm
<b>Usage     </b> : &rev_lines_pdb(\$ARGV[0]);
<b>Version   </b> : 1.0
<b>Warning   </b> : A Biomatic
</pre><hr>

<H3><A href="tally_2_hashes
.pl.html">tally_2_hashes
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//tally_2_hashes
.pl">tally_2_hashes
.pl</A>
<pre>
<b>Argument  </b> : (\%hash1, \%hash2) or optionally (\%hash1, \%hash2, ['n', 'i', 'p', 'a'])
             'n' => normalizing, 'p' => percentage out, 'i' => make int out, 'a'=> averaged
<b>Example   </b> : you put two hash refs. (ass. array) as args (\%hash1, \%hash2)
             The hashes are like; hash1  (name1, 0000011111, name2, 0000122222 );
                                  hash2  (name3, 1324..1341, name4, 13424444.. );

             1) The resulting 1st hash output is (0, 20,   1, 13,     2, 12)
             which means that 0 added up to 24 in the second arg hash positions
                              1 added up to 15 in the second arg hash positions
                              2 added up to 18 in the second arg hash positions
             'p' option only works with 'n' or 'a'
             2) The resulting 2nd hash output is (0, 5,   1, 5)
             which means that 0 occurred 5 times in the first input hash
                              1 occurred 5 times in the first input hash
             'p' option only works with 'n' or 'a'
<b>Function  </b> : Makes hashes of tallied occurances and summed up values for disits in
             positions.
             calculates the occurances or occurance rates of CS rate positions.
             The hashes should have numbers.
<b>Keywords  </b> : tally two hashes of numbers.
<b>Options   </b> : [a n i p]
<b>Returns   </b> : ($ref1, $ref2), ie, two references of hash
             averaging option causes division of 20(added up value)
                                                by 9(occurance) in the above
             for '0' of the first hash, so (0, 2.222,  1, 2.1666,  2, 2.4 )
             Average is the average of numbers
             average value in 0-9 scale (or 0-100 with 'p' option)
             So, if there are
                  seq1 00111110000,   The 'a' value of 0 and 1 as in the seq2
                  seq2 33000040000    is 0-> 6/6, 1-> 4/5, while the 'n'
                                        calc would be, 0-> 6 (60%), 1-> 4(40%)

<b>Usage     </b> : ($ref1, $ref2) = &tally_2_hashes(\%hash1, \%hash2, ['n', 'a', 'p', 'i']);
              %tally_addedup=%{$ref1};    '0' position had addedup value of 1000
              %tally_occurances=%{$ref2}; '0' position had occurred 100 times,
                                          '0' on average had 10 in its
                                              corresponding hash positions
<b>Version   </b> : 1.2
</pre><hr>

<H3><A href="superpose_seq_hash
.pl.html">superpose_seq_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//superpose_seq_hash
.pl">superpose_seq_hash
.pl</A>
<pre>
<b>Argument  </b> : 2 refs. for hash of identical keys and value length and gap_chr.
<b>Function  </b> : (name1 000000112324)+(name1  ABC..AD..EFD ) => (name1 000..01..324)
             uses the second hash a template for the first sequences. gap_char is
             '-' or '.'
             To insert gaps rather than overlap, use insert_gaps_in_seq_hash
<b>Keywords  </b> : overlay sequence, overlay alphabet, superpose sequence,
<b>Returns   </b> : one hash ref.
<b>Usage     </b> : %out =%{&superpose_seq_hash(\%hash1, \%hash2)};
<b>Version   </b> : 1.0
<b>Warning   </b> : Accepts only two HASHes and many possible gap_chr. Default gap is '-'
</pre><hr>

<H3><A href="overlay_seq_hash
.pl.html">overlay_seq_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//overlay_seq_hash
.pl">overlay_seq_hash
.pl</A>
<pre>
<b>Argument  </b> : 2 refs. for hash of identical keys and value length and gap_chr.
<b>Function  </b> : (name1 000000112324)+(name1  ABC..AD..EFD ) => (name1 000..01..324)
             uses the second hash a template for the first sequences. gap_char is
             '-' or '.'
             To insert gaps rather than overlap, use insert_gaps_in_seq_hash
<b>Returns   </b> : one hash ref.
<b>Usage     </b> : %out =%{&overlay_seq_hash(\%hash1, \%hash2)};
<b>Version   </b> : 1.0
<b>Warning   </b> : Accepts only two HASHes and many possible gap_chr. Default gap is '-'
</pre><hr>

<H3><A href="insert_gaps_in_seq_hash
.pl.html">insert_gaps_in_seq_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//insert_gaps_in_seq_hash
.pl">insert_gaps_in_seq_hash
.pl</A>
<pre>
<b>Argument  </b> : 2 ref for hash of identical keys and value length.
<b>Function  </b> : superpose two hashes of the same sequence or same seq. length sequences,
             but unlike 'superpose_seq_hash', this inserts gaps and extend the
             sequences.
             (name1_sec  hHHHHHH EEEEEEE) +
             (name1_seq  .CDEABC..AD..EFD..EKST) => (name1_ext  .hHHHHH..H...EEE..EEEE)
             In the example, the undefined sec. str. position is replaced as gaps('.')
             Uses the second hash a template for the first sequences. gap_char is
             '-' or '.'
             One rule is that the SECOND hash contains gaps!!
             There are two types of hash input. One is simple seq hash(both args)
              The other is from secondary structure prediction. The hash has contents
              like: $averaged{$position}=[$residue1, $sec_str2, $dif_reliability];
<b>Keywords  </b> : superposing sequences with gaps, interpolate_sequences, interpolate_gaps
<b>Returns   </b> : one hash ref.
<b>Usage     </b> : %out_extended_seq =%{&insert_gaps_in_seq_hash(\%hash1, \%hash2)};
<b>Version   </b> : 1.3
<b>Warning   </b> : coded by A Biomatic
</pre><hr>

<H3><A href="scan_win_get_average
.pl.html">scan_win_get_average
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//scan_win_get_average
.pl">scan_win_get_average
.pl</A>
<pre>
<b>Example   </b> : input hash: ( seq1,  '13241234141234234',      (2 or more sequences accepted)
                           seq2,  '1341324123413241234')
             input winsize : 5;

             output hash; (seq1, 1234123413241234);
             output hash; (seq2, 1344234123412341);
                  The numbers are ratios(compos/seqid) with given window size.
<b>Usage     </b> : %out1 = %{&scan_win_get_av(\%input, \$window_size, \%input2,,,,)};
             The order of the arguments doesn't matter.
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="scan_win_and_get_sc_rate_pairs
.pl.html">scan_win_and_get_sc_rate_pairs
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//scan_win_and_get_sc_rate_pairs
.pl">scan_win_and_get_sc_rate_pairs
.pl</A>
<pre>
<b>Argument  </b> : One ref. for hash, one ref. for a scalar.
<b>Example   </b> : input hash: ( seq1,  'ABCDEFG.HIK',      (2 or more sequences accepted)
                         seq2,  'DFD..ASDFAFS',
                         seq3,  'DDDDD..ASDFAFS' );
             input winsize : 5;

             output hash; (seq1seq2, 1,2,2,2,1,1,2,2); <-- joined by ',';
             output hash; (seq1seq3, 1,2,2,2,1,1,2,2); <-- joined by ',';
                  The numbers are ratios(compos/seqid) with given window size.
<b>Function  </b> : scans input sequences(arg1) in a given(arg2) window size and gets
             each composition and sequence identity rate(sc_rate) of the window.
             sc rate = Sequence Id(%)/ Composition Id(%)
<b>Returns   </b> : a reference of a hash.
<b>Usage     </b> : %out1 = %{&scan_win_and_get_sc_rate_pairs(\%input, \$window_size)};
<b>Version   </b> : 1.1
<b>Warning   </b> : when $seqid is zero  the rate becomes $compos_id/10   !!!
</pre><hr>

<H3><A href="get_windows_sc_rate_array
.pl.html">get_windows_sc_rate_array
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_windows_sc_rate_array
.pl">get_windows_sc_rate_array
.pl</A>
<pre>
<b>Argument  </b> : (\@input, \$window_size);  @input => ('ABCDEFG.HIK', 'DFD..ASDFAFS', 'ASDFASDFASAS');
             Input ar => ( 'ABCDEFG
                'DFD..ASDFAFS'
                'ASDFASDFASAS' )  as the name of  @sequences.
<b>Author    </b> : A Biomatic
<b>Function  </b> : actual working part of scan_windows_and_get_compos_seqid_rate
<b>Returns   </b> : \@ratio_array, \$ratio_whole_seq
<b>Usage     </b> : @out_rate = @{&get_windows_compos_and_seqid_rate_array(\@seq, \$win_size)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="scan_win_and_get_cs_rate_pairs
.pl.html">scan_win_and_get_cs_rate_pairs
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//scan_win_and_get_cs_rate_pairs
.pl">scan_win_and_get_cs_rate_pairs
.pl</A>
<pre>
<b>Argument  </b> : One ref. for hash, one ref. for a scalar.
<b>Example   </b> : input hash: ( seq1,  'ABCDEFG.HIK',      (2 or more sequences accepted)
                               seq2,  'DFD..ASDFAFS',
                               seq3,  'DDDDD..ASDFAFS' );
                 input winsize : 5;

                 output hash; (seq1seq2, 1,2,2,2,1,1,2,2); <-- joined by ',';
                 output hash; (seq1seq3, 1,2,2,2,1,1,2,2); <-- joined by ',';
             The numbers are ratios(compos/seqid) with given window size.

<b>Function  </b> : scans input sequences(arg1) in a given(arg2) window size and gets
             each composition and sequence identity rate(cs_rate) of the window.
             CS rate = Composition Id / Sequence Id
<b>Returns   </b> : a reference of a hash.
             It is getting the entropy of the column and calculates something after.
<b>Usage     </b> : %out1 = %{&scan_win_and_get_cs_rate_pairs(\%input, \$window_size)};
<b>Version   </b> : 1.0
<b>Warning   </b> : when $seqid is zero  the rate becomes $compos_id/10   !!!
</pre><hr>

<H3><A href="get_residue_error_rate
.pl.html">get_residue_error_rate
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_residue_error_rate
.pl">get_residue_error_rate
.pl</A>
<pre>
<b>Argument  </b> : Takes a ref. for hash which have positions of residues of sequences.
<b>Function  </b> : This is the final step in error rate getting.
             gets a ref. of a hash and calculates the absolute position diffs.
<b>Options   </b> : 'L' for limitting the error rate to 9 to make one digit output
$LIMIT becomes 'L' by L, l, -l, -L
<b>Returns   </b> : one ref. for an array of differences of input arrays. array context.
             ---Example input (a hash with sequences); The values are differences after
                                comparion with structural and sequential alignments.
             %diffs =('seq1', '117742433441...000',   <-- input (can be speparated by '' or ','.
                      'seq2', '12222...99999.8888',
                      'seq3', '66222...44444.8822',
                      'seq4', '12262...00666.772.');
             example output;
             seq3_seq4       '0,1,0,0,0,.,.,.,,.,0,,0,0,,0,0,,.,0,,0,0,.'
             seq1_seq2       '0,1,0,1,1,.,.,.,,.,2,,2,2,,2,2,,.,.,,2,2,1'
             seq1_seq3       '0,1,0,1,1,.,.,.,,.,1,,1,1,,0,.,,.,.,,1,1,1'
             seq1_seq4       '0,1,0,,1,1,.,.,.,,.,1,,1,1,0,.,.,,.,1,,2,2'
             seq2_seq3       '0,1,0,,0,0,,.,.,,.,0,,1,0,,0,0,,.,0,,0,0,0'
             seq2_seq4       '0,0,0,,1,0,,.,.,,.,0,,1,0,,0,0,,.,0,,0,0,.'
<b>Usage     </b> : %position_diffs =%{&get_residue_error_rate(\@seq_position1, \@seq_position2)};
<b>Version   </b> : 1.1
<b>Warning   </b> : split and join char is ',';
</pre><hr>

<H3><A href="get_each_posi_diff_hash
.pl.html">get_each_posi_diff_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_each_posi_diff_hash
.pl">get_each_posi_diff_hash
.pl</A>
<pre>
<b>Argument  </b> : Takes a ref. for hash which have positions of residues of sequences.
<b>Function  </b> : This is the final step in error rate getting.
             gets a ref. of a hash and calculates the position diffs.
<b>Options   </b> : 'L' for limitting the error rate to 9 to make one digit output
$LIMIT becomes 'L' by L, l, -l, -L

<b>Returns   </b> : one ref. for an array of differences of input arrays. array context.
             ---Example input (a hash with sequences); The values are differences after
                                comparion with structural and sequential alignments.
             %diffs =('seq1', '117742433441...000',   <-- input (can be speparated by '' or ','.
                      'seq2', '12222...99999.8888',
                      'seq3', '66222...44444.8822',
                      'seq4', '12262...00666.772.');
             example output;
             seq3_seq4       '0,1,0,0,0,.,.,.,,.,0,,0,0,,0,0,,.,0,,0,0,.'
             seq1_seq2       '0,1,0,1,1,.,.,.,,.,2,,2,2,,2,2,,.,.,,2,2,1'
             seq1_seq3       '0,1,0,1,1,.,.,.,,.,1,,1,1,,0,.,,.,.,,1,1,1'
             seq1_seq4       '0,1,0,,1,1,.,.,.,,.,1,,1,1,0,.,.,,.,1,,2,2'
             seq2_seq3       '0,1,0,,0,0,,.,.,,.,0,,1,0,,0,0,,.,0,,0,0,0'
             seq2_seq4       '0,0,0,,1,0,,.,.,,.,0,,1,0,,0,0,,.,0,,0,0,.'
<b>Usage     </b> : %position_diffs =%{&get_each_posi_diff_hash(\@seq_position1, \@seq_position2)};
<b>Version   </b> : 1.0
<b>Warning   </b> : split and join char is ',';
</pre><hr>

<H3><A href="get_posi_rates_hash_out
.pl.html">get_posi_rates_hash_out
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_posi_rates_hash_out
.pl">get_posi_rates_hash_out
.pl</A>
<pre>
<b>Argument  </b> : %{&get_posi_rates_hash_out(\%msfo_file, \%jpo_file)};
             Whatever the names, it takes one TRUE structral and one ALIGNED hash.
<b>Function  </b> : This is to get position specific error rate for line display rather than
             actual final error rate for the alignment.
             Output >>
             seq1_seq2  1110...222...2222
             seq2_seq3  1111....10...1111
             seq1_seq3  1111....0000.0000

<b>Returns   </b> : \%final_posi_diffs;
<b>Usage     </b> : %rate_hash = %{&get_posi_shift_hash(\%hash_msf, \%hash_jp)};
<b>Version   </b> : 1.0
<b>Warning   </b> : split and join char is ','; (space)
</pre><hr>

<H3><A href="get_posi_rates_hash_out_compact
.pl.html">get_posi_rates_hash_out_compact
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_posi_rates_hash_out_compact
.pl">get_posi_rates_hash_out_compact
.pl</A>
<pre>
<b>Argument  </b> : %{&get_posi_rates_hash_out(\%msfo_file, \%jpo_file)};
             Whatever the names, it takes one TRUE structral and one ALIGNED hash.
<b>Function  </b> : This is to get position specific error rate for line display rather than
             actual final error rate for the alignment.
             Output >>  something like below but, without gaps, so final one is;
             seq1_seq2  1110...222...2222     seq1_seq2  11102222222
             seq2_seq3  1111....10...1111  -> seq2_seq3  1111101111
             seq1_seq3  1111....0000.0000     seq1_seq3  111100000000

<b>Returns   </b> : \%final_posi_diffs_compact;  Compare with  'get_posi_rates_hash_out_jp'
<b>Usage     </b> : %rate_hash = %{&get_posi_shift_hash(\%hash_msf, \%hash_jp)};
<b>Version   </b> : 1.0
<b>Warning   </b> : split and join char is ','; (space)
</pre><hr>

<H3><A href="get_posi_rates_hash_out_jp
.pl.html">get_posi_rates_hash_out_jp
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_posi_rates_hash_out_jp
.pl">get_posi_rates_hash_out_jp
.pl</A>
<pre>
<b>Argument  </b> : %{&get_posi_rates_hash_out_jp(\%msfo_file, \%jpo_file)};
             Whatever the names, it takes one TRUE structral and one ALIGNED hash.
<b>Function  </b> : This is to get position specific error rate for line display rather than
             actual final error rate for the alignment. get_posi_rates_hash_out_jp
             results in jp template sequence, while get_posi_rates_hash_out_msf does
             in msf template sequence.
             Output >>
             seq1_seq2  1110...222...2222   <--- the alignment template is JPO's
             seq2_seq3  1111....10...1111        (ie structural)
             seq1_seq3  1111....0000.0000

<b>Returns   </b> : \%final_posi_diffs;
<b>Usage     </b> : %rate_hash = %{&get_posi_shift_hash(\%hash_msf, \%hash_jp)};
<b>Version   </b> : 1.0
<b>Warning   </b> : split and join char is ','; (space)
</pre><hr>

<H3><A href="get_posi_rates_hash_out
.pl.html">get_posi_rates_hash_out
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_posi_rates_hash_out
.pl">get_posi_rates_hash_out
.pl</A>
<pre>
<b>Argument  </b> : %{&get_posi_rates_hash_out(\%msfo_file, \%jpo_file)};
             Whatever the names, it takes one TRUE structral and one ALIGNED hash.
             Output >>
             seq1_seq2  1110...222...2222
             seq2_seq3  1111....10...1111
             seq1_seq3  1111....0000.0000

<b>Function  </b> : This is to get position specific error rate for line display rather than
             actual final error rate for the alignment.
<b>Returns   </b> : \%final_posi_diffs;
<b>Usage     </b> : %rate_hash = %{&get_posi_shift_hash(\%hash_msf, \%hash_jp)};
<b>Version   </b> : 1.0
<b>Warning   </b> : split and join char is ','; (space)
</pre><hr>

<H3><A href="normalize_numbers
.pl.html">normalize_numbers
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//normalize_numbers
.pl">normalize_numbers
.pl</A>
<pre>
<b>Argument  </b> : (\%hash1, %hash2, \%hash3, ....)
<b>Example   </b> : intputhash>                   Outputhash>
             ( '1-2', '12,.,1,2,3,4',     ( '1-2',   '9,.,0,1,2,3',
              '2-3', '12,.,1,5,3,4',       '2-3',   '9,.,0,4,2,3',
              '4-3', '12,3,1,2,3,4',       '3-1',   '9,3,.,.,2,3',
              '3-1', '12,4,.,.,3,4' );     '4-3',   '9,2,0,1,2,3' );
<b>Function  </b> : with given numbers in hashes, it makes a scale of 0-9 and puts
             all the elements in the scale. Also returns the average of the numbs.
<b>Returns   </b> : (\%norm_hash1, \%norm_hash2, \%norm_hash3,.... )

<b>Usage     </b> : %output=%{&normalize_numbers(\%hash1)};
             originally made to normalize the result of get_posi_rates_hash_out
             in   'scan_compos_and_seqid.pl'
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="scan_windows_and_get_compos_seqid_rate
.pl.html">scan_windows_and_get_compos_seqid_rate
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//scan_windows_and_get_compos_seqid_rate
.pl">scan_windows_and_get_compos_seqid_rate
.pl</A>
<pre>
<b>Argument  </b> : One ref. for hash, one ref. for a scalar.
<b>Example   </b> : input hash: ( seq1,  'ABCDEFG.HIK',    (2 or more sequences accepted)
                           seq2,  'DFD..ASDFAFS',
                           seq3,  'DDDDD..ASDFAFS' );
             input winsize : 5;

             output hash; (seq1seq2, 1,2,2,2,1,1,2,2); <-- joined by ',';
                  The numbers are ratios(compos/seqid) with given
                  window size.
<b>Function  </b> : scans input sequences(arg1) in a given(arg2) window size and gets
             each composition and sequence identity rate of the window.
<b>Returns   </b> : a reference of a hash.
<b>Usage     </b> : %out1 =%{&scan_windows_and_get_compos_seqid_rate(\%input, \$window_size)};
<b>Warning   </b> : when $seqid is zero  the rate becomes $compos_id/10   !!!
</pre><hr>

<H3><A href="get_windows_cs_rate_array
.pl.html">get_windows_cs_rate_array
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_windows_cs_rate_array
.pl">get_windows_cs_rate_array
.pl</A>
<pre>
<b>Argument  </b> : (\@input, \$window_size);  @input => ('ABCDEFG.HIK', 'DFD..ASDFAFS', 'ASDFASDFASAS');
             Input ar => ( 'ABCDEFG
                'DFD..ASDFAFS'
                'ASDFASDFASAS' )  as the name of  @sequences.
<b>Function  </b> : actual working part of scan_windows_and_get_compos_seqid_rate
<b>Returns   </b> : \@ratio_array, \$ratio_whole_seq
<b>Usage     </b> : @out_rate = @{&get_windows_cs_rate_array(\@seq, \$win_size)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="read_any_seq_files
.pl.html">read_any_seq_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//read_any_seq_files
.pl">read_any_seq_files
.pl</A>
<pre>
<b>Argument  </b> : one of more ref. for scalar.
<b>Example   </b> : (*out1,  *out2) =&read_any_seq_files(\$input1, \$input2);
             : (@out_ref_array)=@{&read_any_seq_files(\$input1, \$input2)};
             : (%one_hash_out) =%{&read_any_seq_files(\$input1)};
<b>Function  </b> : Tries to find given input regardless it is full pathname, with or
             without extension. If not in pwd, it searches the dirs exhaustively.
<b>Keywords  </b> : open_any_seq_files,
<b>Returns   </b> : 1 ref. for a HASH of sequence ONLY if there was one hash input
             1 array (not REF.) of references for multiple hashes.
<b>Usage     </b> : %out_seq=%{&read_any_seq_files(\$input_file_name)};
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="seq_to_regexp
.pl.html">seq_to_regexp
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//seq_to_regexp
.pl">seq_to_regexp
.pl</A>
<pre>
<b>Function  </b> : given an array and a start and end length,
              return an array of regular expressions, where each element of the original
              array has been expanded to a set of regular expressions that match the
              original exactly num times, for num between the start and end length

<b>Returns   </b> : a ref. of an array for
<b>Version   </b> : 1.0
<b>Warning   </b> : Copyright (C) 1993-1994 by James Tisdall
</pre><hr>

<H3><A href="strip_rotated_seq
.pl.html">strip_rotated_seq
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//strip_rotated_seq
.pl">strip_rotated_seq
.pl</A>
<pre>
<b>Function  </b> : remove all but one string of each set of rotations
             (reverse of rotated_seq )
<b>Returns   </b> : a ref. for
<b>Version   </b> : 1.0
<b>Warning   </b> : Copyright (C) 1993-1994 by James Tisdall
             stolen from Tisdall
</pre><hr>

<H3><A href="rotate_seq
.pl.html">rotate_seq
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//rotate_seq
.pl">rotate_seq
.pl</A>
<pre>
<b>Function  </b> : given a string, return all the rotations of that string
             e.g. given 'abcd', return ('abcd','bcda','cdab','dabc')
<b>Returns   </b> : a ref. for reverse complement
<b>Usage     </b> : @out_array=@{&rotate_seq($string)};
<b>Version   </b> : 1.0
<b>Warning   </b> : Copyright (C) 1993-1994 by James Tisdall
             stolen from Tisdall   ##### RevCom
</pre><hr>

<H3><A href="convert_to_anti_sense
.pl.html">convert_to_anti_sense
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//convert_to_anti_sense
.pl">convert_to_anti_sense
.pl</A>
<pre>
<b>Returns   </b> : a ref. for reverse complement
<b>Version   </b> : 1.0
<b>Warning   </b> : Copyright (C) 1993-1994 by James Tisdall
             stolen from Tisdall   ##### RevCom
</pre><hr>

<H3><A href="convert_rna_to_protein
.pl.html">convert_rna_to_protein
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//convert_rna_to_protein
.pl">convert_rna_to_protein
.pl</A>
<pre>
<b>Argument  </b> : a scalar for RNA sequence data
<b>Function  </b> : translate RNA seq to protein seq.
<b>Keywords  </b> : rna2protein, rna_2_protein, RNA2protein, translate_rna
             dna2protein, convert_RNA_to_protein, RNA_2_PROTEIN, RNA_2_protein
<b>Returns   </b> : a ref. of an array for protein translation
<b>Version   </b> : 1.1
<b>Warning   </b> : Copyright (C) 1993-1994 by James Tisdall
             stolen from Tisdall
</pre><hr>

<H3><A href="convert_dna_to_protein
.pl.html">convert_dna_to_protein
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//convert_dna_to_protein
.pl">convert_dna_to_protein
.pl</A>
<pre>
<b>Argument  </b> : a scalar for DNA sequence data
<b>Function  </b> : translate DNA or RNA seq to protein seq.
<b>Keywords  </b> : dna2protein, dna_2_protein, DNA2protein, translate_dna
             dna2protein, convert_DNA_to_protein, translate_nucleic_acid
             rna2protein, rna_2_protein, RNA2protein, translate_rna
             dna2protein, convert_RNA_to_protein
<b>Returns   </b> : a ref. of an array for protein translation
<b>Version   </b> : 1.2
<b>Warning   </b> : Copyright (C) 1993-1994 by James Tisdall
             stolen from Tisdall
</pre><hr>

<H3><A href="write_staden_file
.pl.html">write_staden_file
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//write_staden_file
.pl">write_staden_file
.pl</A>
<pre>
<b>Returns   </b> : a ref. of an array for  STADEN formatted sequence record
<b>Version   </b> : 1.0
<b>Warning   </b> : Copyright (C) 1993-1994 by James Tisdall
             stolen from Tisdall
</pre><hr>

<H3><A href="write_primer_file
.pl.html">write_primer_file
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//write_primer_file
.pl">write_primer_file
.pl</A>
<pre>
<b>Returns   </b> : a ref. of an array for PRIMER formatted sequence record
<b>Version   </b> : 1.0
<b>Warning   </b> : Copyright (C) 1993-1994 by James Tisdall
             stolen from Tisdall
</pre><hr>

<H3><A href="write_gcg_genbank_file
.pl.html">write_gcg_genbank_file
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//write_gcg_genbank_file
.pl">write_gcg_genbank_file
.pl</A>
<pre>
<b>Returns   </b> : a ref. of an array for GCG-Genbank formatted sequence record
<b>Version   </b> : 1.0
<b>Warning   </b> : Copyright (C) 1993-1994 by James Tisdall
             stolen from Tisdall
</pre><hr>

<H3><A href="write_pir_file
.pl.html">write_pir_file
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//write_pir_file
.pl">write_pir_file
.pl</A>
<pre>
<b>Returns   </b> : a ref. of an array for PIR formatted sequence record
<b>Version   </b> : 1.0
<b>Warning   </b> : Copyright (C) 1993-1994 by James Tisdall
             from Tisdall
</pre><hr>

<H3><A href="write_genbank_file
.pl.html">write_genbank_file
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//write_genbank_file
.pl">write_genbank_file
.pl</A>
<pre>
<b>Argument  </b> : two scalars.
<b>Function  </b> : (This is DNA seq handling routine!)
<b>Returns   </b> : a ref. of an array for Genbank formatted sequence record
<b>Usage     </b> : @out =  @{&write_genbank_file($sequ, $header)};
<b>Version   </b> : 1.0
<b>Warning   </b> : Copyright (C) 1993-1994 by James Tisdall
             stolen from Tisdall
</pre><hr>

<H3><A href="write_gcg_file
.pl.html">write_gcg_file
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//write_gcg_file
.pl">write_gcg_file
.pl</A>
<pre>
<b>Returns   </b> : a ref. of an array for GCG formatted sequence record
<b>Version   </b> : 1.0
<b>Warning   </b> : Copyright (C) 1993-1994 by James Tisdall
</pre><hr>

<H3><A href="write_fasta_array
.pl.html">write_fasta_array
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//write_fasta_array
.pl">write_fasta_array
.pl</A>
<pre>
<b>Argument  </b> : \%input
<b>Example   </b> : @out = (
             $out[0] =>     ">name",
             $out[1] =>     "ABCDEABCDEBCDEABCDEABCDEABCDEABCDEBCDEABCDE",
             $out[2] =>     "TTTTTTTTDEBCDEABCDEABCDEABCDEABCDEBCDEABCDE",
             $out[3] =>     "ABCDEABCDEBCDEABCDEABCDEABCDEABCDEBCDEABCDE",
                 );

<b>Function  </b> : take Single sequence and produce single output array of fasta
<b>Returns   </b> : ref. for an array of FASTA formatted sequence record

<b>Usage     </b> : @output = @{&put_fasta($sequence, $name)};
<b>Version   </b> : 1.0
<b>Warning   </b> : Copyright (C) 1993-1994 by James Tisdall
</pre><hr>

<H3><A href="find_seq_files
.pl.html">find_seq_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//find_seq_files
.pl">find_seq_files
.pl</A>
<pre>
<b>Argument  </b> : (\$input_file_name) while $input_file_name can be  'xxx.xxx', or '/xxx/xxx/xxx/xxy.yyy'
             or just directory name like 'aat' for  /nfs/ind4/ccpe1/people/A Biomatic /jpo/align/aat
             then, it tries to find a file with stored seq file extensions like msf, jp, pir etc
             to make aat.msf, aat.jp, aat.pir ... and searches for these files.
<b>Example   </b> : $found_file=${&find_seq_files(\$input_file_name)};
<b>Function  </b> : (similar to find.pl) used in 'read_any_seq_file.pl'
             seeks given test file in pwd, specified dir, default path etc.
             If not found yet, it looks at all the subdirectories of path and pwd.
             PATH environment dirs, then returns full path file name.
<b>Keywords  </b> : find_anyj_seq_files, find any seq files, find seq files
<b>Returns   </b> : return( \$final );
<b>Usage     </b> : $found_file = ${&find_seq_files(\$input_file_name)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="search_files_in_subdir
.pl.html">search_files_in_subdir
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//search_files_in_subdir
.pl">search_files_in_subdir
.pl</A>
<pre>
<b>Argument  </b> : gets a ref. of a scaler (dir name) and returns nothing(void).
<b>Function  </b> : open dir and process all files in the dir if you wish,
             and then go in any other sub
             if any file(dir) is linked, it skips that file.
<b>Usage     </b> : 
                     $inputdir='/nfs/ind4/ccpe1/people/A Biomatic /jpo/align';
<b>Version   </b> : 1.0
<b>Warning   </b> : the final var $found_from_search_files_in_subdir mustn't be 'my'ed.
</pre><hr>

<H3><A href="find_seq_file_old
.pl.html">find_seq_file_old
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//find_seq_file_old
.pl">find_seq_file_old
.pl</A>
<pre>
<b>Argument  </b> : one ref. for SCALAR
<b>Function  </b> : seeks text file in pwd. If not found it looks at
             PATH environment dirs
<b>Returns   </b> : one ref. for SCALAR of a full path filename.
<b>Usage     </b> : $found_file=${&find_seq_file_old(\$input_file_name)};
<b>Version   </b> : 1.0
<b>Warning   </b> : << This is READABLE old version of  find_seq_file
</pre><hr>

<H3><A href="open_sst_files_with_gap
.pl.html">open_sst_files_with_gap
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_sst_files_with_gap
.pl">open_sst_files_with_gap
.pl</A>
<pre>
<b>Argument  </b> : a ref. for scaler of "jp file name"
<b>Example   </b> : jp file  ==  seq1 ABDSF--DSFSDFS   <- true sequence
                              seq2 T--kdf-GAGGGASF     (aligned)

                 sst files ==> 'seq1.sst', 'seq2.sst' (in the same dir)

             original sst format:  seq1 hHHHHHttEEEE  <-- No gaps!
                                  seq2 hHHHHHHEEhh
             After this sub ==>
             (final out hash =   (  seq1 hHHHH--HttEEEE  <-- inserted
                                  seq2 h--HHH-HHHEEEhh  )     gaps

<b>Function  </b> : gets the name of a file(jp file) with its absolute dir path
             reads the sequence names in the jp file and looks up all
             the sst files in the same directory. Puts sst sequences
             in a hash with keys of sequence names.

<b>Returns   </b> : a ref. for a hash
<b>Usage     </b> : %out_sst_hash =%{&open_sst_files_with_gap(\$jp_file_dir_and_name)};
<b>Version   </b> : 1.0
<b>Warning   </b> : $jp_file_dir_and_name should be absolute dir and file name
             >> This gets JP file not SST file as input !!!!
</pre><hr>

<H3><A href="put_gaps_in_hash
.pl.html">put_gaps_in_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//put_gaps_in_hash
.pl">put_gaps_in_hash
.pl</A>
<pre>
<b>Argument  </b> : 2 hash references.
<b>Returns   </b> : one hash reference.
<b>Usage     </b> : %out=%{&put_gaps_in_hash(\%hash_with_gap, \%hash_sans_gap)};

             %hash1=('1ctx',  '111111111111111',      <-- hash input without gaps
                     '2ctx',  '2222222222222222',
                     '3ctx',  '3333333333');

             %hash2=('1ctx',  'AAA--AAAAAAAAAAAA-',   <-- hash input with template gaps
                     '2ctx',  'BBBBBBBBBBBB-BBBB',
                     '3ctx',  'CCCCCC----CCCC');

             >> resulting out hash;

             %hash3=('1ctx',     '111--111111111111-',
                     '2ctx',     '222222222222-2222',
                     '3ctx',     '333333----3333 );

<b>Version   </b> : 1.0
<b>Warning   </b> : The keys for hashes should be the same and the two sequences
             should be identical.
</pre><hr>

<H3><A href="get_gap_positions
.pl.html">get_gap_positions
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_gap_positions
.pl">get_gap_positions
.pl</A>
<pre>
<b>Argument  </b> : 1 ref. of array eg)=( ABCDE--EF--GH ) while '-' is for gap.
<b>Example   </b> : for a string '--iu--sdf-j--', it will output  -2 -1 2 3 7 9 10
<b>Function  </b> : gets gap positions of seq. and stores in an array
<b>Keywords  </b> : get_gap_positions_in_seq, get_seq_gap_positions get_gap_positions_in_array
<b>Options   </b> : p for all positive gaps numbering. No negatives for '---STRING--'

<b>Returns   </b> : 1 ref. of array eg)=(2,3,7,8,10,100,122);
<b>Usage     </b> : @gap_pos=@{&get_gap_positions(\@string1)}; <- ('A','C','D','E')
             @gap_pos=@{&get_gap_positions(\$string1)}; <- ( ACDE )
<b>Version   </b> : 1.4
<b>Warning   </b> : uses References.
</pre><hr>

<H3><A href="make_pairs_from_hash
.pl.html">make_pairs_from_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//make_pairs_from_hash
.pl">make_pairs_from_hash
.pl</A>
<pre>
<b>Argument  </b> : one ref. of hash
<b>Example   </b> : @output=($ref1, $ref2, ....$refn)
             each $ref is the reference of a hash of a pair of sequence
             >>  %pair1 = %{$ref1}; %pair2 = %{$ref2}; %pair3 = %{$ref3};

             %pair1 is like;       %pair1 is like;       %pair3 is like;

             seq1  ABCDEFAD     seq1  ABCDEFAD        seq2  SDFSFSDF
             seq2  SDFSFSDF     seq3  SDFSFSDF        seq3  SDFSFSDF

<b>Function  </b> : returns all the possible pairs of a set of sequences in
             an array of references;

<b>Returns   </b> : one ref. of array for references for hashes.
<b>Usage     </b> : @output =@{&make_pairs_from_hash(\%input_sequence_hash);
             Input example
             %input =  seq1  ABCDEFAD
              seq2  SDFSFSDF
              seq3  SDFSFSDF

<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="mail_it
.pl.html">mail_it
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//mail_it
.pl">mail_it
.pl</A>
<pre>
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="read_fssp_files
.pl.html">read_fssp_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//read_fssp_files
.pl">read_fssp_files
.pl</A>
<pre>
<b>Function  </b> : read hssp file and put sequences in a hash
<b>Usage     </b> : %anyarray = %{&read_fssp_files(\$any_sequence_file_fssp_form)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_posi_shift_rms_whole
.pl.html">get_posi_shift_rms_whole
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_posi_shift_rms_whole
.pl">get_posi_shift_rms_whole
.pl</A>
<pre>
<b>Argument  </b> : takes 2 refs. of scalars for dir name (protein group name)
             and threshold for rms
<b>Example   </b> : (0.284994272623139   0.166781214203895)
             The first figure is for error rate with out rms consideration
             The second is for after applying threshold.
<b>Returns   </b> : two refs. of scalar values (rates)
<b>Usage     </b> : just type   get_posi_shift_rms_whole.pl
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="write_jp
.pl.html">write_jp
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//write_jp
.pl">write_jp
.pl</A>
<pre>
<b>Function  </b> : gets a ref(s) for hash and prints the content in lines of 60 char
<b>Returns   </b> : Nothing, i.e. STDOUT
<b>Usage     </b> : &write_jp(\%input_hash1,\%input_hash2, \%input_hash3.... );
<b>Version   </b> : 1.0
<b>Warning   </b> : derived from  print_in_block
</pre><hr>

<H3><A href="convert_num_to_0_or_1_hash
.pl.html">convert_num_to_0_or_1_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//convert_num_to_0_or_1_hash
.pl">convert_num_to_0_or_1_hash
.pl</A>
<pre>
<b>Argument  </b> : two references, one for hash one for scaler for threshold

<b>Example   </b> : A hash =>  name1  10012924729874924792742749748374297
                        name2  10012924729874924792710012924729874
             A threshold => 4
             !! if numbers are smaller than 4, they become 1 (or true).
             Outputhash  =>  name1  11111011011111011111011011110101111
                        name2  11111011010001011001011010010101100

             ($ref1, $ref2)=&convert_num_to_0_or_1_hash(\%hash, \%hash, \$threshold);
             above is the example when with more than 2 input hashes.
<b>Function  </b> : changes all the numbers into 0 or 1 according to threshold given.
             convert_num_0_or_1_hash converts threshold and bigger nums. to
             '0' while convert_num_0_or_1_hash_opposite converts to '1'.
<b>Usage     </b> : with a variable for threshold ->

             %out = %{&convert_num_to_0_or_1_hash(\%input_hash, \$threshold, \%input_hash2..)};

<b>Version   </b> : 1.0
<b>Warning   </b> : Threshold value is set to 0 as well as all values smaller than that.
</pre><hr>

<H3><A href="convert_num_0_or_1_hash_opposite
.pl.html">convert_num_0_or_1_hash_opposite
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//convert_num_0_or_1_hash_opposite
.pl">convert_num_0_or_1_hash_opposite
.pl</A>
<pre>
<b>Argument  </b> : two references, one for hash one for scaler for threshold

<b>Example   </b> : A hash =>  name1  10012924729874924792742749748374297
                        name2  10012924729874924792710012924729874
             A threshold => 4
             !! if numbers are smaller than 4, they become 1 (or true).
             Outputhash  =>  name1  11111011011111011111011011110101111
                        name2  11111011010001011001011010010101100

             ($ref1, $ref2)=&convert_num_to_0_or_1_hash(\%hash, \%hash, \$threshold);
             above is the example when with more than 2 input hashes.
<b>Function  </b> : changes all the numbers into 0 or 1 according to threshold given.
             convert_num_0_or_1_hash converts threshold and bigger nums. to
             '0' while convert_num_0_or_1_hash_opposite converts to '1'.
<b>Usage     </b> : with a variable for threshold ->

               %out = %{&convert_num_0_or_1_hash_opposite(\%input_hash, \$threshold)};

<b>Version   </b> : 1.0
<b>Warning   </b> : Threshold value is set to 0 as well as all values smaller than that.
</pre><hr>

<H3><A href="convert_char_to_0_or_1_hash
.pl.html">convert_char_to_0_or_1_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//convert_char_to_0_or_1_hash
.pl">convert_char_to_0_or_1_hash
.pl</A>
<pre>
<b>Argument  </b> : one reference of HASH.

<b>Example   </b> : A hash =>  name1  ABCDSSFDSF..ASDFSD.....ADFASDF...AA
                        name2  ASDFSD.....ADFBCDSSFDSF..ASASDF...A

             Outputhash  => name1  00000000001100000011111000000011100
                            name2  00000011111000000000000110000001110

<b>Function  </b> : changes all the chars into 1, gaps are to 0
<b>Keywords  </b> : convert_char, translate_char, convert_char_to_digit,
             convert_char_to_number
<b>Returns   </b> : A ref. of a hash
<b>Usage     </b> : with a variable for threshold ->

               %out = %{&convert_char_0_or_1_hash(\%input_hash)};

<b>Version   </b> : 1.2
</pre><hr>

<H3><A href="digitize_char
.pl.html">digitize_char
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//digitize_char
.pl">digitize_char
.pl</A>
<pre>
<b>Argument  </b> : one reference of HASH.

<b>Example   </b> : A hash =>  name1  ABCDSSFDSF..ASDFSD.....ADFASDF...AA
                        name2  ASDFSD.....ADFBCDSSFDSF..ASASDF...A

             Outputhash  => name1  00000000001100000011111000000011100
                            name2  00000011111000000000000110000001110

<b>Function  </b> : changes all the chars into 1, gaps are to 0
<b>Keywords  </b> : convert_char, translate_char, convert_char_to_digit,
             convert_char_to_number, digitize_sequence, digitize_char
             digitize_hash
<b>Returns   </b> : A ref. of a hash
<b>Usage     </b> : with a variable for threshold ->

               %out = %{&digitize_char(\%input_hash)};

<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="get_posi_diff_and_rms_hash
.pl.html">get_posi_diff_and_rms_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_posi_diff_and_rms_hash
.pl">get_posi_diff_and_rms_hash
.pl</A>
<pre>
<b>Argument  </b> : Takes two ref. for hash
<b>Function  </b> : gets two ref. of hashes and calculates the position diffs.
<b>Returns   </b> : one ref. for an array of differences of input arrays. array context.
             ---Example input (a hash with numbers); The values are differences after comparion
                                            with structural and sequential alignments.
             %diffs =('seq1', '112342431111
             'seq2', '12222...09011.1122',
             'seq3', '13222...00011.1122',
             'seq4', '12262...00011.112.');

             %rms_corrected_0_or_1 => seq1_seq2  0111011111011101011110100101101010011
                           seq1_seq3  01111.....111110111111111111100001011
             example output;
             seq3_seq4       01040...00000.000.
             seq1_seq2       01012...1810...122
             seq1_seq3       02012...1110...122
             seq1_seq4       01032...1110...12.
             seq2_seq3       01000...09000.0000
             seq2_seq4       00040...09000.000.

<b>Usage     </b> : %position_diffs =\{&get_posi_diff_hash(\%diffs, \%rms_corrected)};
<b>Version   </b> : 1.0
<b>Warning   </b> : split and join char is ",";
</pre><hr>

<H3><A href="get_posi_shift_rms_hash
.pl.html">get_posi_shift_rms_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_posi_shift_rms_hash
.pl">get_posi_shift_rms_hash
.pl</A>
<pre>
<b>Argument  </b> : takes 4 hash REFERENCES for (one seq. and one struc. alignment(2nd arg)
<b>Returns   </b> : two refs. for scalar values of shift rate of positions for proteins.
              frirst scalar is rate without correcting rms deviation
              second scalar is rate with    correcting rms deviation
             >> example of xx

             1cdg            APDTSVSNKQ NFSTDVIYQI FTDRFSDGNP ANNPTGAAFD GTC.TNLRLY
             2aaa            ......LSAA SWRTQSIYFL LTDRFGR... ....TDNSTT ATCNTGNEIY

             >> example of xx

             2aaa       ------lsaasWrtqSIYFLLTDRFGrtdns-------ttatCntgneiy
             1cdg       apdtsvsnkqnFSTDVIYQIFTDRFsdgnpannptgaafdgtCtn-lrly

             >> example of xx

             1cdg         APDTSVSNKQ NFSTDVIYQI FTDRFSDGNP ANNPTGAAFD GTCTN-LRLY
             2aaa         ------LSAA SWRTQSIYFL LTDRFGRTDN S-------TT ATCNTGNEIY
             1cdg_2aaa    ------7774 2221210000 0000000148 9-------99 41114-4000
             1cdg_6taa    ------8674 2232220000 0000011059 9-------99 52114-3000

<b>Usage     </b> : ($rate1_ref,$rate2_ref) =${&get_posi_shift_rms_hash(\%msf_hash, \%jp_hash,
                                                                 \%rms_file_hash, \$threshold)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="open_rms_files
.pl.html">open_rms_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_rms_files
.pl">open_rms_files
.pl</A>
<pre>
<b>Argument  </b> : takes one ref. for a file.
<b>Function  </b> : open rms files and put sequences in a hash
             Example of rms (aa
             1cdg         APDTSVSNKQ NFSTDVIYQI FTDRFSDGNP ANNPTGAAFD GTCTN-LRLY
             2aaa         ------LSAA SWRTQSIYFL LTDRFGRTDN S-------TT ATCNTGNEIY
             6taa         ------ATPA DWRSQSIYFL LTDRFARTDG S-------TT ATCNTADQKY
             1cdg_2aaa    ------7774 2221210000 0000000148 9-------99 41114-4000
             1cdg_6taa    ------8674 2232220000 0000011059 9-------99 52114-3000
             2aaa_6taa    ------1000 0000000000 0000000010 0-------00 0000000000

             Example output hash;
             1nor        LECHNQQSSQPPTTKTCS-GETNCYKKWWSDH----RGTIIERGFFC--GCPKVK-PGVNLNCCRT-DRCNN-------
             1cdg        APDTSVSNKQNFSTDVASISGLVTSLP-QGSYNDVLGGLLNGNTLSVGSGGAASNFTLAAGGTAVWQYTAATATPTIGH
             1cdg_2aaa   ------777002112111-----343333---431127----5433234-72354541131211111176899999999

<b>Returns   </b> : a ref. of a hash
<b>Usage     </b> : %anyarray = {&open_rms_files(\$any_sequence_file_msf_form)};
<b>Version   </b> : 1.0
<b>Warning   </b> : xxx.rms files are Tim Hubbard's 'msarms' program's output.
</pre><hr>

<H3><A href="open_rms_files2
.pl.html">open_rms_files2
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_rms_files2
.pl">open_rms_files2
.pl</A>
<pre>
<b>Argument  </b> : takes one ref. for a file.
<b>Function  </b> : same as open rms files but returns two hashes.
             Example of rms (aa
             1cdg         APDTSVSNKQ NFSTDVIYQI FTDRFSDGNP ANNPTGAAFD GTCTN-LRLY
             2aaa         ------LSAA SWRTQSIYFL LTDRFGRTDN S-------TT ATCNTGNEIY
             1cdg_2aaa    ------7774 2221210000 0000000148 9-------99 41114-4000
             1cdg_6taa    ------8674 2232220000 0000011059 9-------99 52114-3000

             Example output 2 hashes;
             1nor        LECHNQQSSQPPTTKTCS-GETNCYKKWWSDH----RGTIIERGFFC--GCPKVK-PGVNLNCCRT-DRCNN-------
             1cdg        APDTSVSNKQNFSTDVASISGLVTSLP-QGSYNDVLGGLLNGNTLSVGSGGAASNFTLAAGGTAVWQYTAATATPTIGH

             1cdg_2aaa   ------777002112111-----343333---431127----5433234-72354541131211111176899999999
             1cdg_2taa   ------777002112111-----343333---431127----5433234-72354541131211111176899999999

<b>Returns   </b> : return(@out); while @out is (\%hash_rms, \%hash_jp)
<b>Usage     </b> : ($hash_for_jp, $hash_for_rms) = &open_rms_files(\$any_sequence_file_msf_form);
<b>Version   </b> : 1.0
<b>Warning   </b> : xxx.rms files are Tim Hubbard's 'msarms' program's output.
</pre><hr>

<H3><A href="steve_permute_array
.pl.html">steve_permute_array
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//steve_permute_array
.pl">steve_permute_array
.pl</A>
<pre>
<b>Argument  </b> : upto 3 arg. 1st one is for the ref. of an array. 2nd for min
             element no. 3rd for max element no. 2nd and 3rd are optional.
<b>Returns   </b> : a ref. of a hash.
<b>Usage     </b> : %final_out_hash=%{&steve_permute_array(\@list, \2, \4)};
                         Above is for pairs, 3 seqs, and 4 seqs.
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="opendir_and_go_in_and_do_something
.pl.html">opendir_and_go_in_and_do_something
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//opendir_and_go_in_and_do_something
.pl">opendir_and_go_in_and_do_something
.pl</A>
<pre>
<b>Argument  </b> : gets a ref. of a scaler (dir name) and returns nothing(void).
<b>Example   </b> : as in my 'indexing.pl' for perl file indexer.
<b>Function  </b> : open dir and process all files in the dir if you wish,
             and then go in any other sub
             if any file(dir) is linked, it skips that file.
<b>Keywords  </b> : open_dir_and_go_in_and_do_something,
             go in there do something, get into subdir and do something.
             go_in_subdir_and_do_something, recursive execution
<b>Usage     </b> : &opendir_and_go_in_and_do_something(\$input_dir);
                     $inputdir='/nfs/ind4/ccpe1/people/A Biomatic /jpo/align';
<b>Version   </b> : 1.1
<b>Warning   </b> : Seems to work fine., !! Change the name of this sub to shorter one
                                  !! for your own purpose.
</pre><hr>

<H3><A href="open_subdir_and_go_in_and_do
.pl.html">open_subdir_and_go_in_and_do
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_subdir_and_go_in_and_do
.pl">open_subdir_and_go_in_and_do
.pl</A>
<pre>
<b>Argument  </b> : gets a ref. of a scaler (dir name) and returns nothing(void).
<b>Example   </b> : as in my 'indexing.pl' for perl file indexer.
<b>Function  </b> : open dir and process all files in the dir if you wish,
             and then go in any other sub
             if any file(dir) is linked, it skips that file.
<b>Usage     </b> : &opendir_and_go_in_and_do_something(\$input_dir);
                     $inputdir='/nfs/ind4/ccpe1/people/A Biomatic /jpo/align';
<b>Version   </b> : 1.0
<b>Warning   </b> : Seems to work fine., !! Change the name of this sub to shorter one
                                  !! for your own purpose.
</pre><hr>

<H3><A href="get_occurances_of_shift_type_hash
.pl.html">get_occurances_of_shift_type_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_occurances_of_shift_type_hash
.pl">get_occurances_of_shift_type_hash
.pl</A>
<pre>
<b>Argument  </b> : Two references of hashes.
<b>Returns   </b> : one reference  of hash. (eg, 0=>1000, 1=>888, 2=>83, ...
                                          0,1,2... are position shift types
                                          1000, 888, 83... are occurances in
                                          the comparision between str. and seq.
                                          alignments.)
<b>Usage     </b> : for single protein group
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_occurances_of_shift_type_hash_all
.pl.html">get_occurances_of_shift_type_hash_all
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_occurances_of_shift_type_hash_all
.pl">get_occurances_of_shift_type_hash_all
.pl</A>
<pre>
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_occurances_of_char
.pl.html">get_occurances_of_char
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_occurances_of_char
.pl">get_occurances_of_char
.pl</A>
<pre>
<b>Argument  </b> : one ref. of hash (seq1 alsdfjlsj
                               seq2 asldfjsld
                               seq3 owiurouou);
<b>Function  </b> : gets the numbers of occurances for 1, 2, 3 ... position shifts.
             If hash is given, it only looks at the values.
             If multiple string, array, hash or combinations of these
              are given, it will add up to one single result
<b>Keywords  </b> : composition of chars, composition table making,
             make_composition, make composition table
             occurances_of_char, get_char_occurances, occurances
             get_percentage_occurances_of_char, percentage_occurances_of_char
<b>Options   </b> : 'p' for percentage output of the char among others
             'n' for NO name option when HASH input is given
<b>Returns   </b> : one ref. of hash  (a =>5, b=>6, c=>4,,,,,)
<b>Usage     </b> : %occurances_shft_type=%{&get_occurances_of_char(\%final_posi_diffs)};
             %char_occur=%{&get_occurances_of_char(\@ref_array_of_chars)};
             %char_occur=%{&get_occurances_of_char(\$ref_string_of_chars)};
             %char_occur=%{&get_occurances_of_char($string_of_chars)};

<b>Version   </b> : 1.3
</pre><hr>

<H3><A href="make_composition_table
.pl.html">make_composition_table
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//make_composition_table
.pl">make_composition_table
.pl</A>
<pre>
<b>Argument  </b> : one ref. of hash (seq1 alsdfjlsj
                               seq2 asldfjsld
                               seq3 owiurouou);
<b>Function  </b> : gets the numbers of occurances for 1, 2, 3 ... position shifts.
<b>Keywords  </b> : composition of chars, composition table making, make composition table
             make_composition_table, get_composition, get_amino_acid_composition
             protein_composition, make_aa_composition_tablem, aa_composition
<b>Returns   </b> : one ref. of hash  (a =>5, b=>6, c=>4,,,,,)
<b>Usage     </b> : %occurances=%{&make_compos_table(\%key_and_value_for_seq)};
<b>Version   </b> : 1.2
</pre><hr>

<H3><A href="make_composition_ratio_table_simple
.pl.html">make_composition_ratio_table_simple
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//make_composition_ratio_table_simple
.pl">make_composition_ratio_table_simple
.pl</A>
<pre>
<b>Argument  </b> : one ref. of hash (seq1 alsdfjlsj
                               seq2 asldfjsld
                               seq3 owiurouou);
<b>Function  </b> : gets ratio of the numbers of occurances for any chars.
<b>Keywords  </b> : composition table, composition of chars, composition table making,
             make composition table, make_composition_table
<b>Returns   </b> : one ref. of hash  (a =>0.05, b=>0.06, c=>0.04,,,,,)
<b>Usage     </b> : %occurances=%{&make_compos_ratio_table(\%final_posi_diffs)};
<b>Version   </b> : 1.0
<b>Warning   </b> : This pools all the sequences, to not distinct seq composition if
              you put more than one seq.
</pre><hr>

<H3><A href="make_composition_ratio_table
.pl.html">make_composition_ratio_table
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//make_composition_ratio_table
.pl">make_composition_ratio_table
.pl</A>
<pre>
<b>Argument  </b> : one or more ref. of hash (seq1 alsdfjlsj
                                       seq2 asldfjsld
                                       seq3 owiurouou);
<b>Function  </b> : gets ratio of the numbers of occurances for any chars.
<b>Keywords  </b> : composition table, composition of chars, composition table making,
             make composition table, make_composition_table
             aa_composition_ratio, composition_ratio, protein_composition,
             get_composition_ratio, get_aa_composition_ratio
<b>Returns   </b> : one ref. of hash  ('seq_name', { a =>0.05, b=>0.06, c=>0.04,,,,, } )
<b>Usage     </b> : %rate=%{&make_compos_ratio_table(\%hash1, \%hash2, ,,,)};
<b>Version   </b> : 1.3
<b>Warning   </b> : This produces each composition ration table for each seq
</pre><hr>

<H3><A href="get_position_shift_rate
.pl.html">get_position_shift_rate
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_position_shift_rate
.pl">get_position_shift_rate
.pl</A>
<pre>
<b>Argument  </b> : %{&get_position_shift_rate(\%msfo_file, \%jpo_file)};
             Whatever the names, it takes one TRUE structral and one ALIGNED hash.
<b>Example   </b> : my(%error_rate)=%{&get_position_shift_rate(\%input, \%input2)};
<b>Function  </b> : This is to get position specific error rate for line display rather than
             actual final error rate for the alignment. Takes two file names of seq.
             Output >>
             seq1_seq2  1110...222...2222
             seq2_seq3  1111....10...1111
             seq1_seq3  1111....0000.0000

<b>Options   </b> : 'ss' for secondary structure regions(Helix and Beta region only
                 calculation for error rate). There is specialized sub called
              get_segment_shift_rate for sec. str. only handling.

    $ss_opt            becomes    ss by  ss, SS, -ss, -SS     #  for secondary structure only
    $H                 =         'H' by   -H or -h or H       # to retrieve only H segment
    $S                 becomes   'S' by   -S or  S            # to retrieve only S segment
    $E                 becomes   'E' by   -E or  E            # to retrieve only E segment
    $T                 becomes   'T' by   -T or -t or T or t  # to retrieve only T segment
    $I                 becomes   'I' by   -I or  I            # to retrieve only I segment
    $G                 becomes   'G' by   -G or -g or G or g  # to retrieve only G segment
    $B                 becomes   'B' by   -B or -b or B or b  # to retrieve only B segment
    $HELP              becomes    1  by   -help   # for showing help
    $simplify          becomes    1  by   -p or P or -P, p
    $simplify          becomes    1  by   -simplify or simplify, Simplify SIMPLIFY
    $comm_col          becomes   'C' by   -C or C or common
    $LIMIT             becomes    L  by   -L, L               # to limit the error rate to 9 .

<b>Returns   </b> : \%final_posi_diffs;
<b>Usage     </b> : %rate_hash = %{&get_position_shift_rate(\%hash_msf, \%hash_jp)};
<b>Version   </b> : 1.5
<b>Warning   </b> : split and join char is ','; (space)
</pre><hr>

<H3><A href="get_posi_rates_hash_out
.pl.html">get_posi_rates_hash_out
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_posi_rates_hash_out
.pl">get_posi_rates_hash_out
.pl</A>
<pre>
<b>Argument  </b> : %{&get_posi_rates_hash_out(\%msfo_file, \%jpo_file)};
             Whatever the names, it takes one TRUE structral and one ALIGNED hash.
<b>Function  </b> : This is to get position specific error rate for line display rather than
             actual final error rate for the alignment.
             Output >>
             seq1_seq2  1110...222...2222
             seq2_seq3  1111....10...1111
             seq1_seq3  1111....0000.0000

<b>Returns   </b> : \%final_posi_diffs;
<b>Usage     </b> : %rate_hash = %{&get_posi_shift_hash(\%hash_msf, \%hash_jp)};
<b>Version   </b> : 1.0
<b>Warning   </b> : split and join char is ','; (space)
</pre><hr>

<H3><A href="get_posi_diff_hash
.pl.html">get_posi_diff_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_posi_diff_hash
.pl">get_posi_diff_hash
.pl</A>
<pre>
<b>Argument  </b> : Takes a ref. for hash which have positions of residues of sequences.
<b>Function  </b> : gets a ref. of a hash and calculates the position diffs.
<b>Returns   </b> : one ref. for an array of differences of input arrays. array context.
             ---Example input (a hash with sequences); The values are differences after comparion
                                            with structural and sequential alignments.
             %diffs =('seq1', '112342431111
             'seq2', '12222...09011.1122',
             'seq3', '13222...00011.1122',
             'seq4', '12262...00011.112.');
             example output;
             seq3_seq4       01040...00000.000.
             seq1_seq2       01012...1810...122
             seq1_seq3       02012...1110...122
             seq1_seq4       01032...1110...12.
             seq2_seq3       01000...09000.0000
             seq2_seq4       00040...09000.000.
<b>Usage     </b> : %position_diffs =\{&get_posi_diff_hash(\@seq_position1, \@seq_position2)};
<b>Version   </b> : 1.0
<b>Warning   </b> : split and join char is ',';    # used in 'get_posi_shift_hash'
</pre><hr>

<H3><A href="get_posi_shift_hash
.pl.html">get_posi_shift_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_posi_shift_hash
.pl">get_posi_shift_hash
.pl</A>
<pre>
<b>Argument  </b> : takes two hash REFERENCES for (one seq. and one struc. alignment(2nd arg)
<b>Returns   </b> : One scalar value of shift rate of position for proteins.
<b>Usage     </b> : $rate_final = ${&get_posi_shift_hash(\%hash_msf, \%hash_jp)};
<b>Version   </b> : 1.1
<b>Warning   </b> : split and join char is ','; (space)
</pre><hr>

<H3><A href="print_seq_in_block_with_print
.pl.html">print_seq_in_block_with_print
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//print_seq_in_block_with_print
.pl">print_seq_in_block_with_print
.pl</A>
<pre>
<b>Function  </b> : gets a ref(s) for hash and prints the content in lines of 60 char
<b>Returns   </b> : Nothing, STDOUT
<b>Usage     </b> : &print_seq_in_block (\%input_hash1,\%input_hash2, \%input_hash3.... );
<b>Version   </b> : 1.0
<b>Warning   </b> : derived from  print_in_block
</pre><hr>

<H3><A href="fill_ending_space
.pl.html">fill_ending_space
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//fill_ending_space
.pl">fill_ending_space
.pl</A>
<pre>
<b>Argument  </b> : (\%input1, \%input2, \%input3.....);
<b>Function  </b> : fills the ending gaps or space of sequences (shorter ones)
<b>Returns   </b> : (\%hash1,..... )
<b>Usage     </b> : (*out, *out2, *out3)=&fill_ending_space(\%input1, \%input2, \%input3);
             &print_seq_in_block(\%out,\%out2,\%out3); <-- if you want printout.
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="print_seq_in_block_old
.pl.html">print_seq_in_block_old
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//print_seq_in_block_old
.pl">print_seq_in_block_old
.pl</A>
<pre>
<b>Argument  </b> : one or more refs. for hash
               if there are more than one array input it makes such outputs

             Name1    THIS.IS.from.hash.one
             Name2    This

             Name1    THIS
             Name2    This.is.from.hash.two

<b>Function  </b> : gets a ref(s) for hash (single key and value)
             and prints the content in lines of 60 char
<b>Returns   </b> : Nothing, STDOUT
<b>Usage     </b> : &print_seq_in_block_old (\%input_hash1,\%input_hash2, \%input_hash3.... );
<b>Version   </b> : 1.0
<b>Warning   </b> : This is more or less for debugging. Use  print_seq_in_block
</pre><hr>

<H3><A href="print_in_block
.pl.html">print_in_block
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//print_in_block
.pl">print_in_block
.pl</A>
<pre>
<b>Argument  </b> : one or more refs. for array
               if there are more than one array input it makes such outputs
             Example out)
               THIS.IS.from.array.one
             This.is.from.array.two

              THIS.IS.from.array.one
               This.is.from.array.two

<b>Function  </b> : gets a ref(s) for array and prints the content in lines of 60 char
<b>Returns   </b> : Nothing, STDOUT
<b>Usage     </b> : &print_in_block (\@input_array,\@input_array2, \@input_array3.... );
<b>Version   </b> : 1.0
<b>Warning   </b> : This is more or less for debugging. Use  print_seq_in_block
</pre><hr>

<H3><A href="get_posi_diff
.pl.html">get_posi_diff
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_posi_diff
.pl">get_posi_diff
.pl</A>
<pre>
<b>Argument  </b> : Takes two ref. for arrays which have positions of residues.
<b>Example   </b> : @compacted_posi_dif =(1 ,2, 1, 1, '.' ,2,  1,  1, '.');
             @compacted_posi_dif2=(4 ,2, 1, 1, ,2,  1, '.' ,3,  1);
             output ==> ( 3 0 0 0 . 1 . 2 .)   (it ignores positions which have non digits.
             output ==> (-3 0 0 0 . 1 .-2 .) when abs is not used.
<b>Returns   </b> : one ref. for an @array of differences of input arrays. array context.
<b>Usage     </b> : @position_diffs =&get_posi_diff(\@seq_position1,\@seq_position2);
<b>Version   </b> : 1.4
</pre><hr>

<H3><A href="get_posi_diff_abs
.pl.html">get_posi_diff_abs
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_posi_diff_abs
.pl">get_posi_diff_abs
.pl</A>
<pre>
<b>Argument  </b> : Takes two ref. for arrays which have positions of residues.
<b>Example   </b> : @compacted_posi_dif =(1 ,2, 1, 1, '.' ,2,  1,  1, '.');
             @compacted_posi_dif2=(4 ,2, 1, 1, ,2,  1, '.' ,3,  1);
             output ==> ( 3 0 0 0 . 1 . 2 .)   (it ignores positions which have non digits.
             output ==> (-3 0 0 0 . 1 .-2 .) when abs is not used.
<b>Returns   </b> : one ref. for an @array of differences of input arrays. array context.
<b>Usage     </b> : @position_diffs =&get_posi_diff_abs(\@seq_position1,\@seq_position2);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="put_position_back_to_str_seq
.pl.html">put_position_back_to_str_seq
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//put_position_back_to_str_seq
.pl">put_position_back_to_str_seq
.pl</A>
<pre>
<b>Argument  </b> : takes two refs for arrays (one for char the other for digits
<b>Example   </b> : @string_from_struct=('X', 'T', 'A' ,'B' , '.' ,'F',  'G', '.' , 'O' ,'P', '.');
             @compacted_posi_dif=(1 ,2, 1, 1, ,2, 1, 1, 1);
<b>Returns   </b> : a ref. for an array
<b>Usage     </b> : @result =@{&put_position_back_to_str_seq(\@string_from_struct, \@compacted_posi_dif)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_posi_shift_hash_rms
.pl.html">get_posi_shift_hash_rms
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_posi_shift_hash_rms
.pl">get_posi_shift_hash_rms
.pl</A>
<pre>
<b>Function  </b> : caculates the error rate of seq after filtering according to
                rms deviation.
<b>Usage     </b> : $result=${&get_posi_shift_hash_rm(\%h1, \%h2, \%h3)};
<b>Version   </b> : 1.0
<b>Warning   </b> : Not complete yet.
</pre><hr>

<H3><A href="open_fil_file
.pl.html">open_fil_file
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_fil_file
.pl">open_fil_file
.pl</A>
<pre>
<b>Function  </b> : reads xxx.fil file which shows whether I have to discard
             regions of sequences due to too big RMS deviation.
<b>Returns   </b> : a ref. for a hash(associative array).
<b>Usage     </b> : %out = %{&open_fil_file(\$input_seq_file)};
<b>Version   </b> : 1.0
<b>Warning   </b> : !!! not yet complete !!!
</pre><hr>

<H3><A href="send_mail
.pl.html">send_mail
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//send_mail
.pl">send_mail
.pl</A>
<pre>
<b>Example   </b> : 
             send_mail ( $to, $subject, @lines );
             #-# i -- $to      = email address
             #-# i -- $subject = string to be put in the Subject: line
             #-# i -- @lines   = lines to be mailed - must not have \n
             -- DISCUSSION:

             Uses /usr/lib/sendmail to mail a bunch of lines to the email address
             specified. The @lines should not have terminating \n characters: they
             will be supplied.

             -- EXAMPLE:
             &P10::mail ( 'schip@lmsc.lockheed.com', 'Test 34', @mylines );
             -- END
             : Could some one share their knowledge of how to mail a message from
             :  within a Perl script with a novice Perl user?
<b>Function  </b> : mail a bunch of @lines to a user
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="rand_word
.pl.html">rand_word
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//rand_word
.pl">rand_word
.pl</A>
<pre>
<b>Function  </b> : This sub routine should return an alphabet string of
             length specified by  an argument.
<b>Keywords  </b> : randomize words, makes random words, scramble_word,
              shuffle_words,
<b>Usage     </b> : $word = ${&rand_word(7)};
             print "sub rand_word gives $word\n";
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="opendir_and_go_rand_fasta_and_clustal
.pl.html">opendir_and_go_rand_fasta_and_clustal
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//opendir_and_go_rand_fasta_and_clustal
.pl">opendir_and_go_rand_fasta_and_clustal
.pl</A>
<pre>
<b>Example   </b> : $inputdir='/nfs/ind4/ccpe1/people/A Biomatic /jpo/align';
             &opendir_and_go($inputdir);
<b>Function  </b> : open dir and process all files if you wish, and then go in any sub
             dir of it. Using recursion. created by A Biomatic
             if any file is linked, it skips that file.
<b>Usage     </b> : &opendir_and_go_rand_fasta_and_clustal(\$input_dir); #$inputdir='/nfs/ind4/ccpe1/people/A Biomatic /jpo/align';
<b>Version   </b> : 1.0
<b>Warning   </b> : Seems to work fine.
</pre><hr>

<H3><A href="opendir_and_go_rand_fasta
.pl.html">opendir_and_go_rand_fasta
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//opendir_and_go_rand_fasta
.pl">opendir_and_go_rand_fasta
.pl</A>
<pre>
<b>Example   </b> : $inputdir='/nfs/ind4/ccpe1/people/A Biomatic /jpo/align';
             &opendir_and_go($inputdir);
<b>Function  </b> : open dir and process all files if you wish, and then go in any sub
             dir of it. Using recursion. created by A Biomatic
             if any file is linked, it skips that file.
<b>Usage     </b> : &opendir_and_go_rand_fasta(\$input_dir); #$inputdir='/nfs/ind4/ccpe1/people/A Biomatic /jpo/align';
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="reverse_sequences
.pl.html">reverse_sequences
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//reverse_sequences
.pl">reverse_sequences
.pl</A>
<pre>
<b>Argument  </b> : hash, eg(1, 'skdfj', 2, 'kdfjkdj', 3, 'kdfjk');
             Input example:
             ..
             >HI0256
             FLSANVLPIAPIINGGRTAVDNITQSVSDKPFVKDIGTKIKEAIALSKYSTQPQYISTTN
             >HI0094
             DILRTFVKMETGLKFPKKFKLKANLALFMNRRNKRPDTIMTAVADAGQKISEAKLNTTAK
             ..

             Output example: (Reversed :-)
             ..
             >HI0256_rv   <<-- note the added extension
             ALDJFLKAJFJALSDJFLAJSLFJAKLSDFJLASJDFLAJSLDFJASJDFLJSDFJSDLJ
             >HI0094_rv
             LASJDFLKAJFJALSDJFLKSDJLFAJLKDJFLASJDFLKDFJKDJFKDJFKDJFKJDLJ
             ..

<b>Function  </b> : gets ref. of strings, reverses the elems.
<b>Keywords  </b> : reverse_sequence, reverse_sequence_hash, rev_sequence_hash
<b>Returns   </b> : one or more hash references.
<b>Usage     </b> : %out = %{&rev_sequence_one_hash(\%input_seq_hash, \%hash2,...)};
<b>Version   </b> : 1.4
</pre><hr>

<H3><A href="rev_sequence_mul_array
.pl.html">rev_sequence_mul_array
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//rev_sequence_mul_array
.pl">rev_sequence_mul_array
.pl</A>
<pre>
<b>Function  </b> : gets a ref. of an string, reverses the elems.
<b>Returns   </b> : one ref. of  mul_array, eg. ('jfkdj', 'kdfjsdj', 'jjjkk')
<b>Usage     </b> : @out = @{&rev_sequence_mul_array(\@input_mul_seq_array)};
<b>Version   </b> : 1.0
<b>Warning   </b> : This reverses sequences!
</pre><hr>

<H3><A href="scramble_sequences
.pl.html">scramble_sequences
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//scramble_sequences
.pl">scramble_sequences
.pl</A>
<pre>
<b>Argument  </b> : ref. of hash, eg(1, 'skdfj', 2, 'kdfjkdj', 3, 'kdfjk');

             Input example:
             ..
             >HI0256
             FLSANVLPIAPIINGGRTAVDNITQSVSDKPFVKDIGTKIKEAIALSKYSTQPQYISTTN
             >HI0094
             DILRTFVKMETGLKFPKKFKLKANLALFMNRRNKRPDTIMTAVADAGQKISEAKLNTTAK
             ..

             Output example: (scrambled :-)
             ..
             >HI0256_sc   <<-- note the added extension
             ALDJFLKAJFJALSDJFLAJSLFJAKLSDFJLASJDFLAJSLDFJASJDFLJSDFJSDLJ
             >HI0094_sc
             LASJDFLKAJFJALSDJFLKSDJLFAJLKDJFLASJDFLKDFJKDJFKDJFKDJFKJDLJ
             ..
<b>Function  </b> : gets ref. of strings, scambles the elems.
<b>Keywords  </b> : make_scrambled_seq, make_scrambled_sequence, scramble_seq_hash,
              scramble_sequences, shuffle_sequences, shuffle_seq
<b>Returns   </b> : (\%hashout), or (\%hash1, \%hash2,,,,,)
<b>Usage     </b> : %out = %{&scramble_seq_hash(\%input_seq_hash)};
<b>Version   </b> : 1.5
</pre><hr>

<H3><A href="scramble_array
.pl.html">scramble_array
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//scramble_array
.pl">scramble_array
.pl</A>
<pre>
<b>Function  </b> : shuffles the elements of array
<b>Keywords  </b> : randomise_array, randomize_array, shuffle_array
<b>Usage     </b> : @in=@{&scramble_array(\@in)};
<b>Version   </b> : 1.4
</pre><hr>

<H3><A href="rand_sequence_mul_array
.pl.html">rand_sequence_mul_array
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//rand_sequence_mul_array
.pl">rand_sequence_mul_array
.pl</A>
<pre>
<b>Argument  </b> : one ref. of mul_array, eg. ('lsjdfj', 'kdfjsdj', 'jjjkk')
<b>Function  </b> : gets a ref. of an string, scambles the elem.
<b>Keywords  </b> : scramble_sequence_mul_array, shuffle_sequence_mul_array
<b>Returns   </b> : one ref. of  mul_array, eg. ('jfkdj', 'kdfjsdj', 'jjjkk')
<b>Usage     </b> : @out = @{&rand_sequence_mul_array(\@input_mul_seq_array)};
<b>Version   </b> : 1.1
<b>Warning   </b> : This scrambles sequences!!
</pre><hr>

<H3><A href="rand_sequence_one_string
.pl.html">rand_sequence_one_string
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//rand_sequence_one_string
.pl">rand_sequence_one_string
.pl</A>
<pre>
<b>Argument  </b> : one ref. of string, eg ( 'ldkfjlsdjfsdjflj' )
<b>Function  </b> : gets a ref. of a  string, scambles the elem.
<b>Returns   </b> : one ref. of string,
<b>Usage     </b> : @out = @{&rand_sequence_one_string(\$input_seq_string)};
<b>Version   </b> : 1.0
<b>Warning   </b> : This scrambles sequences!!
</pre><hr>

<H3><A href="rand_sequence_one_array
.pl.html">rand_sequence_one_array
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//rand_sequence_one_array
.pl">rand_sequence_one_array
.pl</A>
<pre>
<b>Argument  </b> : one ref. of array, eg ('e', 'b', 'c', 'd')
<b>Function  </b> : gets a ref. of an array, scambles the elem.
<b>Returns   </b> : one ref. of array,
<b>Usage     </b> : @out = @{&rand_sequence_one_array(\@input_seq_array)};
<b>Version   </b> : 1.0
<b>Warning   </b> : This scrambles sequences!!
</pre><hr>

<H3><A href="make_random_sequence
.pl.html">make_random_sequence
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//make_random_sequence
.pl">make_random_sequence
.pl</A>
<pre>
<b>Argument  </b> : 1 200 [-p] [@array_of_array_refs]
             1 = num of seq, 200=leng of seq, -p =option, @arr.. = option
             You can optionally give amino acid matrices
<b>Example   </b> : $out=${&make_random_sequence(@ARGV)};  While @ARGV can be '1 200 -p'
<b>Function  </b> : gets one or more numbers for seq length and makes random sequences
             It can handle proportional random sequenes according to the
             amino acid occurance matrix.
<b>Keywords  </b> : scramble_sequence, make_scrambled_sequence, shuffle_sequence
             random_sequence, make_random_sequence, generate_random_protein_seq
             create_random_sequene create_random_aa_sequence
<b>Options   </b> : 'p' for proportional random sequence option
             'f' for fastsa format output (returns one ref. of HASH)
<b>Returns   </b> : one or more scalar references according to the input numbers.
<b>Usage     </b> : $protein = ${&make_random_sequence(1, 400)};
<b>Version   </b> : 1.4
</pre><hr>

<H3><A href="rand_DNA_seq_generate
.pl.html">rand_DNA_seq_generate
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//rand_DNA_seq_generate
.pl">rand_DNA_seq_generate
.pl</A>
<pre>
<b>Argument  </b> : (343) or (\$length)
<b>Function  </b> : gets one or more numbers for seq length and makes random sequences
<b>Returns   </b> : one or more scalar references according to the input numbers.
<b>Usage     </b> : $DNA = ${&rand_DNA_seq_generate(400)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="rand_RNA_seq_generate
.pl.html">rand_RNA_seq_generate
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//rand_RNA_seq_generate
.pl">rand_RNA_seq_generate
.pl</A>
<pre>
<b>Argument  </b> : (343) or (\$length)
<b>Function  </b> : gets one or more numbers for seq length and makes random sequences
<b>Returns   </b> : one or more scalar references according to the input numbers.
<b>Usage     </b> : $DNA = ${&rand_RNA_seq_generate(400)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="replace_text
.pl.html">replace_text
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//replace_text
.pl">replace_text
.pl</A>
<pre>
<b>Argument  </b> : reference of one array of file names in pwd
<b>Function  </b> : finds patterns of text and replaces them in multiple input files
<b>Keywords  </b> : replace_txt, change_text,
<b>Returns   </b> : nothing
<b>Usage     </b> : &replace_text(\@input_array_of_filenames);
<b>Version   </b> : 1.4
<b>Warning   </b> : This produces a temporary file and rename it...
</pre><hr>

<H3><A href="get_av_seq_length
.pl.html">get_av_seq_length
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_av_seq_length
.pl">get_av_seq_length
.pl</A>
<pre>
<b>Argument  </b> : one hash reference for sequences.
<b>Function  </b> : gets hash of sequence, compares lengths, and outs av.
<b>Returns   </b> : one ref. for scaler digit.
<b>Usage     </b> : $std_devi_of_lengths = &get_av_seq_length(\%hash_ref);
<b>Version   </b> : 1.0
<b>Warning   </b> : uses a sub  &array_average(\@lengths);
</pre><hr>

<H3><A href="get_sd_of_length_diff
.pl.html">get_sd_of_length_diff
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_sd_of_length_diff
.pl">get_sd_of_length_diff
.pl</A>
<pre>
<b>Argument  </b> : gets one hash reference,
<b>Returns   </b> : one scaler digit
<b>Usage     </b> : $result = &get_sd_of_length_diff(\%input);
<b>Version   </b> : 1.0
<b>Warning   </b> : removes all non-char(.-, space....) in the input string
</pre><hr>

<H3><A href="get_av_and_sd_seq_length
.pl.html">get_av_and_sd_seq_length
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_av_and_sd_seq_length
.pl">get_av_and_sd_seq_length
.pl</A>
<pre>
<b>Argument  </b> : Two hash references for sequences.
<b>Function  </b> : gets ref of hash of sequence, compares lengths, and outs av.
<b>Returns   </b> : Two scaler digit.
<b>Usage     </b> : $get_av_and_sd_seq_length= &get_av_seq_length(\%hash_ref);
<b>Version   </b> : 1.0
<b>Warning   </b> : uses a sub  &array_average(\@lengths);
</pre><hr>

<H3><A href="get_seq_hash_sans_gaps.pl.html">get_seq_hash_sans_gaps</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_seq_hash_sans_gaps.pl">get_seq_hash_sans_gaps.pl</A>
<pre>
<b>Usage     </b> : ($ref_out1, $ref_out2)=&get_seq_hash_sans_gaps(\%hash, \%hash);
              %out=%{&get_seq_hash_sans_gaps(\%hash)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_posi_sans_gaps
.pl.html">get_posi_sans_gaps
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_posi_sans_gaps
.pl">get_posi_sans_gaps
.pl</A>
<pre>
<b>Argument  </b> : one scalar variable input of sequence string.
<b>Returns   </b> : the positions of residues after removing gaps(but keeps pos).
               used for analysis of shifted positions of seq. comparison.
<b>Usage     </b> : @seq_position1 = &get_posi_sans_gaps($string1);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_posi_shift_rate
.pl.html">get_posi_shift_rate
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_posi_shift_rate
.pl">get_posi_shift_rate
.pl</A>
<pre>
<b>Argument  </b> : takes two file names for seq. and struc. alignment.
             : Assumes the files are in the pwd.
<b>Returns   </b> : one ref. for scalar value of shift rate of position for proteins.
<b>Usage     </b> : $rate_final = &get_posi_shift_rate("perl.msf", "perl.jp");
<b>Version   </b> : 1.0
<b>Warning   </b> : sub  hash_common was unstable.
</pre><hr>

<H3><A href="read_hssp_no_inserts
.pl.html">read_hssp_no_inserts
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//read_hssp_no_inserts
.pl">read_hssp_no_inserts
.pl</A>
<pre>
<b>Function  </b> : read hssp file and put sequences in a hash
<b>Usage     </b> : %anyarray = &read_hssp_no_inserts ($any_sequence_file_hssp_form);
<b>Version   </b> : 1.0
<b>Warning   </b> : It produces incomplete sequences when hssp seqs. have insertions.
</pre><hr>

<H3><A href="open_pdbg_files
.pl.html">open_pdbg_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_pdbg_files
.pl">open_pdbg_files
.pl</A>
<pre>
<b>Example   </b> : %out = %{&open_pdbg_files(@ARGV)};
             while @ARGV at prompt was: 'pdb_40.pdbg'
<b>Function  </b> : open pdb group files and put scopclass in a hash.
             PDB group file format is like this;

  >d1bia_1 1.4.3.1.1 (1-63) Biotin repressor, N-terminal domain [Escherichia coli]
  >d1baba_ 1.1.1.1.15 Hemoglobin, alpha-chain [human (Homo sapiens)]
  >d1cpcb_ 1.1.1.2.1 C-phycocyanin [cyanobacterium (Fremyella diplosiphon)]
  >d1fcdc2 1.3.1.3.1 (81-174) Flavocytochrome c sulfide dehydrogenase, FCSD, cytochrome subunit [Purple phototrophic bacterium (Cromatium vinosum)]

             This can also return the sizes of sequences rather than seqs.
<b>Keywords  </b> : open_pdbg_files, open_pdb_group_files
<b>Options   </b> : any digit for the minimum seq length
        b  for simple style reading (this reads in the name of pdbg file as it is)

<b>Usage     </b> : %seq=%{&open_pdbg_files($tim_seq_file, ['1fcdc1'], [s] )};
             if you put additional seq name as 1fcdc1 it will
             fetch that scopclass only in the database file.
             Any digit will be used as minimum seq size to be fetched.
<b>Version   </b> : 1.5
</pre><hr>

<H3><A href="open_cel_files
.pl.html">open_cel_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_cel_files
.pl">open_cel_files
.pl</A>
<pre>
<b>Example   </b> : 
  Example INPUT file:
  #  Tabl delimited data file
   X       Y     MeanIn  STDEV  Pixel num

   0       0       200.0   15      16      1.24343
   1       0       200.0   15      16      1.24343
   2       0       200.0   15      16      1.24343
   3       0       200.0   15      16      1.24343

<b>Returns   </b> : 3D array =( [X][Y][0], [X][Y][1], [X][Y][2], [X][Y][3])
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="open_stride_dat_files
.pl.html">open_stride_dat_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_stride_dat_files
.pl">open_stride_dat_files
.pl</A>
<pre>
<b>Author    </b> : jong@salt2.med.harvard.edu
<b>Usage     </b> : @out=@{&open_stride_dat_files(@ARGV)};
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="open_fasta_files
.pl.html">open_fasta_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_fasta_files
.pl">open_fasta_files
.pl</A>
<pre>
<b>Example   </b> : %out = %{&open_fasta_files(@ARGV)};
             %out2=%{&open_fasta_files('seq.fa', \%index)};
             %out3=%{&open_fasta_files('seq.fa', \%range)};
             %seq=%{&open_fasta_files($PDB40_FASTA, \@seq_to_fetch)};

             while @ARGV at prompt was: 'GMJ.pep MJ0084'

<b>Function  </b> : open fasta files and put sequences in a hash
              If hash(es) is put which has sequence names and seek position
              of the index file, it searches the input FASTA file to
              fetch at that seek position. This is useful for Big fasta DBs
             If the seq name has ranges like  XXXXXX_1-30, it will only
              return 1-30 of XXXXXX sequence.

             FASTA sequence file format is like this;

             > 1st-seq
             ABCDEFGHIJKLMOPABCDEFGHIJKLMOPABCDEFGHIJKLMOPABCDEFG
             > 2nd.sequ
             ABCDEFGHIJKLMOYYUIUUIUIYIKLMOPABCDEFGHIJKLMOPABCDEFG
             >owl|P04439|1A03_HUMAN HLA CLASS I HISTOCOMPATIBILITY ANTIGEN, A-3 ALPHA CHAIN PRECURSOR....
             MARGDQAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRFIAVGYVDDT

             This can also return the sizes of sequences rather than seqs.

             This ignores any dup entrynames coming later.

<b>Keywords  </b> : open_fasta, open_fa_files, open_FASTA_files,
<b>Options   </b> : Seq name to fetch the specified seq only.
             as open_fasta_files.pl MY_SEQ_NAME Swissprot.fasta
            -d  for giving back desc as well as the name. so it
                gives  'HI0002 This is the description part'
                as the key
             If you put hash which is like('seq_name', ['20-30', '30-44',..])
              it will produce hash which has got:
              ( seq_name_20-30 'asdfasdfasdfasdfasd',
                seq_name_30-44 'kljkljkjkjljkjljkll',
                ....           .... )
            -s for returning sequence size only
   $reverse_seq=r by r ## to reverse seq.
<b>Usage     </b> : %fasta_seq=%{&open_fasta_files($fasta_file, ['MJ0084'])};
             if you put additional seq name as MJ0084 it will
             fetch that sequence only in the database file.

             %out=%{&open_fasta_files(@ARGV, \%index)};
               while  %index has (seq indexpos seq2 indexpos2,,,)
               In this case, the fasta file should have xxxx.fa format

<b>Version   </b> : 4.1
</pre><hr>

<H3><A href="msf_permute_hash_write
.pl.html">msf_permute_hash_write
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//msf_permute_hash_write
.pl">msf_permute_hash_write
.pl</A>
<pre>
<b>Function  </b> : gets 2 references (one for %hash the other for group $name)
             uses &msf_permute_array_write(\%hash, \$group_name)
             the second arg is for output file name. can be anything.
<b>Usage     </b> : &msf_permute_hash_write(\%hash, $group_name); # void
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="msf_permute_array_write
.pl.html">msf_permute_array_write
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//msf_permute_array_write
.pl">msf_permute_array_write
.pl</A>
<pre>
<b>Argument  </b> : gets 2 references
<b>Function  </b> : 
             the second arg is for output file name. can be anything.
             used in &msf_permu_hash_write
<b>Usage     </b> : &msf_permu_array_write(\%hash, \$group_name); # void
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="pir_permute_hash_write
.pl.html">pir_permute_hash_write
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//pir_permute_hash_write
.pl">pir_permute_hash_write
.pl</A>
<pre>
<b>Function  </b> : gets a reference of hash which has names and sequences as keys and values.
             uses &pir_permute_array_write
             the second arg is for output file name. can be anything.
<b>Usage     </b> : &pir_permute_hash_write($hash_ref, $group_name); # void
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="fasta_permute_hash_write
.pl.html">fasta_permute_hash_write
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//fasta_permute_hash_write
.pl">fasta_permute_hash_write
.pl</A>
<pre>
<b>Function  </b> : gets a reference of a hash which has names and sequences as keys and values.
             uses &fasta_permute_array_write
             the second arg is for output file name. can be anything.
<b>Usage     </b> : &fasta_permute_hash_write($hash_ref, $group_name); # void
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="fasta_permute_array_write
.pl.html">fasta_permute_array_write
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//fasta_permute_array_write
.pl">fasta_permute_array_write
.pl</A>
<pre>
<b>Function  </b> : gets a reference of an array which has names and sequences as keys and values.
             the second arg is for output file name. can be anything.
             used in &fasta_permu_hash_write
<b>Usage     </b> : &fasta_permu_array_write($hash_ref, $group_name); # void
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="ssp_permute_hash_write
.pl.html">ssp_permute_hash_write
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//ssp_permute_hash_write
.pl">ssp_permute_hash_write
.pl</A>
<pre>
<b>Function  </b> : gets a reference of hash which has names and sequences as keys and values.
             uses &ssp_permute_array_write
             the second arg is for output file name. can be anything.
<b>Usage     </b> : &ssp_permute_hash_write($hash_ref, $group_name); # void
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="pir_permute_array_write
.pl.html">pir_permute_array_write
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//pir_permute_array_write
.pl">pir_permute_array_write
.pl</A>
<pre>
<b>Function  </b> : gets a reference of hash which has names and sequences as keys and values.
             the second arg is for output file name. can be anything.
             used in &pir_permu_hash_write
<b>Usage     </b> : &pir_permu_array_write($hash_ref, $group_name); # void
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="ssp_permute_array_write
.pl.html">ssp_permute_array_write
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//ssp_permute_array_write
.pl">ssp_permute_array_write
.pl</A>
<pre>
<b>Function  </b> : gets a reference of hash which has names and sequences as keys and values.
             the second arg is for output file name. can be anything.
             used in &ssp_permu_hash_write
             ssp file is for PHD secondary structure prediction service.
<b>Usage     </b> : &ssp_permu_array_write($hash_ref, $group_name); # void
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="permute
.pl.html">permute
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//permute
.pl">permute
.pl</A>
<pre>
<b>Function  </b> : gets permutated array elements except single char elements.
             fastest
<b>Usage     </b> : &permute(\@array);
<b>Version   </b> : 1.0
<b>Warning   </b> : from : Kenneth Albanowski <kjahds@kjahds.com> CIS: 70705,126)
</pre><hr>

<H3><A href="permute_binary
.pl.html">permute_binary
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//permute_binary
.pl">permute_binary
.pl</A>
<pre>
<b>Example   </b> : &permute_binary(@array);
<b>Function  </b> : outs permutated array elements
<b>Version   </b> : 1.0
<b>Warning   </b> : from : silly@ugcs.caltech.edu
</pre><hr>

<H3><A href="ssp_write
.pl.html">ssp_write
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//ssp_write
.pl">ssp_write
.pl</A>
<pre>
<b>Example   </b> : &ssp_write($hash_pointer, $out_file_name);
<b>Function  </b> : writes multiple seqs. in fasta format (takes one or more than one seq.!!)
             ssp is PHD server format.
<b>Usage     </b> : two argments:  $seq_hash_reference  and $output_file_name
             takes a hash which has got names keys and sequences values.
             uses Perl5 pointers(references).
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="pir_write
.pl.html">pir_write
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//pir_write
.pl">pir_write
.pl</A>
<pre>
<b>Example   </b> : &pir_write($hash_pointer, $out_file_name);
<b>Function  </b> : writes multiple seqs. in fasta format (takes one or more than one seq.!!)
             pir is PHD server format.
<b>Usage     </b> : two argments:  $seq_hash_reference  and $output_file_name
             takes a hash which has got names keys and sequences values.
             uses Perl5 pointers(references).
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="pir_write
.pl.html">pir_write
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//pir_write
.pl">pir_write
.pl</A>
<pre>
<b>Example   </b> : &pir_write($hash_pointer, $out_file_name);
<b>Function  </b> : writes multiple seqs. in fasta format (takes one or more than one seq.!!)
<b>Usage     </b> : two argments:  $seq_hash_reference  and $output_file_name
             takes a hash which has got names keys and sequences values.
             uses Perl5 pointers(references).
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="make_singlet_list_from_pdb_entries
.pl.html">make_singlet_list_from_pdb_entries
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//make_singlet_list_from_pdb_entries
.pl">make_singlet_list_from_pdb_entries
.pl</A>
<pre>
<b>Example   </b> : 
: &make_singlet_list_from_pdb_entries(\@files);
 Input>
   >d2sn3__ 7.3.6.1.1 scorpion toxin [Centruroides sculpturatus ewing, variant 3]
   KEGYLVKKSDGCKYGCLKLGENEGCDTECKAKNQGGSYGYCYAFACWCEGLPESTPTYPL

 OUTPUT>
   >d2cmd_1 3.18.1.5.2 (1-145) Malate dehydrogenase [Escherichia coli]
   >d2naca2 3.18.1.4.1 (148-335) Formate dehydrogenase [Pseudomonas sp. 101]

<b>Function  </b> : 
: gets the classificaiton of scop in pdb40d.fa like file and
             produces pdb40d.pdbs file.

             1.1.1.1.4  means: Class.Fold.Superfamily.Family.Protein

             Compare with make_groups_from_pdb_entries
<b>Keywords  </b> : 
: make_singlet_list_from_pdb40d, make_singlet_list_from_scop, make_superfamilies
             write_pdbs_files, make_pdbs_files, make_pdb_group_files,
             write_pdbs, make_singlet_list_from_pdb_entries
<b>Options   </b> : _  for debugging.
             #  for debugging.
: 
<b>Usage     </b> : 
: &write_pdbs_files(\@files);
<b>Version   </b> : 1.0
: 1.0
<b>Warning   </b> : You MUST NOT delete '# options : ..' entry
              as it is read  by various subroutines.
</pre><hr>

<H3><A href="write_pdbg_files
.pl.html">write_pdbg_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//write_pdbg_files
.pl">write_pdbg_files
.pl</A>
<pre>
<b>Example   </b> : 
 Input>
   >d2sn3__ 7.3.6.1.1 scorpion toxin [Centruroides sculpturatus ewing, variant 3]
   KEGYLVKKSDGCKYGCLKLGENEGCDTECKAKNQGGSYGYCYAFACWCEGLPESTPTYPL

 OUTPUT>
   >d2cmd_1 3.18.1.5.2 (1-145) Malate dehydrogenase [Escherichia coli]
   >d2naca2 3.18.1.4.1 (148-335) Formate dehydrogenase [Pseudomonas sp. 101]

<b>Function  </b> : gets the classificaiton of scop in pdb40d.fa like file and
             produces pdb40d.pdbg file.

             1.1.1.1.4  means: Class.Fold.Superfamily.Family.Protein

<b>Keywords  </b> : make_groups_from_pdb40d, make_groups_from_scop, make_superfamilies
             write_pdbg_files, make_pdbg_files, make_pdb_group_files,
             write_pdbg, make_groups_from_pdb_entries
<b>Usage     </b> : &write_pdbg_files(\@files);
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="write_msp3_files
.pl.html">write_msp3_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//write_msp3_files
.pl">write_msp3_files
.pl</A>
<pre>
<b>Example   </b> : &write_msp3_files(\@files);  # while @files has G*.pdbg
<b>Function  </b> : opens two files. Gx.msp_1 and Gx.msp_2 to create Gx.msp3 file
              you can set the msp3 file extension by e= option,
              for example, e=interm will make  G1.interm instead of G1.msp3

<b>Keywords  </b> : make_msp3_files, create_msp3_files
<b>Options   </b> : 
    $upper_expect_limit2= by u2=  # u2 is for msp_2 files (eg, 0.0006)
    $upper_expect_limit1= by u1=  # u1 is for msp_1 files (eg, 0.081 )
    $lower_expect_limit1= by l1=
    $lower_expect_limit2= by l2=
    R for NOT adding ranges in seq names.
    e= for  extension name
    n  for  no sort by columns in output
    e  for  sorting columns by E values (first first and then second)
    E  for  sorting columns by E values but reverse order

<b>Returns   </b> : returns the names of msp3 files
<b>Usage     </b> : &write_msp3_files(\@files);
<b>Version   </b> : 1.8
</pre><hr>

<H3><A href="write_parf_files
.pl.html">write_parf_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//write_parf_files
.pl">write_parf_files
.pl</A>
<pre>
<b>Function  </b> : takes xxxx.msp files and writes xxxx.parf file
<b>Keywords  </b> : write_parf
<b>Options   </b> : 
  $pdbd_seq_only=d by d -d
  $sam_571_seq_only=571 by 571 -571
$pdb95d_2092_seq =2092 by 2092 -2092
  $ISS_2nd_Eval_factor= by E= ## "E=$eval"
  $PDB40D_935_FASTA= 935 by 935
  $use_raw_score=r
  $use_eval_but_show_raw_score=e by e -e  ## eval order but only raw score is shown.
                                            This is to make a special graph
                                           requested by David Haussler
<b>Usage     </b> : &write_parf_files(@ARGV);
<b>Version   </b> : 2.4
</pre><hr>

<H3><A href="write_mprf_files
.pl.html">write_mprf_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//write_mprf_files
.pl">write_mprf_files
.pl</A>
<pre>
<b>Keywords  </b> : mean_position_rank_file, mean_rank_position_file
<b>Usage     </b> : &write_mprf_files(@files);
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="write_evss_files
.pl.html">write_evss_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//write_evss_files
.pl">write_evss_files
.pl</A>
<pre>
<b>Author    </b> : jong@salt2.med.harvard.edu
<b>Function  </b> : This produces EVSS file(Error VS Score) from PARF file
<b>Keywords  </b> : get_score_vs_error_from_parf_files.pl
<b>Options   </b> : 
  d=$query_number    for dividing the errors by all the query number
   $negate_score=n by n -n  # to make sign change for PSI scores
   $error_per_query=q by q -q # divide the error by query number
   $log_of_errors=l by l -l
   $log_of_evalue_or_score=e by e -e
   $get_log_base_10=t by t -t

<b>Usage     </b> : @files_produced=@{write_evss_files(\@files)};
<b>Version   </b> : 1.5
</pre><hr>

<H3><A href="write_fasta
.pl.html">write_fasta
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//write_fasta
.pl">write_fasta
.pl</A>
<pre>
<b>Argument  </b> : 
   $sort_seq_names=s by s  ## in writing sorted sequences are written
   $write_rv_seq_as_well=R by R  # write reverse seq as well as forward seq
<b>Example   </b> : &write_fasta(\%in1, \$out_file_name, \%in2, \%in3,..., );
             << The order of the hash and scalar ref. doesn't matter. >>
<b>Function  </b> : writes multiple seqs. in fasta format (takes one or more seq.!!)
             This needs hash which have 'name' 'actual sequence as value'

             To print out each fasta seq into each single file, use write_fasta_seq_by_seq
             This can rename seq names

<b>Keywords  </b> : write_fasta_file, print_fasta_file, write fasta file, fasta_write
             show_fasta, write_sequence_fasta, write_fasta_files,
<b>Options   </b> : v for STD out.
             r for rename the sequences so that Clustalw would not complain with 10 char limit
               so result wuld be:  0 ->ASDFASDF, 1->ASDFASFASF, 2->ADSFASDFA
       $write_pure_seq_only=o by o -o  ## writing only the seq (no gap chars or space)
<b>Usage     </b> : many argments:  $seq_hash_reference  and $output_file_name
             takes a hash which has got names keys and sequences values.
<b>Version   </b> : 3.0
<b>Warning   </b> : The default output file name is 'default_out.fa' if you do not
             specify output file name.
             OUTput file should have xxxxx.fa or xxxx.any_ext NOT just 'xxxxx'
</pre><hr>

<H3><A href="write_fasta_seq_by_seq
.pl.html">write_fasta_seq_by_seq
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//write_fasta_seq_by_seq
.pl">write_fasta_seq_by_seq
.pl</A>
<pre>
<b>Example   </b> : with >xxxx
                  ASDFASDFASDFASDFASDFASDFASDF
                  >yyyy
                  ASDFASDFASDFASDFASDFASDFSDAFSD

             You will get two files (xxxx.fa, yyyy.fa)
<b>Function  </b> : accepts one hash of multiple sequences and writes many files
             of single sequences by using the names as file names.
             If $extension is provided, it writes an output as in
             the below example (seq1_sc.fasta). If not, it just attach
             'fa' to files.
             This needs, hash of 'name', 'actual sequence as value'
<b>Keywords  </b> : write_each_fasta, write_single_fasta, write_fasta_single
             single_fasta_write, write_fasta_files_seq_by_seq,
             write_single_fasta_files,
<b>Options   </b> : can specify extension name.
             e  for checking fasta file exists or not and skipps if so
             r for rename the sequences so that Clustalw would not complain with 10 char limit
               so result wuld be:  0 ->ASDFASDF, 1->ASDFASFASF, 2->ADSFASDFA
   $write_rv_seq_as_well=R by R  # write reverse seq as well as forward seq
   $extension= by E=
<b>Returns   </b> : nothing. default OUTPUT file name is '$key.fa' !!
<b>Usage     </b> : &write_fasta_seq_by_seq(\%hash, [$extension], [\$output_filename]);
<b>Version   </b> : 2.1
</pre><hr>

<H3><A href="write_rdif_files
.pl.html">write_rdif_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//write_rdif_files
.pl">write_rdif_files
.pl</A>
<pre>
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="write_rdif2_files
.pl.html">write_rdif2_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//write_rdif2_files
.pl">write_rdif2_files
.pl</A>
<pre>
<b>Usage     </b> : &write_rdif2_files(@files);
<b>Version   </b> : 1.2
</pre><hr>

<H3><A href="write_ardf_files
.pl.html">write_ardf_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//write_ardf_files
.pl">write_ardf_files
.pl</A>
<pre>
<b>Usage     </b> : &write_ardf_files(@files);
<b>Version   </b> : 1.2
</pre><hr>

<H3><A href="write_nhco_files
.pl.html">write_nhco_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//write_nhco_files
.pl">write_nhco_files
.pl</A>
<pre>
<b>Author    </b> : jpark@rascal.med.harvard.edu, NHCO stands for Nomolog, Homolog Column Output file
<b>Function  </b> : writes nhco files with each class(4 of them) file nhco as well.
<b>Keywords  </b> : write homology column file, write_nomol_homol_column_files,
             write_homol_col_files
<b>Version   </b> : 1.5
</pre><hr>

<H3><A href="write_c3ss_files
.pl.html">write_c3ss_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//write_c3ss_files
.pl">write_c3ss_files
.pl</A>
<pre>
<b>Author    </b> : jong@salt2.med.harvard.edu
<b>Version   </b> : 1.2
</pre><hr>

<H3><A href="write_pred_files
.pl.html">write_pred_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//write_pred_files
.pl">write_pred_files
.pl</A>
<pre>
<b>Example   </b> : &write_pred_files(\%gapped_av_for_back_pred, $final_output_pred_name,
                 $graphical_rep_of_str, "$put_reliability_line");

<b>Keywords  </b> : write_predator_short_out_file, write_pred_file, write_prd_file
<b>Options   </b> : 
   $put_reliability_line=r by r
   $omit_coil_region=c by c
   $protein_name= by n=
   $graphical_rep_of_str=g by g
   $show_on_screen_only=s by s
   $seq_block_size= by b=
<b>Version   </b> : 1.4
</pre><hr>

<H3><A href="write_prdl_files
.pl.html">write_prdl_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//write_prdl_files
.pl">write_prdl_files
.pl</A>
<pre>
<b>Keywords  </b> : write_predator_long_out_file
<b>Options   </b> : 
   $protein_name= by n=
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="show_in_fasta
.pl.html">show_in_fasta
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//show_in_fasta
.pl">show_in_fasta
.pl</A>
<pre>
<b>Example   </b> : &show_in_fasta(\%hash);
<b>Function  </b> : shows multiple seqs. in fasta format (takes one or more seq.!!)
<b>Usage     </b> : &show_hash_in_fasta(\%in1, \%in2, \%in3, .... );
             takes a hash which has got names keys and sequences values.
             uses Perl5 pointers(references).
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="One_To_Three_Letter
.pl.html">One_To_Three_Letter
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//One_To_Three_Letter
.pl">One_To_Three_Letter
.pl</A>
<pre>
<b>Function  </b> : a hash of one letter to 3 letter amino acid code , returns a hash
<b>Keywords  </b> : 1_to_3
<b>Usage     </b> : %one_letter  = %{&One_To_Three_Letter};   # takes no arguments (void).
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="ONE_TO_THREE_LETTER
.pl.html">ONE_TO_THREE_LETTER
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//ONE_TO_THREE_LETTER
.pl">ONE_TO_THREE_LETTER
.pl</A>
<pre>
<b>Function  </b> : a hash of one letter to 3 letter amino acid code , returns a hash
<b>Keywords  </b> : 1_to_3, convert_1_to_3_letter
<b>Usage     </b> : %one_letter  = %{&ONE_TO_THREE_LETTER };   # takes no arguments (void).
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="one_to_three_letter
.pl.html">one_to_three_letter
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//one_to_three_letter
.pl">one_to_three_letter
.pl</A>
<pre>
<b>Function  </b> : a hash of one letter to 3 letter amino acid code , returns a hash
<b>Usage     </b> : %one_letter  = %{&one_to_three_letter};   # takes no arguments (void).
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="three_to_one_letter
.pl.html">three_to_one_letter
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//three_to_one_letter
.pl">three_to_one_letter
.pl</A>
<pre>
<b>Function  </b> : a hash of one letter to 3 letter amino acid code , returns a hash
<b>Keywords  </b> : 321, 3to1 3_to_1 THREE_TO_ONE_LETTER Three_To_One_Letter
             convert_3_to_1, convert_3_to_1_aa_name
<b>Usage     </b> : %three_letter  = &three_to_one_letter ;   # takes no arguments (void).
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="convert_3_to_1_letter
.pl.html">convert_3_to_1_letter
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//convert_3_to_1_letter
.pl">convert_3_to_1_letter
.pl</A>
<pre>
<b>Function  </b> : a hash of one letter to 3 letter amino acid code , returns a hash
<b>Keywords  </b> : 321, 3to1 3_to_1 THREE_TO_ONE_LETTER Three_To_One_Letter
             convert_3_to_1, convert_3_to_1_aa_name
<b>Usage     </b> : %three_letter  = &three_to_one_letter ;   # takes no arguments (void).
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="convert_1_to_3_letter
.pl.html">convert_1_to_3_letter
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//convert_1_to_3_letter
.pl">convert_1_to_3_letter
.pl</A>
<pre>
<b>Function  </b> : a hash of one letter to 3 letter amino acid code , returns a hash
<b>Keywords  </b> : 123, 1to3 1_to_3 one_TO_three_LETTER One_To_Three_Letter
             convert_1_to_3, convert_1_to_3_aa_name
<b>Usage     </b> : %three_letter  = &three_to_one_letter ;   # takes no arguments (void).
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="amino_acid_compos_id_percent
.pl.html">amino_acid_compos_id_percent
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//amino_acid_compos_id_percent
.pl">amino_acid_compos_id_percent
.pl</A>
<pre>
<b>Argument  </b> : hash of at least 2 sequences.
<b>Function  </b> : gets amino acid composition identity of any given
             number of sequences(at least 2).
<b>Keywords  </b> : get_amino_acid_composition, get_protein_composition, composition
<b>Usage     </b> : $percent = &amino_acid_compos_id_percent (%any_hash_with_sequences);
             The way identity(composition) is derived is;

<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="seq_id_percent_array
.pl.html">seq_id_percent_array
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//seq_id_percent_array
.pl">seq_id_percent_array
.pl</A>
<pre>
<b>Function  </b> : produces amino acid composition identity of any given number of sequences.
<b>Keywords  </b> : get_percent_composition_identity, seq_composition_identity,
             percent_sequence_composition_id

<b>Usage     </b> : $percent = &seq_id_percent_array(@any_array_sequences);
             The way identity(pairwise) is derived is;

<b>Version   </b> : 1.0
<b>Warning   </b> : This can handle 'common gaps' in the sequences
</pre><hr>

<H3><A href="compos_id_percent_array
.pl.html">compos_id_percent_array
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//compos_id_percent_array
.pl">compos_id_percent_array
.pl</A>
<pre>
<b>Function  </b> : produces amino acid composition identity of any given number of sequences.
<b>Usage     </b> : $percent = &compos_id_percent_array(@any_array_sequences);
             The way identity(composition) is derived is;
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="compos_id_percent_hash
.pl.html">compos_id_percent_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//compos_id_percent_hash
.pl">compos_id_percent_hash
.pl</A>
<pre>
<b>Function  </b> : gets amino acid composition identity of any given number of sequences.
<b>Keywords  </b> : get_amino_acid_composiiton
<b>Usage     </b> : $percent = &compos_id_percent_hash(%any_hash_with_sequences);
             The way identity(composition) is derived is;

<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="common_compos_id_hash
.pl.html">common_compos_id_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//common_compos_id_hash
.pl">common_compos_id_hash
.pl</A>
<pre>
<b>Argument  </b> : two references of hash of seqeunces.
<b>Example   </b> : ('A', 200, 'C', 191, D, 99)
                  ('A', 290, 'C', 199, D, 100)
             uses only two sequences.
<b>Function  </b> : actual calculation of identity
<b>Returns   </b> : ref. of a scaler (in percent)  eg)  95
<b>Usage     </b> : %hash = &common_compos_hash(\%any_hash1, \%any_hash1);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="calc_compos_id_hash
.pl.html">calc_compos_id_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//calc_compos_id_hash
.pl">calc_compos_id_hash
.pl</A>
<pre>
<b>Argument  </b> : two references of hash of seqeunces.
<b>Example   </b> : ('A', 200, 'C', 191, D, 99)
                  ('A', 290, 'C', 199, D, 100)
             uses only two sequences.
<b>Function  </b> : actual calculation of identity
<b>Returns   </b> : ref. of a scaler (in percent)  eg)  95
<b>Usage     </b> : %hash = &calc_compos_hash(\%any_hash1, \%any_hash1);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_percentage
.pl.html">get_percentage
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_percentage
.pl">get_percentage
.pl</A>
<pre>
<b>Argument  </b> : ref. for Scalar string or Array of chars or Hash  AND 'the target char'
<b>Example   </b> : if the string is  'seq  ABCDEEEEEFFEFE' given in a hash
             if you put 'A' as one argument, it counts the occurances of 'A'
             and gets the percentage of it.
<b>Function  </b> : calculates the percentage content of any single char over the whole
             length of strings in it.
<b>Keywords  </b> : get_percentage_of_char
<b>Options   </b> : None yet.
<b>Returns   </b> : Numerical Percentage
<b>Usage     </b> : %out= %{&get_percentage(\%result, '1')};
<b>Version   </b> : 1.0
<b>Warning   </b> : This converts array and string input as ref. into arbitrary hash and
             returns hash
             programmed by A Biomatic
</pre><hr>

<H3><A href="pairwise_percent_id
.pl.html">pairwise_percent_id
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//pairwise_percent_id
.pl">pairwise_percent_id
.pl</A>
<pre>
<b>Function  </b> : takes a ref. of a hash of names and sequences, returns
             percent identity.
<b>Usage     </b> : $identity = ${&pairwise_percent_id(%arrayinput)};

<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_seq_identity
.pl.html">get_seq_identity
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_seq_identity
.pl">get_seq_identity
.pl</A>
<pre>
<b>Argument  </b> : hash(es) of sequences.
<b>Function  </b> : takes a ref. of a hash of names and sequences, returns
             percent identity. NOT composition identity.
<b>Keywords  </b> : get_sequence_identity
<b>Usage     </b> : $identity = ${&get_seq_identity(%arrayinput)};

<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_correct_percent_alignment_rate
.pl.html">get_correct_percent_alignment_rate
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_correct_percent_alignment_rate
.pl">get_correct_percent_alignment_rate
.pl</A>
<pre>
<b>Argument  </b> : two sequence files which have identical sequence names.
<b>Function  </b> : accepts two files and prints out the sequence identities of the alignment.
<b>Options   </b> : h  # for help
             v  # for verbose printouts(prints actual sequences)
<b>Returns   </b> : reference of Scalar for percentage correct alignment(for already
             aligned sequences)
<b>Usage     </b> : &get_correct_percent_alignment_rate(\$file1, \$file2);
<b>Warning   </b> : Alpha version,  A Biomatic , made for Bissan
</pre><hr>

<H3><A href="amino_acid_compos_id_percent_trend
.pl.html">amino_acid_compos_id_percent_trend
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//amino_acid_compos_id_percent_trend
.pl">amino_acid_compos_id_percent_trend
.pl</A>
<pre>
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="composition_table
.pl.html">composition_table
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//composition_table
.pl">composition_table
.pl</A>
<pre>
<b>Function  </b> : returns a table of alphabet with occurances.
             can handle any char, this converts char to upper case.
<b>Returns   </b> : %hash1 = ('A',3, 'C',2, 'D',1, 'Q',2, 'S',1), %hash2,,,
<b>Usage     </b> : %output = %{&compos_table(@input_array1, @input_array2,,,,)};
             example input

<b>Warning   </b> : converts all SMALL letters to Capital letters before counting!!
</pre><hr>

<H3><A href="common_compos_2_hash
.pl.html">common_compos_2_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//common_compos_2_hash
.pl">common_compos_2_hash
.pl</A>
<pre>
<b>Argument  </b> : two references of hash of seqeunces.
<b>Example   </b> : common gaps means only '.' (dots, not alphabets!!)
             AAA....BBCB
             AAAB..B.BCC  --> A.A.....BC. (as in an array)
             A.AAA...BCA
<b>Returns   </b> : a hash (string1, number1, string2, number2, string3, number3, ...)
<b>Usage     </b> : %hash = &common_compos_hash(\%any_hash1, \%any_hash1);
</pre><hr>

<H3><A href="pair_percent_id_trend
.pl.html">pair_percent_id_trend
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//pair_percent_id_trend
.pl">pair_percent_id_trend
.pl</A>
<pre>
<b>Example   </b> : common gaps means only '.' (dots, not alphabets!!)
             AAA....BBCB
             AAAB..B.BCC  --> A.A.....BC. (as in an array)
             A.AAA...BCA
             The resulting array XXXXX..XXXX is literally like so.
             This is to detect absurd gaps in the above.


<b>Usage     </b> : @array = &pair_percent_id_trend (%arrayinput);
</pre><hr>

<H3><A href="smaller_one
.pl.html">smaller_one
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//smaller_one
.pl">smaller_one
.pl</A>
<pre>
<b>Example   </b> : will return   5   with  &smaller_one(5, 50);
<b>Function  </b> : gets smaller value of the two inputs
<b>Usage     </b> : $smaller = & smaller_one($var, $var2);
<b>Warning   </b> : gets only digits!!
</pre><hr>

<H3><A href="count_num_of_char
.pl.html">count_num_of_char
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//count_num_of_char
.pl">count_num_of_char
.pl</A>
<pre>
<b>Function  </b> : takes only ARRAY and counts the number of char. Each elem should be
             a single char.
<b>Usage     </b> : $num_char = &count_num_of_char(@input_array_of_single_char);
</pre><hr>

<H3><A href="remov_com_column2
.pl.html">remov_com_column2
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//remov_com_column2
.pl">remov_com_column2
.pl</A>
<pre>
<b>Argument  </b> : accepts reference for a hash.
<b>Example   </b> : seq1  ABCDE------DDD         seq1  ABCDE--DDD
             seq2  ABCDEE-----DD-  ==>    seq2  ABCDEE-DD-
             seq3  ---DEE----DDE-         seq3  ---DEEDDE-
                         ^^^^
             from above the 4 columns of gap will be removed
             To remove absurd gaps in multiple sequence alignment
<b>Returns   </b> : a ref. of a hash.

               <input hash>                   <out hash>

<b>Usage     </b> : %new_string = %{&remov_com_column2(\%input_hash)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_common_column
.pl.html">get_common_column
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_common_column
.pl">get_common_column
.pl</A>
<pre>
<b>Argument  </b> : 2 or more ref for hash of identical keys and value length.
             One optional arg for replacing space char to the given one.
<b>Author    </b> : jong@salt2.med.harvard.edu
<b>Class     </b> : get_common_column, get_common_column_in_seq, get common column in sequence
             for secondary structure only representation.
<b>Example   </b> : %out =%{&get_common_column(\%hash1, \%hash2, '-')};
             output> with 'E' option >>> "name1     --HHH--1232-"
   Following input will give;
  %hash1 = ('s1', '--EHH-CHHEE----EHH--HHEE----EHH--HHEE----EHH-CHHEE--');
  %hash2 = ('s2', '--EEH-CHHEE----EEH-CHHEE----EEH-CHHEE----EEH-CHHEE--');
  %hash3 = ('s3', '-KEEH-CHHEE-XX-EEH-CHHEE----EEH-CHHEE----EEH-CHHEE--');
  %hash4 = ('s4', '-TESH-CHEEE-XX-EEH-CHHEE----EEH-CHHEE----EEH-CHHEE--');

    s1_s2_s3_s4    --E-H-CH-EE----E-H--HHEE----E-H--HHEE----E-H-CHHEE--

<b>Function  </b> : (name1         --EHH--HHEE-- )
             (name2         --HHH--EEEE-- ) ==> result is;

             (name1_name2   -- HH--  EE-- )
             to get the identical chars in hash strings of sequences.

<b>Keywords  </b> : Overlap, superpose hash, overlay identical chars, superpose_seq_hash
             get_common_column, get_com_column, get_common_sequence,
             get_common_seq_region, multiply_seq_hash, get_common_column_in_sequence
<b>Returns   </b> : one hash ref. of the combined key name (i.e., name1_name2). Combined by '_'
<b>Usage     </b> : %out =%{&get_common_column(\%hash1, \%hash2, '-')};
<b>Version   </b> : 1.6
<b>Warning   </b> : This gets more than 2 hashes. Not more than that!

</pre><hr>

<H3><A href="overlay_seq_for_identical_chars
.pl.html">overlay_seq_for_identical_chars
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//overlay_seq_for_identical_chars
.pl">overlay_seq_for_identical_chars
.pl</A>
<pre>
<b>Argument  </b> : 2 ref for hash of identical keys and value length. One optional arg for
             replacing space char to the given one.
<b>Example   </b> : %out =%{&overlay_seq_for_identical_chars(\%hash1, \%hash2, '-')};
             output> with 'E' option >>> "name1     --HHH--1232-"
<b>Function  </b> : (name1         --EHH--HHEE-- )
             (name2         --HHH--EEEE-- ) ==> result is;

             (name1_name2   -- HH--  EE-- )
             to get the identical chars in hash strings of sequences.

<b>Keywords  </b> : Overlap, superpose hash, overlay identical chars, superpose_seq_hash
<b>Returns   </b> : one hash ref. of the combined key name (i.e., name1_name2). Combined by '_'
<b>Usage     </b> : %out =%{&overlay_seq_for_identical_chars(\%hash1, \%hash2, '-')};
<b>Version   </b> : 1.0
<b>Warning   </b> : Works only for 2 sequence hashes.
</pre><hr>

<H3><A href="remov_com_column
.pl.html">remov_com_column
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//remov_com_column
.pl">remov_com_column
.pl</A>
<pre>
<b>Argument  </b> : accepts reference for hash(es) and array(s).
<b>Function  </b> : removes common gap column in seq.
<b>Keywords  </b> : remove_com_column, remove_common_column,
             remove_common_gap_column, remov_common_gap_column,
             remove com column
<b>Returns   </b> : a ref. of  hash(es) and array(s).

             name1   ABCDE....DDD       name1  ABCDE..DDD
             name2   ABCDEE..DD..  -->  name2  ABCDEEDD..
             name3   ...DEE..DDE.       name3  ...DEEDDE.

             (ABC....CD, ABCD...EE) --> (ABC.CD, ABCDEE)
             from above the two column of dot will be removed
             To remove absurd gaps in multiple sequence alignment. for nt6-hmm.pl
<b>Usage     </b> : %new_string = %{&remov_com_column(\%hashinput)};
             @out=@{&remov_com_column(\@array3)};
</pre><hr>

<H3><A href="remov_common_gap
.pl.html">remov_common_gap
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//remov_common_gap
.pl">remov_common_gap
.pl</A>
<pre>
<b>Example   </b> : XXX...XXX with AAA.....BBBB, The common positions of 3,4,5 deleted
             XXX...XXX will be removed in AAA.....BBBB --> AAA..BBBB
             XXX...XXX is an @array, while AAA.....BBBB is a value of the input hash
<b>Function  </b> : XXX...XXX, and an hash input. removes all the common gap(dots) in targets.
<b>Usage     </b> : %result = &remov_common_gap (*common_pos_arr, *target_hash_of_sequence);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="com_gap_pos_hash
.pl.html">com_gap_pos_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//com_gap_pos_hash
.pl">com_gap_pos_hash
.pl</A>
<pre>
<b>Argument  </b> : gets a ref. of a hash of sequences
<b>Example   </b> : common gaps means only '.' (dots, not alphabets!!)
             AAA....BBBB
             AABB....BBC  --> XXXXX..XXXX (as in an array)
             ..AAA...BCA
             This is to detect absurd gaps in the above.
<b>Function  </b> : returns X...XXXX, as an array. '.' means common elements.
<b>Keywords  </b> : common_gap_pos_hash
<b>Usage     </b> : @array = @{&com_elem_pos_hash(%arrayinput)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="pairwise_iden_pos
.pl.html">pairwise_iden_pos
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//pairwise_iden_pos
.pl">pairwise_iden_pos
.pl</A>
<pre>
<b>Example   </b> : common gaps means only '.' (dots, not alphabets!!)
             AAA....BBCB
             AAAB..B.BCC  --> A.A.....BC. (as in an array)
             A.AAA...BCA
             The resulting array XXXXX..XXXX is literally like so.
             This is to detect absurd gaps in the above.
<b>Usage     </b> : @array = &pairwise_iden_pos(%arrayinput);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="open_pdb_files
.pl.html">open_pdb_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_pdb_files
.pl">open_pdb_files
.pl</A>
<pre>
<b>Argument  </b> : one ref. for an inputfile (absolute
             >>> PDB example >>>
             SEQRES   1 A  284  MET ASP ALA ILE LYS LYS LYS MET GLN MET LEU LYS LEU  2TMA  51
             SEQRES   2 A  284  ASP LYS GLU ASN ALA LEU ASP ARG ALA GLU GLN ALA GLU  2TMA  52

<b>Function  </b> : Convert a PDB structure file to FASTA format sequences.
<b>Keywords  </b> : read_pdb_files{, read pdb files, open pdb files
<b>Returns   </b> : One ref. for a hash of sequences(DNA, RNA, PROTEIN (IN diff chains)
             If the two chains are identical, it rids of one of them and returns
             a name with out chain note-->  2tma, not 2tmaA and 2tmaB
<b>Usage     </b> : %out = %{&open_pdb_files(\$VAR)};
<b>Version   </b> : 1.7
<b>Warning   </b> : (read the sequences only)
</pre><hr>

<H3><A href="open_brk_files
.pl.html">open_brk_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_brk_files
.pl">open_brk_files
.pl</A>
<pre>
<b>Argument  </b> : one ref. for an inputfile (absolute
             >>> PDB example >>>

             SEQRES   1 A  284  MET ASP ALA ILE LYS LYS LYS MET GLN MET LEU LYS LEU  2TMA  51
             SEQRES   2 A  284  ASP LYS GLU ASN ALA LEU ASP ARG ALA GLU GLN ALA GLU  2TMA  52
             SEQRES   3 A  284  ALA ASP LYS LYS ALA ALA GLU ASP ARG SER LYS GLN LEU  2TMA  53

<b>Function  </b> : Convert a PDB structure file to FASTA format sequences.
<b>Returns   </b> : One ref. for a hash of sequences(DNA, RNA, PROTEIN (IN diff chains)
             If the two chains are identical, it rids of one of them and returns
             a name with out chain note-->  2tma, not 2tmaA and 2tmaB
<b>Usage     </b> : %out = %{&open_brk_files(\$VAR)};
</pre><hr>

<H3><A href="open_msf_jp_files
.pl.html">open_msf_jp_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_msf_jp_files
.pl">open_msf_jp_files
.pl</A>
<pre>
<b>Function  </b> : makes two hashes from  ...msf and ..jp files. %array1 is for msf
<b>Usage     </b> : &open_msf_jp_files($file1, $file2);
<b>Warning   </b> : !!! not very general bettter not use.
             msf file is meant to be seq
             jp file is meant to be structural alignment (correct seq

             msf format is

             cofi_human  ATFVKM
             ici2_horvu  RVRLFVDKLD NIA
             ici3_horvu  RVRLFVDRLD NIA

             jp format is;

             ycah_ecoli  RNVEIV----VID-GVRRFGNIA
             icis_vicfa  RVRLYVDESNKVV-RAAPIGNIA
             ier1_lyces  RVRLFVNLLDIVV-QTPKVGNIA
</pre><hr>

<H3><A href="scoring
.pl.html">scoring
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//scoring
.pl">scoring
.pl</A>
<pre>
<b>Warning   </b> : not general, !!!
</pre><hr>

<H3><A href="sort_files_by_time
.pl.html">sort_files_by_time
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//sort_files_by_time
.pl">sort_files_by_time
.pl</A>
<pre>
<b>Function  </b> : sorts files by creation time. Oldest the first
<b>Keywords  </b> : sort_by_time, sort_files_chronically
<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Usage     </b> : @files = @{&sort_files_by_time(\@files)};
<b>Version   </b> : 1.0
<b>Warning   </b> : You MUST NOT delete '# options : ..' entry
              as it is read  by various subroutines.
</pre><hr>

<H3><A href="sort_hash_by_value_and_make_array
.pl.html">sort_hash_by_value_and_make_array
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//sort_hash_by_value_and_make_array
.pl">sort_hash_by_value_and_make_array
.pl</A>
<pre>
<b>Function  </b> : sorts any hash by its values and returns ref. of sorted hash values
             with keys attached. So, if the input key value were
             key1 value1, the result will be an element 'value1 key1' as
             a string
<b>Keywords  </b> : sort_hash_by_value, sort_hash, sort_by_values,
<b>Options   </b> : -n  for numerical sort(not working yet)
<b>Usage     </b> : @values_sorted =@{&sort_hash_by_value_and_make_array(\%assoc)};
<b>Version   </b> : 1.1
<b>Warning   </b> : The same values will be overwritten.
</pre><hr>

<H3><A href="sort_by_hash_values
.pl.html">sort_by_hash_values
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//sort_by_hash_values
.pl">sort_by_hash_values
.pl</A>
<pre>
<b>Function  </b> : sorts any hash by its values and returns ref. of sorted hash values
<b>Keywords  </b> : sort_hash_by_value, sort_hash, sort_by_values, sort_by_value
<b>Usage     </b> : @values_sorted =@{sort_by_by_values(\%assoc)};
<b>Version   </b> : 1.1
<b>Warning   </b> : The same values will be overwritten.
</pre><hr>

<H3><A href="sort_by_keys
.pl.html">sort_by_keys
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//sort_by_keys
.pl">sort_by_keys
.pl</A>
<pre>
<b>Function  </b> : sorts any hash by its values and returns ref. of sorted hash values
<b>Keywords  </b> : sort_hash_by_keys, sort_hash, key_sort
<b>Usage     </b> : @values_sorted =@{sort_by_by_values(\%assoc)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="sort_hash_by_keys
.pl.html">sort_hash_by_keys
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//sort_hash_by_keys
.pl">sort_hash_by_keys
.pl</A>
<pre>
<b>Function  </b> : sorts any hash by its values and returns ref. of sorted hash values
<b>Keywords  </b> : sort_hash_by_keys, sort_hash, key_sort
<b>Usage     </b> : @values_sorted =@{sort_by_values(\%assoc)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="sort_hash_by_value
.pl.html">sort_hash_by_value
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//sort_hash_by_value
.pl">sort_hash_by_value
.pl</A>
<pre>
<b>Function  </b> : sorts any hash by its values and returns ref. of sorted hash values
<b>Keywords  </b> : sort_hash_by_value, sort_hash, value_sort,
<b>Usage     </b> : @values_sorted =@{sort_hash_by_values(\%assoc)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="by_values
.pl.html">by_values
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//by_values
.pl">by_values
.pl</A>
<pre>
<b>Usage     </b> : for $key(sort by_values(values %assoc)){print $assoc{$key},"\n";}
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="sort_string_by_length
.pl.html">sort_string_by_length
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//sort_string_by_length
.pl">sort_string_by_length
.pl</A>
<pre>
<b>Function  </b> : sorts strings in array according to their sizes
             bigger comes first.
<b>Keywords  </b> : sort_array_by_length, sort_str_by_length, sort_array_string_by
             sort_string_by_leng, sort_by_length, sort_by_leng,
             sort_array_by_string_length, sort_array_elements_by_string_length
<b>Options   </b> : -r  reverse the order
<b>Usage     </b> : @output = @{&sort_string_by_length(@any_input_strings, [-r], @more)};
<b>Version   </b> : 1.2
</pre><hr>

<H3><A href="get_host_by_addr
.pl.html">get_host_by_addr
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_host_by_addr
.pl">get_host_by_addr
.pl</A>
<pre>
<b>Example   </b> : ($name,$aliases,$addrtype,$length,@addrs)=&get_host_by_addr($var); while $var = "13.13.12.12";
<b>Keywords  </b> : get_host_by_address, get_hostname_by_address
<b>Usage     </b> : ($name,$aliases,$addrtype,$length,@addrs)=&get_host_by_addr('131.111.137.11'); or
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="hostname
.pl.html">hostname
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//hostname
.pl">hostname
.pl</A>
<pre>
<b>Keywords  </b> : get_hostname
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_host_by_name
.pl.html">get_host_by_name
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_host_by_name
.pl">get_host_by_name
.pl</A>
<pre>
<b>Example   </b> : ($name,$aliases,$addrtype,$length,@addrs)=&get_host_by_name($var);
             while $var = "ind4";
<b>Usage     </b> : ($name,$aliases,$addrtype,$length,@addrs)=&get_host_by_name('ind4'); or
<b>Version   </b> : 1.0
<b>Warning   </b> : ! not working yet.
</pre><hr>

<H3><A href="word_wrap
.pl.html">word_wrap
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//word_wrap
.pl">word_wrap
.pl</A>
<pre>
<b>Returns   </b> : 
             The string with newlines replacing spaces in appropriate places.
<b>Usage     </b> : &word_wrap($line_to_format)
<b>Version   </b> : 1.0
<b>Warning   </b> : 
             The following subroutine does word wrapping on a text string
</pre><hr>

<H3><A href="show_array
.pl.html">show_array
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//show_array
.pl">show_array
.pl</A>
<pre>
<b>Example   </b> : Output:      item1
             Output:      item2
             Output:      item3
<b>Function  </b> : for debugging purpose. Shows any array elem line by line.
<b>Options   </b> : -h  for horizontal display of elements
             c   for compact (do not put new line between array chunk)
             s   for putting new line between arrays
<b>Usage     </b> : &show_array(\@input_array);
<b>Version   </b> : 2.4
<b>Warning   </b> : can handle scalar ref, too.
</pre><hr>

<H3><A href="array_most_occur
.pl.html">array_most_occur
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//array_most_occur
.pl">array_most_occur
.pl</A>
<pre>
<b>Argument  </b> : \@array
<b>Keywords  </b> : median_array, get_median_array, get_array_median, array_median
<b>Returns   </b> : \$median
<b>Usage     </b> : $median = ${&array_most_occur(\@array)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="array_least_occur
.pl.html">array_least_occur
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//array_least_occur
.pl">array_least_occur
.pl</A>
<pre>
<b>Argument  </b> : \@array
<b>Keywords  </b> : median_array, get_median_array, get_array_median, array_median
<b>Returns   </b> : \$median
<b>Usage     </b> : $median = ${&array_least_occur(\@array)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="show_hash
.pl.html">show_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//show_hash
.pl">show_hash
.pl</A>
<pre>
<b>Example   </b> : Output:      item1
             Output:      item2
             Output:      item3
<b>Function  </b> : for debugging purpose. Shows any array elem line by line.
             the line is 60 elements long (uses recursion)
<b>Options   </b> : -s or -S or s or S for spaced output. Eg)
             seq1       1 1 1 1 1 1 1 1 1 1 1 1

             instead of
             seq1       111111111111

             -h or -H or h or H for horizontal line of '---------...'

<b>Usage     </b> : &show_hash(\@input_array);
<b>Version   </b> : 1.7
<b>Warning   </b> : There is a global variable:  $show_hash_option
             It tries to detect any given sting which is joined by ','
</pre><hr>

<H3><A href="open_predator_files
.pl.html">open_predator_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_predator_files
.pl">open_predator_files
.pl</A>
<pre>
<b>Example   </b> : 
  There are 2 types of output.  The short output:>

  > MOZ_HUMAN_part
                .         .         .         .         .
  1    LDHKTLYYDVEPFLFYVLTQNDVKGCHLVGYFSKEKHCQQKYNVSCIMIL   50
       ___EEEEEE__HHHHHHH_______EEE____________EEEEEEEEE_

 ((-l option for long output )
  NAME MOZ_HUMAN_part
  HEADER  |- Residue -|  Pred  Rel      NAli   Asn
  PRED    1    MET    M  c     0.000    0      ?
  PRED    2    ALA    A  c     0.000    0      ?

<b>Function  </b> : gets sec. str. prediction of predator and puts in hash
             If 's' option is given, it also gives sequence hash ref
             as the second output ref. This can handle the 2 types
             of output format of predator. So, the output can will
             be different according to inputs.
<b>Keywords  </b> : open_prd_files, open_pred_files, predator, open_prdl_files
             open_pre_files, secondary structure prediction file
<b>Options   </b> : 's' for sequence output as well (\%sec_str, \%seq)
             'p' for percentage of the sec. str.
             'a' for accumulated percentage. This will
                  set 'p' automatically
             'n' for NO name when outputing Percentage of chars with
                 HASH input to get_occurances_of_char sub.
      $reverse_residue_order=r by r
<b>Version   </b> : 1.8
</pre><hr>

<H3><A href="open_phd_files
.pl.html">open_phd_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_phd_files
.pl">open_phd_files
.pl</A>
<pre>
<b>Argument  </b> : one or more file names and options. Files should be PHD server's result.
<b>Function  </b> : open phd files and put sequences in a hash(s) (run open_phd_files.pl to
             get some ideas on how this works. type  'open_phd_files.pl xxx.phdo s',
             it will produce 5 different hashes of secondary structure pred.
<b>Options   </b> : $secondary, $access, $PHD_sec, $Rel_sec, $prH_sec, $prE_sec, $prL_sec,
                  $prL_sec, $SUB_sec, $P_3_acc, $PHD_acc, $Rel_acc, $SUB_acc);
   $attach_class_info_in_seq_name=c by c ## this makes seq_name   seq_name_PHD_s
   $simple_seq_with_name_hash=s by s

<b>Returns   </b> : one or more hashes(ref.) secondary structure prediction of PHD server
             --- The PHD secondary server output which are read by open_phd_files -----
             1 =>       PHD sec |         HHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHH     HHHHHHH|
             2 =>       Rel sec |987544342178899999999987678999998478999999999995679771688999|
             3 =>       prH sec |001222323478899999999987778999998678999999999986110115788999
             4 =>       prE sec |000010000101000000000000010000000000000000000000000000010000
             5 =>       prL sec |987666565410000000000001110000001211000000000002789774100000
             6 =>       SUB sec |LLLL
             7 =>       P_3 acc |eeeeeeeeee bbeeebbbebbbbebeeee b bbebbebb eebeebe eee eebbeb|
             8 =>       PHD acc |988787787630066600060000606667515007007005760671847885760160
             9 =>       Rel acc |979685546222352421667053233245604127749164753790316552446141
             0 =>       SUB acc |eeeeeeeee
             types of PHD output, like 1 for 'PHD sec', 2 for 'Rel sec' etc.
<b>Usage     </b> : &open_phd_files(\$file_name, $options,,,,,);
<b>Version   </b> : 1.6
<b>Warning   </b> : All the spaces are converted to '_'
</pre><hr>

<H3><A href="open_swissprot_seq_files
.pl.html">open_swissprot_seq_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_swissprot_seq_files
.pl">open_swissprot_seq_files
.pl</A>
<pre>
<b>Function  </b> : open swiss files and puts ONLY the sequences in a hash(s)
<b>Keywords  </b> : open_swiss_seq_files, open_swiss_seq, read_swissprot_seq_files,
            read_swiss_seq, get_swissprot_seq, take_swissprot_seq,
<b>Options   </b> : 'v' for STDOUT printout as well.
<b>Version   </b> : 1.2
<b>Warning   </b> : ONLY the seq.
</pre><hr>

<H3><A href="open_clu_files
.pl.html">open_clu_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_clu_files
.pl">open_clu_files
.pl</A>
<pre>
<b>Example   </b> : Clu file eg)

  Cluster 7360103
    1  1 SLL1058         7-255       2   Origin: 3   736   Sub:3
    1  1 MJ0422          17-283      2   Origin: 3   736   Sub:3
    1  1 HI1308          3-245       2   Origin: 3   736   Sub:3

<b>Keywords  </b> : open_cluster_files,
<b>Options   </b> : _  for debugging.
             #  for debugging.
             b  for to get just names ($simple_clu_reading)
             r  for adding ranges in the names
             U  for makeing sequence names upppercase

<b>Returns   </b> : a ref of hash of $clus{"$clus_size\-$id"}.=$m."\n";
             Actual content:
             3-133 => 'HI00111 HI00222 MG1233 '
<b>Usage     </b> : %clus=%{&open_clu_files(\$input)};
<b>Version   </b> : 1.9
<b>Warning   </b> : You MUST NOT delete '# options : ..' entry
              as it is read  by various subroutines.
              This automatically converts lower to upper letters
</pre><hr>

<H3><A href="open_msf_files
.pl.html">open_msf_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_msf_files
.pl">open_msf_files
.pl</A>
<pre>
<b>Argument  </b> : (\$inputfile1, \$inputfile2, .... )};
<b>Function  </b> : open msf files and put sequences in a hash(s)
<b>Options   </b> : 
   $no_gap_char_included=n by n  ## to remove gaps noted by '.'
   $reverse_seq=r by r
   $produce_seq_oder_info=o by o
<b>Returns   </b> : (*out, *out2)  or (@out_array_of_refs)
<b>Usage     </b> : (*out, *out2) = @{&open_msf_files(\$inputfile1, \$inputfile2)};
             : %hash_seq = %{&open_msf_files(\$inputfile1)};
             : (@out)        = @{&open_msf_files(\$inputfile1, \$inputfile2)};
             ---------- Example of MSF ---
             PileUp

             MSF:   85  Type: P    Check:  5063   ..

<b>Version   </b> : 1.7
</pre><hr>

<H3><A href="open_hmmls_files
.pl.html">open_hmmls_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_hmmls_files
.pl">open_hmmls_files
.pl</A>
<pre>
<b>Function  </b> : hmmls matches the full length model to target seq. while, hmmfs
             does for fragments as well.
<b>Options   </b> : 
   t=$thresh  for bits score threshold
   e=$evalue_thresh  for bits score threshold
    r for adding ranges
    m for making MSP file format output
    E=Enguiry_name    for specifying enquiry seq name rather than 'HMM', the default
<b>Usage     </b> : %out=%{&open_hmmls_files(\@file)};
<b>Version   </b> : 1.5
</pre><hr>

<H3><A href="open_hmmfs_files
.pl.html">open_hmmfs_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_hmmfs_files
.pl">open_hmmfs_files
.pl</A>
<pre>
<b>Options   </b> : 
   "t=$thresh"  for bits score threshold
<b>Usage     </b> : %out=%{&open_hmmfs_files(\@file, "t=$thresh", $attch_ranges)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="open_seq_files
.pl.html">open_seq_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_seq_files
.pl">open_seq_files
.pl</A>
<pre>
<b>Example   </b> : %out = %{&open_seq_files(@ARGV)};
                    while @ARGV at prompt was: 'pdb_40.seq'
             %seq=%{&open_seq_files(@ARGV, '1cgpa_140-197')};
                    to fetch 1cgbA but in range of 140-197 only
<b>Function  </b> : open seq files and put sequences in a hash
             seq sequence file format is like this;

 1l94   162 MNIFEMLRIDEGLRLKIYKDTEGYYTIGIGHLLTKSPSLNAAKSELDKAIGRTFRTGTWDAYK
 1lye   162 MNIFEMLRIDEGLRLKIYKDTEGYYTIGIGHLLTKSPSLNAAKSELDKAIGRTFRTGTWDAYK
 1lyj   162 MNIFEMLRIDEGLRLKIYKDTEGYYTIGIGHLLTKSPSLNAAKSELDKAIGRTFRTGTWDAYK
 1mngA  203 PYPFKLPDLGYPYEALEPHIDAKTMEIHHQKHHGAYVTNLNAALEKYPYLHGVLNWDVAEEFFKKA

             This can also return the sizes of sequences rather than seqs.
<b>Keywords  </b> : open_pdbs_files
<b>Options   </b> : any digit for the minimum seq length
<b>Usage     </b> : %seq=%{&open_seq_files($tim_seq_file, ['MJ0084'], [15] )};
             if you put additional seq name as MJ0084 it will
             fetch that sequence only in the database file.
             Any digit will be used as minimum seq size to be fetched.
<b>Version   </b> : 1.6
</pre><hr>

<H3><A href="open_sso_files
.pl.html">open_sso_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_sso_files
.pl">open_sso_files
.pl</A>
<pre>
<b>Example   </b> : 
  717    0         0.343  16    373    EC1260_16-373              74    434    YBL6_YEAST_74-434
  348    9e-16     0.500  113   233    EC1260_113-233             27    146    YDBG_ECOLI_27-146
  472    2.9e-08   0.271  13    407    EC1260_13-407              148   567    YHJ9_YEAST_148-567
  459    1.9e-22   0.260  1     407    EC1260_1-407               65    477    YLQ6_CAEEL_65-477
  452    4.5e-14   0.275  1     407    EC1260_1-407               103   537    YSCPUT2_103-537
  1131   0         0.433  1     407    EC1260_1-407               112   519    ZMU43082_112-519

  Input SSO file example)-> below

   >>MG032 ATP-dependent nuclease (addA) {Bacillus subtilis  (666 aa)
    Z-score: 88.3 expect()  1.9
   Smith-Waterman score: 77;  27.143% identity in 70 aa overlap

           30        40        50        60        70        80
   MJ0497 RSAGSKGVDLIAGRKGEVLIFECKTSSKTKFYINKEDIEKLISFSEIFGGKPYLAIKFNG
                                        : .. ...  . .:.:::. :: : ..:
   MG032  HDKVRYAFEVKFNIALVLSINKSNVDFDFDFILKTDNFSDIENFNEIFNRKPALQFRFYT
        200       210       220       230       240       250

           90       100             110       120       130
   MJ0497 EMLFINPFLLSTNGK------NYVIDERIKAIAIDFYEVIGRGKQLKIDDLI
          .   ::   :: ::.      : ....... . ::. . :
   MG032  K---INVHKLSFNGSDSTYIANILLQDQFNLLEIDLNKSIYALDLENAKERFDKEFVQPL
        260          270       280       290       300       310

 Parseable form -m 10 option =========================================
   >>>MJ0497.fa, 133 aa vs GMG.fa library
   ; pg_name: Smith-Waterman (PGopt)
   ; pg_ver: 3.0 June, 1996
   ; pg_matrix: BL50
   ; pg_gap-pen: -12 -2
   >>MG032 ATP-dependent nuclease (addA) {Bacillus subtilis
   ; sw_score:  77
   ; sw_z-score: 88.3
   ; sw_expect    1.9
   ; sw_ident: 0.271
   ; sw_overlap: 70
   >MJ0497 ..
   ; sq_len: 133
   ; sq_type: p
   ; al_start: 58
   ; al_stop: 121
   ; al_display_start: 28

<b>Function  </b> : This reads the parseable( -m 10 option)
              and non-parseable form of ssearch program output
             If you give 5 files, it produces 5 hashes as a ref of array.
             This understands xxxx.gz files.
             This reads FASTA -m 10 output, too.
<b>Keywords  </b> : open_ssearch_output_files, ssearch_output, ssearch, FASTA,
<b>Options   </b> : _  for debugging.
             #  for debugging.
             u= for upper E value limit
             l= for lower E value limit
             r  for attaching ranges to out seq names (eg> HI0001_1-20 as a key)
             U  for making the matched seqname to upppercase
             L  for making the matched seqname to lowercase
             R  for attaching ranges to out seq names for both TARGET and MATCH
             n  for new format (msp2)
             a  for getting alignments of the pair

<b>Usage     </b> : @sso=@{&open_sso_files(@file, $add_range, $add_range2, "u=$upper_expect_limit",
                               "l=$lower_expect_limit", "m=$margin", $new_format)};
<b>Version   </b> : 4.5
<b>Warning   </b> : By default, the SW score comes to the first
             If expect value is not found, it becomes '0'
             By default, the offset of seq match with a seq name like seq_30-40
               will be 30 not 1.
             It ignores special chars like , : .prot in the name (eg, AADF_FASDF: will be AADF_FASDF)
</pre><hr>

<H3><A href="open_msp_files
.pl.html">open_msp_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_msp_files
.pl">open_msp_files
.pl</A>
<pre>
<b>Example   </b> : Example output(with 'n' opt):
   d1bi6h1         d1bi6h1_1-24     IBR1_ANACO_20-42  IBR2_ANACO_19-42
   e1bi6.1h1       IBR1_ANACO_38-52 e1bi6.1h1_1-18    IBR2_ANACO_38-52

<b>Function  </b> : opens Erik Sonhammer's MSPcrunch file output(default).
             This looks up xxxxx.fa files in the pwd (with S opt) and see
             if it can get the sequences as well.
             With 'n' option you can just get the matched sequence
              names with ranges.
<b>Keywords  </b> : exchange_msp_file_columns,
<b>Options   </b> : 
          s -s  for size return only
          S -S  for the sequences are fetched if equivalent xxxx.fa files are in pwd
          n -n  for matched seq NAMEs with ranges only (eg: HI0001_1-12,,), hash ref is out
          R     for NO range attachment in Name only return option (n)
          e=    for evalue threshhold, if e=1, ignores all which are over 1
          t=    for score threshhold if t=100, ignores all which are less 100
          l=    for match length threshold.
          x     for exchange query with matched seqs. eg)      12 0.09 1 30 QUERY  1 29 MATCH
                                                       becomes 12 0.09 1 30 MATCH  1 29 QUERY
                This returns the same lines as input only with exchanged query and match seqs

<b>Usage     </b> : %seq=%{&open_msp_files(@file, $names_only)};
<b>Version   </b> : 2.8
</pre><hr>

<H3><A href="open_dssp_files
.pl.html">open_dssp_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_dssp_files
.pl">open_dssp_files
.pl</A>
<pre>
<b>Argument  </b> : files names like (6taa, 6taa.dssp) If you put just '6taa' without extension, it
             searches if there is a '6taa.dssp' in both PWD and $DSSP env. set directory.
             ---------- Example of dssp ---
             **** SECONDARY STRUCTURE DEFINITION BY THE PROGRAM DSSP, VERSION JUL
             REFERENCE W
             HEADER    RIBOSOME-INACTIVATING PROTEIN           01-JUL-94   1MRG
             COMPND    ALPHA-MOMORCHARIN COMPLEXED WITH ADENINE
             SOURCE    BITTER GOURD (CUCURBITACEAE MOMORDICA CHARANTIA) SEEDS
             AUTHOR    Q
             246  1  0  0  0 TOTAL NUMBER OF RESIDUES, NUMBER OF CHAINS, NUMBER OF SS-BRIDGES(TOTAL,INTRACHAIN,INTERCHAIN)                .
             112 95.0   ACCESSIBLE SURFACE OF PROTEIN (ANGSTROM**2)                                                                         .
             171 69.5   TOTAL NUMBER OF HYDROGEN BONDS OF TYPE O(I)-->H-N(J)  , SAME NUMBER PER 100 RESIDUES                              .
             12   4.9   TOTAL NUMBER OF HYDROGEN BONDS IN     PARALLEL BRIDGES, SAME NUMBER PER 100 RESIDUES                              .
             36  14.6   TOTAL NUMBER OF HYDROGEN BONDS IN ANTIPARALLEL BRIDGES, SAME NUMBER PER 100 RESIDUES                              .
             1    0.4   TOTAL NUMBER OF HYDROGEN BONDS OF TYPE O(I)-->H-N(I-5), SAME NUMBER PER 100 RESIDUES                              .
             1    0.4   TOTAL NUMBER OF HYDROGEN BONDS OF TYPE O(I)-->H-N(I-4), SAME NUMBER PER 100 RESIDUES                              .
             74  30.1   TOTAL NUMBER OF HYDROGEN BONDS OF TYPE O(I)-->H-N(I+4), SAME NUMBER PER 100 RESIDUES                              .
             5    2.0   TOTAL NUMBER OF HYDROGEN BONDS OF TYPE O(I)-->H-N(I+5), SAME NUMBER PER 100 RESIDUES                              .
             1    2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30     *** HISTOGRAMS OF ***           .
             0    0  0  0  1  1  0  2  0  0  1  0  0  1  0  0  0  0  0  2  0  0  0  0  0  0  0  0  0  0    RESIDUES PER ALPHA HELIX         .
             1    0  0  2  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0    PARALLEL BRIDGES PER LADDER      .
             2    0  1  2  0  1  1  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0    ANTIPARALLEL BRIDGES PER LADDER  .
             2    0  0  0  1  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0    LADDERS PER SHEET                .
             #   RESIDUE AA STRUCTURE BP1 BP2  ACC   N-H-->O  O-->H-N  N-H-->O  O-->H-N    TCO  KAPPA ALPHA  PHI   PSI    X-CA   Y-CA   Z-CA
             1    1   D              0   0  132    0, 0.0   2,-0.3   0, 0.0  49,-0.2   0.000 360.0 360.0 360.0 153.4   44.0   96.9  -23.8
             2    2   V  E     -a   50   0A  10   47,-1.5  49,-2.8   2, 0.0   2,-0.3  -0.889 360.0-163.3-115.9 151.4   43.1  100.4  -22.5
             3    3   S  E     -a   51   0A  63   -2,-0.3   2,-0.3  47,-0.2  49,-0.2  -0.961  10.3-172.8-131.0 152.3   44.8  103.7  -23.4
             4    4   F  E     -a   52   0A   8   47,-2.2  49,-2.3  -2,-0.3   2,-0.4  -0.985   6.9-161.2-143.2 139.5   45.0  107.2  -22.0
             5    5   R  E     -a   53   0A 144   -2,-0.3   4,-0.2  47,-0.2  49,-0.2  -0.993   9.7-156.0-121.0 125.9   46.6  110.2  -23.6
             6    6   L  S    S+     0   0    1   47,-2.3   2,-0.5  -2,-0.4   3,-0.4   0.644  73.2  90.9 -73.3 -22.4   47.5  113.2  -21.4
             7    7   S  S    S+     0   0   81   47,-0.3   3,-0.1   1,-0.2  -2,-0.1  -0.695 106.0   5.2 -75.5 121.0   47.4  115.6  -24.4
             8    8   G  S    S+     0   0   72   -2,-0.5  -1,-0.2   1,-0.3   5,-0.1   0.269  97.6 147.8  90.2 -10.7   43.9  117.0  -24.7
             9    9   A        +     0   0   10   -3,-0.4  -1,-0.3  -4,-0.2  -3,-0.1  -0.256  16.8 166.8 -58.8 142.4   42.9  115.2  -21.5
             (\$inputfile1, \$inputfile2, .... )};
<b>Function  </b> : open dssp files and put sequences in a hash(s)
              It can take options for specific secondary structure types. For example,
              if you put an option $H in the args of the sub with the value of 'H'
              open_dssp_files will only read secondary structure whenever it sees 'H'
              in xxx.dssp file ignoring any other sec. str. types.
              If you combine the options of 'H' and 'E', you can get only Helix and long
              beta strand sections defined as segments. This is handy to get sec. str. segments
              from any dssp files to compare with pdb files etc.
             With 'simplify' option, you can convert only all the 'T', 'G' and 'I' sec. to
              'H' and 'E'.
<b>Options   </b> : H, S, E, T, I, G, B, P, C, -help
 $H        =        'H' by   -H or -h or H or h  # to retrieve 4-helix (alpha helical)
 $S        becomes  'S' by   -S or -s or S or s  # to retrieve Extended strand, participates in B-ladder
 $E        becomes  'E' by   -E or -e or E or e  # to retrieve residue in isolated Beta-bridge
 $T        becomes  'T' by   -T or -t or T or t  # to retrieve H-bonded turn
 $I        becomes  'I' by   -I or -i or I or i  # to retrieve 5-helix (Pi helical) segment output
 $G        becomes  'G' by   -G or -g or G or g  # to retrieve 3-helix (3-10 helical)
 $B        becomes  'B' by   -B or -b or B or b  # to retrieve only B segment
 $simplify becomes   1  by   -p or P or -P, p
 $comm_col becomes  'c' by   -c or c or C or -C or common
 $HELP     becomes   1  by   -help   # for showing help

<b>Returns   </b> : (*out, *out2)  or (@out_array_of_refs)
<b>Usage     </b> : (*out, *out2) = @{&open_dssp_files(\$inputfile1, \$inputfile2, \$H, \$S,,,,)};
             (@out)        = @{&open_dssp_files(\$inputfile1, \$inputfile2, \$H, \$S,,,,)};
<b>Version   </b> : 2.9
$debug feature has been added to make it produce error messages with '#' option.
<b>Warning   </b> : 6taa.dssp  and 6taa are regarded as the same.
</pre><hr>

<H3><A href="open_dna_files
.pl.html">open_dna_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_dna_files
.pl">open_dna_files
.pl</A>
<pre>
<b>Argument  </b> : (\$inputfile1, \$inputfile2, .... )};
<b>Function  </b> : open dna files and put sequences in a hash(s)
<b>Returns   </b> : (@out_array_of_refs)
<b>Usage     </b> : ($out, $out2) = @{&open_dna_files(\$inputfile1, \$inputfile2)};
             : (@out)        = @{&open_dna_files(\$inputfile1, \$inputfile2)};
             ---------- Example of dna file --- dna files are genbank file format


             1 ggatcttgct gaatacatgg tggcacaatt gaaattagat ccgcgaattt
               tcatcaaaac
             61 agcgggatta tggtcaacaa atccgtaaaa atgaaaagcc tgtcttgcga
               caggcttttt
             121 tatttgaatg taatcctcac tggtaaacgt ttaacgccaa agacaaaggg
               actagggatc
             181 gcttcaagct tttcatcatg agcagctttt tcgatacaag ctgacattga

</pre><hr>

<H3><A href="open_tem_files
.pl.html">open_tem_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_tem_files
.pl">open_tem_files
.pl</A>
<pre>
<b>Argument  </b> : (\$inputfile1, \$inputfile2, .... )};
<b>Function  </b> : opens JPO's xxxx.tem file, stores in 5 hashes. (usually one tem file)
<b>Options   </b> : -n, n, or N for removing any gaps in the sequences.
             -s, s, or S for getting only the sequences.
<b>Returns   </b> : ($r1, $r2, $r3, $r4, $r5) <= these are references for hashes.
<b>Usage     </b> : ($r1, $r2, $r3, $r4, $r5)=&open_tem_files(\$infile1, \$inputfile2..)};
             ---------- Example of xxxx
             >P1;1cdg
             sequence
             APDTSVSNKQNFSTDVIYQIFTDRFSDGNPANNPTGAAFDGTCTN-LRLYCGGDWQGIINKINDGYLTGMGVTAI
             >P1;1cdg
             secondary structure and phi angle
             CCCCCCCCCCCCCCCCEEECCHHHHCCCCHHHCCCPHHCCCCPCC-CCCCCPCCHHHHHHHHHCPHHHHHPCCEE
             >P1;1cdg
             solvent accessibility
             TTTTTTTTTTTFFFFFFFFFFFFFFTTTTTTTTTTTTTTTTTFTT-TTTTFFFFFTFFTTTFTTTFFTTFTFTFF
             >P1;1cdg
             DSSP
             CCCCCCCCCCCCCCCCEEECCHHHHCCCCGGGCCCGGGCCCCCCC-CCCCCCCCHHHHHHHHHCCHHHHHCCCEE
             >P1;1cdg
             percentage accessibility
             67523272360000000000000002213792129b722248085-14110000030015105660028040200
             2ltn           ----TETTSFLITKFSPDQQNLIFQGDGYTT-KEKLTLTK------AVKNTVGRALYSSP
             1loe           ----TETTSFSITKFGPDQQNLIFQGDGYTT-KERLTLTK------AVRNTVGRALYSSP

             2ltn           ----CEEEEEEECCCCCCCCCEEEEPCCEEP-PPCEEEEC------CCCPCEEEEEECCC
             1loe           ----CEEEEEEECCCCCCCCCEEEEPCCEEE-PPEEEEEC------CCCPCEEEEEECCC

             2ltn           ----TTTTTTTTTTFTTTTTTFTTTTTFTFT-TTTFTFFT------TTTTTTFFFFTTTT
             1loe           ----TTTTTTTTTTFTTTTTTFTTTTTFTFT-TTTFFFFT------TTTTTTFFFFTTTT

             2ltn           ----CEEEEEEECCCCCCCCCEEEEECCEEC-CCCEEEEC------CCCCCEEEEEECCC
             1loe           ----CEEEEEEECCCCCCCCCEEEEECCEEE-CCEEEEEC------CCCCCEEEEEECCC

             2ltn           ----543251b16504681c50422650502-75201006------35681200001453
             1loe           ----6532e1508a07981b50422750404-8a200006------36672200001453
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="open_hlx_files
.pl.html">open_hlx_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_hlx_files
.pl">open_hlx_files
.pl</A>
<pre>
<b>Function  </b> : 
             Example of hlx file (For Bo Nielson)
             Residue Frame Score Probability
             1 M   a  1.00563E+00 2.05479E-03
             2 T   b  1.01814E+00 2.52053E-03
             3 R   c  1.01814E+00 2.52053E-03
<b>Returns   </b> : list of ref. for hash(es)
</pre><hr>

<H3><A href="open_jp_files
.pl.html">open_jp_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_jp_files
.pl">open_jp_files
.pl</A>
<pre>
<b>Function  </b> : reads jp files and stores results in a hash.
<b>Returns   </b> : a reference of a hash for names and  their sequences.
<b>Usage     </b> : %out_hash=%{&open_jp_files(\$file_name)};
<b>Version   </b> : 1.1
<b>Warning   </b> : All the spaces  '-' !!!
</pre><hr>

<H3><A href="open_ali_files
.pl.html">open_ali_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_ali_files
.pl">open_ali_files
.pl</A>
<pre>
<b>Function  </b> : open fasta files and put sequences in a hash
             FASTA sequence file format is like this;

             >P1;1abp
             structureX:1abp:   1 : : 306 : :L-arabinose-binding protein:Escherichia coli: 2.40:-1.00
             ENLKLGFLVKQPEEPWFQTEWKFADKAGKDLG-FEVIKIAV-PDGEKTLNAIDSLAASGAKGFVICTPDPKLGSA
             TEGQGFKAADIIGIGINGVDAVSELSKAQATGFYGSLLPSPDVHGYKSSEMLYNWVAK--------DVEPPKFTE
             VTDVVLITRDNFKEELEKKGLGGK*
             >P1;2gbp
             structureX:2gbp:   1 : : 309 : :D-galactose/D-glucose-bind:Escherichia coli: 1.90:14.60
             ADTRIGVTIYKYDDNFMSVVRKAIEQDAKAAPDVQLLMNDSQNDQSKQNDQIDVLLAKGVKALAINLVDPAAAGT
             LKAHNKS-SIP-VFGVDA--LPEALALVKSGALAGTVLNDANNQAKATFDLAKNLADGKGAADGTNWKIDNKVVR
             VP-YVGVDKDNLAEFSKK------*

<b>Usage     </b> : %anyhash = %{&open_ali_files(\$filename)};
</pre><hr>

<H3><A href="open_pir_files
.pl.html">open_pir_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_pir_files
.pl">open_pir_files
.pl</A>
<pre>
<b>Function  </b> : open fasta files and put sequences in a hash
             FASTA sequence file format is like this;

             >P1;1abp
             structureX:1abp:   1 : : 306 : :L-arabinose-binding protein:Escherichia coli: 2.40:-1.00
             ENLKLGFLVKQPEEPWFQTEWKFADKAGKDLG-FEVIKIAV-PDGEKTLNAIDSLAASGAKGFVICTPDPKLGSA
             VTDVVLITRDNFKEELEKKGLGGK*
             >P1;2gbp
             structureX:2gbp:   1 : : 309 : :D-galactose/D-glucose-bind:Escherichia coli: 1.90:14.60
             LKAHNKS-SIP-VFGVDA--LPEALALVKSGALAGTVLNDANNQAKATFDLAKNLADGKGAADGTNWKIDNKVVR
             VP-YVGVDKDNLAEFSKK------*

<b>Usage     </b> : %anyhash = &open_pir_files($any_sequence_file_fasta_form);
<b>Version   </b> : 1.2
</pre><hr>

<H3><A href="open_aln_files
.pl.html">open_aln_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_aln_files
.pl">open_aln_files
.pl</A>
<pre>
<b>Function  </b> : reads CLUSTALW aln files and stores results in a hash.
<b>Returns   </b> : a reference of a hash for names and  their sequences.
<b>Usage     </b> : %out_hash=%{&open_aln_files(\$file_name)};
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="open_seq_alignment_files
.pl.html">open_seq_alignment_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_seq_alignment_files
.pl">open_seq_alignment_files
.pl</A>
<pre>
<b>Argument  </b> : (\$inputfile1, \$inputfile2, .... )};
<b>Function  </b> : open various sequence alignment files and put sequences in a hash(s)
<b>Returns   </b> : (*out, *out2)  or (@out_array_of_refs)
<b>Usage     </b> : (*out, *out2) = @{&open_seq_alignment_files(\$inputfile1, \$inputfile2)};
           : %hash_seq = %{&open_seq_alignment_files(\$inputfile1)};
           : (@out)        = @{&open_seq_alignment_files(\$inputfile1, \$inputfile2)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="open_sst_files
.pl.html">open_sst_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_sst_files
.pl">open_sst_files
.pl</A>
<pre>
<b>Argument  </b> : a ref. for scaler of "jp file name"
<b>Example   </b> : jp file  ==  seq1 ABDSF--DSFSDFS   <- true sequence
                              seq2 lkdf-jlsjlsjf

                 sst files == seq1.sst, seq2.sst

                 output hash == seq1 hHHHHHHHttEEEEEEEE
                                seq2 hHHHHHHHHHEEEEEEhh

<b>Function  </b> : gets the name of a file(jp file) with its absolute dir path
             reads the sequence names in the jp file and looks up all
             the sst files in the same directory. Puts sst sequences
             in a hash with keys of sequence names.

<b>Returns   </b> : a ref. for a hash
<b>Usage     </b> : %out_sst_hash =%{&open_sst_files(\$jp_file_dir_and_name)};
<b>Warning   </b> : $jp_file_dir_and_name should be absolute dir and file name
</pre><hr>

<H3><A href="read_sst_files
.pl.html">read_sst_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//read_sst_files
.pl">read_sst_files
.pl</A>
<pre>
<b>Argument  </b> : a ref. for scaler of "jp file name"
<b>Example   </b> : jp file  ==  seq1 ABDSF--DSFSDFS   <- true sequence
                              seq2 lkdf-jlsjlsjf

                 sst files == seq1.sst, seq2.sst

                 output hash == seq1 hHHHHHHHttEEEEEEEE
                                seq2 hHHHHHHHHHEEEEEEhh

<b>Function  </b> : gets the name of a file(jp file) with its absolute dir path
             reads the sequence names in the jp file and looks up all
             the sst files in the same directory. Puts sst sequences
             in a hash with keys of sequence names.

<b>Returns   </b> : a ref. for a hash
<b>Usage     </b> : %out_sst_hash =%{&read_sst_files(\$jp_file_dir_and_name)};
<b>Warning   </b> : $jp_file_dir_and_name should be absolute dir and file name
</pre><hr>

<H3><A href="open_slx_files
.pl.html">open_slx_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_slx_files
.pl">open_slx_files
.pl</A>
<pre>
<b>Argument  </b> : takes one ref. for a file.
<b>Example   </b> : selex file (foo.slx) looks like this:

         #=SQ GLB_TUBTU  5.9393 - - 0..0::0 -
         #=SQ GGZLB      20.9706 - - 0..0::0 -
         #=RF        x.....x.xxxx.xxx.xxxxxx....xxxxxxxxxxxxxxx.xxxx
         HAHU        ......VLSPADKTNVKAAWGKVGA......HAGEYGAEALERMFLS
         HBA3_PANTR  ......VLSPADKTNVKAAWGKVGA......HAGZYGAEALERMFLS

<b>Function  </b> : open slx files and put sequences in a hash
<b>Returns   </b> : a ref. of a hash
<b>Usage     </b> : %anyarray = &open_slx_files(\$any_sequence_file_slx_form);
<b>Version   </b> : 1.0
<b>Warning   </b> : The slx FORMAT SHOULD BE AT LEAST 30 residue long
</pre><hr>

<H3><A href="open_out_files
.pl.html">open_out_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//open_out_files
.pl">open_out_files
.pl</A>
<pre>
<b>Argument  </b> : takes one ref. for a file.
             >>Out file looks like this===>

             3aat         mfe   aapadp----adlfraderpGk   gigvY--etgktpvltS
             1ama       sswwshvemgppdp  krdtns--kkMnLG---YrddngkpyvLnC-

<b>Function  </b> : open out files and put their sequences in a hash
<b>Returns   </b> : a ref. of a hash
             Output example in a hash(fills the space)

             3aat       --mfe---aapadp----adlfraderpGk---gigvY--etgktpvltS
             1ama       ---eamiaakkmdkeylpiaGladFtraSA----eAfksgryVTV

<b>Usage     </b> : %anyarray = &open_out_files(\$any_out_file);
<b>Warning   </b> : well tested. It skips lines starting with blank, lines with '-' in them.
</pre><hr>

<H3><A href="package
.pl.html">package
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//package
.pl">package
.pl</A>
<pre>
<b>Function  </b> : Roman.pm : Roman <-> Arabic conversion package
<b>Version   </b> : 1.0
<b>Warning   </b> : From: ozawa@prince.pe.u-tokyo.ac.jp (OZAWA Sakuro)
</pre><hr>

<H3><A href="time_date
.pl.html">time_date
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//time_date
.pl">time_date
.pl</A>
<pre>
<b>Function  </b> : returns current time & date as 05/15/95 23:22:41
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="sep
.pl.html">sep
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//sep
.pl">sep
.pl</A>
<pre>
<b>Function  </b> : separater. \n#________________________________\n
<b>Keywords  </b> : separating_line
<b>Usage     </b> : &sep;
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="diff_dates
.pl.html">diff_dates
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//diff_dates
.pl">diff_dates
.pl</A>
<pre>
<b>Function  </b> : gets number of days between two dates ( "05/15/94" )
<b>Usage     </b> : $output = &diff_dates("05/15/1994", "05/15/1995")
<b>Version   </b> : 1.0
<b>Warning   </b> : modified (originally from reb@serf.nsc.com (Edward Brown))
</pre><hr>

<H3><A href="fromJulian
.pl.html">fromJulian
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//fromJulian
.pl">fromJulian
.pl</A>
<pre>
<b>Example   </b> : print &fromJulian(34469), "\n";
<b>Function  </b> : taking the days between two dates.
<b>Version   </b> : 1.0
<b>Warning   </b> : got from reb@serf.nsc.com (Edward Brown)
             require "julian
             $Value1 = &toJulian("05/15/1994");        # Assign $Value1 a Julian Day
             print "$Value1\n";
             $Value2 = &toJulian("05/20/1994");        # Assign Value2 a Julian Day
             print "$Value2\n";
             $Days = $Value2 - $Value1;              #Difference in Days
             print "$Days\n";
             print &fromJulian(34469), "\n";         # Give a Julian Day, give the date
             print &fromJulian(34474), "\n";
             What is the Date 25 Days from Today?  (You can get format from `date`)

             $Value = &toJulian("05/16/1995");
             $Value +=  25;
             print &fromJulian($Value), "\n";
</pre><hr>

<H3><A href="toJulian
.pl.html">toJulian
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//toJulian
.pl">toJulian
.pl</A>
<pre>
<b>Example   </b> : $Value1 = &toJulian("05/15/94"); print "$Value1\n";
<b>Function  </b> : taking the days between two dates.
<b>Version   </b> : 1.0
<b>Warning   </b> : got from reb@serf.nsc.com (Edward Brown)
</pre><hr>

<H3><A href="opendir_and_go
.pl.html">opendir_and_go
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//opendir_and_go
.pl">opendir_and_go
.pl</A>
<pre>
<b>Example   </b> : as in my 'indexing.pl' for perl file indexer.
<b>Function  </b> : open dir and process all files if you wish, and then go in any sub
             dir of it. Using recursion. created by A Biomatic
             if any file is linked, it skips that file.
<b>Usage     </b> : &opendir_and_go($input_dir); #$inputdir='/nfs/ind4/ccpe1/people/A Biomatic /jpo/align';
<b>Version   </b> : 1.0
<b>Warning   </b> : Seems to work fine.
</pre><hr>

<H3><A href="occurances
.pl.html">occurances
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//occurances
.pl">occurances
.pl</A>
<pre>
<b>Function  </b> : this is for sort, to sort things according to the higher num. of occu.
<b>Usage     </b> : sort occurances (@any_array_with_repeating_element);
<b>Version   </b> : 1.0
<b>Warning   </b> : This is from 21 DAYS book, page 373.
</pre><hr>

<H3><A href="extract_ori_seq
             nt5
.pl.html">extract_ori_seq
             nt5
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//extract_ori_seq
             nt5
.pl">extract_ori_seq
             nt5
.pl</A>
<pre>
<b>Function  </b> : extract seqs. which are from struc. alignment only. to be analysed.
             after mul. alignment with added seq. you can extract original str.
             sequ. by using this. The output always has ...msff  ext.
             *array_ali is the JPO's or true alignment hash.
<b>Usage     </b> : &extract_ori_seq($input_file, $output_file, $out_seq_no, *array2);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_pair_homol_array
.pl.html">get_pair_homol_array
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_pair_homol_array
.pl">get_pair_homol_array
.pl</A>
<pre>
<b>Function  </b> : get pair wise seq. !! Number of pair identical residues.
<b>Usage     </b> : $hom_out_count = ${&get_pair_homol_array(\@any_array_of_2_elem)};= @ar=(ABCDE..., CDEGA..)
<b>Version   </b> : 1.0
<b>Warning   </b> : reliable, but input seq. strings shouldn't contain spaces.
</pre><hr>

<H3><A href="get_percent_homol_arr
.pl.html">get_percent_homol_arr
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_percent_homol_arr
.pl">get_percent_homol_arr
.pl</A>
<pre>
<b>Function  </b> : get pair wise seq. identity of any two strings, outputs a scalar (%)
<b>Usage     </b> : $homology_out = ${&get_pair_homol(\@any_array_of_2_elem)};= @ar=(ABCDE..., CDEGA..)
<b>Version   </b> : 1.0
<b>Warning   </b> : reliable, but input seq. strings shouldn't contain spaces.
</pre><hr>

<H3><A href="get_pair_homol_hash
.pl.html">get_pair_homol_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_pair_homol_hash
.pl">get_pair_homol_hash
.pl</A>
<pre>
<b>Function  </b> : get pair wise seq. identity as a scalar count
<b>Usage     </b> : $homology_out = & get_pair_homol (%any_hash); , eg) %hash = (name1, ABCDE..., name2, CDEGA..)
<b>Version   </b> : 1.0
<b>Warning   </b> : reliable, but input seq. strings shouldn't contain spaces.
</pre><hr>

<H3><A href="get_percent_homo_hash
.pl.html">get_percent_homo_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_percent_homo_hash
.pl">get_percent_homo_hash
.pl</A>
<pre>
<b>Function  </b> : get pair wise seq. identity(%) of any two strings put in as a hash
<b>Usage     </b> : $homology_out = &get_pair_homol_hash(%any_hash); , eg) %hash = (name1, ABCDE..., name2, CDEGA..)
<b>Version   </b> : 1.0
<b>Warning   </b> : reliable, but input seq. strings shouldn't contain spaces.
</pre><hr>

<H3><A href="file_size
.pl.html">file_size
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//file_size
.pl">file_size
.pl</A>
<pre>
<b>Function  </b> : returns the size of any single testing file
<b>Usage     </b> : $outputfilesize = &file_size($input_file_name);
<b>Version   </b> : 1.0
<b>Warning   </b> : Q is for quality of this sub. This can't be wrong.
</pre><hr>

<H3><A href="seq_comp_percent2
.pl.html">seq_comp_percent2
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//seq_comp_percent2
.pl">seq_comp_percent2
.pl</A>
<pre>
<b>Function  </b> : get string seq COMPOSITION identities(a to z). gets array
             of strings and outs array of % numbers
<b>Usage     </b> : @outarray = &seq_comp_percent2(@any_input_string_array);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_full_file_name
.pl.html">get_full_file_name
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_full_file_name
.pl">get_full_file_name
.pl</A>
<pre>
<b>Function  </b> : returns full directory path (= pwd ), eg.  /nfs/ind4/ccpe1/people/A Biomatic
<b>Keywords  </b> : get_long_path_name, get_complete_path_name
<b>Usage     </b> : $any_path = ${&get_full_dir_path($any_directory)}; or &dir_path('.') for pwd.
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="dir_path
.pl.html">dir_path
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//dir_path
.pl">dir_path
.pl</A>
<pre>
<b>Function  </b> : returns directory path (= pwd ), eg.  /nfs/ind4/ccpe1/people/A Biomatic
<b>Usage     </b> : $any_path = &dir_path($any_directory); or &dir_path('.') for pwd.
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="full_pwd_path
.pl.html">full_pwd_path
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//full_pwd_path
.pl">full_pwd_path
.pl</A>
<pre>
<b>Function  </b> : returns full directory path (= pwd ), eg.  /nfs/ind4/ccpe1/people/A Biomatic
<b>Keywords  </b> : get_long_path_name, get_complete_path_name
<b>Usage     </b> : $any_path = ${&full_dir_path($any_directory)}; or &dir_path('.') for pwd.
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_full_pwd_path
.pl.html">get_full_pwd_path
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_full_pwd_path
.pl">get_full_pwd_path
.pl</A>
<pre>
<b>Function  </b> : returns full directory path (= pwd ), eg.  /nfs/ind4/ccpe1/people/A Biomatic
<b>Keywords  </b> : get_long_path_name, get_complete_path_name
<b>Usage     </b> : $any_path = ${&get_full_dir_path($any_directory)}; or &dir_path('.') for pwd.
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_whole_pwd_path
.pl.html">get_whole_pwd_path
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_whole_pwd_path
.pl">get_whole_pwd_path
.pl</A>
<pre>
<b>Function  </b> : returns full directory path (= pwd ), eg.  /nfs/ind4/ccpe1/people/A Biomatic
<b>Keywords  </b> : get_long_path_name, get_complete_path_name
<b>Usage     </b> : $any_path = ${&get_whole_dir_path($any_directory)}; or &dir_path('.') for pwd.
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="pwd_path
.pl.html">pwd_path
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//pwd_path
.pl">pwd_path
.pl</A>
<pre>
<b>Function  </b> : returns directory path (= pwd ), eg.  /nfs/ind4/ccpe1/people/A Biomatic
<b>Usage     </b> : $any_path = ${&dir_path($any_directory)}; or &dir_path('.') for pwd.
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_pwd_dir
.pl.html">get_pwd_dir
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_pwd_dir
.pl">get_pwd_dir
.pl</A>
<pre>
<b>Function  </b> : returns present working dir base
<b>Usage     </b> : $dir = &get_pwd_dir($anydir); # to return say,  'perl' .
<b>Version   </b> : 1.0
<b>Warning   </b> : well tested.
</pre><hr>

<H3><A href="dir_name
.pl.html">dir_name
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//dir_name
.pl">dir_name
.pl</A>
<pre>
<b>Function  </b> : returns present working dir base
<b>Usage     </b> : $dir = &pwd_dir($anydir); # to return say,  'perl' .
<b>Version   </b> : 1.0
<b>Warning   </b> : well tested.
</pre><hr>

<H3><A href="pwd_dir_name
.pl.html">pwd_dir_name
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//pwd_dir_name
.pl">pwd_dir_name
.pl</A>
<pre>
<b>Example   </b> : returns 'jong' with the input of '/nfs/ind5/A Biomatic '
<b>Function  </b> : returns present working dir name
<b>Usage     </b> : $dir = &pwd_dir($any_absolute_path_dir);
<b>Version   </b> : 1.0
<b>Warning   </b> : well tested.
</pre><hr>

<H3><A href="get_pwd_dir_name
.pl.html">get_pwd_dir_name
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_pwd_dir_name
.pl">get_pwd_dir_name
.pl</A>
<pre>
<b>Function  </b> : returns present working dir name
<b>Usage     </b> : $dir = &get_pwd_dir($any_absolute_path_dir);
<b>Version   </b> : 1.0
<b>Warning   </b> : well tested.
</pre><hr>

<H3><A href="get_full_path_dir_names
.pl.html">get_full_path_dir_names
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_full_path_dir_names
.pl">get_full_path_dir_names
.pl</A>
<pre>
<b>Example   </b> : with 'jong' it gives '/nfs/ind5/jong', '/nfs/ind4/ccep1/people/A Biomatic '...
             when 'jong' is in /nfs/ind4/jong/Perl, it returns /nfs/ind4/A Biomatic
<b>Function  </b> : returns full path dir names with given short dir names.
<b>Usage     </b> : @full_path_dirs = @{&get_full_path_dir_names(@short_dir_name)};
<b>Version   </b> : 1.0
<b>Warning   </b> : when 'jong' is in /nfs/ind4/jong/Perl, it returns /nfs/ind4/A Biomatic
</pre><hr>

<H3><A href="get_extension_names
.pl.html">get_extension_names
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_extension_names
.pl">get_extension_names
.pl</A>
<pre>
<b>Keywords  </b> : get_file_extension, get_extension, get_file_ext, get_ext_names
             get_file_extensions
<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Usage     </b> : @ext=@{&get_file_extensions(\@file)}  or
             $ext=${&get_file_extensions(\$file)}
<b>Version   </b> : 1.2
<b>Warning   </b> : You MUST NOT delete '# options : ..' entry
              as it is read  by various subroutines.
</pre><hr>

<H3><A href="get_file_extensions
.pl.html">get_file_extensions
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_file_extensions
.pl">get_file_extensions
.pl</A>
<pre>
<b>Keywords  </b> : get_file_extension, get_extension, get_file_ext, get_ext_names
             get_extension_names
<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Usage     </b> : @ext=@{&get_file_extensions(\@file)}  or
             $ext=${&get_file_extensions(\$file)}
<b>Version   </b> : 1.2
<b>Warning   </b> : You MUST NOT delete '# options : ..' entry
              as it is read  by various subroutines.
</pre><hr>

<H3><A href="get_base_names
.pl.html">get_base_names
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_base_names
.pl">get_base_names
.pl</A>
<pre>
<b>Argument  </b> : handles both ref and non-ref.
<b>Example   </b> : $base => 'test'  with 'test.txt' or '/home/dir/of/mine/text.txt'
<b>Function  </b> : produces the file base name(eg, "evalign"  out of "evalign.pl" ).
              when xxxx.xx.gz form file is given, it removes gz as well

<b>Keywords  </b> : get_base_name{, base_name, file_base_name ,  get_file_base_name
             get_basename, basename, get_root_name, base , root, get_file_root
<b>Usage     </b> : $base =${&get_base_names(\$file_name)};
             :   or @bases = &get_base_names(\@files);  # <-- uses `pwd` for abs directory
<b>Version   </b> : 1.5
</pre><hr>

<H3><A href="read_file_names_only
.pl.html">read_file_names_only
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//read_file_names_only
.pl">read_file_names_only
.pl</A>
<pre>
<b>Example   </b> : @all_files=@{&read_file_names_only(\$abs_path_dir_name, ..)};
             @all_files=@{&read_file_names_only(\$dir1, '.pl', '.txt')};
             @all_files=@{&read_file_names_only(\$dir1, '.', \$dir2, \$dir3, 'e=pl')};
             @all_files=@{&read_file_names_only(\$abs_path_dir_name, 'G1_*.txt')};
             @all_files=@{&read_file_names_only(\$abs_path_dir_name, \@target_file_names)};

<b>Function  </b> : read any file names and REMOVES the '.', '..' and dir entries.
             And then put in array.  This checks if anything is a real file.
             You can use 'txt' as well as '.txt' as extension
             You can put multiple file extension (txt, doc, ....)
               and multiple dir path (/usr/Perl, /usr/local/Perl....)
               It will fetch all files wanted in all the direc specified

             It can handle file glob eg)
             @all_files=@{&read_file_names_only(\$abs_path_dir_name, 'G1_*.txt')};
               for all txt files starting with 'G1_'

<b>Keywords  </b> : filename only, filename_only, read_files_only, read files
             get_file_names_only, get_files_only, read_files_only
<b>Options   </b> : "extension name". If you put , 'pl' as an option, it will show
             files only with '.pl' extension.
  '-p'      for path also included resulting in '/path/path/file.ext'
              rather than 'file.ext' in output @array
  '-s'      for sorting the results
  e='xxx'  for extention xxx
  '.pl'    for files extended by '.pl'
  'pl'     for files extended by 'pl', same as above
  D=       for dir name input

<b>Usage     </b> : @all_files=@{&read_file_names_only(<dir>, [extension])};
<b>Version   </b> : 2.8
<b>Warning   </b> : This does not report '.', '..'
             Only file names are reported. Compare with &read_any_dir
             extension size should be less than 15 char.
             It sorts the results!
</pre><hr>

<H3><A href="read_file_extension_names_only
.pl.html">read_file_extension_names_only
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//read_file_extension_names_only
.pl">read_file_extension_names_only
.pl</A>
<pre>
<b>Function  </b> : reads only extension names. It returns the ext as keys
             and occurrances of them as values of the keys.
<b>Keywords  </b> : read_file_ext_only, read_file_ext_names_only, read_ext_names_only,
             read_ext_only
<b>Usage     </b> : %file_ext=%{&read_file_extension_names_only('.')};
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="read_dir_names_only
.pl.html">read_dir_names_only
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//read_dir_names_only
.pl">read_dir_names_only
.pl</A>
<pre>
<b>Argument  </b> : takes one or more scaler references. ('.', \$path, $path, ... )
<b>Example   </b> : @files=@{&read_dir_names_only('n', "s=1", '.')};
<b>Function  </b> : read any dir names and and then put in array. If no argument
             for the target directory, it opens PWD automatically
             You can specify the length of dir names to choose.
<b>Keywords  </b> : read_dir_only, get_dir_names, get_dir_names_only, get_subdir_names,
<b>Options   </b> : n   for names only reading(not the full path) , default is full path
             s=  for the size of dirs name. If you want all the dir names
                   with a size of 1 char, s=1
<b>Returns   </b> : one ref. of array.
<b>Usage     </b> : @all_dirs_list = @{&read_dir_names_only(\$absolute_path_dir_name, ....)};
<b>Version   </b> : 3.4
<b>Warning   </b> : This does not report '.', '..'
             Only file names are reported. Compare with &read_any_dir
</pre><hr>

<H3><A href="take_file_name
.pl.html">take_file_name
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//take_file_name
.pl">take_file_name
.pl</A>
<pre>
<b>Example   </b> : will return file.name  from /dir/dir/file.name

<b>Function  </b> : takes file name portion from long dir/filename
<b>Keywords  </b> : get_file_name_only, extract_file_name, take_file_name_only
<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Usage     </b> : $base_portion =${&take_file_name(\'/dir/file.name')};
<b>Version   </b> : 1.0
<b>Warning   </b> : You MUST NOT delete '# options : ..' entry
              as it is read  by various subroutines.
</pre><hr>

<H3><A href="get_file_dir_name
.pl.html">get_file_dir_name
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_file_dir_name
.pl">get_file_dir_name
.pl</A>
<pre>
<b>Example   </b> : /dir/file.name
             =>  /dir/
<b>Function  </b> : returns the dir portion of long filename.
             If file does not have dir portion it returns './'
<b>Keywords  </b> : get_file_dir_name, take_file_dir_name, take_file_dir_names
<b>Options   </b> : _  for debugging.
             #  for debugging.
<b>Version   </b> : 1.0
<b>Warning   </b> : You MUST NOT delete '# options : ..' entry
              as it is read  by various subroutines.
</pre><hr>

<H3><A href="get_dir_names_only
.pl.html">get_dir_names_only
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_dir_names_only
.pl">get_dir_names_only
.pl</A>
<pre>
<b>Argument  </b> : takes one or more scaler references. ('.', \$path, $path, ... )
<b>Function  </b> : read any dir names and and then put in array.
<b>Returns   </b> : one ref. of array.
<b>Usage     </b> : @all_dirs_list = @{&get_dir_names_only(\$absolute_path_dir_name, ....)};
<b>Version   </b> : 3.0
<b>Warning   </b> : This does not report '.', '..'
             Only file names are reported. Compare with &read_any_dir
</pre><hr>

<H3><A href="get_subdir_names
.pl.html">get_subdir_names
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_subdir_names
.pl">get_subdir_names
.pl</A>
<pre>
<b>Argument  </b> : takes one or more scaler references. ('.', \$path, $path, ... )
<b>Example   </b> : <output example with   "get_subdir_names.pl /A Biomatic /Perl/Bio_Seq" at prompt
                  It is an array shown in lines.

             /A Biomatic /Perl/Bio_Seq/Evalign
             /A Biomatic /Perl/Bio_Seq/Seq_hash_handling
             /A Biomatic /Perl/Bio_Seq/Apple
             /A Biomatic /Perl/Bio_Seq/Genome
             /A Biomatic /Perl/Bio_Seq/Open_xxx_files
             /A Biomatic /Perl/Bio_Seq/Open_xxx_files/Evalign

             tk
             ch1
             sub2
             sub3
             xxxx.cong
             whatever_sub_dir
<b>Function  </b> : Gets all subdir and subsubsub...dir names in absolute path names.
<b>Returns   </b> : one ref. of array. (NOT full path names), refer  'read_full_dir_names'
<b>Usage     </b> : @all_sub_dirs_list = @{&read_dir_names_only(\$absolute_path_dir_name, ....)};
<b>Version   </b> : 3.0
<b>Warning   </b> : This does not report '.', '..' ,  Also, this does not show full path
             Only file names are reported. Compare with &read_any_dir
</pre><hr>

<H3><A href="read_full_dir_names
.pl.html">read_full_dir_names
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//read_full_dir_names
.pl">read_full_dir_names
.pl</A>
<pre>
<b>Argument  </b> : takes one or more scaler references. ('.', \$abs_path, $path, ... )
<b>Example   </b> : input>> &read_full_dir_names('/nfs/ind4/ccpe1/people/A Biomatic /perl');
             /tmp_mnt/nfs/ind4/ccpe1/people/A Biomatic /perl/code
             /tmp_mnt/nfs/ind4/ccpe1/people/A Biomatic /perl/tk
             /tmp_mnt/nfs/ind4/ccpe1/people/A Biomatic /perl/ch1
             /tmp_mnt/nfs/ind4/ccpe1/people/A Biomatic /perl/sub2
             /tmp_mnt/nfs/ind4/ccpe1/people/A Biomatic /perl/sub3
             /tmp_mnt/nfs/ind4/ccpe1/people/A Biomatic /perl/xxxx.cong
             /tmp_mnt/nfs/ind4/ccpe1/people/A Biomatic /perl/whatever
<b>Returns   </b> : one ref. of array.
<b>Usage     </b> : @all_files_list = @{&read_full_dir_names(\$absolute_path_dir_name, ....)};
<b>Version   </b> : 1.0
<b>Warning   </b> : This does not report '.', '..'
             Only file names are reported. Compare with &read_any_dir
</pre><hr>

<H3><A href="get_full_dir_names
.pl.html">get_full_dir_names
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_full_dir_names
.pl">get_full_dir_names
.pl</A>
<pre>
<b>Argument  </b> : takes one or more scaler references. ('.', \$abs_path, $path, ... )
<b>Example   </b> : input>> &get_full_dir_names('/nfs/ind4/ccpe1/people/A Biomatic /perl');
             /tmp_mnt/nfs/ind4/ccpe1/people/A Biomatic /perl/code
             /tmp_mnt/nfs/ind4/ccpe1/people/A Biomatic /perl/tk
             /tmp_mnt/nfs/ind4/ccpe1/people/A Biomatic /perl/ch1
             /tmp_mnt/nfs/ind4/ccpe1/people/A Biomatic /perl/sub2
             /tmp_mnt/nfs/ind4/ccpe1/people/A Biomatic /perl/sub3
             /tmp_mnt/nfs/ind4/ccpe1/people/A Biomatic /perl/xxxx.cong
             /tmp_mnt/nfs/ind4/ccpe1/people/A Biomatic /perl/whatever
<b>Returns   </b> : one ref. of array.
<b>Usage     </b> : @all_files_list = @{&read_full_dir_names(\$absolute_path_dir_name, ....)};
<b>Version   </b> : 1.0
<b>Warning   </b> : This does not report '.', '..'
             Only file names are reported. Compare with &read_any_dir
</pre><hr>

<H3><A href="read_any_dir_simple
.pl.html">read_any_dir_simple
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//read_any_dir_simple
.pl">read_any_dir_simple
.pl</A>
<pre>
<b>Argument  </b> : takes one scaler reference.
<b>Function  </b> : read any dir and REMOVES the '.' and '..' entries. And then put in array.
<b>Returns   </b> : one ref. of array.
<b>Usage     </b> : @file_list = @{&read_any_dir(\$absolute_path_dir_name)};
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="read_any_dir
.pl.html">read_any_dir
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//read_any_dir
.pl">read_any_dir
.pl</A>
<pre>
<b>Argument  </b> : takes one scaler reference.
<b>Function  </b> : read any dir and REMOVES the '.' and '..' entries.
             And then put in array.
<b>Returns   </b> : one ref. of array. for the files in the given directory.
<b>Usage     </b> : @file_list = @{&read_any_dir(\$absolute_path_dir_name)};
<b>Version   </b> : 1.2
</pre><hr>

<H3><A href="read_any_dir2
.pl.html">read_any_dir2
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//read_any_dir2
.pl">read_any_dir2
.pl</A>
<pre>
<b>Argument  </b> : takes one or more scaler references.
<b>Function  </b> : read any dir and REMOVES the '.' and '..' entries. And then put in array.
<b>Returns   </b> : one ref. of array.
<b>Usage     </b> : @file_list = @{&read_any_dir(\$absolute_path_dir_name, ....)};
<b>Version   </b> : 1.0
<b>Warning   </b> : This does not report '.', '..', '#xxxx', ',xxxx', etc. only legitimate
             file and dir names are reported.
</pre><hr>

<H3><A href="max_str_value_hash
.pl.html">max_str_value_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//max_str_value_hash
.pl">max_str_value_hash
.pl</A>
<pre>
<b>Function  </b> : gets the largest 'string' length in values of any one hash
<b>Usage     </b> : $largest_str_length_of_values = &max_value_hash(%any_hash);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_max_hash_by_value
.pl.html">get_max_hash_by_value
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_max_hash_by_value
.pl">get_max_hash_by_value
.pl</A>
<pre>
<b>Function  </b> : gets the largest 'string' length in values of any one hash
<b>Keywords  </b> : get_max_hash_value, get_largest_hash_value, get_max_hash_key_value
             get_max_hash_num_value, max_hash_value
<b>Usage     </b> : $largest_str_length_of_values = &max_value_hash(%any_hash);
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="max_str_key_hash
.pl.html">max_str_key_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//max_str_key_hash
.pl">max_str_key_hash
.pl</A>
<pre>
<b>Function  </b> : gets the largest 'string' length in keys of any one hash
<b>Keywords  </b> : largest key length,
<b>Usage     </b> : $largest_str_length_of_values = &max_value_hash(%any_hash);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="min_string_value_hash
.pl.html">min_string_value_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//min_string_value_hash
.pl">min_string_value_hash
.pl</A>
<pre>
<b>Function  </b> : gets the smallest 'string' length in values of any one hash
<b>Usage     </b> : $small_str_length_of_values = &min_str_value_hash(%any_hash);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="min_str_key_hash
.pl.html">min_str_key_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//min_str_key_hash
.pl">min_str_key_hash
.pl</A>
<pre>
<b>Function  </b> : gets the smallest 'string' length in values of any one hash
<b>Usage     </b> : $small_str_length_of_values = &min_str_value_hash(%any_hash);
</pre><hr>

<H3><A href="min_string_key_hash
.pl.html">min_string_key_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//min_string_key_hash
.pl">min_string_key_hash
.pl</A>
<pre>
<b>Function  </b> : gets the smallest 'string' length in values of any one hash
<b>Usage     </b> : $small_str_length_of_values = &min_str_value_hash(%any_hash);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="fasta_append
.pl.html">fasta_append
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//fasta_append
.pl">fasta_append
.pl</A>
<pre>
<b>Function  </b> : append addtional one fasta format sequence.
<b>Usage     </b> : &fasta_append($name, $string, $output_file);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="fasta_output
.pl.html">fasta_output
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//fasta_output
.pl">fasta_output
.pl</A>
<pre>
<b>Function  </b> : prints fasta format output which is using $mul_factor
$seq is the whole sequence number(largest).
             $dir.$mul_factor.fasta can be any output name,
<b>Usage     </b> : &fasta_output($dir.$mul_factor.fasta,  $whole_seq, *array_ali, *array1);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="fasta_out_seq_no
.pl.html">fasta_out_seq_no
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//fasta_out_seq_no
.pl">fasta_out_seq_no
.pl</A>
<pre>
<b>Function  </b> : prints fasta format output with specified seq no from whole seq. no.
$seq is the whole sequence number(largest). $out_seq_no is the target
<b>Usage     </b> : &fasta_out_seq_no($dir, $out_seq_no, $seq, *array2, *array1);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="ctime
.pl.html">ctime
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//ctime
.pl">ctime
.pl</A>
<pre>
<b>Example   </b> : $Date = &ctime(time);
<b>Function  </b> : a simple Perl emulation for the well known ctime(3C) function.
<b>Usage     </b> : $Date = &ctime(time);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_time
.pl.html">get_time
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_time
.pl">get_time
.pl</A>
<pre>
<b>Example   </b> : "Nov30 4:37 1995"
<b>Function  </b> : a simple Perl emulation for the well known ctime(3C) function.
<b>Usage     </b> : $Date = &get_time(time);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_date
.pl.html">get_date
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_date
.pl">get_date
.pl</A>
<pre>
<b>Example   </b> : 30-Nov-1995
<b>Function  </b> : returns date: $date6d (6 digit format) and
             $datec (dd-mmm-yyyy format), Tim's version is 'getdate' in th_lib.pl
<b>Keywords  </b> : get_present_date,
<b>Returns   </b> : ref of an array for (1-May-1995 and 010595)
<b>Usage     </b> : @outformat = &get_date;  eg result >  (010595 1-May-1995)
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="if_file_older_than_x_days
.pl.html">if_file_older_than_x_days
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//if_file_older_than_x_days
.pl">if_file_older_than_x_days
.pl</A>
<pre>
<b>Function  </b> : checks the date of last modi of file given and compares with
             present time. Substracts diff and returns the actual diff days.
<b>Keywords  </b> : how_old_file, how_old, is_file_older_than_x_days, file_age,
             file_age_in_days, if_older_than_x_days,
<b>Returns   </b> : the actual days older, so NON-ZERO, otherwise, 0
<b>Usage     </b> : if( ${&if_file_older_than_x_days($ARGV[0], $days)} > 0){
<b>Version   </b> : 1.3
</pre><hr>

<H3><A href="array_chk
.pl.html">array_chk
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//array_chk
.pl">array_chk
.pl</A>
<pre>
<b>Argument  </b> : gets on ref. of array.
<b>Example   </b> : This is used only with subs which accepts array inputs.
<b>Function  </b> : checks if any inputting array is empty or with one element.
<b>Keywords  </b> : array_check
<b>Returns   </b> : nothing, prints out messages to STDOUT
<b>Usage     </b> : &array_chk(\@any_array_to_chk);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="hash_chk
.pl.html">hash_chk
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//hash_chk
.pl">hash_chk
.pl</A>
<pre>
<b>Function  </b> : checks hash input of any subroutine.
<b>Keywords  </b> : hash_check
<b>Usage     </b> : &hash_chk(\%input_hash);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="hash_output_chk
.pl.html">hash_output_chk
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//hash_output_chk
.pl">hash_output_chk
.pl</A>
<pre>
<b>Function  </b> : checks hash output of any subroutine.
<b>Usage     </b> : &hash_output_chk(\%outing_hash);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="n
.pl.html">n
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//n
.pl">n
.pl</A>
<pre>
<b>Function  </b> : puts one single new line
<b>Usage     </b> : &n;
</pre><hr>

<H3><A href="cls
.pl.html">cls
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//cls
.pl">cls
.pl</A>
<pre>
<b>Function  </b> : clears screen
<b>Usage     </b> : &cls;
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="seq_comp_percent1
.pl.html">seq_comp_percent1
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//seq_comp_percent1
.pl">seq_comp_percent1
.pl</A>
<pre>
<b>Argument  </b> : one ref. of an array
<b>Function  </b> : get string seq identities(a to z). gets array of strings and outs array of % numbers
<b>Returns   </b> : one ref. of an array
<b>Usage     </b> : @outarray = &seq_comp_percent1(@any_input_string_array);
</pre><hr>

<H3><A href="get_id_among_2_1
.pl.html">get_id_among_2_1
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_id_among_2_1
.pl">get_id_among_2_1
.pl</A>
<pre>
<b>Function  </b> : gets the % id of any two sequences, returns in  100.0% format.
<b>Usage     </b> : $id = &get_id_among_2(*charcount1, *charcount2) <- hashes
</pre><hr>

<H3><A href="get_id_among_2_2
.pl.html">get_id_among_2_2
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_id_among_2_2
.pl">get_id_among_2_2
.pl</A>
<pre>
<b>Function  </b> : gets the % id of any two sequences, returns in  100.0% format.
<b>Usage     </b> : $id = &get_id_among_2(*charcount1, *charcount2) <- hashes
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="array_average
.pl.html">array_average
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//array_average
.pl">array_average
.pl</A>
<pre>
<b>Argument  </b> : takes one array reference.
<b>Function  </b> : (the same as average_array)
<b>Keywords  </b> : get_array_average, av_array, average_array, get_average_array
             average_of_array, average_array
<b>Returns   </b> : single scaler digit.
<b>Usage     </b> : $output = &array_average(\@any_array);
<b>Version   </b> : 1.2
<b>Warning   </b> : If divided by 0, it will automatically replace it with 1
</pre><hr>

<H3><A href="average_array
.pl.html">average_array
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//average_array
.pl">average_array
.pl</A>
<pre>
<b>Argument  </b> : takes one array reference.
<b>Function  </b> : (the same as array_average)
<b>Returns   </b> : single scaler digit.
<b>Usage     </b> : $output = &average_array(\@any_array);
<b>Version   </b> : 1.0
<b>Warning   </b> : If divided by 0, it will automatically replace it with 1
</pre><hr>

<H3><A href="average_of_array
.pl.html">average_of_array
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//average_of_array
.pl">average_of_array
.pl</A>
<pre>
<b>Argument  </b> : takes one array reference.
<b>Options   </b> : -int to make the resultant numbers shown in integer
<b>Returns   </b> : single scaler digit.
<b>Usage     </b> : $output = &average_of_array(\@any_array);
<b>Version   </b> : 2.0
<b>Warning   </b> : If divided by 0, it will automatically replace it with 1
             '$item == 0' does not work !!! in the following
</pre><hr>

<H3><A href="hash_average
.pl.html">hash_average
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//hash_average
.pl">hash_average
.pl</A>
<pre>
<b>Example   </b> : %in=(1, "13242442", 2, "92479270", 3, "2472937439");
<b>Returns   </b> : %out =(1, 2.13242, 2, 5.2702, 3, 1.72937439); <-- somethins like
             numbers. So, undefined array element is not counted
             This is more correct.
<b>Usage     </b> : %out=%{&hash_average(\%in)};  or
             ($out1, $out2)=&hash_average(\%in,\%in2);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_hash_value_average
.pl.html">get_hash_value_average
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_hash_value_average
.pl">get_hash_value_average
.pl</A>
<pre>
<b>Keywords  </b> : get_values_average, get_average_hash_value, get_average_value
<b>Returns   </b> : %out =(1, 2.13242, 2, 5.2702, 3, 1.72937439); <-- somethins like
             numbers. So, undefined array element is not counted
             This is more correct.
<b>Usage     </b> : %out=%{&get_hash_value_average(\%in)};  or
             ($out1, $out2)=&hash_average(\%in,\%in2);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="hash_stat_for_all
.pl.html">hash_stat_for_all
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//hash_stat_for_all
.pl">hash_stat_for_all
.pl</A>
<pre>
<b>Example   </b> : %in =(1, "13242442", 2, "92479270", 3, "2472937439");
             %in2=(1, "28472", 2, "23423240", 3, "123412342423439");

             %in =(name1, "1,3,2,4,2,4,4,2", name2, "9,2,4,7,9,2,7,0");

<b>Function  </b> : gets the min, max, av, sum for the whole values of ALL the
             hashes put in. (grand statistics)
<b>Returns   </b> : normal array of ($min, $max, $sum, $av)
             Example  out:>                 |  min max sum  av
                            -----------------------------------
                            of the whole    |   0   9  110   6
<b>Usage     </b> : %out=%{&hash_average(\%in, \%in2,..)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="min
.pl.html">min
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//min
.pl">min
.pl</A>
<pre>
<b>Function  </b> : accepts ref of array, scalar and normal digits to
             find the min. Only gets numbers. If you put something
             like 'H333333', it gets digits '333333' only and returns it.
             this uses RECURSION.
<b>Usage     </b> : $min = &min (37, 24, 3,1,5, \@array, @array2, \$arr_ref);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="max
.pl.html">max
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//max
.pl">max
.pl</A>
<pre>
<b>Function  </b> : accepts ref of array, scalar and normal digits to
             find the min. Only gets numbers. If you put something
             like 'H333333', it gets digits '333333' only and returns it.
             this uses RECURSION.
<b>Usage     </b> : $max = &max (37, 24, 3,1,5, \@array, @array2, \$arr_ref);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_longest_str_size
.pl.html">get_longest_str_size
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_longest_str_size
.pl">get_longest_str_size
.pl</A>
<pre>
<b>Argument  </b> : gets one reference of an array of strings.
<b>Function  </b> : get_longest_str_size in an array. eg. get ABCDE among (A, CAB, CDE, ABCDE)
             When hash is given it processes the values of it.
<b>Keywords  </b> : get_the_largest_string_size{, get_largest_string_size,
             get_largest_str_size{,largest_string_size{, get_largest_string_size_hash
             get_long_str_size, get_longest_string_size, lonest_string_size
<b>Usage     </b> : $long_str_size = ${&get_long_str_size (\@any_array_of_string)};
             $long_str_size = ${&get_long_str_size (\@any_array_of_string)};
<b>Version   </b> : 1.2
</pre><hr>

<H3><A href="get_shortest_str_size
.pl.html">get_shortest_str_size
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_shortest_str_size
.pl">get_shortest_str_size
.pl</A>
<pre>
<b>Argument  </b> : gets one reference of an array of strings.
<b>Function  </b> : get_shortest_str_size in an array. eg. get A among (A, CAB, CDE, ABCDE)
<b>Keywords  </b> : get_short_str_size, get_short_string_size, shortest_string_size,
<b>Usage     </b> : $short_str_size = &get_short_str_size (\@any_array_of_string);
<b>Version   </b> : 1.0
<b>Warning   </b> : once debugged. 1st May/95
</pre><hr>

<H3><A href="get_id_among_2
.pl.html">get_id_among_2
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_id_among_2
.pl">get_id_among_2
.pl</A>
<pre>
<b>Argument  </b> : gets two references of hashes of chars and their occurances.
<b>Example   </b> : %hash1=('A', 30, 'B', 99, 'C', 15 .....)
<b>Function  </b> : gets the % id of any two sequences
<b>Usage     </b> : $id = &get_id_among_2(\%charcount1, \%charcount2) <- hashes
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="extract_num_to_array
.pl.html">extract_num_to_array
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//extract_num_to_array
.pl">extract_num_to_array
.pl</A>
<pre>
<b>Function  </b> : extract only numbers(including negatives) from a string and put into an array
<b>Usage     </b> : @my_outarray = &extract_num_to_array($any_input_string);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="weighted_average
.pl.html">weighted_average
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//weighted_average
.pl">weighted_average
.pl</A>
<pre>
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="weighted_av
.pl.html">weighted_av
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//weighted_av
.pl">weighted_av
.pl</A>
<pre>
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="sum_digits_in_string
.pl.html">sum_digits_in_string
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//sum_digits_in_string
.pl">sum_digits_in_string
.pl</A>
<pre>
<b>Author    </b> : Randal
<b>Keywords  </b> : add_digits_in_string,
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="sum_array
.pl.html">sum_array
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//sum_array
.pl">sum_array
.pl</A>
<pre>
<b>Argument  </b> : ref. of an array of numbers.
<b>Function  </b> : sum of all the  elements of an array .
<b>Keywords  </b> : get_array_sum get_sum_array, get sum of array
<b>Returns   </b> : a ref. of a scaler.
<b>Usage     </b> : $out =  ${&sum_array(\@anyarray)};
</pre><hr>

<H3><A href="sum_of_array
.pl.html">sum_of_array
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//sum_of_array
.pl">sum_of_array
.pl</A>
<pre>
<b>Argument  </b> : ref. of an array of numbers.
<b>Function  </b> : sum of all the  elements of an array .
<b>Options   </b> : -int for integerised output.
<b>Returns   </b> : a ref. of a scaler.
<b>Usage     </b> : $out =  ${&sum_of_array(\@anyarray)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="array_sum
.pl.html">array_sum
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//array_sum
.pl">array_sum
.pl</A>
<pre>
<b>Argument  </b> : ref. of an array of numbers.
<b>Function  </b> : sum of all the  elements of an array .
<b>Returns   </b> : a ref. of a scaler.
<b>Usage     </b> : $out =  ${&sum_array(\@anyarray)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="sum_hash_values_of_string
.pl.html">sum_hash_values_of_string
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//sum_hash_values_of_string
.pl">sum_hash_values_of_string
.pl</A>
<pre>
<b>Example   </b> : %hashinput= ( name1, '12..3e',
                            name2, '...234');
             $result = 1+2+3+2+3+4 = 15 (from above example)
<b>Function  </b> : sum of all the  numbers in valuse of a hash
<b>Keywords  </b> : sum_hash_string_values, get_sum_hash_string_values, get_hash_value_sum
<b>Usage     </b> : $out = &sum_hash_values_of_string(\%anyhash);
<b>Version   </b> : 1.1
<b>Warning   </b> : It only gets digits in the input strings and sums them up.
</pre><hr>

<H3><A href="sum_hash_values
.pl.html">sum_hash_values
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//sum_hash_values
.pl">sum_hash_values
.pl</A>
<pre>
<b>Example   </b> : %hashinput= ( name1, '12..3e',
                            name2, '...234');
             $result = 1+2+3+2+3+4 = 15 (from above example)
<b>Function  </b> : sum of all the  numbers in valuse of a hash
<b>Keywords  </b> : sum_hash_number_values, get_sum_hash_values, get_hash_value_sum
<b>Usage     </b> : $out = &sum_hash_values(%anyhash);
<b>Version   </b> : 1.0
<b>Warning   </b> : It only gets digits in the input strings and sums them up.
</pre><hr>

<H3><A href="key_ready
.pl.html">key_ready
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//key_ready
.pl">key_ready
.pl</A>
<pre>
<b>Function  </b> : detects keyboard input without reading it
<b>Returns   </b> : 
             You should check out the Frequently Asked Questions list in
             comp.unix.* for things like this: the answer is
             essentially the same.
             It's very system dependent.  Here's one solution that
             works on BSD systems:
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="round
.pl.html">round
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//round
.pl">round
.pl</A>
<pre>
<b>Function  </b> : gives rounded numbers
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="round_number
.pl.html">round_number
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//round_number
.pl">round_number
.pl</A>
<pre>
<b>Function  </b> : gives rounded integer numbers. 9.5 will be 10, 9.4 will be 9
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="round_numbers
.pl.html">round_numbers
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//round_numbers
.pl">round_numbers
.pl</A>
<pre>
<b>Function  </b> : gives rounded integer numbers. 9.5 will be 10, 9.4 will be 9
<b>Usage     </b> : @output=@{&round_numbers(\@input_numbs)};
             or  $output=${&round_numbers(\$input_numbs)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="trim_numbers
.pl.html">trim_numbers
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//trim_numbers
.pl">trim_numbers
.pl</A>
<pre>
<b>Example   </b> : given num array( 1.33333, 3.555242424, 0.2342324, 4.9234723747)
             >>>            (1.33,  3.56,  0.23,  4.92 )

<b>Function  </b> : gives trimmed numbers (not rounded)
<b>Usage     </b> : @output=@{&trim_numbers(\@input_numbs, \$size_of_posi)};
<b>Version   </b> : 1.0
<b>Warning   </b> : If you put '1' with trimming value of 2 it will be '1.00'
</pre><hr>

<H3><A href="min_elem_array
.pl.html">min_elem_array
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//min_elem_array
.pl">min_elem_array
.pl</A>
<pre>
<b>Argument  </b> : numerical arrays
<b>Function  </b> : gets the smallest element of any array of numbers.
<b>Returns   </b> : one or more ref. for scalar numbers.
<b>Usage     </b> : ($out1, $out2)=@{&min_elem_array(\@array1, \@array2)};
             ($out1)       =${&min_elem_array(\@array1)          };
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="max_elem_array
.pl.html">max_elem_array
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//max_elem_array
.pl">max_elem_array
.pl</A>
<pre>
<b>Argument  </b> : numerical arrays
<b>Function  </b> : gets the largest element of any array of numbers.
<b>Returns   </b> : one or more ref. for scalar numbers.
<b>Usage     </b> : ($out1, $out2)=@{&max_elem_array(\@array1, \@array2)};
             ($out1)       =${&max_elem_array(\@array1)          };
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="max_elem_string_array
.pl.html">max_elem_string_array
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//max_elem_string_array
.pl">max_elem_string_array
.pl</A>
<pre>
<b>Argument  </b> : numerical arrays
<b>Function  </b> : gets the largest string length of element of any array of numbers.
<b>Keywords  </b> : largest string length of array
<b>Returns   </b> : one or more ref. for scalar numbers.
<b>Usage     </b> : ($out1, $out2)=@{&max_elem_array(\@array1, \@array2)};
             ($out1)       =${&max_elem_array(\@array1)          };
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="min_elem_string_array
.pl.html">min_elem_string_array
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//min_elem_string_array
.pl">min_elem_string_array
.pl</A>
<pre>
<b>Argument  </b> : numerical arrays
<b>Function  </b> : gets the largest string length of element of any array of numbers.
<b>Keywords  </b> : shortest string length of array
<b>Returns   </b> : one or more ref. for scalar numbers.
<b>Usage     </b> : ($out1, $out2)=@{&max_elem_array(\@array1, \@array2)};
             ($out1)       =${&max_elem_array(\@array1)          };
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="maximum
.pl.html">maximum
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//maximum
.pl">maximum
.pl</A>
<pre>
<b>Function  </b> : another way of finding maximum
<b>Keywords  </b> : get_maximum, get_bigger, get_largest
<b>Usage     </b> : $biggest = &maximum(37, 24);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="minimum
.pl.html">minimum
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//minimum
.pl">minimum
.pl</A>
<pre>
<b>Function  </b> : another way of finding minimum
<b>Usage     </b> : $biggest = &maximise(37, 24);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_largest_element
.pl.html">get_largest_element
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_largest_element
.pl">get_largest_element
.pl</A>
<pre>
<b>Function  </b> : If strings are given, it gets the largest string elem(by leng)
             If numbers are given, it gets the largest number elem
             It automatically checks if string is given
<b>Keywords  </b> : get_largest_value, get_biggest_value,
             get_maximum_element, get_largest_number,
             get_largest_number_element, get_longest_element,
             get_longest_string
<b>Options   </b> : _  for debugging.
             #  for debugging.
             s  for string input (as the second input argument!)

<b>Usage     </b> : $max=${&get_largest_element(\@array_input)};
<b>Version   </b> : 1.3
</pre><hr>

<H3><A href="sqrt_array
.pl.html">sqrt_array
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//sqrt_array
.pl">sqrt_array
.pl</A>
<pre>
<b>Function  </b> : sqrt all elements of an array
</pre><hr>

<H3><A href="square_array
.pl.html">square_array
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//square_array
.pl">square_array
.pl</A>
<pre>
<b>Function  </b> : converts all the elements of an array to squared values.
</pre><hr>

<H3><A href="sum_of_squared_array
.pl.html">sum_of_squared_array
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//sum_of_squared_array
.pl">sum_of_squared_array
.pl</A>
<pre>
<b>Function  </b> : sum of all the squared elements of an array .
<b>Usage     </b> : $out = &sum_of_squared_array(@anyarray);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="x_mul_y_arrays
.pl.html">x_mul_y_arrays
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//x_mul_y_arrays
.pl">x_mul_y_arrays
.pl</A>
<pre>
<b>Function  </b> : multiplies each item of two arrays .
<b>Usage     </b> : @out_array = &x_mul_y_arrays(*array1,*array2);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="sum_x_mul_y_arrays
.pl.html">sum_x_mul_y_arrays
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//sum_x_mul_y_arrays
.pl">sum_x_mul_y_arrays
.pl</A>
<pre>
<b>Function  </b> : sums up multiplied items of two arrays .
             one to one multiplication(elem 1 of array 1 x elem 1 of array2)
<b>Usage     </b> : $out = &sum_x_mul_y_arrays(*array1,*array2);
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="corelation_coefficient
.pl.html">corelation_coefficient
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//corelation_coefficient
.pl">corelation_coefficient
.pl</A>
<pre>
<b>Function  </b> : gets corelation_coefficient of two equal length arrays
<b>Keywords  </b> : cc, get_cc, get_corelation_coefficient
<b>Usage     </b> : $cc = &corelation_coefficient(\@array_not_hash1, \@array_not_hash2);
<b>Version   </b> : 1.0
<b>Warning   </b> : uses references for ARRAY.
</pre><hr>

<H3><A href="cc
.pl.html">cc
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//cc
.pl">cc
.pl</A>
<pre>
<b>Function  </b> : synonmym of  corelation_coefficient
<b>Usage     </b> : $cc = &cc(\@array_not_hash1, \@array_not_hash2);
<b>Version   </b> : 1.0
<b>Warning   </b> : uses references for ARRAY
</pre><hr>

<H3><A href="sd
.pl.html">sd
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//sd
.pl">sd
.pl</A>
<pre>
<b>Argument  </b> : array references are accepted. outputs scalar single val.
<b>Keywords  </b> : standard deviation, get_standard_deviation,
             standard_deviation, get_SD, get_sd, stdev
<b>Returns   </b> : a ref. of a scaler
<b>Usage     </b> : $sd=${&sd(\@array_of_numbers)};
<b>Version   </b> : 1.2
</pre><hr>

<H3><A href="se
.pl.html">se
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//se
.pl">se
.pl</A>
<pre>
<b>Argument  </b> : ref. for an array.
<b>Function  </b> : gets standard error of any given array
<b>Keywords  </b> : standard error, get_standard_error, sterr
<b>Usage     </b> : $se=${&se(\@array_of_numbers)};
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="remove_non_char
.pl.html">remove_non_char
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//remove_non_char
.pl">remove_non_char
.pl</A>
<pre>
<b>Function  </b> : removes non chars on any input string. (scaler context)
<b>Usage     </b> : $outstring = &remove_non_char($input_string);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="numerically
.pl.html">numerically
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//numerically
.pl">numerically
.pl</A>
<pre>
<b>Function  </b> : sorts elements by nemerical size.
<b>Usage     </b> : sort numerically (@array);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="abs_numerically
.pl.html">abs_numerically
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//abs_numerically
.pl">abs_numerically
.pl</A>
<pre>
<b>Keywords  </b> : numerically_abs, numerically_absolutely
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="rev_abs_numerically
.pl.html">rev_abs_numerically
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//rev_abs_numerically
.pl">rev_abs_numerically
.pl</A>
<pre>
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="randomise_lines
.pl.html">randomise_lines
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//randomise_lines
.pl">randomise_lines
.pl</A>
<pre>
<b>Function  </b> : 
             outs line numbers with lines
<b>Usage     </b> : To randomize th_lib.pl just type &random_lines(300,500,"th_lib.pl");
             &random_lines(300, 50, "th_lib.pl"); <-- to get 300 lines
                                                      from 50 numbers
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="pick_random_hash_pairs
.pl.html">pick_random_hash_pairs
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//pick_random_hash_pairs
.pl">pick_random_hash_pairs
.pl</A>
<pre>
<b>Example   </b> : in signature roation or FVWM rc file menu color rotation.
<b>Function  </b> : randomly pick any num of pairs of hash elements.
             outs line numbers with lines
             Default pick number is 1.
<b>Keywords  </b> : choose_random_hash_pairs
<b>Returns   </b> : ARRAY ref not HASH  ref
<b>Usage     </b> : @array = @{&pick_random_hash_pairs(\%hash1, \$xx)};
<b>Version   </b> : 1.3
</pre><hr>

<H3><A href="pick_random_files
.pl.html">pick_random_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//pick_random_files
.pl">pick_random_files
.pl</A>
<pre>
<b>Example   </b> : @array=@{&pick_random_files(\@files, \$num_of_pick)};
<b>Function  </b> : randomly pick any num of files given.
<b>Keywords  </b> : choose_random_files pick_files_randomly
<b>Returns   </b> : ARRAY ref not HASH  ref
<b>Usage     </b> : @array = @{&pick_random_files(\@files, \$num_of_pick)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="hash_substract_by_keys
.pl.html">hash_substract_by_keys
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//hash_substract_by_keys
.pl">hash_substract_by_keys
.pl</A>
<pre>
<b>Example   </b> : %hash1 = %hash1 - %hash2, ==> (4,4)=(2,2, 4,4) - (2,2)
<b>Function  </b> : removes overlapping entries in hashes.
<b>Keywords  </b> : substract_hash, substract_hash_by_value, hash_substract
<b>Usage     </b> : %hash1 = %{&hash_substract_by_keys(\%hash1, \%hash2)};
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="substract_hash_by_keys
.pl.html">substract_hash_by_keys
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//substract_hash_by_keys
.pl">substract_hash_by_keys
.pl</A>
<pre>
<b>Example   </b> : %hash1 = %hash1 - %hash2, ==> (4,4)=(2,2, 4,4) - (2,2)
<b>Function  </b> : removes overlapping entries in hashes.
<b>Keywords  </b> : substract_hash, substract_hash_by_keys
<b>Usage     </b> : %hash1 = %{&substract_hash(\%hash1, \%hash2)};
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="substract_hash_by_values
.pl.html">substract_hash_by_values
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//substract_hash_by_values
.pl">substract_hash_by_values
.pl</A>
<pre>
<b>Example   </b> : %hash1 = %hash1 - %hash2, ==> (4,4)=(2,2, 4,4) - (2,2)
<b>Function  </b> : removes overlapping value entries in hashes.
<b>Keywords  </b> : substract_hash, substract_hash_by_values
<b>Usage     </b> : %hash1 = %{&substract_hash_by_values(\%hash1, \%hash2)};
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="substract_array
.pl.html">substract_array
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//substract_array
.pl">substract_array
.pl</A>
<pre>
<b>Example   </b> : Following will produce (A K C);
  @array1= qw( A B K B B C);
    @array2= qw( B E D);
    @subs = @{&substract_array(\@array1, \@array2)};
<b>Function  </b> : removes any occurances of certain elem. of the first
             input array with second input array.
<b>Keywords  </b> : array_subtract, substract_array, ary1_minus_ary2
<b>Usage     </b> : @subs = @{&substract_array(\@array1, \@array2)};
<b>Version   </b> : 1.6
</pre><hr>

<H3><A href="hash_catenate
.pl.html">hash_catenate
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//hash_catenate
.pl">hash_catenate
.pl</A>
<pre>
<b>Function  </b> : removes overlapping entries in hashes.
<b>Usage     </b> : %output = %{&hash_catenate(\%hash1, \%hash2)};
</pre><hr>

<H3><A href="merge_hash
.pl.html">merge_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//merge_hash
.pl">merge_hash
.pl</A>
<pre>
<b>Function  </b> : removes overlapping entries in hashes.
<b>Keywords  </b> : merge_hash_elements,add_hash, merge two hashes.
             merge hashes, merge_hashes.
<b>Usage     </b> : %output = %{&merge_hash(\%hash1, \%hash2)};
<b>Version   </b> : 1.2
<b>Warning   </b> : one bug caught.
</pre><hr>

<H3><A href="superpose_hash
.pl.html">superpose_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//superpose_hash
.pl">superpose_hash
.pl</A>
<pre>
<b>Function  </b> : superpose hash keys and values to another hash. %target
             is the superposing hash(new ones will have the values of
             this target hash. For example, if you superpose
                (1, 123, 2, 343)
             to (1, 111, 2, 2222, 3, 3333), you will get
                (1, 123, 2, 343,  3, 3333) as the result.
             Template provide blank key entries.
<b>Usage     </b> : %output = %{superpose_hash(\%template, \%target));
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="hash_common2
.pl.html">hash_common2
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//hash_common2
.pl">hash_common2
.pl</A>
<pre>
<b>Argument  </b> : accepts only two references of hashes
<b>Example   </b> : %hashout= %hash1 - %hash2, ==> (4,4)=(2,2, 4,4) - (2,2)
<b>Returns   </b> : a ref of a hash.
<b>Usage     </b> : %output = &hash_common($ref1, $ref2);
<b>Version   </b> : 1.0
<b>Warning   </b> : NOT working
</pre><hr>

<H3><A href="remove_dup_in_hash
.pl.html">remove_dup_in_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//remove_dup_in_hash
.pl">remove_dup_in_hash
.pl</A>
<pre>
<b>Argument  </b> : one or more hash ref.
<b>Example   </b> : If %input was
                (1,1, 2,1, 3,1);
              The values are the same, so the last key value (3 1) will
              be the result.
             If %input was
                (1,1, 2,1, 3,1, 4,2, 5,2)
              result=(3 1, 5 2)

<b>Function  </b> : removes the duplicate  values of any hashes
<b>Keywords  </b> : remove_dupplicate_values_in_hash, remove_duplicate_values,
             remov_hash_dup, remove_duplication_in_hash
<b>Returns   </b> : one or more hash ref.
<b>Usage     </b> : %out=%{&remove_dup_in_hash(\%input_hash)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="reverse_hash
.pl.html">reverse_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//reverse_hash
.pl">reverse_hash
.pl</A>
<pre>
<b>Argument  </b> : one or more hash ref.
<b>Function  </b> : exchanges the value and key of any hashes
<b>Keywords  </b> : invert_hash, inverse_hash
<b>Returns   </b> : one or more hash ref.
<b>Usage     </b> : %out=%{&reverse_hash(\%input_hash)};
<b>Version   </b> : 1.0
<b>Warning   </b> : Takes ALIGNED sequences.
</pre><hr>

<H3><A href="hash_common
.pl.html">hash_common
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//hash_common
.pl">hash_common
.pl</A>
<pre>
<b>Returns   </b> : the VALUES OF THE FIRST HASH which occur in later hashes
             are returned
<b>Usage     </b> : %hash1_value = %{&hash_common(\%hash1, \%hash2,...)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="hash_common_by_keys
.pl.html">hash_common_by_keys
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//hash_common_by_keys
.pl">hash_common_by_keys
.pl</A>
<pre>
<b>Returns   </b> : the VALUES OF THE FIRST HASH which occur in later hashes
             are returned
<b>Usage     </b> : %hash1_value = %{&hash_common_by_keys(\%hash1, \%hash2,...)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_common_hash_keys
.pl.html">get_common_hash_keys
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_common_hash_keys
.pl">get_common_hash_keys
.pl</A>
<pre>
<b>Function  </b> : gets the common hash keys of two hashes.
<b>Usage     </b> : @output_array_of_keys = @{&get_common_hash_keys(\%hash1, \%hash2)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="hash_no_common
.pl.html">hash_no_common
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//hash_no_common
.pl">hash_no_common
.pl</A>
<pre>
<b>Example   </b> : %hashout= %hash1 - %hash2, ==> (4,4)=(2,2, 4,4) - (2,2)
<b>Function  </b> : removes overlapping entries in hashes.
<b>Usage     </b> : %output = &hash_catenate(*hash1, *hash2);
<b>Version   </b> : 1.0
<b>Warning   </b> : surely working, This grep version is faster than for and defined loop.
</pre><hr>

<H3><A href="beep
.pl.html">beep
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//beep
.pl">beep
.pl</A>
<pre>
<b>Usage     </b> : &beep;
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="capitalize_word.pl.html">capitalize_word</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//capitalize_word.pl">capitalize_word.pl</A>
<pre>
<b>Keywords  </b> : capitalise word,  capitalise_word
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="capitalize_sentence.pl.html">capitalize_sentence</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//capitalize_sentence.pl">capitalize_sentence.pl</A>
<pre>
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="shift_word_recursively
.pl.html">shift_word_recursively
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//shift_word_recursively
.pl">shift_word_recursively
.pl</A>
<pre>
<b>Argument  </b> : SCALAR or ARRAY refs. and delimitor ('/', '.', '-'.....)
             delimitor can be multi line => '#$%/=.'
             default delimiter is space ' ';
<b>Example   </b> : @new_lines=shift_word_recursively(\@lines, '/-', 2); to chop lines
             off two words with the two delimiters of '/' and '-'.
             /jong1/perl-jong2/perl-jong3  will become   /perl-jong2/perl-A Biomatic 3
             /bin/-kkk/-jjj/-jj will become  /-kkk/-jjj/-jj
             @out=@{&shift_word_recursively($testline, '/-', 2)};
             You can use perl regexp patterns for  $delimiter as it is directly
             used in a pattern matching in the sub. So, you canuse '\W'
<b>Function  </b> : shift lines word by word. This needs delimiter like '/' or '.'
             and stores the resulting arrays. This is to get all the possible
             directories.
             For example, with /nfs/A Biomatic /perl/temp/here  input, you get
             (  /A Biomatic /perl/temp/here,   /perl/temp/here ,
             temp/here, /here, )  in an array.

<b>Usage     </b> : @new_lines=shift_word_recursively(\@lines, '/'); or
             @new_lines=shift_word_recursively(\@lines, '\W'); or
             @new_lines=shift_word_recursively(\@lines, 'a-zA-Z'); or
             @new_lines=shift_word_recursively(\@lines, '/', 2); <--- for multiple chop unit
             or $new_line = shift_word_recursively(\$line, '.'); <--- for scalar input.
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="shift_word
.pl.html">shift_word
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//shift_word
.pl">shift_word
.pl</A>
<pre>
<b>Argument  </b> : SCALAR or ARRAY refs. and delimitor ('/', '.', '-'.....)
             delimitor can be multi line => '#$%/=.'
             default delimiter is space ' ';
<b>Example   </b> : @new_lines=shift_word(\@lines, '/-', 2); to shift off lines two words
             with the two delimiters of '/' and '-'.
             /jong1/perl-jong2/perl-jong3  will become   /jong1/perl-A Biomatic 2
             /bin/-kkk/-jjj/-jj will become  /jong1/perl-A Biomatic 2 by
             @out=@{&shift_word($testline, '/-', 2)};
             You can use perl regexp patterns for  $delimiter as it is directly
             used in a pattern matching in the sub. So, you canuse '\W'
<b>Function  </b> : shift lines word by word. This needs delimiter like '/' or '.'
<b>Usage     </b> : @new_lines=shift_word(\@lines, '/'); or
             @new_lines=shift_word(\@lines, '\W'); or
             @new_lines=shift_word(\@lines, 'a-zA-Z'); or
             @new_lines=shift_word(\@lines, '/', 2); <--- for multiple chop unit
             or $new_line = shift_word(\$line, '.'); <--- for scalar input.
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="chop_word
.pl.html">chop_word
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//chop_word
.pl">chop_word
.pl</A>
<pre>
<b>Argument  </b> : SCALAR or ARRAY refs. and delimitor ('/', '.', '-'.....)
             delimitor can be multi line => '#$%/=.'
             default delimiter is space ' ';
<b>Example   </b> : @new_lines=chop_word(\@lines, '/-', 2); to chop off lines two words
             with the two delimiters of '/' and '-'.
             /jong1/perl-jong2/perl-jong3  will become   /jong1/perl-A Biomatic 2
             /bin/-kkk/-jjj/-jj will become  /jong1/perl-A Biomatic 2 by
             @out=@{&chop_word($testline, '/-', 2)};
             You can use perl regexp patterns for  $delimiter as it is directly
             used in a pattern matching in the sub. So, you canuse '\W'
<b>Function  </b> : chop lines word by word. This needs delimiter like '/' or '.'
<b>Keywords  </b> : chop_word_recursively, remove_word, chop_word_one_by_one
<b>Options   </b> : -w, w, Word, etc,  for getting the chopped off word(s) rather
             than the original lines minus the word.
<b>Usage     </b> : @new_lines=chop_word(\@lines, '/'); or
             @new_lines=chop_word(\@lines, '\W'); or
             @new_lines=chop_word(\@lines, 'a-zA-Z'); or
             @new_lines=chop_word(\@lines, '/', 2); <--- for multiple chop unit
             or $new_line = chop_word(\$line, '.'); <--- for scalar input.
<b>Version   </b> : 2.0
<b>Warning   </b> : The returning value is not the chopped off word.
</pre><hr>

<H3><A href="get_median
.pl.html">get_median
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_median
.pl">get_median
.pl</A>
<pre>
<b>Argument  </b> : \@array
<b>Keywords  </b> : median_array, get_median_array, get_array_median, array_median
<b>Returns   </b> : \$median
<b>Usage     </b> : $median = ${&get_median(\@array)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="array_median
.pl.html">array_median
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//array_median
.pl">array_median
.pl</A>
<pre>
<b>Argument  </b> : \@array
<b>Keywords  </b> : median_array, get_median_array, get_array_median, array_median
<b>Returns   </b> : \$median
<b>Usage     </b> : $median = ${&array_median(\@array)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_median
.pl.html">get_median
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_median
.pl">get_median
.pl</A>
<pre>
<b>Argument  </b> : \@array
<b>Keywords  </b> : median_array, get_median_array, get_array_median, array_median
<b>Returns   </b> : \$median
<b>Usage     </b> : $median = ${&get_median(\@array)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="push_if_not_already
.pl.html">push_if_not_already
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//push_if_not_already
.pl">push_if_not_already
.pl</A>
<pre>
<b>Argument  </b> : two references. The first should be an array ref. The 2nd can be either
             scalar or array reference.
<b>Function  </b> : returns ref. of an array for a list of non-repetitive entry.
<b>Returns   </b> : a ref. of an array.
<b>Usage     </b> : @out=@{&push_if_not_already(@mother_array, @adding_array )};
             @out=@{&push_if_not_already(@mother_array, $adding_scalar)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="replace_lines
.pl.html">replace_lines
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//replace_lines
.pl">replace_lines
.pl</A>
<pre>
<b>Function  </b> : replace_lines in any txt files
<b>Usage     </b> : &replace_lines(@files, 'removing_string', 'match_str' );
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="insert_lines_anywhere
.pl.html">insert_lines_anywhere
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//insert_lines_anywhere
.pl">insert_lines_anywhere
.pl</A>
<pre>
<b>Function  </b> : insert lines anywhere in any txt files. Without any
              position options(Before, After), it attaches the line
<b>Keywords  </b> : insert_text, insert_lines, insert_something,
             attach_lines_in_text, attach_lines, insert_text_lines
<b>Options   </b> : 
   $adding_line= by a=
   $pattern_match_line= by p=
   $option_before_or_after= by o=

<b>Usage     </b> : &insert_lines_anywhere(\@files, \$inst_str,'after', \@match_str);
<b>Version   </b> : 1.4
<b>Warning   </b> : Case Insensitive by default.
</pre><hr>

<H3><A href="get_all_dirs_from_ENV
.pl.html">get_all_dirs_from_ENV
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_all_dirs_from_ENV
.pl">get_all_dirs_from_ENV
.pl</A>
<pre>
<b>Argument  </b> : NONE
<b>Example   </b> : my(@default_env_dirs) = @{&get_all_dirs_from_ENV}; in handle_arguments
<b>Function  </b> : extracts all the directories from %ENV  setting.
<b>Options   </b> : None
<b>Returns   </b> : a ref. of an array of directories.
<b>Usage     </b> : my(@default_env_dirs) = @{&get_all_dirs_from_ENV};
<b>Version   </b> : 1.0
<b>Warning   </b> : produces repetitive pathes (ie, can output identical path several times)
</pre><hr>

<H3><A href="get_path_dirs_from_ENV
.pl.html">get_path_dirs_from_ENV
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_path_dirs_from_ENV
.pl">get_path_dirs_from_ENV
.pl</A>
<pre>
<b>Argument  </b> : NONE
<b>Example   </b> : my(@default_env_dirs) = @{&get_path_dirs_from_ENV}; in handle_arguments
<b>Function  </b> : extracts path directories from %ENV  setting.
<b>Options   </b> : None
<b>Returns   </b> : a ref. of an array of directories.
<b>Usage     </b> : my(@default_env_dirs) = @{&get_path_dirs_from_ENV};
<b>Version   </b> : 1.0
<b>Warning   </b> : Replaces '.' to $pwd.
</pre><hr>

<H3><A href="handle_arguments_old
.pl.html">handle_arguments_old
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//handle_arguments_old
.pl">handle_arguments_old
.pl</A>
<pre>
<b>Argument  </b> : one single ref. (\@input_args);
<b>Function  </b> : Sub argument handling for opening files with options. General
             form of 'handle_arguments_xxxx', while xxxx can be files, hashes, arrays,,,,
<b>Options   </b> : None yet, extendable by adding refs. of something.
<b>Returns   </b> : an array of refs for file names, hashes, arrays and  the opion string
<b>Usage     </b> : my(@in)=&handle_arguments_old(\@input_args);   Do not dereference it.
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="mv
.pl.html">mv
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//mv
.pl">mv
.pl</A>
<pre>
<b>Argument  </b> : 2 references of file name or  2 file names.
<b>Author    </b> : Larry Wall, Jong
<b>Example   </b> : mv("mv.pl", *STDOUT);  # This will print mv.pl contents to your screen.
<b>Function  </b> : moves files fast, replacement of 'system("mv xxx xxxx"); '
<b>Keywords  </b> : move files fast. mv_file, mv_files, move_files, move_file
<b>Usage     </b> : &mv( \$srcFile, \$dstFile); or  &mv( $srcFile, $dstFile);
             or &mv(FILEHANDLE1, FILEHANDLE2),  or  &mv(FILEHANDLE1, $output)
<b>Version   </b> : 1.4
<b>Warning   </b> : 27 times slower than 'mv' at prompt.  using system is 32 times slower
</pre><hr>

<H3><A href="cp
.pl.html">cp
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//cp
.pl">cp
.pl</A>
<pre>
<b>Argument  </b> : 2 references of file name or  2 file names.
<b>Author    </b> : Larry Wall, Jong
<b>Example   </b> : cp("cp.pl", *STDOUT);  # This will print cp.pl contents to your screen.
<b>Function  </b> : copies files fast, replacement of 'system("cp xxx xxxx"); '
<b>Keywords  </b> : copy files fast. cp_file, cp_files, copy_files, copy_file
<b>Usage     </b> : &cp( \$srcFile, \$dstFile); or  &cp( $srcFile, $dstFile);
             or &cp(FILEHANDLE1, FILEHANDLE2),  or  &cp(FILEHANDLE1, $output)
<b>Version   </b> : 1.4
<b>Warning   </b> : 27 times slower than 'cp' at prompt.  using system is 32 times slower
</pre><hr>

<H3><A href="wh
.pl.html">wh
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//wh
.pl">wh
.pl</A>
<pre>
<b>Function  </b> : shows the path for a file you want
             similar to which in UNIX
</pre><hr>

<H3><A href="condense_script
.pl.html">condense_script
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//condense_script
.pl">condense_script
.pl</A>
<pre>
<b>Argument  </b> : one or more files.
<b>Example   </b> : condense_script.pl th_lib.pl th-test.pl xxx xxxx ....
<b>Function  </b> : makes compact size subroutines of developed perl codes
<b>Options   </b> : None
<b>Returns   </b> : xxxxxx.pl.out  but sub routines condensed.
<b>Usage     </b> : condense_script.pl  xxxxxx.pl
<b>Version   </b> : 1.0
<b>Warning   </b> : The only condition is that you need to have 'sub xxxxx' from the
             first column and the last '}' should be again at the first column
             This is due to the pattern matching for any sub routines.
</pre><hr>

<H3><A href="initialize_code
.pl.html">initialize_code
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//initialize_code
.pl">initialize_code
.pl</A>
<pre>
<b>Argument  </b> : None
<b>Function  </b> : initialize all developing codes by putting Header section infor
<b>Returns   </b> : None
<b>Usage     </b> : &initialze_code;
<b>Version   </b> : 1.0
<b>Warning   </b> : This writes over the program you run (itself). temp file is ini_code.temp
</pre><hr>

<H3><A href="parse_arguments
.pl.html">parse_arguments
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//parse_arguments
.pl">parse_arguments
.pl</A>
<pre>
<b>Argument  </b> : uses @ARGV
<b>Example   </b> : &parse_arguments(1);
             @files=@{&parse_arguments(1)};
<b>Function  </b> : Parse and assign any types of arguments on prompt in UNIX to
             the various variables inside of the running program.
             This is more visual than getopt and easier.
             just change the option table_example below for your own variable
             setttings. This program reads itself and parse the arguments
             according to the setting you made in this subroutine or
             option table in anywhere in the program.
             It also imports the ENV variables to your program.

<b>Keywords  </b> : pass_arguments
<b>Options   </b> : '0'  to specify that there is no argument to sub, use
              &parse_arguments(0);
             parse_arguments itself does not have any specific option.
             '#' at prompt will make a var  $debug set to 1. This is to
              print out all the print lines to make debugging easier.

             'e=xxxx' for filtering input files by extension xxxx

<b>Returns   </b> : Filenames in a reference of array
             and input files in an array (file1, file2)=@{&parse_arguments};
<b>Usage     </b> : &parse_arguments; or  (file1, file2)=@{&parse_arguments};
<b>Version   </b> : 2.1
<b>Warning   </b> : HASH and ARRAY mustn't be like = (1, 2,3) or (1,2 ,3)
</pre><hr>

<H3><A href="assign_options_to_variables
.pl.html">assign_options_to_variables
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//assign_options_to_variables
.pl">assign_options_to_variables
.pl</A>
<pre>
<b>Argument  </b> : None.
<b>Example   </b> : When you want to set 'a' char to a variable called '$dummy' in
             the program, you put a head box commented line
             '#  $dummy    becomes  a  by  -a '
             Then, the parse_arguments and this sub routine will read the head
             box and assigns 'a' to $dummy IF you put an argument of '-a' in
             the prompt.
<b>Function  </b> : Assigns the values set in head box to the variables used in
             the programs according to the values given at prompt.
             This produces global values.
             When numbers are given at prompt, they go to @num_opt
              global variable. %vars global option will be made

<b>Options   </b> : '#' at prompt will make a var  $debug set to 1. This is to
              print out all the print lines to make debugging easier.
<b>Returns   </b> : Some globaly used variables according to prompt options.
             @num_opt,

<b>Usage     </b> : &assign_options_to_variables(\$input_line);
<b>Version   </b> : 2.8
<b>Warning   </b> : This is a global vars generator!!!
</pre><hr>

<H3><A href="read_head_box
.pl.html">read_head_box
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//read_head_box
.pl">read_head_box
.pl</A>
<pre>
<b>Argument  </b> : One or None. If you give an argu. it should be a ref. of an ARRAY
              or a filename, or ref. of a filename.
             If no arg is given, it reads SELF, ie. the program itself.
<b>Example   </b> : Output is something like
             ('Title', 'read_head_box', 'Tips', 'Use to parse doc', ...)
<b>Function  </b> : Reads the introductory header box(the one you see on top of sub routines of
             Jong's programs.). Make a hash(associative array) to put entries
             and descriptions of the items. The hash values have new lines '\n' are
             attached, so that later write_head_box just sorts Title to the top
             and prints without much calculation.
             This is similar to read_head_box, but
             This has one long straight string as value(no \n inside)
             There are two types of ending line one is Jong's #---------- ...
             the other is Astrid's  #*************** ...
<b>Keywords  </b> : open_head_box, open_headbox, read_headbox
<b>Options   </b> : 'b' for remove blank lines. This will remove all the entries
             with no descriptions
<b>Returns   </b> : A hash ref.
<b>Usage     </b> : %entries = %{&read_head_box([\$file_to_read, \@BOXED ] )};
<b>Version   </b> : 2.7
</pre><hr>

<H3><A href="read_first_head_box
.pl.html">read_first_head_box
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//read_first_head_box
.pl">read_first_head_box
.pl</A>
<pre>
<b>Function  </b> : Reads the header box(the one you see on top of sub routines of
             Jong's programs.)
             There are two types of ending line one is Jong's #---------- ...
             the other is Astrid's  #*************** ...
<b>Usage     </b> : %entries = %{&read_first_head_box(\$file_to_read )};
<b>Version   </b> : 2.0
</pre><hr>

<H3><A href="read_head_boxes
.pl.html">read_head_boxes
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//read_head_boxes
.pl">read_head_boxes
.pl</A>
<pre>
<b>Argument  </b> : one or more filenames
<b>Example   </b> : @hashes = @{&read_head_boxes(@ARGV)};
             $num_of_sub = @hashes;
             print "\n Number of subs was $num_of_sub\n";
<b>Function  </b> : Reads the introductory header box(the one you see on top of sub routines of
             Jong's programs.). Make a hash(associative array) to put entries
             and descriptions of the items.
<b>Returns   </b> : A hash ref.
<b>Usage     </b> : %entries = %{&read_head_box(\$file_to_read, ,,, )};
<b>Version   </b> : 1.2
</pre><hr>

<H3><A href="read_head_box2
.pl.html">read_head_box2
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//read_head_box2
.pl">read_head_box2
.pl</A>
<pre>
<b>Argument  </b> : One or None. If you give an argu. it should be a ref. of an ARRAY
              or a filename, or ref. of a filename.
             If no arg is given, it reads SELF, ie. the program itself.
<b>Example   </b> : Output is something like
             ('Title', 'read_head_box', 'Tips', 'Use to parse doc', ...)
<b>Function  </b> : Reads the header box(the one you see on top of sub routines of
             Jong's programs.). This is similar to read_head_box, but
             This has one long straight string as value(no \n inside)
             There are two types of ending line one is Jong's #---------- ...
             the other is Astrid's  #*************** ...
<b>Options   </b> : 'b' for remove blank lines. This will remove all the entries
             with no descriptions
<b>Returns   </b> : A hash ref.
<b>Usage     </b> : %entries = %{&read_head_box(\$file_to_read )};
<b>Version   </b> : 1.5
</pre><hr>

<H3><A href="read_all_head_boxes
.pl.html">read_all_head_boxes
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//read_all_head_boxes
.pl">read_all_head_boxes
.pl</A>
<pre>
<b>Function  </b> : Reads the header boxes(the one you see on top of sub routines of
             Jong's programs.)
             There are two types of ending line one is Jong's #---------- ...
             the other is Astrid's  #*************** ...
<b>Usage     </b> : %entries = %{&read_all_head_box(\$file_to_read )};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="correct_head_box
.pl.html">correct_head_box
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//correct_head_box
.pl">correct_head_box
.pl</A>
<pre>
<b>Argument  </b> : a filename
<b>Example   </b> : correct_head_box.pl Bio.pl
<b>Function  </b> : Makes headbox in right and updated format. The most
            updated headbox format is very this headbox. So, to
            change all other headbox format, change this first.
<b>Usage     </b> : just type correct_head_box.pl with a file name.
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="read_correct_head_box
.pl.html">read_correct_head_box
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//read_correct_head_box
.pl">read_correct_head_box
.pl</A>
<pre>
<b>Function  </b> : This reads correct_head_box only.
<b>Keywords  </b> : read_update_head_box, read update headbox
<b>Options   </b> : v  for verbose message printing.
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="write_head_box
.pl.html">write_head_box
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//write_head_box
.pl">write_head_box
.pl</A>
<pre>
<b>Function  </b> : gets a hash ref. and writes the head box for a subroutine
<b>Keywords  </b> : write_headbox
<b>Options   </b> : v  for verbose representation. This will print boxes on STDOUT
            n  for no '#' leader.
<b>Version   </b> : 2.2
</pre><hr>

<H3><A href="read_option_table
.pl.html">read_option_table
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//read_option_table
.pl">read_option_table
.pl</A>
<pre>
<b>Function  </b> : Reads the option table made by Jong in any perl script. The
             option table is a box with separators.
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="show_default_help
.pl.html">show_default_help
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//show_default_help
.pl">show_default_help
.pl</A>
<pre>
<b>Example   </b> : &show_default_help2; &show_default_help2(\$arg_num_limit);   &show_default_help2( '3' );
             1 scalar digit for the minimum number of arg (optional),
             or its ref. If this defined, it will produce exit the program
             telling the minimum arguments.
<b>Function  </b> : Prints usage information and others when invoked. You need to have
             sections like this explanation box in your perl code. When invoked,
             show_default_help routine reads the running perl code (SELF READING) and
             displays what you have typed in this box.
             After one entry names like # Function :, the following lines without
             entry name (like this very line) are attached to the previous entry.
             In this example, to # Function : entry.
<b>Keywords  </b> : default_help
<b>Returns   </b> : formated information
<b>Usage     </b> : &show_default_help2;  usually with 'parse_arguments' sub.
<b>Version   </b> : 3.4
<b>Warning   </b> : this uses format and references
</pre><hr>

<H3><A href="default_help
.pl.html">default_help
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//default_help
.pl">default_help
.pl</A>
<pre>
<b>Example   </b> : &show_default_help2; &show_default_help2(\$arg_num_limit);   &show_default_help2( '3' );
             1 scalar digit for the minimum number of arg (optional),
             or its ref. If this defined, it will produce exit the program
             telling the minimum arguments.
<b>Function  </b> : Prints usage information and others when invoked. You need to have
             sections like this explanation box in your perl code. When invoked,
             show_default_help routine reads the running perl code (SELF READING) and
             displays what you have typed in this box.
             After one entry names like # Function :, the following lines without
             entry name (like this very line) are attached to the previous entry.
             In this example, to # Function : entry.
<b>Keywords  </b> : default_help
<b>Returns   </b> : formated information
<b>Usage     </b> : &show_default_help2;  usually with 'parse_arguments' sub.
<b>Version   </b> : 3.4
<b>Warning   </b> : USE show_default_help, This is not action oriented
</pre><hr>

<H3><A href="show_default_help_old
.pl.html">show_default_help_old
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//show_default_help_old
.pl">show_default_help_old
.pl</A>
<pre>
<b>Argument  </b> : 1 scalar digit for the minimum number of arg (optional),
             or its ref. If this defined, it will produce exit the program
             telling the minimum arguments.
<b>Example   </b> : &show_default_help; &show_default_help(\$arg_num_limit);   &show_default_help( '3' );
<b>Function  </b> : prints usage information and others when invoked. You need to have
             sections like this explanation box in your perl code. When invoked,
             show_default_help routine reads the running perl code (self reading) and
             displays what you have typed in this box.
             After one entry names like # Function :, the following lines without
             entry name (like this very line) are attached to the previous entry.
             In this example, to # Function : entry.
<b>Returns   </b> : formated information
<b>Usage     </b> : &show_default_help;  usually with 'parse_arguments' sub.
<b>Version   </b> : 2.0
<b>Warning   </b> : this uses format and references
</pre><hr>

<H3><A href="print_seq_in_block
.pl.html">print_seq_in_block
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//print_seq_in_block
.pl">print_seq_in_block
.pl</A>
<pre>
<b>Argument  </b> : many refs  for hash (one for bottm, one for top, etc,top hash is usually
               to denote certain caculations or results of the bottom one
<b>Example   </b> : If there are 3 hashes output will be; (in the order of \%hash3, \%hash2, \%hash1)
             >> 1st Hash        >> 2nd Hash         >> 3rd Hash
             Name1  THIS-IS-    Name123  eHHHHHHH   Name123  12222223

             You will get;
                            Name1    THIS-IS-
                            Name123  eHHHHHHH
                            Name123  12222223

             Example of ( no option, DEFAULT )  # Example of ('i' or 'I' option,
                                                                INTERLACE )
             6taa           ----ATPADWRSQSIY    #   6taa       ------ATPADWRSQSIY
             2aaa           ------LSAASWRTQS    #   6taa       ------CCHHHHCCCCEE
             1cdg           APDTSVSNKQNFSTDV    #   6taa       ------563640130000

             6taa           ------CCHHHHCCCC    #   2aaa       ------LSAASWRTQSIY
             2aaa           ------CCHHHHCCCC    #   2aaa       ------CCHHHHCCCCEE
             1cdg           CCCCCCCCCCCCCCCC    #   2aaa       ------271760131000

             6taa           ------5636401300    #   1cdg       APDTSVSNKQNFSTDVIY
             2aaa           ------2717601310    #   1cdg       CCCCCCCCCCCCCCCCEE
             1cdg           6752327236000000    #   1cdg       675232723600000000

             Example of('s' or 'S' option,SORT) # Example of ('o' or 'O' option,
                                                        ORDERED by input hashes )
             1cdg           APDTSVSNKQNFSTDV    #   6taa       ------ATPADWRSQSIY
             2aaa           ------LSAASWRTQS    #   2aaa       ------LSAASWRTQSIY
             6taa           ------ATPADWRSQS    #   1cdg       APDTSVSNKQNFSTDVIY

             1cdg           CCCCCCCCCCCCCCCC    #   6taa       ------CCHHHHCCCCEE
             2aaa           ------CCHHHHCCCC    #   2aaa       ------CCHHHHCCCCEE
             6taa           ------CCHHHHCCCC    #   1cdg       CCCCCCCCCCCCCCCCEE

             1cdg           6752327236000000    #   6taa       ------563640130000
             2aaa           ------2717601310    #   2aaa       ------271760131000
             6taa           ------5636401300    #   1cdg       675232723600000000

<b>Function  </b> : gets many refs  for one scalar  or hashes and prints
               the contents in lines of \$block_leng(the only scalar ref. given) char.
<b>Keywords  </b> : print_sequence_in_block print_alignment_in_block
<b>Options   </b> : 'o' or 'O' => ordered hash print,
             'n' or'N' => no space between blocks.
             's' or 'S' => printout sorted by seq names.
             'i' or 'I' => interlaced print.(this requires identical names in hashes)
             'v' or 'V' => show sequence start number at each line
             'g' or 'G' => with gap chars between  aa residues
              l= for block length. Default is 60 char
              t= for specifying the length of seq names shown
              t  for truncating seq names shwn to 12 chars.
              f= for file output  eg. f=XXXXXX.issa
              r=digit-digit  (eg. 10-70) to take only the defined region of sequences
            digit-digit  (eg. 10-70) to take only the defined region of sequences

            just digit  for block length

             (all options can be like \$sort
             while $sort has 's' as value. naked number like 100 will be the
             block_length. 'i' or 'I' => interlaced print.(this requires
             identical names in hashes)
<b>Usage     </b> : &print_seq_in_block (\$block_leng, 'i',\%h1, 'sort', \%h2, \%hash3,,,);
<b>Version   </b> : 1.6
</pre><hr>

<H3><A href="print_seq_in_columns
.pl.html">print_seq_in_columns
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//print_seq_in_columns
.pl">print_seq_in_columns
.pl</A>
<pre>
<b>Argument  </b> : many refs  for hash (one for bottm, one for top, etc,top hash is usually
               to denote certain caculations or results of the bottom one
<b>Example   </b> : With command 'print_seq_in_columns.pl c2 s2', you get:

      name1 11111111  name1 22222
      name2 11        name2 2222222
      name3 1111111   name3 22222
      name4 11111     name4 2222
      name5 11111     name5 222

      name1 3333      name1 4444
      name2 3333      name2 444
      name3 333       name3 4
      name4 333       name4 4444
      name5 3333      name5 4444444

<b>Function  </b> : gets many refs  for one scalar  or hashes and prints
               the contents in lines of \$block_leng(the only scalar ref. given) char.
<b>Options   </b> : c, i, s
<b>Usage     </b> : &print_seq_in_block (\$block_leng, 'i',\%h1, 'sort', \%h2, \%hash3,,,);
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="convert_arr_and_str_2_hash
.pl.html">convert_arr_and_str_2_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//convert_arr_and_str_2_hash
.pl">convert_arr_and_str_2_hash
.pl</A>
<pre>
<b>Argument  </b> : one or more ref. of arrays
<b>Example   </b> : &print_seq_in_block(&convert_arr_and_str_2_hash(\@input,\@input2,\@input3 ));
             &convert_arr_and_str_2_hash(\$input1,\$input2, '2' );
             results in; (ordering starts from the given '2')
                          array_2       input1 arraystring
                          array_3       input2 arraystring

             one more exam
                          string_6       This is st                  and 3 strings)
                          string_10      This is st
                          array_2        111233434242
                          array_6        111233434242
                          array_10       111243424224
<b>Function  </b> : makes hash(es) out of array(s)
             if ordering digit(s) is put, it orders the keys according to it.
             if ordering digit is not increased by one, the difference is used
             as the increasing factor. No option results in
             array_1, array_2, array_3...

<b>Returns   </b> : one or more ref. of hashes.
<b>Usage     </b> : ($hash1, $hash2)=&convert_arr_and_str_2_hash(\$input, \$input2, '1', '2'.. );
             * This is the combination of convert_string_to_hash & convert_array_to_hash
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="convert_string_to_hash
.pl.html">convert_string_to_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//convert_string_to_hash
.pl">convert_string_to_hash
.pl</A>
<pre>
<b>Argument  </b> : one or more ref. of arrays
<b>Example   </b> : &print_seq_in_block(&convert_string_to_hash(\$input,\$input2,\$input3 ));
             &convert_string_to_hash(\$input1,\$input2, '2' );
             results in; (ordering starts from the given '2')
                          string_2       input1 string
                          string_3       input2 string

<b>Function  </b> : makes hash(es) out of string(s)
             if ordering digit(s) is put, it orders the keys according to it.
             if ordering digit is not increased by one, the difference is used
             as the increasing factor. No option results in
             string_1, string_2, string_3...

<b>Returns   </b> : one or more ref. of hashes.
<b>Usage     </b> : ($hash1, $hash2)=&convert_string_to_hash(\$input, \$input2, '1', '2'.. );
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="convert_array_to_hash
.pl.html">convert_array_to_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//convert_array_to_hash
.pl">convert_array_to_hash
.pl</A>
<pre>
<b>Argument  </b> : one or more ref. of arrays
<b>Example   </b> : &print_seq_in_block(&convert_array_to_hash(\@input,\@input2,\@input3 ));
             &convert_array_to_hash(\$input1,\$input2, '2' );
             results in; (ordering starts from the given '2')
                          array_2       input1 arraystring
                          array_3       input2 arraystring

<b>Function  </b> : makes hash(es) out of array(s)
             if ordering digit(s) is put, it orders the keys according to it.
             if ordering digit is not increased by one, the difference is used
             as the increasing factor. No option results in
             array_1, array_2, array_3...

<b>Returns   </b> : one or more ref. of hashes.
<b>Usage     </b> : ($hash1, $hash2)=&convert_array_to_hash(\$input, \$input2, '1', '2'.. );
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="remove_dup_in_array
.pl.html">remove_dup_in_array
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//remove_dup_in_array
.pl">remove_dup_in_array
.pl</A>
<pre>
<b>Argument  </b> : one or more refs for arrays or one array.
<b>Example   </b> : (1,1,1,1,3,3,3,3,4,4,4,3,3,4,4);  --> (1,3,4);
<b>Function  </b> : removes duplicate entries in an array. You can sort the
             result if you wish by 's' opt. Otherwise, result will keep
             the original order
<b>Keywords  </b> : merge array elements, remove_repeting_elements,
             remove_same_array_elements, remove_redundancy, remove_redundant_elements
             remove_duplication_in_array
<b>Options   </b> : 
   s  for sorting the array output
<b>Returns   </b> : one or more references.
<b>Usage     </b> : @out2 = @{&remove_dup_in_array(\@input1, \@input2,,,,)};
             @out1 = &remove_dup_in_array(\@input1 );
<b>Version   </b> : 1.6
</pre><hr>

<H3><A href="remove_text
.pl.html">remove_text
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//remove_text
.pl">remove_text
.pl</A>
<pre>
<b>Argument  </b> : reference of one array of file names in pwd
<b>Author    </b> : jong
<b>Function  </b> : finds patterns of text and replaces them in multiple input files
<b>Returns   </b> : nothing
<b>Usage     </b> : &remove_text(\@input_array_of_filenames);
<b>Version   </b> : 1.3
<b>Warning   </b> : This produces a temporary file and rename it...
</pre><hr>

<H3><A href="remove_elements_by_pattern
.pl.html">remove_elements_by_pattern
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//remove_elements_by_pattern
.pl">remove_elements_by_pattern
.pl</A>
<pre>
<b>Argument  </b> : one or more refs for arrays. The first array is always the
             only target.
<b>Example   </b> : @TARGET=qw(1 % $ ^ # A B 4444 44 4 4 3 33 3 11 A 3 4 4 7 AB);
              @remove=qw(\W);  # removes all the non word stuff
              @remove2=qw(\d );
              @out=@{&remove_elements_by_pattern(\@TARGET, \@remove,\@remove2)};
<b>Function  </b> : removes elements by pattern in the array
<b>Keywords  </b> : remove_this_elements, remove_these_elements, remove_elements
             remove_elements_by_position, kill_array_elements, kill_elements
             take_away_elements, remove_array_elements
<b>Returns   </b> : one or more references.
<b>Usage     </b> : @out2 = @{&remove_elements_by_pattern(\@input1, \@input2,,,,)};
             @out1 = @{&remove_elements_by_pattern(\@input1 )};
<b>Version   </b> : 1.2
</pre><hr>

<H3><A href="remove_elements_by_name
.pl.html">remove_elements_by_name
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//remove_elements_by_name
.pl">remove_elements_by_name
.pl</A>
<pre>
<b>Argument  </b> : one or more refs for arrays. The first array is always the
             only target. The removing elements can be scalar ref or
             just scalar.
<b>Example   </b> : ( two input:  (1,2,3,4,4,4,5,5,6,7), (1,3,4)  --> (2,5,5,6,7);
<b>Function  </b> : removes elements by name in the array
<b>Keywords  </b> : remove_this_elements, remove_these_elements, remove_elements
             remove_elements_by_position, kill_array_elements, kill_elements
             take_away_elements
<b>Returns   </b> : one or more references.
<b>Usage     </b> : @out2 = @{&remove_elements_by_name(\@input1, \@input2)};
             @out1 = @{&remove_elements_by_name(\@input1, \$name )};
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="remove_elements_by_position
.pl.html">remove_elements_by_position
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//remove_elements_by_position
.pl">remove_elements_by_position
.pl</A>
<pre>
<b>Argument  </b> : one or more refs for arrays. The first array is always the
             only target.
<b>Example   </b> : ( two input:  (1,2,3,4,5,6,7), (1,3,4)  --> (2 5 6 7);
<b>Function  </b> : removes elements by name in the array
<b>Keywords  </b> : remove_this_elements, remove_these_elements, remove_elements
             remove_elements_by_position, kill_array_elements, kill_elements
             take_away_elements
<b>Returns   </b> : one or more references.
<b>Usage     </b> : @out2 = @{&remove_elements_by_position(\@input1, \@input2,,,,)};
             @out1 = @{&remove_elements_by_position(\@input1 )};
<b>Version   </b> : 1.1
<b>Warning   </b> : Position 1 means $array[0]
</pre><hr>

<H3><A href="merge_array
.pl.html">merge_array
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//merge_array
.pl">merge_array
.pl</A>
<pre>
<b>Argument  </b> : one or more refs for arrays.
<b>Example   </b> : (1,1,1,1,3,3,3,3,4,4,4,3,3,4,4);  --> (1,3,4);
<b>Function  </b> : removes duplicate entries in an array. If you put
             more than one array as inputs, it will produce references of
             arrays merged singly. Each resulting array is independant.
             CF. merge_many_arrays
<b>Keywords  </b> : merge array elements, merge_array_elements,
<b>Returns   </b> : one or more references.
<b>Usage     </b> : @out2 = @{&merge_array(\@input1, \@input2,,,,)};
             @out1 = @{&merge_array(\@input1 )};
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="make_one_array
.pl.html">make_one_array
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//make_one_array
.pl">make_one_array
.pl</A>
<pre>
<b>Argument  </b> : Two or more refs for arrays.
<b>Example   </b> : (1,2,3,4,5),(6,7,8,9,10)-----> (1,2,3,4,5,6,7,8,9,10)
<b>Function  </b> : makes one array from several
<b>Keywords  </b> : make_one_array, make_one_from_several
<b>Returns   </b> : An array reference
<b>Usage     </b> : @array_one=@{&make_one_array(\@input_array_1, \@input_array_2)};
<b>Version   </b> : 1.0
<b>Warning   </b> : This does not remove duplicate entries.
</pre><hr>

<H3><A href="get_multiple_array_entry
.pl.html">get_multiple_array_entry
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_multiple_array_entry
.pl">get_multiple_array_entry
.pl</A>
<pre>
<b>Argument  </b> : one or more refs for arrays.
<b>Example   </b> : (1,1,1,1,3,3,3,3,4,4,4,3,3,4,4);  --> (1,3,4);
             if you put two arrays(1,1,1,3,3, 100) and (2,2, 4,4, 100), you will get
             references of arrays( 1,3) and (2,4) ignoring single array entries.
<b>Function  </b> : Gets any multiple array entry in a given array. If more than
             one array is given, each array will have a reference return.
<b>Keywords  </b> : multiple entry array, get_common_entry_array
<b>Returns   </b> : one or more references.
<b>Usage     </b> : @out2 = @{&merge_array(\@input1, \@input2,,,,)};
             @out1 = @{&merge_array(\@input1 )};
</pre><hr>

<H3><A href="get_common_array_entry
.pl.html">get_common_array_entry
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_common_array_entry
.pl">get_common_array_entry
.pl</A>
<pre>
<b>Argument  </b> : one or more refs for arrays.
<b>Example   </b> : (1,1,1,2,3,3,3,4)                 --> (1,3);
             (1,2,3) (1,2,3,4,5)               --> (1,2,3);
             (1,2,3,4,5) (1,2,3,4,5) (3,4,5,6) --> (4,5);
<b>Function  </b> : Gets any common array entry in given arrays. If one single array
             is given, mutiply occurring entries in the array will be returned.
<b>Keywords  </b> : multiple entry array, get_common_entry_array, multiply array,
             get_common_array_elements, get_common_array_element,
             get_dup_array_elements,
<b>Returns   </b> : one or more references.
<b>Usage     </b> : @out2 = @{&get_common_array_entry(\@input1, \@input2,,,,)};
             @out1 = @{&get_common_array_entry(\@input1 )};
<b>Version   </b> : 1.1
<b>Warning   </b> : accepts only references of arrays(others are ignored).
</pre><hr>

<H3><A href="merge_many_arrays
.pl.html">merge_many_arrays
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//merge_many_arrays
.pl">merge_many_arrays
.pl</A>
<pre>
<b>Argument  </b> : one or more refs for arrays. or just arrays.
<b>Example   </b> : (1,1,1,1,3,3), (1,3,3,4,4,4,3,3,4,4);  --> (1,3,4);
<b>Function  </b> : removes duplicate entries in multiple array inputs.
<b>Keywords  </b> : merge array elements from multiple arrays. merge_array_elements
<b>Returns   </b> : one reference.
<b>Usage     </b> : @out2 = @{&merge_many_arrays(\@input1, @inputX, \@input2,,,,)};
             @out1 = @{&merge_many_arrays(\@input1 )};
<b>Version   </b> : 1.0
<b>Warning   </b> : synonym of  remove_dup_in_array
</pre><hr>

<H3><A href="remove_repetitives_in_array
.pl.html">remove_repetitives_in_array
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//remove_repetitives_in_array
.pl">remove_repetitives_in_array
.pl</A>
<pre>
<b>Argument  </b> : one or more refs for arrays.
<b>Example   </b> : (1,1,1,1,3,3,3,3,4,4,4,3,3,4,4);  --> (1,3,4);
<b>Function  </b> : removes duplicate entries in an array. If you put
             more than one array as inputs, it will produce references of
             arrays merged singly. Each resulting array is independant.
             CF. merge_many_arrays
<b>Keywords  </b> : remove_dup_in_array, merge array elements, remove_duplicates,
<b>Returns   </b> : one or more references.
<b>Usage     </b> : @out2 = @{&remove_repetitives_in_array(\@input1, \@input2,,,,)};
             @out1 = @{&remove_repetitives_in_array(\@input1 )};
<b>Version   </b> : 1.0
<b>Warning   </b> : synonym of  remove_dup_in_array
</pre><hr>

<H3><A href="filter_hash_by_num_value
.pl.html">filter_hash_by_num_value
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//filter_hash_by_num_value
.pl">filter_hash_by_num_value
.pl</A>
<pre>
<b>Function  </b> : returns hash refs. after filtering with threshold value.
<b>Usage     </b> : ($ref1, $ref2, $ref3)=&filter_hash_by_num_value(\%h1, \$thres,...);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="dir_search_single
.pl.html">dir_search_single
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//dir_search_single
.pl">dir_search_single
.pl</A>
<pre>
<b>Argument  </b> : One Ref. for a scalar.
<b>Function  </b> : With given full path or single name for a dir. it returns
             the full path dir name. If it fails to find in pwd or given
             specified path, it tries to search PATH, HOME etc..
<b>Returns   </b> : one Ref. for an array.
<b>Usage     </b> : $output_best_possible_dir = ${&dir_search_single(\$input_name)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="dir_search
.pl.html">dir_search
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//dir_search
.pl">dir_search
.pl</A>
<pre>
<b>Argument  </b> : One Ref. for a scalar.
<b>Function  </b> : With given full path or single name for a dir. it returns
             the full path dir name. If it fails to find in pwd or given
             specified path, it tries to search PATH, HOME etc..
<b>Returns   </b> : one Ref. for an array.
<b>Usage     </b> : @output_possible_dirs = @{&dir_search(\$input_name)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="break_down_clu_file
.pl.html">break_down_clu_file
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//break_down_clu_file
.pl">break_down_clu_file
.pl</A>
<pre>
<b>Example   </b> : INPUT looks like this:>

 Cluster 1330106       # Ori:9    Sub:6    From:133
   1  1 EC2987          1-332       1
   1  1 HI0530          1-338       1
   1  1 MJ1130          5-302       1
   1  1 SLR0807         3-333       1
   1  1 D09_ORF319      6-312       1
   1  1 MG046           6-307       1
 Cluster 1330203       # Ori:9    Sub:3    From:133
   1  1 SLL1063         16-209      1
   1  1 HI0388          4-236       1
   1  1 EC1764          2-224       1
 Cluster 1330302       # Ori:9    Sub:2    From:133
   1  1 EC2987          1-118       1
   1  1 HI0388          4-104       1

<b>Function  </b> : breaks down the single linkage cluster into smaller clusters
<b>Keywords  </b> : split_clu_files
<b>Options   </b> : 

  pattern= 'Cluster'

<b>Usage     </b> : &break_down_clu_file(\@file);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="exchange_query_with_match_in_msp
.pl.html">exchange_query_with_match_in_msp
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//exchange_query_with_match_in_msp
.pl">exchange_query_with_match_in_msp
.pl</A>
<pre>
<b>Keywords  </b> : exchange_msp_columns, open_and_exchange_query_with_match_in_msp,
             open_msp_files_with_exchange_of_columns
<b>Options   </b> : 
          R     for NO range attachment in Name only return option (n)
          e=    for evalue threshhold, if e=1, ignores all which are over 1
          s=    for score threshhold if t=100, ignores all which are less 100
<b>Usage     </b> : @exchanged_msp=@{&exchange_query_with_match_in_msp(\@file)};
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="run_fasta_sequence_search
.pl.html">run_fasta_sequence_search
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//run_fasta_sequence_search
.pl">run_fasta_sequence_search
.pl</A>
<pre>
<b>Author    </b> : Jong Park, jong@salt2.med.harvard.edu, for commercial use, ask me.
<b>Keywords  </b> : run_ssearch_sequence_search, do_fasta_sequence_search
<b>Options   </b> : 
             Query_seqs=  for enquiry sequences eg)  "Query_seqs=$ref_of_hash"
             DB=   for target DB  "DB=$DB_used"
             File= to get file base(root) name.  "File=$file[0]"
             i= to get file base(root) name. same as File=
             m  for MSP format directly from FASTA or Ssearch result than through sso_to_msp to save mem
             s  for the big single output (msp file output I mean)
             s= for the single big msp file name
             O= for Out file name, same as s=
             o  for overwrite existing xxxx.fa files for search
             c  for create SSO file (sequence search out file)
             d  for very simple run and saving the result in xxxx.gz format in sub dir starting with one char
             r  for reverse the query sequence
             R  for attaching ranges of sequences
             k= for k-tuple value. default is 1 (ori. FASTA prog. default is 2)
             u= for $upper_expect_limit
             l= for $lower_expect_limit
             a= for choosing either fasta or ssearch algorithm
             d= for defining the size of subdir made. 2 means it creates
                    eg, DE while 1 makes D
             d  for $make_gz_in_sub_dir_opt, putting resultant sso files in gz format and in single char subdir
             D  for $make_msp_in_sub_dir_opt, convert sso to msp and put in sub dir like /D/, /S/
             n  for new format to create new msp file format with sso_to_msp routine
          PVM=  for PVM run of FASTA (FASTA only)
             M  for machine readable format -m 10 option
             M= for machine readable format -m 10 option
             N  for 'NO' do not do any processing but, do the searches only.
       FILE_AGE for defining the age of file in days to be overwritten.
<b>Usage     </b> : $gzipped_msp_file=${&run_fasta_sequence_search("a=$algorithm",
                        "O=$out_file_msp_name", "File=$temp_file_name", "e=$E_val",
                        "DB=$sequence_DB", "k=$k_tuple", "$machine_readable")};

<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="do_self_blastp_search
.pl.html">do_self_blastp_search
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//do_self_blastp_search
.pl">do_self_blastp_search
.pl</A>
<pre>
<b>Keywords  </b> : run_blastp, run_blastp_seq_search, blastp_seq_search, blastp_search,
             do_blast_search
<b>Options   </b> : 
   r  for reverseing enquiry sequences
   T= for Blastp T param
   S= for Blastp S param
   B= for Blastp B param
   V= for Blastp V param
   E= for Blastp E param

<b>Usage     </b> : &do_blastp_search(\@file);
<b>Version   </b> : 1.3
</pre><hr>

<H3><A href="show_options
.pl.html">show_options
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//show_options
.pl">show_options
.pl</A>
<pre>
<b>Keywords  </b> : display_options, show_help_options, show_argument_options,
             show_options_in_headbox, show_prompt_options
<b>Usage     </b> : &show_options;  usually with 'parse_arguments' sub.
<b>Version   </b> : 1.2
</pre><hr>

<H3><A href="self_self_search
.pl.html">self_self_search
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//self_self_search
.pl">self_self_search
.pl</A>
<pre>
<b>Example   </b> : &self_self_search(\@file, $over_write, $msp_directly_opt, $create_sso, $single_big_msp);
<b>Function  </b> : self_to_self input database search with reverse query as an option
<b>Keywords  </b> : do_self_self_search, self_self_sequence_search, self_self_seq_search,
             self_to_self_search, self_to_rev_self_search, self_to_reversed_self_search
             search_self, search_self_seq, search_self
<b>Options   </b> : 
             Query_seqs=  for enquiry sequences eg)  "Query_seqs=$ref_of_hash"
             DB=   for target DB  "DB=$DB_used"
             File= to get file base(root) name.  "File=$file[0]"
             m  for MSP format directly from FASTA or Ssearch result than through sso_to_msp to save mem
             s  for the big single output (msp file output I mean)
             o  for overwrite existing xxxx.fa files for search
             c  for create SSO file (sequence search out file)
             r  for reverse the query sequence
             R  for attaching ranges of sequences
             b  for doing in batch. Reads all the seqs in memory at one time
             m10 for machine readable form
             k= for k-tuple value. default is 1 (ori. FASTA prog. default is 2)
             u= for $upper_expect_limit
             l= for $lower_expect_limit
             a= for choosing either fasta or ssearch algorithm
             d= for defining the size of subdir made. 2 means it creates
                    eg, DE while 1 makes D
             d  for $make_gz_in_sub_dir_opt, putting resultant sso files in gz format and in single char subdir
             D  for $make_msp_in_sub_dir_opt, convert sso to msp and put in sub dir like /D/, /S/
             n  for new format (msp2 format)
<b>Usage     </b> : &self_self_search(\@file, $over_write, $msp_directly_opt, $create_sso, $single_big_msp);
<b>Version   </b> : 2.2
</pre><hr>

<H3><A href="search_self
.pl.html">search_self
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//search_self
.pl">search_self
.pl</A>
<pre>
<b>Example   </b> : &search_self(\@file, $over_write, $msp_directly_opt,
                                 $create_sso, $single_big_msp);
<b>Function  </b> : self_to_self input database search with reverse query as an option
<b>Keywords  </b> : do_search_self, self_self_sequence_search, self_self_seq_search,
             self_to_self_search, self_to_rev_self_search, self_to_reversed_self_search
             search_self, search_self_seq
<b>Options   </b> : 
             Query_seqs=  for enquiry sequences eg)  "Query_seqs=$ref_of_hash"
             DB=   for target DB  "DB=$DB_used"
             File= to get file base(root) name.  "File=$file[0]"
             m  for MSP format directly from FASTA or Ssearch result
                       than through sso_to_msp to save mem
             s  for the big single output (msp file output I mean)
             o  for overwrite existing xxxx.fa files for search
             c  for create SSO file (sequence search out file)
             r  for reverse the query sequence
             R  for attaching ranges of sequences
             b  for doing in batch. Reads all the seqs in memory at one time
             m10 for machine readable form
             k= for k-tuple value. default is 1 (ori. FASTA prog. default is 2)
             u= for $upper_expect_limit
             l= for $lower_expect_limit
             a= for choosing either fasta or ssearch algorithm
             d= for defining the size of subdir made. 2 means it creates
                    eg, DE while 1 makes D
             d  for $make_gz_in_sub_dir_opt, putting resultant sso files
                       in gz format and in single char subdir
             D  for $make_msp_in_sub_dir_opt, convert sso to msp and
                      put in sub dir like /D/, /S/
             n  for new format (msp2 format)
<b>Usage     </b> : &search_self(\@file, $over_write, $msp_directly_opt,
                                      $create_sso, $single_big_msp);
<b>Version   </b> : 2.5
</pre><hr>

<H3><A href="make_cdf_file
.pl.html">make_cdf_file
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//make_cdf_file
.pl">make_cdf_file
.pl</A>
<pre>
<b>Function  </b> : gets all the clu files and producesf one xxxx.cdf file
             CDF file is a fasta database file with all the clu domains are
<b>Keywords  </b> : make_cdf_file_with_clu, clu_to_cdf, clu_2_cdf
<b>Usage     </b> : @file=@{&parse_arguments(1)};
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="make_seq_index_file
.pl.html">make_seq_index_file
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//make_seq_index_file
.pl">make_seq_index_file
.pl</A>
<pre>
<b>Function  </b> : creates xxxx.fa.idx file and makes a link to pwd. If @file contains
              names with .idx extension already, it will not put another idx
              index to it.
<b>Keywords  </b> : make_fasta_seq_index_file, create_seq_index_file, make_idx_file,
             create_idx_file, create_seq_idx_file, make_index_file, create_index_file
             make_sequence_index_file, create_sequene_index_file
<b>Usage     </b> : @idx_files_made=@{&make_seq_index_file(\@file)};
<b>Version   </b> : 1.4
</pre><hr>

<H3><A href="randomise_file_contents
.pl.html">randomise_file_contents
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//randomise_file_contents
.pl">randomise_file_contents
.pl</A>
<pre>
<b>Options   </b> : 
   V for NON-verbose
<b>Usage     </b> : &randomise_file_contents(@ARGV);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="filter_by_string_length
.pl.html">filter_by_string_length
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//filter_by_string_length
.pl">filter_by_string_length
.pl</A>
<pre>
<b>Keywords  </b> : find_palindromes, get_palindromes, find_palindrome, GetPalindrom
: filter_hash_by_string_length
<b>Options   </b> : 
  min=  for miniumum palindrome size
   p    for putting the position of the start of the palindrome
: 
  cutoff_min= by cutoff_min=
  c=  by c=     # the same as cutoff_min

<b>Usage     </b> : search_palindromes(\%seq, [\%seq2]);
: %out=%{&filter_by_string_length(\%hash, [100], ["cutoff=100"])};
<b>Version   </b> : 1.2
: 1.0
</pre><hr>

<H3><A href="filter_seq_DB_by_seq_length
.pl.html">filter_seq_DB_by_seq_length
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//filter_seq_DB_by_seq_length
.pl">filter_seq_DB_by_seq_length
.pl</A>
<pre>
<b>Keywords  </b> : filter_seq_file_by_seq_length
<b>Options   </b> : 
  cutoff_min= by cutoff_min=
  c=  by c=     # the same as cutoff_min
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="remove_dup_seq_entry
.pl.html">remove_dup_seq_entry
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//remove_dup_seq_entry
.pl">remove_dup_seq_entry
.pl</A>
<pre>
<b>Keywords  </b> : remove_duplicated_sequence_entries, remove_dup_sequence_entry
             remove_dup_seq_entries, remove_dup_sequences
<b>Usage     </b> : &remove_dup_seq_entry(\@file);  # while @file has 'xxx.fa'
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="sort_files_by_size
.pl.html">sort_files_by_size
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//sort_files_by_size
.pl">sort_files_by_size
.pl</A>
<pre>
<b>Function  </b> : sort files by size and returns the ref of the array
<b>Keywords  </b> : sort_file_by_size
<b>Usage     </b> : @sorted=@{&sort_files_by_size(\@files)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="make_fasta_files_from_msp_1_files
.pl.html">make_fasta_files_from_msp_1_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//make_fasta_files_from_msp_1_files
.pl">make_fasta_files_from_msp_1_files
.pl</A>
<pre>
<b>Example   </b> : &make_fasta_files_from_msp_1_files(\@files, "E=0.081", "l=0");
             &make_fasta_files_from_msp_1_files(\@files,
                    "E=$E_thresh",
                    "Seq_Source_DB=$seq_source_db",
                    "l=$lower_expect_limit",
                    "i=$cut_off_increase_factor",
                     $over_write_file);
<b>Function  </b> : creates fasta files for each query seq in xxxx.msp_1 file
<b>Options   </b> : 
   $seq_source_db= by "Seq_Source_DB=xxxxx.fa"
   $E_thresh = by E=  #  E value cutoff
   u= for $upper_expect_limit
   l= for $lower_expect_limit
   $over_write_file=o by o -o
   $cut_off_increase_factor = by i=
   s for selfless fasta out put (removes the original self seq among intermediates)

<b>Usage     </b> : &make_fasta_files_from_msp_1_files(\@files);

<b>Version   </b> : 1.2
</pre><hr>

<H3><A href="remove_dup_match_in_msp_files
.pl.html">remove_dup_match_in_msp_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//remove_dup_match_in_msp_files
.pl">remove_dup_match_in_msp_files
.pl</A>
<pre>
<b>Function  </b> : removes the exact duplicates in MSP files
<b>Keywords  </b> : remove_redundancy_in_msp_files, remove_redundant_matches_in_msp
             remove_redundant_matches, make_non_redundant_msp_files
<b>Usage     </b> : @out=@{&remove_dup_match_in_msp_files(\@file)};
<b>Version   </b> : 1.3
</pre><hr>

<H3><A href="filter_intermediates_by_E_value
.pl.html">filter_intermediates_by_E_value
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//filter_intermediates_by_E_value
.pl">filter_intermediates_by_E_value
.pl</A>
<pre>
<b>Function  </b> : filters intermediate sequences according to the E value
              thresholds and returns the lines in an array
<b>Usage     </b> : @filtered_msp3=@{&filter_intermediates_by_E_value(\@msp3,
                                           "E1=$E_value1", "E2=$E_value2")};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="make_intermediate_sequence_library
.pl.html">make_intermediate_sequence_library
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//make_intermediate_sequence_library
.pl">make_intermediate_sequence_library
.pl</A>
<pre>
<b>Function  </b> : extracts intermediate sequences from OWL fasta database to
             make intermediate seq library
             This looks for /gn0/jong/DB/PDB/PDB95D_against_OWL/E/$msp_file_gz
                and         /gn0/jong/DB/PDB/PDB95D_against_OWL/D/$msp_file_gz
<b>Keywords  </b> : make_interm_library_for_each_group, make_interm_lib,
             make_intermediate_library, compile_interm_library, create_interm_library,
<b>Options   </b> : 
      'FASTA_DB' for sequence source fasta file  eg:  "FASTA_DB=$source_db_fasta"
      o  for overwrite option (overwrites 1.2.3.fa like file)
      MSP_DIR= for msp seq file result directory
      m=       for msp seq file result direc (same as MSP_DIR)
      e=       for E value thresh
   $pdbg_file= by p=
      E=       for E value thresh
      s=       for score thresh

<b>Usage     </b> : &make_intermediate_sequence_library(\@files, "FASTA_DB=$owl_db_fasta");
               while @files have either pdbs or pdbg file (PDB grouping file)
<b>Version   </b> : 1.7
</pre><hr>

<H3><A href="read_machine_readable_sso_lines
.pl.html">read_machine_readable_sso_lines
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//read_machine_readable_sso_lines
.pl">read_machine_readable_sso_lines
.pl</A>
<pre>
<b>Keywords  </b> : read_m10_sso_lines read_msso_lines
<b>Options   </b> : a c r r2 n
             u= for upper E value limit
             l= for lower E value limit
<b>Usage     </b> : @out_refs=@{&read_machine_readable_sso_lines(\@SSO, $get_alignment,
                           $create_sso, $upper_expect_limit,$new_format, $lower_expect_limit,
                           $attach_range_in_names, $attach_range_in_names2)};
<b>Version   </b> : 1.5
</pre><hr>

<H3><A href="read_sso_lines
.pl.html">read_sso_lines
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//read_sso_lines
.pl">read_sso_lines
.pl</A>
<pre>
<b>Function  </b> : Main subroutine for open_sso_files. This calls either machine
              readable or unreadable form parsing subroutine
<b>Keywords  </b> : read_sso_lines_in_array
<b>Options   </b> : a c r r2 n
             u= for upper E value limit
             l= for lower E value limit
<b>Usage     </b> : &read_sso_lines([@sso], $create_sso, $attach_range_in_names,
                 $attach_range_in_names2, $new_format, $get_alignment) );
<b>Version   </b> : 1.4
</pre><hr>

<H3><A href="read_machine_unreadable_sso_lines
.pl.html">read_machine_unreadable_sso_lines
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//read_machine_unreadable_sso_lines
.pl">read_machine_unreadable_sso_lines
.pl</A>
<pre>
<b>Example   </b> : output will look=>
  ZFH1_DROME 60 d1ad3a_ 446 d1ad3a_ 0.9 71 34 3-37 253-287
  ZFH1_DROME 60 d1ahdp_ 68 d1ahdp_ 0.00018 100 56 2-58 3-59
  ZFH1_DROME 60 d1crka2 282 d1crka2 8.4 58 50 5-55 73-123
  ZFH1_DROME 60 d1dkza_ 215 d1dkza_ 4.9 59 40 1-41 112-152
  ZFH1_DROME 60 d1ecra_ 305 d1ecra_ 3.2 63 47 9-56 201-248

<b>Keywords  </b> : read_normal_sso_lines
<b>Options   </b> : a c r r2 n
             u= for upper E value limit
             l= for lower E value limit
<b>Usage     </b> : @out_refs=@{&read_machine_unreadable_sso_lines(\@SSO, $get_alignment,
                           $create_sso, $upper_expect_limit,$new_format, $lower_expect_limit,
                           $attach_range_in_names, $attach_range_in_names2)};
<b>Version   </b> : 1.3
</pre><hr>

<H3><A href="make_seq_alignment_length_even
.pl.html">make_seq_alignment_length_even
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//make_seq_alignment_length_even
.pl">make_seq_alignment_length_even
.pl</A>
<pre>
<b>Example   </b> : 
    seq3     ---lasdkfjklsdjfkldjklfj----
    seq4     dfasdfasdfadsfsadfsaas

  will result in
    seq3     ---lasdkfjklsdjfkldjklfj----
    seq4     dfasdfasdfadsfsadfsaas------

<b>Function  </b> : creates hashes with values of equal lengths.
<b>Keywords  </b> : make_alignment_length_even, equalise_seq_alignments
<b>Usage     </b> : @out=@{&make_seq_alignment_length_even(\%hash1, \%hash2)};
             @out=@{&make_seq_alignment_length_even(\%hash1)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="tempname
.pl.html">tempname
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//tempname
.pl">tempname
.pl</A>
<pre>
<b>Function  </b> : Returns a unique temporary filename.
             Reasonably robust but not completely immune to race conditions
             with other processes simultaneously requesting a tempname.
<b>Usage     </b> : $tmp=&tempname;
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="fasta_kt1_search
.pl.html">fasta_kt1_search
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//fasta_kt1_search
.pl">fasta_kt1_search
.pl</A>
<pre>
<b>Author    </b> : Sarah A. Teichmann
<b>Date      </b> : 19th September 1997
<b>Example   </b> : &fasta_kt1_search ($qdb_main, $tdb_main, $fastaver_main);
<b>Function  </b> : to search one database against the other using fasta
                ktup=1 (default is simply "fasta"). The results are stored in sub dirs
                which are from the 2 first chars of the query sequence.
<b>Keywords  </b> : fasta_search, fasta_database_search
<b>Usage     </b> : &fasta_kt1_search($query_database, $target_database, $fasta_version_to_use
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="msp_single_link_hash
.pl.html">msp_single_link_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//msp_single_link_hash
.pl">msp_single_link_hash
.pl</A>
<pre>
<b>Author    </b> : Sarah A. Teichmann with thanks to Alex Bateman
<b>Function  </b> : To make a hash with all the genes in the msp files as the keys,
             which are linked at or below the E-value threshhold,
             with the values denoting the cluster number
<b>Keywords  </b> : single_linkage, msp_single_linkage, msp_single_linkage_hash
<b>Usage     </b> : %hash=%{&msp_single_link_hash(\@msp_files, E-value);
<b>Version   </b> : 1.3
</pre><hr>

<H3><A href="print_clusfile_from_hash
.pl.html">print_clusfile_from_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//print_clusfile_from_hash
.pl">print_clusfile_from_hash
.pl</A>
<pre>
<b>Author    </b> : Sarah A. Teichmann
<b>Function  </b> : To print out a file in cluster file format from an input hash containing the genes as keys and the cluster number as values.
<b>Keywords  </b> : print_single_linkage_cluster, print_cluster_file
<b>Usage     </b> : &print_clusfile_from_hash(\%hash)
<b>Version   </b> : 1.2
</pre><hr>

<H3><A href="make_clustering_summary
.pl.html">make_clustering_summary
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//make_clustering_summary
.pl">make_clustering_summary
.pl</A>
<pre>
<b>Author    </b> : Sarah A. Teichmann
<b>Date      </b> : 19th September 1997
<b>Function  </b> : to make a summary file of a sorted cluster file
<b>Keywords  </b> : summary, make_cluster_summary, subclustering summary
<b>Usage     </b> : &make_summ($sorted_cluster_file)
<b>Version   </b> : 1.5
</pre><hr>

<H3><A href="create_sorted_cluster
.pl.html">create_sorted_cluster
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//create_sorted_cluster
.pl">create_sorted_cluster
.pl</A>
<pre>
<b>Author    </b> : Sarah A. Teichmann, modified by Jong
<b>Date      </b> : 19th September 1997
<b>Function  </b> : to make a "sorted_cluster_file" from the .clu files in a directory
<b>Keywords  </b> : make_cluster_file, sort_clu_files
<b>Usage     </b> : &create_sorted_cluster
<b>Version   </b> : 1.7
</pre><hr>

<H3><A href="interm_lib_search
.pl.html">interm_lib_search
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//interm_lib_search
.pl">interm_lib_search
.pl</A>
<pre>
<b>Example   </b> : &interm_lib_search(\@file, $over_write, $msp_directly_opt, $create_sso, $single_big_msp);
<b>Function  </b> : self_to_self input database search with reverse query as an option
<b>Keywords  </b> : do_interm_lib_search, self_self_sequence_search, self_self_seq_search,
             self_to_self_search, self_to_rev_self_search, self_to_reversed_self_search,
<b>Options   </b> : 
             Query_seqs=  for enquiry sequences eg)  "Query_seqs=$ref_of_hash"
             DB=   for target DB  "DB=$DB_used"
             File= to get file base(root) name.  "File=$file[0]"
             m  for MSP format directly from FASTA or Ssearch result than through sso_to_msp to save mem
             s  for the big single output (msp file output I mean)
             o  for overwrite existing xxxx.fa files for search
             c  for create SSO file (sequence search out file)
             r  for reverse the query sequence
             b  for doing in batch. Reads all the seqs in memory at one time
             m10 for machine readable form
             k= for k-tuple value. default is 1 (ori. FASTA prog. default is 2)
             u= for $upper_expect_limit
             l= for $lower_expect_limit
             a= for choosing either fasta or ssearch algorithm
             d  for $make_gz_in_sub_dir_opt, putting resultant sso files in gz format and in single char subdir
             D  for $make_msp_in_sub_dir_opt, convert sso to msp and put in sub dir like /D/, /S/
             n  for new format (msp2 format)
       FILE_AGE for defining the age of file in days to be overwritten.

<b>Usage     </b> : &interm_lib_search(\@file, $over_write, $msp_directly_opt, $create_sso, $single_big_msp);
             &interm_lib_search(\%seq,  $over_write, $msp_directly_opt, $create_sso, $single_big_msp);
<b>Version   </b> : 1.8
</pre><hr>

<H3><A href="geanfammer
.pl.html">geanfammer
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//geanfammer
.pl">geanfammer
.pl</A>
<pre>
<b>Author    </b> : Sarah A Teichmann, Jong Park, sat@mrc-lmb.cam.ac.uk,
                                      jong@salt2.med.harvard.edu
<b>Example   </b> : geanfammer.pl E_gnme.fa             # simplest form
            geanfammer.pl E_gnme.fa a=ssearch   # use SSEARCH
            geanfammer.pl E_gnme.fa o           # for overwriting
                                                   when you want a
                                                   fresh run ovr old
            geanfammer.pl E_gnme.fa c         # For keeping
                                                 SSO files
                                                 (fasta output)
            geanfammer.pl E_gnme.fa k=2       # changing default
                                                 k tuple for
                                                 FASTA to 2
            geanfammer.pl E_gnme.fa E=0.01     # set the E value
                                                 for initial single
                                                 linkage clustering
            geanfammer.pl E_gnme.fa e=0.01    # set the E value
                                                for domain level linkage
       -->  geanfammer.pl E_gnme.fa e=0.01 E=0.01 # set the 2 E values
                                                    separately (no need
                                                    to do this)

<b>Function  </b> : Creates a domain level clustering file from a given
              FASTA format sequence DB. It has been used for complete
              genome sequence analysis.

              ------------ USAGE INFORMATION -------------------
             The parameters you put are important for the meaningful
               protein family maker.
             The most important one is the E and e options (Mostly,
               they can have same value).
             Large E is for setting the threshold for the single
               linkage clustering.
             This means, any sequence hit BELOW the threshold
               (which is good ) will be linked.
             For example, if Seq1 matched with Seq2 with E value
              of FASTA search:
              0.001, and you set the threshold 0.1, then YOU
              ordered the geanfammer to regard them a family.

             The second small e option is for the dividing a complex
              and wrong cluster into correct more correct
              duplication modules. This is necessary as a
              lot of multidomain proteins can be clustered together
              WRONGLY by single linkage.
             At this stage, the e value is irrelevant to E value
              and you can set a higher or lower one. Or you can set
              the same as E.

             Rough guide from our experience for E and e values:
              We know that with 1000 sequence database, 0.01
              produces around 1% error in grouping sequences
              according to the evalue.
              With 180,000, 0.081 gave us less than 1% error.
             Evalue of FASTA and SSEARCH is DEPENDENT on DB size,
              so you need to play a little bit to know the best
              E value for your database or genome.
             The best approach is :
               1) You run geanfammer.pl with any of your target DB
                  with certain E value you like
               2) Check sequence families which are clustered
                  in the final resultant file xxxx.gclu and decide
                  if the E value is low or high. Lower evalues will
                  make sure you do not make wrong clusters while
                  high evalue will include more probable sequence
                  family members.
               3) Put all the xxxx.msp files in subdirectory(s)
                  created by geanfammer and run divclus.pl (which
                  is accompanied in the package) with different
                  Evalues. Divclus will not run any search algorithm
                  etc, so it can be done fairly quickly.

<b>Keywords  </b> : genome_analysis_and_protein_family_maker,
             genome_ana_protein_fam_maker
<b>Options   </b> : 
             o  for overwrite existing xxxx.fa files for search
             c  for create SSO file (sequence search out file)
             d  for very simple run and saving the result in
                    xxxx.gz format in sub dir starting with one char
             N
             s
             m
             v
             z
             D
             y  for dynamic factor
             L  for Lean output(removes all the intermediate
                                     outputs to save space)
             u  for making separate summary file (redundant now)

             DB=
             File=
             k= for k-tuple value. default is 1 (ori. FASTA prog.
                                                   default is 2)
             a= for choosing either fasta or ssearch algorithm
             E= for Evalue cutoff for single linkage clustering
                    $E_cut_main
             e= for Evalue cutoff for divide_clusters subroutine.
             u=
             l=
             d=

   !! Do not remove the following lines down to # Author line.
                This program parses them

  $Lean_output=L           by L -L
  $dynamic_factor=y        by y  Y -y -Y
  $over_write=o            by o -o
  $create_sso_file=c       by c -c
  $k_tuple=                by k=
  $upper_expect_limit=     by u=
  $lower_expect_limit=     by l=
  $algorithm=              by a=
  $No_processing=N         by N -N
  $single_msp=s            by s -s
  $sequence_db_fasta=      by DB=
  $query_file=             by File=
  $machine_readable=M      by M -M
  $make_subdir_out=D       by D
  $make_subdir_gzipped=d   by d -d
  $direct_MSP_conversion=m by m -m
  $verbose=v               by v -v
  $sub_dir_size=           by d=
  $Evalue_cut_single_link= by E=
  $Evalue_cut_divclus=     by e=
  $optimize=z              by z -z
  $make_separate_summary=u by u -u
  $length_thresh=       by T=

<b>Usage     </b> : &geanfammer(\@your_genome_or_db_to_analyse_file,
                          $verbose);

<b>Version   </b> : 2.5
</pre><hr>

<H3><A href="ISS_server
.pl.html">ISS_server
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//ISS_server
.pl">ISS_server
.pl</A>
<pre>
<b>Function  </b> : This does ISS and makes HTML file to return to HTTPD server
<b>Keywords  </b> : intermdediate_sequence_search_server
<b>Options   </b> : z S s
<b>Usage     </b> : &ISS_server(\%seq, "e=$E_val", "k=$ktuple", "a=$algorithm", "t=$leng_thresh",
             "$which_score", $show_raw_result, $segged_ISSL );
<b>Version   </b> : 1.3
</pre><hr>

<H3><A href="import_ENV_vars
.pl.html">import_ENV_vars
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//import_ENV_vars
.pl">import_ENV_vars
.pl</A>
<pre>
<b>Function  </b> : You can use any ENV set variables directly in your
             program. So, you can say $USER instead of $ENV{'USER'}
<b>Keywords  </b> : import_Env_vars, import_ENV_variables
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="read_any_dir_for_dir
.pl.html">read_any_dir_for_dir
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//read_any_dir_for_dir
.pl">read_any_dir_for_dir
.pl</A>
<pre>
<b>Argument  </b> : takes one or more scaler references.
<b>Function  </b> : read any dir and REMOVES the '.' and '..' entries. And
            then put in array.
<b>Keywords  </b> : read_any_dir_for_dir_command
<b>Returns   </b> : one ref. of array.
<b>Usage     </b> : @file_list = @{&read_any_dir(\$absolute_path_dir_name, ....)};
<b>Version   </b> : 1.1
<b>Warning   </b> : This does not report '.', '..', '#xxxx', ',xxxx', etc. only legitimate
: file and dir names are reported.
</pre><hr>

<H3><A href="produce_random_numbers
.pl.html">produce_random_numbers
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//produce_random_numbers
.pl">produce_random_numbers
.pl</A>
<pre>
<b>Example   </b> : @rand_nums=@{&produce_random_numbers($how_many, $range)};
              while $how_many->10, $range->10
<b>Keywords  </b> : generate_random_numbers, make_random_numbers, create_random_numbers
             random_numbers, get_random_numbers
<b>Usage     </b> : @rand_nums=@{&produce_random_numbers($how_many, $range)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="read_seq_matrix_files
.pl.html">read_seq_matrix_files
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//read_seq_matrix_files
.pl">read_seq_matrix_files
.pl</A>
<pre>
<b>Function  </b> : Makes similarrity matrix hash(reflexive, so it has AT as well as TA)
             %matrix looks like this:  $matrix{X}{Y}= 4
<b>Keywords  </b> : get_2D_aa_matrix, read_seq_matrix
<b>Options   </b> : 
     $reflexive_combi=r by r -r
<b>Usage     </b> : %matrix=%{&read_seq_matrix_files(\@file)};
<b>Version   </b> : 1.2
</pre><hr>

<H3><A href="attach_classification_to_pdb_seq
.pl.html">attach_classification_to_pdb_seq
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//attach_classification_to_pdb_seq
.pl">attach_classification_to_pdb_seq
.pl</A>
<pre>
<b>Keywords  </b> : attach_scop_classificaion_to_pdb_seq, add_scop_classification_to_pdb_seq
<b>Usage     </b> : &attach_classification_to_pdb_seq(\%hash_classification, \%correcting_pairs, \@files, $over_write);
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="put_slash_before_special_chars
.pl.html">put_slash_before_special_chars
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//put_slash_before_special_chars
.pl">put_slash_before_special_chars
.pl</A>
<pre>
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="split_file_by_string
.pl.html">split_file_by_string
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//split_file_by_string
.pl">split_file_by_string
.pl</A>
<pre>
<b>Keywords  </b> : split_file_by_string  divide_file_by_string
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="check_input_file_extension
.pl.html">check_input_file_extension
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//check_input_file_extension
.pl">check_input_file_extension
.pl</A>
<pre>
<b>Example   </b> : @file=@{&check_input_file_extension('msp', \@file)};
<b>Usage     </b> : @file=@{&check_input_file_extension('msp', \@file)};
             or @file=@{&check_input_file_extension('msp,nhco', \@file)};
                for multiple extension allowance
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="geanfammer_main
.pl.html">geanfammer_main
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//geanfammer_main
.pl">geanfammer_main
.pl</A>
<pre>
<b>Function  </b> : The main sub of geanfammer
<b>Usage     </b> : &geanfammer_main;
<b>Version   </b> : 1.7
</pre><hr>

<H3><A href="encrypt_passwd
.pl.html">encrypt_passwd
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//encrypt_passwd
.pl">encrypt_passwd
.pl</A>
<pre>
<b>Author    </b> : --Mark Henderson, modified by Jong
<b>Example   </b> : $crypted = ${&encrypted_passwd( $plaintext, $salt )};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="detect_file_format_type
.pl.html">detect_file_format_type
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//detect_file_format_type
.pl">detect_file_format_type
.pl</A>
<pre>
<b>Author    </b> : jong
<b>Example   </b> : $crypted = ${&encrypted_passwd( $plaintext, $salt )};
<b>Usage     </b> : $file_type=${&detect_file_format_type($file[$i])};
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="get_first_seq_in_alignment
.pl.html">get_first_seq_in_alignment
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_first_seq_in_alignment
.pl">get_first_seq_in_alignment
.pl</A>
<pre>
<b>Author    </b> : jong@salt2.med.harvard.edu
<b>Usage     </b> : $seq_name=${&get_first_seq_in_alignment($file)};
<b>Version   </b> : 1.1
</pre><hr>

<H3><A href="find_program_in_path
.pl.html">find_program_in_path
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//find_program_in_path
.pl">find_program_in_path
.pl</A>
<pre>
<b>Author    </b> : Jong, jong@salt2.med.harvard.edu
<b>Keywords  </b> : which, whence
<b>Version   </b> : 1.2
</pre><hr>

<H3><A href="do_psi_blast_search
.pl.html">do_psi_blast_search
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//do_psi_blast_search
.pl">do_psi_blast_search
.pl</A>
<pre>
<b>Options   </b> : 
    $source_DB_file= by d= s=
    $input_seq_file= by i=
    $Eval_limit= by E=
    $iteration_limit= by j=
    $step_evalue= by h= e=
    $over_write=o by o
    $make_msp_in_sub_dir_opt=D by D
<b>Usage     </b> : &do_psi_blast_search(\@files, "d=$source_DB_file",
                     "i=$input_seq_file",
                     $over_write,
                     $make_msp_in_sub_dir_opt);
<b>Version   </b> : 1.3
</pre><hr>

<H3><A href="merge_superfam_fasta_files_for_ISL
.pl.html">merge_superfam_fasta_files_for_ISL
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//merge_superfam_fasta_files_for_ISL
.pl">merge_superfam_fasta_files_for_ISL
.pl</A>
<pre>
<b>Author    </b> : jong@salt2.med.harvard.edu
<b>Keywords  </b> : compile_superfam_fasta_files_for_ISL
<b>Usage     </b> : &merge_superfam_fasta_files_for_ISL;
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="get_total_memory_size_in_linux
.pl.html">get_total_memory_size_in_linux
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//get_total_memory_size_in_linux
.pl">get_total_memory_size_in_linux
.pl</A>
<pre>
<b>Author    </b> : jong@salt2.med.harvard.edu
<b>Example   </b> : The /proc/meminfo file looks like this:>>>>
           total:    used:    free:  shared: buffers:  cached:
   Mem:  395735040 233975808 161759232 65953792 111476736 41345024
   Swap:  7319552   147456  7172096
   MemTotal:    386460 kB
   MemFree:     157968 kB
   MemShared:    64408 kB
   Buffers:     108864 kB
   Cached:       40376 kB
   SwapTotal:     7148 kB
   SwapFree:      7004 kB

<b>Keywords  </b> : get_memory_size_in_linux, get_mem_size
<b>Usage     </b> : $mem=${&get_total_memory_size_in_linux};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="check_file_exists_in_path
.pl.html">check_file_exists_in_path
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//check_file_exists_in_path
.pl">check_file_exists_in_path
.pl</A>
<pre>
<b>Author    </b> : jong@salt2.med.harvard.edu
<b>Category  </b> : File
<b>Function  </b> : checks if file exists in UNIX path
<b>Usage     </b> : $exist=&check_file_exists_in_path("hmmbuild");
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="compress_files_by_gzip
.pl.html">compress_files_by_gzip
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//compress_files_by_gzip
.pl">compress_files_by_gzip
.pl</A>
<pre>
<b>Author    </b> : jong@salt2.med.harvard.edu
<b>Usage     </b> : &compress_files_by_gzip('file_name_to_be_compressed');
<b>Version   </b> : 1.3
</pre><hr>

<H3><A href="check_if_sec_str_form_hash
.pl.html">check_if_sec_str_form_hash
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//check_if_sec_str_form_hash
.pl">check_if_sec_str_form_hash
.pl</A>
<pre>
<b>Author    </b> : jong@salt2.med.harvard.edu
<b>Usage     </b> : $check_sec_str_form_hash=${&check_if_sec_str_form_hash(\%sec)};
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="check_if_defined
.pl.html">check_if_defined
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//check_if_defined
.pl">check_if_defined
.pl</A>
<pre>
<b>Author    </b> : jong@salt2.med.harvard.edu
<b>Function  </b> : checks if all the args are defined
<b>Usage     </b> : $defined=&check_if_defined($var, $file);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="check_if_files_exist
.pl.html">check_if_files_exist
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//check_if_files_exist
.pl">check_if_files_exist
.pl</A>
<pre>
<b>Author    </b> : jong@salt2.med.harvard.edu
<b>Function  </b> : checks if all the args are defined
<b>Keywords  </b> : check_if_exists, check_if_file_exist
<b>Usage     </b> : $defined=&complain_if_not_defined($var, $file);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="die_if_file_not_present
.pl.html">die_if_file_not_present
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//die_if_file_not_present
.pl">die_if_file_not_present
.pl</A>
<pre>
<b>Author    </b> : jong@salt2.med.harvard.edu
<b>Function  </b> : checks if all the args are present
<b>Keywords  </b> : die_unless_present, die_unless_file_present
<b>Usage     </b> : &die_if_file_not_present($var, $file);
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="ask_for_ENV_vars
.pl.html">ask_for_ENV_vars
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//ask_for_ENV_vars
.pl">ask_for_ENV_vars
.pl</A>
<pre>
<b>Author    </b> : jong@salt2.med.harvard.edu, On commercial use issue, Email me.
<b>Function  </b> : asks for env var and write the env var to appropriate shell
             UNIX only RC file
<b>Keywords  </b> : write_ENV_vars, write_env_vars
<b>Usage     </b> : &ask_for_ENV_vars('BLAST_DIR');
<b>Version   </b> : 1.0
</pre><hr>

<H3><A href="reset_shell_environment
.pl.html">reset_shell_environment
</A></H3>Download <A href="ftp://cyrah.med.harvard.edu/pub/Perl//reset_shell_environment
.pl">reset_shell_environment
.pl</A>
<pre>
<b>Author    </b> : jong@salt2.med.harvard.edu, On commercial use issue, Email me.
<b>Version   </b> : 1.0
</pre><hr>

</html>
